// @instructure/ui-motion/es/components/Transition/index.js
var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _dec, _class, _class2, _temp;

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import React, { Component } from 'react';
import PropTypes from 'prop-types';
import themeable from '@instructure/ui-themeable';
import ms from '@instructure/ui-utils/lib/ms';

import BaseTransition from './BaseTransition';

var styles = {
  template: function template(theme) {
    var tmpl = function tmpl() {
      return '/*  imported from styles.css  */\n\n.WWxSZ4D {\n  /* trigger hardware acceleration for smoother animation */\n  transform: translate3d(0, 0, 0);\n}\n\n/* Animation type: fade */\n\n._3IJnweT {\n  transition:\n    opacity ' + theme.duration + ' ' + theme.timing + ',\n    transform ' + theme.duration + ' ' + theme.timing + ';\n}\n\n._3WQUS3q,\n._2-QUk43 {\n  opacity: 0.01;\n}\n\n._2kFy8hs,\n._2B204HA {\n  opacity: 1;\n}\n\n/* Animation type: scale */\n\n._2NMEcn7 {\n  transition:\n    opacity ' + theme.duration + ' ' + theme.timing + ',\n    transform ' + theme.duration + ' ' + theme.timing + ';\n}\n\n._13z9WE7,\n._3C1TKcM {\n  transform: scale(0.01) translate3d(0, 0, 0);\n  opacity: 0.01;\n}\n\n._2mKMJJJ,\n._2Xc85g8 {\n  transform: scale(1) translate3d(0, 0, 0);\n  opacity: 1;\n}\n\n/* Animation type: slide */\n\n/*\n  Note: the transitions for slide are:\n  from EXITED to ENTERING and from ENTERED to EXITING.\n\n  ENTERED and EXITED will be set directly when enter or exit transitions are disabled and they reset the \'stage\'\n  for the next transition.\n\n  The base transition class enables/disables transitions from one state to another, so transitions should be\n  set there.\n*/\n\n._1szxJ4A,\n._1yuOGoF,\n.SoLogKz,\n.wJ1xcbL {\n  transition:\n    opacity ' + theme.duration + ' ' + theme.timing + ',\n    transform ' + theme.duration + ' ' + theme.timing + ';\n}\n\n._2tcuZNi,\n._2EQLtb2,\n._3A8n1tt,\n._3BiLdu_ {\n  opacity: 0.01;\n}\n\n.h5WHqVv,\n._2tcuZNi {\n  transform: translate3d(100%, 0, 0);\n}\n\n._3EP0PBG,\n._2EQLtb2 {\n  transform: translate3d(-100%, 0, 0);\n}\n\n._2ah23I_,\n._3A8n1tt {\n  transform: translate3d(0, -100%, 0);\n}\n\n._31fS9Up,\n._3BiLdu_ {\n  transform: translate3d(0, 100%, 0);\n}\n\n._1w1tIs0,\n._36H-sw4,\n.E86cpKk,\n._2GcsBWE,\n._2OLQ9h7,\n._2qjbx_k,\n._1_JpX2G,\n._19mKDPN {\n  transform: translate3d(0, 0, 0);\n  opacity: 1;\n}\n';
    };

    return tmpl.call(theme, theme);
  },
  'root': 'WWxSZ4D',
  'fade': '_3IJnweT',
  'fade--exiting': '_3WQUS3q',
  'fade--exited': '_2-QUk43',
  'fade--entering': '_2kFy8hs',
  'fade--entered': '_2B204HA',
  'scale': '_2NMEcn7',
  'scale--exiting': '_13z9WE7',
  'scale--exited': '_3C1TKcM',
  'scale--entering': '_2mKMJJJ',
  'scale--entered': '_2Xc85g8',
  'slide-left': '_1szxJ4A',
  'slide-right': '_1yuOGoF',
  'slide-up': 'SoLogKz',
  'slide-down': 'wJ1xcbL',
  'slide-right--exited': '_2tcuZNi',
  'slide-left--exited': '_2EQLtb2',
  'slide-up--exited': '_3A8n1tt',
  'slide-down--exited': '_3BiLdu_',
  'slide-right--exiting': 'h5WHqVv',
  'slide-left--exiting': '_3EP0PBG',
  'slide-up--exiting': '_2ah23I_',
  'slide-down--exiting': '_31fS9Up',
  'slide-left--entering': '_1w1tIs0',
  'slide-right--entering': '_36H-sw4',
  'slide-up--entering': 'E86cpKk',
  'slide-down--entering': '_2GcsBWE',
  'slide-left--entered': '_2OLQ9h7',
  'slide-right--entered': '_2qjbx_k',
  'slide-up--entered': '_1_JpX2G',
  'slide-down--entered': '_19mKDPN'
};

import theme from './theme';

/**
---
category: components/utilities
---
**/
var Transition = (_dec = themeable(theme, styles), _dec(_class = (_temp = _class2 = function (_Component) {
  _inherits(Transition, _Component);

  function Transition() {
    _classCallCheck(this, Transition);

    return _possibleConstructorReturn(this, (Transition.__proto__ || Object.getPrototypeOf(Transition)).apply(this, arguments));
  }

  _createClass(Transition, [{
    key: 'render',
    value: function render() {
      var _props = this.props,
          type = _props.type,
          children = _props.children,
          props = _objectWithoutProperties(_props, ['type', 'children']);

      var duration = ms(this.theme.duration);

      var classNames = type ? {
        'exited': styles[type + '--exited'],
        'exiting': styles[type + '--exiting'],
        'entering': styles[type + '--entered'],
        'entered': styles[type + '--entering']
      } : {};

      return React.createElement(
        BaseTransition,
        Object.assign({}, props, {
          enterDelay: duration,
          exitDelay: duration,
          transitionClassName: styles[type],
          exitedClassName: classNames.exited,
          exitingClassName: classNames.exiting,
          enteredClassName: classNames.entering,
          enteringClassName: classNames.entered
        }),
        children
      );
    }
  }]);

  Transition.displayName = 'Transition'
  ;
  return Transition;
}(Component), _class2.propTypes = {
  type: PropTypes.oneOf(['fade', 'scale', 'slide-down', 'slide-up', 'slide-left', 'slide-right']),
  /**
   * A single element to animate in and out
   */
  children: PropTypes.element,
  /**
   * Show the component; triggers the enter or exit animation
   */
  in: PropTypes.bool,
  /**
   * Unmount the component (remove it from the DOM) when it is not shown
   */
  unmountOnExit: PropTypes.bool,
  /**
   * Run the enter animation when the component mounts, if it is initially
   * shown
   */
  transitionOnMount: PropTypes.bool,
  /**
   * Run the enter animation
   */
  transitionEnter: PropTypes.bool,
  /**
   * Run the exit animation
   */
  transitionExit: PropTypes.bool,
  /**
   * Callback fired when transitioning to the next state
   */
  onTransition: PropTypes.func,
  /**
   * Callback fired before the "entering" classes are applied
   */
  onEnter: PropTypes.func,
  /**
   * Callback fired after the "entering" classes are applied
   */
  onEntering: PropTypes.func,
  /**
   * Callback fired after the "enter" classes are applied
   */
  onEntered: PropTypes.func,
  /**
   * Callback fired before the "exiting" classes are applied
   */
  onExit: PropTypes.func,
  /**
   * Callback fired after the "exiting" classes are applied
   */
  onExiting: PropTypes.func,
  /**
   * Callback fired after the "exited" classes are applied
   */
  onExited: PropTypes.func
}, _class2.defaultProps = {
  type: 'fade',
  in: false,
  unmountOnExit: false,
  transitionOnMount: false,
  transitionEnter: true,
  transitionExit: true,

  onEnter: function onEnter() {},
  onEntering: function onEntering() {},
  onEntered: function onEntered() {},
  onExit: function onExit() {},
  onExiting: function onExiting() {},
  onExited: function onExited() {},
  onTransition: function onTransition(toState, fromState) {}
}, _class2.states = BaseTransition.states, _temp)) || _class);
export { Transition as default };
