// @instructure/ui-utils/es/index.js
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

export * from './react';
export * from './dom';
export * from './i18n';

import _Browser from './Browser';
export { _Browser as Browser };
import _capitalizeFirstLetter from './capitalizeFirstLetter';
export { _capitalizeFirstLetter as capitalizeFirstLetter };
import _debounce from './debounce';
export { _debounce as debounce };
import _Decimal from './Decimal';
export { _Decimal as Decimal };
import _deepEqual from './deepEqual';
export { _deepEqual as deepEqual };
import _isEmpty from './isEmpty';
export { _isEmpty as isEmpty };
import _mergeDeep from './mergeDeep';
export { _mergeDeep as mergeDeep };
import _ms from './ms';
export { _ms as ms };
import _parseUnit from './parseUnit';
export { _parseUnit as parseUnit };
import _px from './px';
export { _px as px };
import _shallowEqual from './shallowEqual';
export { _shallowEqual as shallowEqual };
import _uid from './uid';
export { _uid as uid };
import _warning from './warning';
export { _warning as warning };
import _within from './within';
export { _within as within };
