// @instructure/ui-buttons/es/components/Button/index.js
var _typeof = typeof Symbol === "function" && typeof Symbol.iterator === "symbol" ? function (obj) { return typeof obj; } : function (obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _dec, _class, _class2, _temp2;

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import keycode from 'keycode';

import themeable from '@instructure/ui-themeable';
import CustomPropTypes from '@instructure/ui-utils/lib/react/CustomPropTypes';
import ThemeablePropTypes from '@instructure/ui-themeable/lib/utils/ThemeablePropTypes';
import { omitProps } from '@instructure/ui-utils/lib/react/passthroughProps';
import getElementType from '@instructure/ui-utils/lib/react/getElementType';
import isActiveElement from '@instructure/ui-utils/lib/dom/isActiveElement';
import findDOMNode from '@instructure/ui-utils/lib/dom/findDOMNode';
import hasVisibleChildren from '@instructure/ui-a11y/lib/utils/hasVisibleChildren';
import warning from '@instructure/ui-utils/lib/warning';

import View from '@instructure/ui-layout/lib/components/View';
import Flex, { FlexItem } from '@instructure/ui-layout/lib/components/Flex';

var styles = {
  template: function template(theme) {
    var tmpl = function tmpl() {
      return '/*  imported from styles.css  */\n\n/* stylelint-disable selector-max-class, no-descending-specificity */\n\n/*\n\nNOTE ABOUT ICON STYLES: If you make changes to the Button icon styles,\nbe sure you are not breaking the following deprecated use cases where\nthe icon is inserted as a child:\n\n<Button>\n  <SomeIcon />\n  Button label\n</Button>\n\n<Button variant=[circle-primary, circle-danger, icon, icon-inverse]>\n  <SomeIcon />\n</Button>\n\n*/\n\n._2-4jKaA[type="button"],\n._2-4jKaA[type="reset"],\n._2-4jKaA[type="submit"],\n._2-4jKaA {\n  -webkit-appearance: none;\n     -moz-appearance: none;\n          appearance: none;\n}\n\n._2-4jKaA::-moz-focus-inner {\n  border: 0; /* removes default dotted focus outline in Firefox */\n}\n\n._2-4jKaA {\n  animation: none 0s ease 0s 1 normal none running;\n  backface-visibility: visible;\n  background: transparent none repeat 0 0 / auto auto padding-box border-box scroll;\n  border: medium none currentColor;\n  border-collapse: separate;\n  border-image: none;\n  border-radius: 0;\n  border-spacing: 0;\n  bottom: auto;\n  box-shadow: none;\n  box-sizing: content-box;\n  caption-side: top;\n  clear: none;\n  clip: auto;\n  color: #000;\n  columns: auto;\n  column-count: auto;\n  column-fill: balance;\n  column-gap: normal;\n  column-rule: medium none currentColor;\n  column-span: 1;\n  column-width: auto;\n  content: normal;\n  counter-increment: none;\n  counter-reset: none;\n  cursor: auto;\n  direction: ltr;\n  display: inline;\n  empty-cells: show;\n  float: none;\n  font-family: serif;\n  font-size: medium;\n  font-style: normal;\n  font-variant: normal;\n  font-weight: normal;\n  font-stretch: normal;\n  line-height: normal;\n  height: auto;\n  hyphens: none;\n  left: auto;\n  letter-spacing: normal;\n  list-style: disc outside none;\n  margin: 0;\n  max-height: none;\n  max-width: none;\n  min-height: 0;\n  min-width: 0;\n  opacity: 1;\n  orphans: 2;\n  outline: medium none invert;\n  overflow: visible;\n  overflow-x: visible;\n  overflow-y: visible;\n  padding: 0;\n  page-break-after: auto;\n  page-break-before: auto;\n  page-break-inside: auto;\n  perspective: none;\n  perspective-origin: 50% 50%;\n  position: static;\n  right: auto;\n  tab-size: 8;\n  table-layout: auto;\n  text-align: left;\n  text-align-last: auto;\n  text-decoration: none;\n  text-indent: 0;\n  text-shadow: none;\n  text-transform: none;\n  top: auto;\n  transform: none;\n  transform-origin: 50% 50% 0;\n  transform-style: flat;\n  transition: none 0s ease 0s;\n  unicode-bidi: normal;\n  vertical-align: baseline;\n  visibility: visible;\n  white-space: normal;\n  widows: 2;\n  width: auto;\n  word-spacing: normal;\n  z-index: auto;\n  all: initial;\n  box-sizing: border-box;\n  max-width: 100%;\n  display: inline-block;\n  vertical-align: middle;\n  position: relative;\n  overflow: visible;\n  direction: inherit;\n  margin: 0;\n  text-decoration: none; /* for links styled as buttons */\n  cursor: pointer;\n  -webkit-user-select: none;\n     -moz-user-select: none;\n      -ms-user-select: none;\n          user-select: none;\n  touch-action: manipulation;\n  transition: background 0.2s, transform 0.2s;\n  transform: ' + theme.transform + ';\n  font-family: ' + theme.fontFamily + ';\n  font-weight: ' + theme.fontWeight + ';\n  text-transform: ' + theme.textTransform + ';\n  letter-spacing: ' + theme.letterSpacing + ';\n  border-style: ' + theme.borderStyle + ';\n  border-width: ' + theme.borderWidth + ';\n}\n\n._2-4jKaA._1cMAmY6 {\n    cursor: not-allowed;\n    pointer-events: none;\n    opacity: 0.5;\n  }\n\n._2-4jKaA:not(.BOknYR-)::before,\n  ._2-4jKaA.BOknYR-._3HojqHd::before { /* still use focus ring on link variant if just an icon */\n    content: "";\n    position: absolute;\n    top: -0.25rem;\n    left: -0.25rem; /* stylelint-disable-line property-blacklist */\n    right: -0.25rem; /* stylelint-disable-line property-blacklist */\n    bottom: -0.25rem;\n    border: ' + theme.focusBorder + ';\n    opacity: 0;\n    transform: scale(0.01);\n    transition: all 0.2s;\n    pointer-events: none;\n  }\n\n._2-4jKaA:focus {\n    outline: none;\n  }\n\n._2-4jKaA:focus:not(.BOknYR-)::before,\n    ._2-4jKaA:focus.BOknYR-._3HojqHd::before {\n      opacity: 1;\n      transform: scale(1);\n    }\n\n._2-4jKaA:hover {\n    transform: ' + theme.hoverTransform + ';\n  }\n\n._3Bkp0v4 {\n  box-sizing: border-box;\n}\n\n/* Rectangular and square Buttons */\n._2Zxryu5 {\n  border-radius: ' + theme.borderRadius + ';\n}\n._2Zxryu5::before {\n    border-radius: calc(' + theme.focusBorderRadius + ' + 0.125rem);\n  }\n\n/* Circular Buttons */\n._2txJb3K {\n  border-radius: 50%;\n}\n._2txJb3K::before {\n    border-radius: 50%;\n  }\n\n/* Styles for square Buttons w/ only an icon and no visible children */\n._3HojqHd {\n  text-decoration: none;\n  text-align: center;\n  line-height: 1;\n}\n[dir="ltr"] ._3HojqHd {\n  text-align: center;\n}\n[dir="rtl"] ._3HojqHd {\n  text-align: center;\n}\n._3HojqHd:focus,\n  ._3HojqHd:hover,\n  ._3HojqHd:active {\n    text-decoration: none;\n  }\n._3HojqHd._5bPmjnA {\n    height: ' + theme.smallHeight + ';\n    width: ' + theme.smallHeight + ';\n  }\n._3HojqHd._5bPmjnA:not(._2XQ2Xur) {\n      font-size: 1rem;\n    }\n._3HojqHd._5bPmjnA ._2wHIes6 {\n      font-size: ' + theme.iconSizeSmall + ';\n    }\n._3HojqHd._3JoYA6t {\n    height: ' + theme.mediumHeight + ';\n    width: ' + theme.mediumHeight + ';\n  }\n._3HojqHd._3JoYA6t:not(._2XQ2Xur) {\n      font-size: 1.25rem;\n    }\n._3HojqHd._3JoYA6t ._2wHIes6 {\n      font-size: ' + theme.iconSizeMedium + ';\n    }\n._3HojqHd.rTmyQu0 {\n    height: ' + theme.largeHeight + ';\n    width: ' + theme.largeHeight + ';\n  }\n._3HojqHd.rTmyQu0:not(._2XQ2Xur) {\n      font-size: 1.75rem;\n    }\n._3HojqHd.rTmyQu0 ._2wHIes6 {\n      font-size: ' + theme.iconSizeLarge + ';\n    }\n._3HojqHd ._3Bkp0v4 {\n    display: flex;\n    align-items: center;\n    justify-content: center;\n    height: 100%;\n  }\n\n/* Shared styles between auto-width and fluid-width buttons */\n._1TqIZ7J._5bPmjnA, ._1JKri3R._5bPmjnA {\n    font-size: ' + theme.smallFontSize + ';\n    padding-left: ' + theme.smallPadding + '; /* stylelint-disable-line property-blacklist */\n    padding-right: ' + theme.smallPadding + '; /* stylelint-disable-line property-blacklist */\n\n    /* subtract the top and bottom border */\n    line-height: calc(' + theme.smallHeight + ' - (' + theme.borderWidth + ' * 2));\n  }\n._1TqIZ7J._5bPmjnA ._2wHIes6, ._1JKri3R._5bPmjnA ._2wHIes6 {\n      font-size: ' + theme.iconSizeSmall + ';\n    }\n._1TqIZ7J._3JoYA6t, ._1JKri3R._3JoYA6t {\n    font-size: ' + theme.mediumFontSize + ';\n    padding-left: ' + theme.mediumPadding + '; /* stylelint-disable-line property-blacklist */\n    padding-right: ' + theme.mediumPadding + '; /* stylelint-disable-line property-blacklist */\n\n    /* subtract the top and bottom border */\n    line-height: calc(' + theme.mediumHeight + ' - (' + theme.borderWidth + ' * 2));\n  }\n._1TqIZ7J._3JoYA6t ._2wHIes6, ._1JKri3R._3JoYA6t ._2wHIes6 {\n      font-size: ' + theme.iconSizeMedium + ';\n    }\n._1TqIZ7J.rTmyQu0, ._1JKri3R.rTmyQu0 {\n    font-size: ' + theme.largeFontSize + ';\n    padding-left: ' + theme.largePadding + '; /* stylelint-disable-line property-blacklist */\n    padding-right: ' + theme.largePadding + '; /* stylelint-disable-line property-blacklist */\n\n    /* subtract the top and bottom border */\n    line-height: calc(' + theme.largeHeight + ' - (' + theme.borderWidth + ' * 2));\n  }\n._1TqIZ7J.rTmyQu0 ._2wHIes6, ._1JKri3R.rTmyQu0 ._2wHIes6 {\n      font-size: ' + theme.iconSizeLarge + ';\n    }\n._1TqIZ7J ._3Bkp0v4, ._1JKri3R ._3Bkp0v4 {\n    display: block;\n  }\n._1TqIZ7J ._2wHIes6, ._1JKri3R ._2wHIes6 {\n    display: inline-block;\n    vertical-align: middle;\n    margin-bottom: 0.1875rem;\n  }\n/* hide this selector from Buttons with icon prop, without breaking existing icons-as-children */\n._1TqIZ7J:not(._2XQ2Xur) ._3Bkp0v4 > svg, ._1JKri3R:not(._2XQ2Xur) ._3Bkp0v4 > svg { /* stylelint-disable-line selector-max-type */\n    font-size: ' + theme.iconPlusTextFontSize + ';\n    -webkit-margin-end: ' + theme.iconPlusTextMargin + ';\n            margin-inline-end: ' + theme.iconPlusTextMargin + ';\n    -webkit-margin-start: auto;\n            margin-inline-start: auto;\n    display: inline-block;\n    vertical-align: middle;\n    transform: translateY(-0.0625rem);\n  }\n[dir="ltr"] ._1TqIZ7J:not(._2XQ2Xur) ._3Bkp0v4 > svg,\n[dir="ltr"] ._1JKri3R:not(._2XQ2Xur) ._3Bkp0v4 > svg {\n    margin-right: ' + theme.iconPlusTextMargin + ';\n    margin-left: auto;\n  }\n[dir="rtl"] ._1TqIZ7J:not(._2XQ2Xur) ._3Bkp0v4 > svg,\n[dir="rtl"] ._1JKri3R:not(._2XQ2Xur) ._3Bkp0v4 > svg {\n    margin-left: ' + theme.iconPlusTextMargin + ';\n    margin-right: auto;\n  }\n\n._1TqIZ7J ._3Bkp0v4 {\n    white-space: nowrap;\n    overflow: hidden;\n    text-overflow: ellipsis;\n  }\n\n._1JKri3R {\n  display: block;\n  width: 100%;\n  text-align: start;\n}\n\n[dir="ltr"] ._1JKri3R {\n  text-align: left;\n}\n\n[dir="rtl"] ._1JKri3R {\n  text-align: right;\n}\n\n._1JKri3R._5bPmjnA ._3Bkp0v4 {\n    padding-top: 0.375rem;\n    padding-bottom: 0.3125rem;\n  }\n\n._1JKri3R._3JoYA6t ._3Bkp0v4 {\n    padding-top: 0.5625rem;\n    padding-bottom: 0.5625rem;\n  }\n\n._1JKri3R.rTmyQu0 ._3Bkp0v4 {\n    padding-top: 0.6875rem;\n    padding-bottom: 0.6875rem;\n  }\n\n._1JKri3R ._3Bkp0v4 {\n    word-wrap: break-word;\n    -webkit-hyphens: auto;\n        -ms-hyphens: auto;\n            hyphens: auto;\n    line-height: ' + theme.fluidWidthLineHeight + ';\n  }\n\n._3kVGK6p,\n._3zPeRf6 {\n  -webkit-font-smoothing: antialiased;\n  -moz-osx-font-smoothing: grayscale;\n}\n\n._3kVGK6p {\n  background: ' + theme.defaultBackground + ';\n  color: ' + theme.defaultColor + ';\n  border-color: ' + theme.defaultBorderColor + ';\n}\n\n._3kVGK6p:hover,\n  ._3kVGK6p:active {\n    background: ' + theme.defaultHoverBackground + ';\n  }\n\n._3kVGK6p:focus,\n  ._3kVGK6p:hover,\n  ._3kVGK6p:active {\n    color: ' + theme.defaultColor + ';\n  }\n\n._3kVGK6p:active {\n    background: ' + theme.defaultActiveBackground + ';\n    box-shadow: ' + theme.defaultActiveBoxShadow + ';\n  }\n\n._1ugLssK {\n  background: ' + theme.primaryBackground + ';\n  color: ' + theme.primaryColor + ';\n  border-color: ' + theme.primaryBorderColor + ';\n}\n\n._1ugLssK:hover,\n  ._1ugLssK:active {\n    background: ' + theme.primaryHoverBackground + ';\n  }\n\n._1ugLssK:focus,\n  ._1ugLssK:hover,\n  ._1ugLssK:active {\n    color: ' + theme.primaryColor + ';\n  }\n\n._1ugLssK:active {\n    background: ' + theme.primaryActiveBackground + ';\n    box-shadow: ' + theme.primaryActiveBoxShadow + ';\n  }\n\n._3e7OjmI {\n  background: ' + theme.successBackground + ';\n  color: ' + theme.successColor + ';\n  border-color: ' + theme.successBorderColor + ';\n}\n\n._3e7OjmI:hover,\n  ._3e7OjmI:active {\n    background: ' + theme.successHoverBackground + ';\n  }\n\n._3e7OjmI:focus,\n  ._3e7OjmI:hover,\n  ._3e7OjmI:active {\n    color: ' + theme.successColor + ';\n  }\n\n._3e7OjmI:active {\n    background: ' + theme.successActiveBackground + ';\n    box-shadow: ' + theme.successActiveBoxShadow + ';\n  }\n\n._1T4QmGG {\n  background: ' + theme.dangerBackground + ';\n  color: ' + theme.dangerColor + ';\n  border-color: ' + theme.dangerBorderColor + ';\n}\n\n._1T4QmGG:hover,\n  ._1T4QmGG:active {\n    background: ' + theme.dangerHoverBackground + ';\n  }\n\n._1T4QmGG:focus,\n  ._1T4QmGG:hover,\n  ._1T4QmGG:active {\n    color: ' + theme.dangerColor + ';\n  }\n\n._1T4QmGG:active {\n    background: ' + theme.dangerActiveBackground + ';\n    box-shadow: ' + theme.dangerActiveBoxShadow + ';\n  }\n\n._3VNE5ug {\n  background: ' + theme.lightBackground + ';\n  color: ' + theme.lightColor + ';\n  border-color: ' + theme.lightBorderColor + ';\n}\n\n._3VNE5ug:hover,\n  ._3VNE5ug:active {\n    background: ' + theme.lightHoverBackground + ';\n  }\n\n._3VNE5ug:focus,\n  ._3VNE5ug:hover,\n  ._3VNE5ug:active {\n    color: ' + theme.lightColor + ';\n  }\n\n._3VNE5ug:active {\n    background: ' + theme.lightActiveBackground + ';\n    box-shadow: ' + theme.lightActiveBoxShadow + ';\n  }\n\n._3zPeRf6 {\n  background: ' + theme.ghostBackground + ';\n  color: ' + theme.ghostColor + ';\n  border-style: ' + theme.ghostBorderStyle + ';\n  border-color: ' + theme.ghostBorderColor + ';\n}\n\n._3zPeRf6:hover,\n  ._3zPeRf6:active {\n    background: ' + theme.ghostHoverBackground + ';\n  }\n\n._3zPeRf6:focus,\n  ._3zPeRf6:hover,\n  ._3zPeRf6:active {\n    color: ' + theme.ghostColor + ';\n  }\n\n._3zPeRf6:active {\n    background: ' + theme.ghostActiveBackground + ';\n    box-shadow: ' + theme.ghostActiveBoxShadow + ';\n  }\n\n._4tMi6jY {\n  background: ' + theme.ghostInverseBackground + ';\n  color: ' + theme.ghostInverseColor + ';\n  border-style: ' + theme.ghostInverseBorderStyle + ';\n  border-color: ' + theme.ghostInverseBorderColor + ';\n}\n\n._4tMi6jY:hover,\n  ._4tMi6jY:active {\n    background: ' + theme.ghostInverseHoverBackground + ';\n  }\n\n._4tMi6jY:focus,\n  ._4tMi6jY:hover,\n  ._4tMi6jY:active {\n    color: ' + theme.ghostInverseColor + ';\n  }\n\n._4tMi6jY:focus::before {\n      border: ' + theme.ghostInverseFocusBorder + ';\n    }\n\n._4tMi6jY:active {\n    background: ' + theme.ghostActiveBackground + ';\n    box-shadow: ' + theme.ghostInverseActiveBoxShadow + ';\n  }\n\n.BOknYR-,\n._2R-_Ov_ {\n  background: transparent;\n  -webkit-text-decoration: ' + theme.linkTextDecoration + ';\n          text-decoration: ' + theme.linkTextDecoration + ';\n  border-color: ' + theme.linkBorderColor + ';\n}\n\n.BOknYR-:focus,\n  .BOknYR-:hover,\n  .BOknYR-:active,\n  ._2R-_Ov_:focus,\n  ._2R-_Ov_:hover,\n  ._2R-_Ov_:active {\n    text-decoration: underline;\n  }\n\n.BOknYR- {\n  color: ' + theme.linkColor + ';\n}\n\n.BOknYR-:focus,\n  .BOknYR-:hover,\n  .BOknYR-:active {\n    border-color: ' + theme.linkFocusBorderColor + ';\n    color: ' + theme.linkHoverColor + ';\n  }\n\n._2R-_Ov_ {\n  color: ' + theme.linkInverseColor + ';\n}\n\n._2R-_Ov_:focus,\n  ._2R-_Ov_:hover,\n  ._2R-_Ov_:active {\n    border-color: ' + theme.linkInverseFocusBorderColor + ';\n    color: ' + theme.linkInverseColor + ';\n  }\n\n._32_OrzU {\n  background: ' + theme.circleDefaultBackground + ';\n  color: ' + theme.circleDefaultColor + ';\n  border-color: ' + theme.circleDefaultBorderColor + ';\n}\n\n._32_OrzU:hover,\n  ._32_OrzU:active {\n    background: ' + theme.circleDefaultHoverBackground + ';\n  }\n\n._32_OrzU:focus,\n  ._32_OrzU:hover,\n  ._32_OrzU:active {\n    color: ' + theme.circleDefaultColor + ';\n  }\n\n._32_OrzU:active {\n    background: ' + theme.circleDefaultActiveBackground + ';\n    box-shadow: ' + theme.circleDefaultActiveBoxShadow + ';\n  }\n\n.lTucCEp {\n  background: ' + theme.circlePrimaryBackground + ';\n  color: ' + theme.circlePrimaryColor + ';\n  border-color: ' + theme.circlePrimaryBorderColor + ';\n}\n\n.lTucCEp:hover,\n  .lTucCEp:active {\n    background: ' + theme.circlePrimaryHoverBackground + ';\n  }\n\n.lTucCEp:focus,\n  .lTucCEp:hover,\n  .lTucCEp:active {\n    color: ' + theme.circlePrimaryColor + ';\n  }\n\n.lTucCEp:active {\n    background: ' + theme.circlePrimaryActiveBackground + ';\n    box-shadow: ' + theme.circlePrimaryActiveBoxShadow + ';\n  }\n\n._1cLDImX {\n  background: ' + theme.circleDangerBackground + ';\n  color: ' + theme.circleDangerColor + ';\n  border-color: ' + theme.circleDangerBorderColor + ';\n}\n\n._1cLDImX:hover,\n  ._1cLDImX:active {\n    background: ' + theme.circleDangerHoverBackground + ';\n  }\n\n._1cLDImX:focus,\n  ._1cLDImX:hover,\n  ._1cLDImX:active {\n    color: ' + theme.circleDangerColor + ';\n  }\n\n._1cLDImX:active {\n    background: ' + theme.circleDangerActiveBackground + ';\n    box-shadow: ' + theme.circleDangerActiveBoxShadow + ';\n  }\n\n._2K1OaYG,\n._3_gmiN7 {\n  border-color: ' + theme.iconBorderColor + ';\n}\n\n._3_gmiN7 {\n  color: ' + theme.iconInverseColor + ';\n}\n\n._3_gmiN7:focus {\n    color: ' + theme.iconInverseHoverColor + ';\n  }\n\n._3_gmiN7:focus::before {\n      border-color: ' + theme.ghostInverseBorderColor + ';\n    }\n\n._3_gmiN7:hover,\n  ._3_gmiN7:active {\n    color: ' + theme.iconInverseHoverColor + ';\n  }\n\n._2K1OaYG {\n  color: ' + theme.iconColor + ';\n}\n\n._2K1OaYG:focus,\n  ._2K1OaYG:hover,\n  ._2K1OaYG:active {\n    color: ' + theme.iconHoverColor + ';\n  }\n\n/* If icons are passed in as elements instead of functions, guarantee\nthe SVG will display block */\n._2wHIes6 > svg { /* stylelint-disable-line selector-max-type */\n  display: block;\n}\n\n/* stylelint-enable selector-max-class, no-descending-specificity */\n';
    };

    return tmpl.call(theme, theme);
  },
  'root': '_2-4jKaA',
  'disabled': '_1cMAmY6',
  'link': 'BOknYR-',
  'width--icon': '_3HojqHd',
  'content': '_3Bkp0v4',
  'borderRadius--rounded': '_2Zxryu5',
  'borderRadius--circle': '_2txJb3K',
  'small': '_5bPmjnA',
  'has-icon': '_2XQ2Xur',
  'iconSVG': '_2wHIes6',
  'medium': '_3JoYA6t',
  'large': 'rTmyQu0',
  'width--auto': '_1TqIZ7J',
  'width--fluid': '_1JKri3R',
  'default': '_3kVGK6p',
  'ghost': '_3zPeRf6',
  'primary': '_1ugLssK',
  'success': '_3e7OjmI',
  'danger': '_1T4QmGG',
  'light': '_3VNE5ug',
  'ghost-inverse': '_4tMi6jY',
  'link-inverse': '_2R-_Ov_',
  'circle-default': '_32_OrzU',
  'circle-primary': 'lTucCEp',
  'circle-danger': '_1cLDImX',
  'icon': '_2K1OaYG',
  'icon-inverse': '_3_gmiN7'
};

import theme from './theme';

/**
---
category: components
---
**/

var Button = (_dec = themeable(theme, styles), _dec(_class = (_temp2 = _class2 = function (_Component) {
  _inherits(Button, _Component);

  function Button() {
    var _ref;

    var _temp, _this, _ret;

    _classCallCheck(this, Button);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = Button.__proto__ || Object.getPrototypeOf(Button)).call.apply(_ref, [this].concat(args))), _this), _this.handleClick = function (e) {
      var _this$props = _this.props,
          disabled = _this$props.disabled,
          readOnly = _this$props.readOnly,
          onClick = _this$props.onClick;


      if (disabled || readOnly) {
        e.preventDefault();
        e.stopPropagation();
      } else if (typeof onClick === 'function') {
        onClick(e);
      }
    }, _this.handleKeyDown = function (e) {
      var _this$props2 = _this.props,
          disabled = _this$props2.disabled,
          readOnly = _this$props2.readOnly,
          onClick = _this$props2.onClick,
          href = _this$props2.href;

      // behave like a button when space key is pressed

      if (_this.elementType !== 'button' && (e.keyCode === keycode.codes.space || e.keyCode === keycode.codes.enter)) {
        e.preventDefault();
        e.stopPropagation();

        if (typeof onClick === 'function' && !disabled && !readOnly) {
          onClick(e);
        }
        if (href) {
          findDOMNode(_this._button).click(); // eslint-disable-line react/no-find-dom-node
        }
      }
    }, _temp), _possibleConstructorReturn(_this, _ret);
  }

  _createClass(Button, [{
    key: 'focus',
    value: function focus() {
      findDOMNode(this._button).focus(); // eslint-disable-line react/no-find-dom-node
    }
  }, {
    key: 'renderIcon',
    value: function renderIcon() {
      var Icon = this.props.icon;
      if (typeof this.props.icon === 'function') {
        return React.createElement(
          'span',
          { className: styles.iconSVG },
          React.createElement(Icon, { inline: false })
        );
      } else if (Icon) {
        return React.createElement(
          'span',
          { className: styles.iconSVG },
          Icon
        );
      } else {
        return null;
      }
    }
  }, {
    key: 'buttonBorderRadius',
    value: function buttonBorderRadius() {
      var circleVariants = ['circle-primary', 'circle-danger', 'circle-default'];

      if (circleVariants.indexOf(this.props.variant) !== -1) {
        return 'circle';
      } else {
        return 'rounded';
      }
    }
  }, {
    key: 'buttonWidth',
    value: function buttonWidth() {
      var _props = this.props,
          variant = _props.variant,
          fluidWidth = _props.fluidWidth;


      var squareVariants = ['circle-default', 'circle-primary', 'circle-danger', 'icon', 'icon-inverse'];

      if (this.hasOnlyIcon || squareVariants.indexOf(variant) !== -1) {
        return 'icon';
      } else if (fluidWidth) {
        return 'fluid';
      } else {
        return 'auto';
      }
    }
  }, {
    key: 'renderContent',
    value: function renderContent() {
      var _props2 = this.props,
          children = _props2.children,
          icon = _props2.icon;

      // show warning if icon is added as a child

      if (this.hasVisibleChildren) {
        React.Children.forEach(children, function (child) {
          var icon = (typeof child === 'undefined' ? 'undefined' : _typeof(child)) === 'object' && child.type.glyphName !== undefined; // eslint-disable-line no-undefined
          warning(!icon, 'Adding icons to Button as children is deprecated. Please use the icon prop instead.');
        });
      }

      if (this.hasTextAndIcon) {
        return React.createElement(
          Flex,
          { height: '100%', width: '100%' },
          React.createElement(
            FlexItem,
            { padding: '0 x-small 0 0' },
            this.renderIcon()
          ),
          React.createElement(
            FlexItem,
            { grow: true, shrink: true },
            React.createElement(
              'span',
              { className: styles.content },
              children
            )
          )
        );
      } else {
        // all other button layouts (icon only and text only)
        return React.createElement(
          'span',
          { className: styles.content },
          icon && this.renderIcon(),
          children
        );
      }
    }
  }, {
    key: 'render',
    value: function render() {
      var _classes,
          _this2 = this;

      var _props3 = this.props,
          variant = _props3.variant,
          size = _props3.size,
          disabled = _props3.disabled,
          readOnly = _props3.readOnly,
          href = _props3.href,
          type = _props3.type,
          onClick = _props3.onClick,
          buttonRef = _props3.buttonRef,
          margin = _props3.margin,
          icon = _props3.icon;


      var classes = (_classes = {}, _defineProperty(_classes, styles.root, true), _defineProperty(_classes, styles[variant], true), _defineProperty(_classes, styles[size], size), _defineProperty(_classes, styles['width--' + this.buttonWidth()], true), _defineProperty(_classes, styles['borderRadius--' + this.buttonBorderRadius()], true), _defineProperty(_classes, styles.disabled, disabled), _defineProperty(_classes, styles['has-icon'], icon), _classes);

      var props = Object.assign({}, omitProps(this.props, Object.assign({}, Button.propTypes, View.propTypes)), {
        elementRef: function elementRef(c) {
          for (var _len2 = arguments.length, args = Array(_len2 > 1 ? _len2 - 1 : 0), _key2 = 1; _key2 < _len2; _key2++) {
            args[_key2 - 1] = arguments[_key2];
          }

          _this2._button = c;
          buttonRef.apply(_this2, [c].concat(args));
        },
        className: classnames(classes),
        disabled: disabled || readOnly,
        'aria-disabled': disabled || readOnly ? 'true' : null,
        onClick: this.handleClick,
        onKeyDown: this.handleKeyDown,
        href: href,
        type: href ? null : type,
        role: onClick && this.props.as ? 'button' : null,
        tabIndex: onClick && this.props.as ? '0' : null
      });

      var ElementType = this.elementType;

      return React.createElement(
        View,
        Object.assign({}, props, {
          display: null,
          as: ElementType,
          margin: margin
        }),
        this.renderContent()
      );
    }
  }, {
    key: 'elementType',
    get: function get() {
      return getElementType(Button, this.props);
    }
  }, {
    key: 'focused',
    get: function get() {
      return isActiveElement(this._button);
    }
  }, {
    key: 'hasVisibleChildren',
    get: function get() {
      return hasVisibleChildren(this.props.children);
    }
  }, {
    key: 'hasTextAndIcon',
    get: function get() {
      return this.hasVisibleChildren && this.props.icon; // any button with an icon + text label
    }
  }, {
    key: 'hasOnlyIcon',
    get: function get() {
      return !this.hasVisibleChildren && this.props.icon; // any button with just an icon visible
    }
  }]);

  Button.displayName = 'Button'
  ;
  return Button;
}(Component), _class2.propTypes = {
  children: PropTypes.node.isRequired,
  type: PropTypes.oneOf(['button', 'submit', 'reset']),
  variant: PropTypes.oneOf(['default', 'primary', 'success', 'danger', 'light', 'ghost', 'ghost-inverse', 'link', 'link-inverse', 'circle-default', 'circle-primary', 'circle-danger', 'icon', 'icon-inverse']),
  /**
  * provides a reference to the underlying focusable (`button` or `a`) element
  */
  buttonRef: PropTypes.func,
  /**
  * the element type to render as (will be `<a>` if href is provided)
  */
  as: CustomPropTypes.elementType,
  size: PropTypes.oneOf(['small', 'medium', 'large']),
  /**
  * should the `<Button/>` fill the width of its container
  */
  fluidWidth: PropTypes.bool,
  disabled: PropTypes.bool,
  readOnly: PropTypes.bool,
  href: PropTypes.string,
  onClick: PropTypes.func,
  /**
  * Valid values are `0`, `none`, `auto`, `xxx-small`, `xx-small`, `x-small`,
  * `small`, `medium`, `large`, `x-large`, `xx-large`. Apply these values via
  * familiar CSS-like shorthand. For example: `margin="small auto large"`.
  */
  margin: ThemeablePropTypes.spacing,
  /**
  * Add an SVG icon to the button. Do not add icons directly as
  * children.
  */
  icon: PropTypes.oneOfType([PropTypes.func, PropTypes.element])
}, _class2.defaultProps = {
  as: 'button',
  type: 'button',
  variant: 'default',
  size: 'medium',
  margin: '0',
  fluidWidth: false,
  buttonRef: function buttonRef(button) {}
}, _temp2)) || _class);


export default Button;
