// @instructure/ui-forms/lib/components/FormFieldGroup/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _dec, _class, _class2, _temp; /*
                                   * The MIT License (MIT)
                                   *
                                   * Copyright (c) 2015 - present Instructure, Inc.
                                   *
                                   * Permission is hereby granted, free of charge, to any person obtaining a copy
                                   * of this software and associated documentation files (the "Software"), to deal
                                   * in the Software without restriction, including without limitation the rights
                                   * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
                                   * copies of the Software, and to permit persons to whom the Software is
                                   * furnished to do so, subject to the following conditions:
                                   *
                                   * The above copyright notice and this permission notice shall be included in all
                                   * copies or substantial portions of the Software.
                                   *
                                   * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
                                   * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
                                   * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
                                   * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
                                   * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
                                   * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
                                   * SOFTWARE.
                                   */

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames2 = require('classnames');

var _classnames3 = _interopRequireDefault(_classnames2);

var _CustomPropTypes = require('@instructure/ui-utils/lib/react/CustomPropTypes');

var _CustomPropTypes2 = _interopRequireDefault(_CustomPropTypes);

var _Grid = require('@instructure/ui-layout/lib/components/Grid');

var _Grid2 = _interopRequireDefault(_Grid);

var _uiThemeable = require('@instructure/ui-themeable');

var _uiThemeable2 = _interopRequireDefault(_uiThemeable);

var _passthroughProps = require('@instructure/ui-utils/lib/react/passthroughProps');

var _FormField = require('../FormField');

var _FormPropTypes = require('../../utils/FormPropTypes');

var _FormPropTypes2 = _interopRequireDefault(_FormPropTypes);

var _theme = require('./theme');

var _theme2 = _interopRequireDefault(_theme);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var styles = {
  template: function template(theme) {
    var tmpl = function tmpl() {
      return '/*  imported from styles.css  */\n\n._2z_Dkcc {\n  border: ' + theme.borderWidth + ' ' + theme.borderStyle + ' ' + theme.borderColor + ';\n  border-radius: ' + theme.borderRadius + ';\n  display: block;\n}\n\n._2z_Dkcc._2r8Zcdf {\n  border-color: ' + theme.errorBorderColor + ';\n  padding: ' + theme.errorFieldsPadding + ';\n}\n\n._2z_Dkcc._23PofQg {\n  opacity: 0.6;\n  cursor: not-allowed;\n  pointer-events: none;\n}\n';
    };

    return tmpl.call(theme, theme);
  },
  'fields': '_2z_Dkcc',
  'invalid': '_2r8Zcdf',
  'disabled': '_23PofQg'
};


/**
---
category: components/forms
---
**/
var FormFieldGroup = (_dec = (0, _uiThemeable2.default)(_theme2.default, styles), _dec(_class = (_temp = _class2 = function (_Component) {
  _inherits(FormFieldGroup, _Component);

  function FormFieldGroup() {
    _classCallCheck(this, FormFieldGroup);

    return _possibleConstructorReturn(this, (FormFieldGroup.__proto__ || Object.getPrototypeOf(FormFieldGroup)).apply(this, arguments));
  }

  _createClass(FormFieldGroup, [{
    key: 'renderColumns',
    value: function renderColumns() {
      return _react.Children.map(this.props.children, function (child, index) {
        return child ? _react2.default.createElement(
          _Grid.GridCol,
          { width: child.props && child.props.width ? 'auto' : null, key: index },
          child
        ) : null;
      });
    }
  }, {
    key: 'renderChildren',
    value: function renderChildren() {
      return _react2.default.createElement(
        _Grid2.default,
        {
          colSpacing: this.props.colSpacing,
          rowSpacing: this.props.rowSpacing,
          vAlign: this.props.vAlign,
          startAt: this.props.startAt || (this.props.layout === 'columns' ? 'medium' : null)
        },
        _react2.default.createElement(
          _Grid.GridRow,
          null,
          this.renderColumns()
        )
      );
    }
  }, {
    key: 'renderFields',
    value: function renderFields() {
      var _classnames;

      return _react2.default.createElement(
        'span',
        {
          key: 'fields',
          className: (0, _classnames3.default)((_classnames = {}, _defineProperty(_classnames, styles.fields, true), _defineProperty(_classnames, styles.invalid, this.invalid), _defineProperty(_classnames, styles.disabled, this.props.disabled), _classnames))
        },
        this.renderChildren()
      );
    }
  }, {
    key: 'render',
    value: function render() {
      return _react2.default.createElement(
        _FormField.FormFieldLayout,
        Object.assign({}, (0, _passthroughProps.omitProps)(this.props, FormFieldGroup.propTypes), (0, _passthroughProps.pickProps)(this.props, _FormField.FormFieldLayout.propTypes), {
          vAlign: this.props.vAlign,
          layout: this.props.layout === 'inline' ? 'inline' : 'stacked',
          label: this.props.description,
          'aria-disabled': this.props.disabled ? 'true' : null,
          'aria-invalid': this.invalid ? 'true' : null
        }),
        this.renderFields()
      );
    }
  }, {
    key: 'invalid',
    get: function get() {
      return this.props.messages && this.props.messages.findIndex(function (message) {
        return message.type === 'error';
      }) >= 0;
    }
  }]);

  FormFieldGroup.displayName = 'FormFieldGroup'
  ;
  return FormFieldGroup;
}(_react.Component), _class2.propTypes = {
  description: _propTypes2.default.node.isRequired,
  /**
  * the element type to render as
  */
  as: _CustomPropTypes2.default.elementType,
  /**
  * object with shape: `{
  * text: PropTypes.string,
  * type: PropTypes.oneOf(['error', 'hint', 'success', 'screenreader-only'])
  *   }`
  */
  messages: _propTypes2.default.arrayOf(_FormPropTypes2.default.message),
  /**
  * id for the form field messages
  */
  messagesId: _propTypes2.default.string,
  disabled: _propTypes2.default.bool,
  children: _propTypes2.default.node,
  layout: _propTypes2.default.oneOf(['stacked', 'columns', 'inline']),
  rowSpacing: _propTypes2.default.oneOf(['none', 'small', 'medium', 'large']),
  colSpacing: _propTypes2.default.oneOf(['none', 'small', 'medium', 'large']),
  vAlign: _propTypes2.default.oneOf(['top', 'middle', 'bottom']),
  startAt: _propTypes2.default.oneOf(['small', 'medium', 'large', 'x-large', null])
}, _class2.defaultProps = {
  as: 'fieldset',
  disabled: false,
  rowSpacing: 'medium',
  colSpacing: 'small',
  vAlign: 'middle'
}, _temp)) || _class);
exports.default = FormFieldGroup;
