// @instructure/ui-elements/lib/components/Heading/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _dec, _class, _class2, _temp; /*
                                   * The MIT License (MIT)
                                   *
                                   * Copyright (c) 2015 - present Instructure, Inc.
                                   *
                                   * Permission is hereby granted, free of charge, to any person obtaining a copy
                                   * of this software and associated documentation files (the "Software"), to deal
                                   * in the Software without restriction, including without limitation the rights
                                   * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
                                   * copies of the Software, and to permit persons to whom the Software is
                                   * furnished to do so, subject to the following conditions:
                                   *
                                   * The above copyright notice and this permission notice shall be included in all
                                   * copies or substantial portions of the Software.
                                   *
                                   * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
                                   * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
                                   * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
                                   * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
                                   * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
                                   * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
                                   * SOFTWARE.
                                   */


var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames2 = require('classnames');

var _classnames3 = _interopRequireDefault(_classnames2);

var _View = require('@instructure/ui-layout/lib/components/View');

var _View2 = _interopRequireDefault(_View);

var _uiThemeable = require('@instructure/ui-themeable');

var _uiThemeable2 = _interopRequireDefault(_uiThemeable);

var _CustomPropTypes = require('@instructure/ui-utils/lib/react/CustomPropTypes');

var _CustomPropTypes2 = _interopRequireDefault(_CustomPropTypes);

var _ThemeablePropTypes = require('@instructure/ui-themeable/lib/utils/ThemeablePropTypes');

var _ThemeablePropTypes2 = _interopRequireDefault(_ThemeablePropTypes);

var _deprecated = require('@instructure/ui-utils/lib/react/deprecated');

var _getElementType = require('@instructure/ui-utils/lib/react/getElementType');

var _getElementType2 = _interopRequireDefault(_getElementType);

var _passthroughProps = require('@instructure/ui-utils/lib/react/passthroughProps');

var _theme = require('./theme');

var _theme2 = _interopRequireDefault(_theme);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var styles = {
  template: function template(theme) {
    var tmpl = function tmpl() {
      return '/*  imported from styles.css  */\n\n._3yomKhw {\n  font-family: ' + theme.fontFamily + ';\n  line-height: ' + theme.lineHeight + ';\n  text-rendering: optimizeLegibility;\n  margin: 0;\n  -webkit-font-smoothing: antialiased;\n  -moz-osx-font-smoothing: grayscale;\n}\n\n._1TDJHFo {\n  font-size: ' + theme.h1FontSize + ';\n  font-weight: ' + theme.h1FontWeight + ';\n}\n\n._1RMN37K {\n  font-size: ' + theme.h2FontSize + ';\n  font-weight: ' + theme.h2FontWeight + ';\n}\n\n._2r0vFwQ {\n  font-size: ' + theme.h3FontSize + ';\n  font-weight: ' + theme.h3FontWeight + ';\n}\n\n._37A6MF1 {\n  font-size: ' + theme.h4FontSize + ';\n  font-weight: ' + theme.h4FontWeight + ';\n}\n\n._3WzYsEt {\n  font-size: ' + theme.h5FontSize + ';\n  font-weight: ' + theme.h5FontWeight + ';\n}\n\n._1DZsMhk {\n  padding-top: ' + theme.borderPadding + ';\n  border-top: ' + theme.borderWidth + ' ' + theme.borderStyle + ' ' + theme.borderColor + ';\n}\n\n._1JmQ-lq {\n  padding-bottom: ' + theme.borderPadding + ';\n  border-bottom: ' + theme.borderWidth + ' ' + theme.borderStyle + ' ' + theme.borderColor + ';\n}\n\n._2hgF6tY {\n  color: inherit;\n}\n\n.pNdBwuw {\n  color: ' + theme.primaryColor + ';\n}\n\n._3XQF40f {\n  color: ' + theme.secondaryColor + ';\n}\n\n._2vZw0ge {\n  color: ' + theme.primaryInverseColor + ';\n}\n\n.BUZLb9k {\n  color: ' + theme.secondaryInverseColor + ';\n}\n\n._3CEWmYi {\n  color: ' + theme.successColor + ';\n}\n\n._2vHBTMX {\n  color: ' + theme.warningColor + ';\n}\n\n._3P2aKEq {\n  color: ' + theme.brandColor + ';\n}\n\n._1lrl4zJ {\n  color: ' + theme.errorColor + ';\n}\n\n.vauSJH7 {\n  margin: 0;\n  font-size: inherit;\n  font-weight: inherit;\n  line-height: inherit;\n}\n\n._3iCarRG {\n  white-space: nowrap;\n  overflow: hidden;\n  text-overflow: ellipsis;\n\n  /* in case element is span or other inline element */\n  display: block;\n}\n';
    };

    return tmpl.call(theme, theme);
  },
  'root': '_3yomKhw',
  'h1': '_1TDJHFo',
  'h2': '_1RMN37K',
  'h3': '_2r0vFwQ',
  'h4': '_37A6MF1',
  'h5': '_3WzYsEt',
  'border-top': '_1DZsMhk',
  'border-bottom': '_1JmQ-lq',
  'color-inherit': '_2hgF6tY',
  'color-primary': 'pNdBwuw',
  'color-secondary': '_3XQF40f',
  'color-primary-inverse': '_2vZw0ge',
  'color-secondary-inverse': 'BUZLb9k',
  'color-success': '_3CEWmYi',
  'color-warning': '_2vHBTMX',
  'color-brand': '_3P2aKEq',
  'color-error': '_1lrl4zJ',
  'reset': 'vauSJH7',
  'ellipsis': '_3iCarRG'
};


/**
---
category: components/typography
---
**/

var Heading = (_dec = (0, _uiThemeable2.default)(_theme2.default, styles), _dec(_class = (_temp = _class2 = function (_Component) {
  _inherits(Heading, _Component);

  function Heading() {
    _classCallCheck(this, Heading);

    return _possibleConstructorReturn(this, (Heading.__proto__ || Object.getPrototypeOf(Heading)).apply(this, arguments));
  }

  _createClass(Heading, [{
    key: 'render',
    value: function render() {
      var _classnames;

      var _props = this.props,
          border = _props.border,
          children = _props.children,
          color = _props.color,
          level = _props.level,
          ellipsis = _props.ellipsis,
          margin = _props.margin,
          elementRef = _props.elementRef;


      var ElementType = (0, _getElementType2.default)(Heading, this.props, function () {
        if (level === 'reset') {
          return 'span';
        } else {
          return level;
        }
      });

      return _react2.default.createElement(
        _View2.default,
        Object.assign({}, (0, _passthroughProps.omitProps)(this.props, Object.assign({}, Heading.propTypes, _View2.default.propTypes)), {
          className: (0, _classnames3.default)((_classnames = {}, _defineProperty(_classnames, styles.root, true), _defineProperty(_classnames, styles[level], true), _defineProperty(_classnames, styles['color-' + color], color), _defineProperty(_classnames, styles['border-' + border], border !== 'none'), _defineProperty(_classnames, styles.ellipsis, ellipsis), _classnames)),
          as: ElementType,
          margin: margin,
          elementRef: elementRef
        }),
        children
      );
    }
  }]);

  Heading.displayName = 'Heading'
  ;
  return Heading;
}(_react.Component), _class2.propTypes = {
  border: _propTypes2.default.oneOf(['none', 'top', 'bottom']),
  children: _propTypes2.default.node.isRequired,
  color: (0, _deprecated.deprecatePropValues)(_propTypes2.default.oneOf(['primary', 'secondary', 'primary-inverse', 'secondary-inverse', 'inherit']), ['success', 'warning', 'error', 'brand']),
  level: _propTypes2.default.oneOf(['h1', 'h2', 'h3', 'h4', 'h5', 'reset']),
  /**
  * the element type to render as (defaults to the level)
  */
  as: _CustomPropTypes2.default.elementType,
  ellipsis: _propTypes2.default.bool,
  /**
  * Valid values are `0`, `none`, `auto`, `xxx-small`, `xx-small`, `x-small`,
  * `small`, `medium`, `large`, `x-large`, `xx-large`. Apply these values via
  * familiar CSS-like shorthand. For example: `margin="small auto large"`.
  */
  margin: _ThemeablePropTypes2.default.spacing,
  elementRef: _propTypes2.default.func
}, _class2.defaultProps = {
  border: 'none',
  color: 'inherit',
  level: 'h2',
  ellipsis: false
}, _temp)) || _class);
exports.default = Heading;
