// @instructure/ui-utils/lib/react/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.windowMessageListener = exports.safeCloneElement = exports.pickProps = exports.omitProps = exports.matchComponentTypes = exports.getElementType = exports.getDisplayName = exports.ensureSingleChild = exports.deprecated = exports.CustomPropTypes = exports.containerQuery = exports.ComponentIdentifier = undefined;

var _passthroughProps = require('./passthroughProps');

Object.defineProperty(exports, 'omitProps', {
  enumerable: true,
  get: function get() {
    return _passthroughProps.omitProps;
  }
});
Object.defineProperty(exports, 'pickProps', {
  enumerable: true,
  get: function get() {
    return _passthroughProps.pickProps;
  }
});

var _ComponentIdentifier2 = require('./ComponentIdentifier');

var _ComponentIdentifier3 = _interopRequireDefault(_ComponentIdentifier2);

var _containerQuery2 = require('./containerQuery');

var _containerQuery3 = _interopRequireDefault(_containerQuery2);

var _CustomPropTypes2 = require('./CustomPropTypes');

var _CustomPropTypes3 = _interopRequireDefault(_CustomPropTypes2);

var _deprecated2 = require('./deprecated');

var _deprecated3 = _interopRequireDefault(_deprecated2);

var _ensureSingleChild2 = require('./ensureSingleChild');

var _ensureSingleChild3 = _interopRequireDefault(_ensureSingleChild2);

var _getDisplayName2 = require('./getDisplayName');

var _getDisplayName3 = _interopRequireDefault(_getDisplayName2);

var _getElementType2 = require('./getElementType');

var _getElementType3 = _interopRequireDefault(_getElementType2);

var _matchComponentTypes2 = require('./matchComponentTypes');

var _matchComponentTypes3 = _interopRequireDefault(_matchComponentTypes2);

var _safeCloneElement2 = require('./safeCloneElement');

var _safeCloneElement3 = _interopRequireDefault(_safeCloneElement2);

var _windowMessageListener2 = require('./windowMessageListener');

var _windowMessageListener3 = _interopRequireDefault(_windowMessageListener2);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.ComponentIdentifier = _ComponentIdentifier3.default; /*
                                                              * The MIT License (MIT)
                                                              *
                                                              * Copyright (c) 2015 - present Instructure, Inc.
                                                              *
                                                              * Permission is hereby granted, free of charge, to any person obtaining a copy
                                                              * of this software and associated documentation files (the "Software"), to deal
                                                              * in the Software without restriction, including without limitation the rights
                                                              * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
                                                              * copies of the Software, and to permit persons to whom the Software is
                                                              * furnished to do so, subject to the following conditions:
                                                              *
                                                              * The above copyright notice and this permission notice shall be included in all
                                                              * copies or substantial portions of the Software.
                                                              *
                                                              * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
                                                              * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
                                                              * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
                                                              * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
                                                              * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
                                                              * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
                                                              * SOFTWARE.
                                                              */

/* list utils in alphabetical order */

exports.containerQuery = _containerQuery3.default;
exports.CustomPropTypes = _CustomPropTypes3.default;
exports.deprecated = _deprecated3.default;
exports.ensureSingleChild = _ensureSingleChild3.default;
exports.getDisplayName = _getDisplayName3.default;
exports.getElementType = _getElementType3.default;
exports.matchComponentTypes = _matchComponentTypes3.default;
exports.safeCloneElement = _safeCloneElement3.default;
exports.windowMessageListener = _windowMessageListener3.default;
