// @instructure/ui-layout/lib/components/Flex/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.FlexItem = exports.default = undefined;

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _dec, _class, _class2, _temp; /*
                                  * The MIT License (MIT)
                                  *
                                  * Copyright (c) 2015 - present Instructure, Inc.
                                  *
                                  * Permission is hereby granted, free of charge, to any person obtaining a copy
                                  * of this software and associated documentation files (the "Software"), to deal
                                  * in the Software without restriction, including without limitation the rights
                                  * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
                                  * copies of the Software, and to permit persons to whom the Software is
                                  * furnished to do so, subject to the following conditions:
                                  *
                                  * The above copyright notice and this permission notice shall be included in all
                                  * copies or substantial portions of the Software.
                                  *
                                  * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
                                  * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
                                  * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
                                  * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
                                  * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
                                  * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
                                  * SOFTWARE.
                                  */

var _FlexItem = require('./FlexItem');

Object.defineProperty(exports, 'FlexItem', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_FlexItem).default;
  }
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _uiThemeable = require('@instructure/ui-themeable');

var _uiThemeable2 = _interopRequireDefault(_uiThemeable);

var _CustomPropTypes = require('@instructure/ui-utils/lib/react/CustomPropTypes');

var _CustomPropTypes2 = _interopRequireDefault(_CustomPropTypes);

var _ThemeablePropTypes = require('@instructure/ui-themeable/lib/utils/ThemeablePropTypes');

var _ThemeablePropTypes2 = _interopRequireDefault(_ThemeablePropTypes);

var _safeCloneElement = require('@instructure/ui-utils/lib/react/safeCloneElement');

var _safeCloneElement2 = _interopRequireDefault(_safeCloneElement);

var _passthroughProps = require('@instructure/ui-utils/lib/react/passthroughProps');

var _View = require('../View');

var _View2 = _interopRequireDefault(_View);

var _FlexItem2 = _interopRequireDefault(_FlexItem);

var _theme = require('./theme');

var _theme2 = _interopRequireDefault(_theme);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var styles = {
  template: function template(theme) {
    var tmpl = function tmpl() {
      return '/*\n  imported from styles.css \n */\n\n/*\n * The MIT License (MIT)\n *\n * Copyright (c) 2015 - present Instructure, Inc.\n *\n * Permission is hereby granted, free of charge, to any person obtaining a copy\n * of this software and associated documentation files (the "Software"), to deal\n * in the Software without restriction, including without limitation the rights\n * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell\n * copies of the Software, and to permit persons to whom the Software is\n * furnished to do so, subject to the following conditions:\n *\n * The above copyright notice and this permission notice shall be included in all\n * copies or substantial portions of the Software.\n *\n * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR\n * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,\n * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE\n * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER\n * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,\n * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE\n * SOFTWARE.\n */\n\n._2NbWFwH {\n  font-family: ' + theme.fontFamily + ';\n  box-sizing: border-box;\n}\n\n._1xP_l0w {\n  flex-direction: column;\n}\n\n.vEwBUoY {\n  flex-wrap: wrap;\n}\n\n._1sMd4zL {\n  justify-content: flex-start;\n}\n\n._2OAXeOw {\n  justify-content: flex-end;\n}\n\n.RrPUSn3 {\n  justify-content: center;\n}\n\n._3pBr8bk {\n  justify-content: space-around;\n}\n\n.mFOz9rn {\n  justify-content: space-between;\n}\n\n._39O_H8Y {\n  align-items: center;\n}\n\n._35KhVWW {\n  align-items: flex-start;\n}\n\n._3m-0Hv2 {\n  align-items: flex-end;\n}\n\n.EnPQEE- {\n  align-items: stretch;\n}\n\n._1y7Z-o0 {\n  outline: 0.0625rem dashed ' + theme.debugOutlineColor + ';\n}\n';
    };

    return tmpl.call(theme, theme);
  },
  'root': '_2NbWFwH',
  'column': '_1xP_l0w',
  'wrapItems': 'vEwBUoY',
  'justifyItems--start': '_1sMd4zL',
  'justifyItems--end': '_2OAXeOw',
  'justifyItems--center': 'RrPUSn3',
  'justifyItems--space-around': '_3pBr8bk',
  'justifyItems--space-between': 'mFOz9rn',
  'alignItems--center': '_39O_H8Y',
  'alignItems--start': '_35KhVWW',
  'alignItems--end': '_3m-0Hv2',
  'alignItems--stretch': 'EnPQEE-',
  'visualDebug': '_1y7Z-o0'
};


/**
---
category: components/layout
---
**/
var Flex = (_dec = (0, _uiThemeable2.default)(_theme2.default, styles), _dec(_class = (_temp = _class2 = function (_Component) {
  _inherits(Flex, _Component);

  function Flex() {
    _classCallCheck(this, Flex);

    return _possibleConstructorReturn(this, (Flex.__proto__ || Object.getPrototypeOf(Flex)).apply(this, arguments));
  }

  _createClass(Flex, [{
    key: 'renderChildren',
    value: function renderChildren() {
      var _this2 = this;

      return _react.Children.map(this.props.children, function (child) {
        return (0, _safeCloneElement2.default)(child, Object.assign({
          visualDebug: _this2.props.visualDebug
        }, child.props, { /* child visualDebug prop should override parent */
          direction: _this2.props.direction
        }));
      });
    }
  }, {
    key: 'render',
    value: function render() {
      var _classes;

      var props = (0, _passthroughProps.omitProps)(this.props, Flex.propTypes);

      var _props = this.props,
          as = _props.as,
          direction = _props.direction,
          height = _props.height,
          inline = _props.inline,
          margin = _props.margin,
          padding = _props.padding,
          justifyItems = _props.justifyItems,
          alignItems = _props.alignItems,
          textAlign = _props.textAlign,
          visualDebug = _props.visualDebug,
          width = _props.width,
          wrapItems = _props.wrapItems;


      var classes = (_classes = {}, _defineProperty(_classes, styles.root, true), _defineProperty(_classes, styles.visualDebug, visualDebug), _defineProperty(_classes, styles.column, direction === 'column'), _defineProperty(_classes, styles['justifyItems--' + justifyItems], true), _defineProperty(_classes, styles['alignItems--' + alignItems], true), _defineProperty(_classes, styles.wrapItems, wrapItems), _classes);

      return _react2.default.createElement(
        _View2.default,
        Object.assign({}, props, {
          className: (0, _classnames2.default)(classes),
          as: as,
          display: inline ? 'inline-flex' : 'flex',
          width: width,
          height: height,
          margin: margin,
          padding: padding,
          textAlign: textAlign
        }),
        this.renderChildren()
      );
    }
  }]);

  Flex.displayName = 'Flex'
  ;
  return Flex;
}(_react.Component), _class2.propTypes = {
  /**
  * Flex only accepts FlexItem as a child
  */
  children: _CustomPropTypes2.default.Children.oneOf([_FlexItem2.default]),
  /**
  * the element type to render as
  */
  as: _CustomPropTypes2.default.elementType,
  /**
  * Sets the flex-direction to row (horizontal) or column (vertical)
  */
  direction: _propTypes2.default.oneOf(['row', 'column']),
  /**
  * Sets the height of the component's container (optional)
  */
  height: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.number]),
  /**
  * Sets the width of the component's container (optional)
  */
  width: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.number]),
  inline: _propTypes2.default.bool,
  textAlign: _propTypes2.default.oneOf(['start', 'center', 'end']),
  /**
  * Valid values are `0`, `none`, `auto`, `xxx-small`, `xx-small`, `x-small`,
  * `small`, `medium`, `large`, `x-large`, `xx-large`. Apply these values via
  * familiar CSS-like shorthand. For example: `margin="small auto large"`.
  */
  margin: _ThemeablePropTypes2.default.spacing,
  /**
  * Valid values are `0`, `none`, `xxx-small`, `xx-small`, `x-small`,
  * `small`, `medium`, `large`, `x-large`, `xx-large`. Apply these values via
  * familiar CSS-like shorthand. For example: `padding="small x-large large"`.
  */
  padding: _ThemeablePropTypes2.default.spacing,
  alignItems: _propTypes2.default.oneOf(['center', 'start', 'end', 'stretch']),
  justifyItems: _propTypes2.default.oneOf(['center', 'start', 'end', 'space-around', 'space-between']),
  /**
  * Places dashed lines around the component's borders to help debug your layout
  */
  visualDebug: _propTypes2.default.bool,
  wrapItems: _propTypes2.default.bool
}, _class2.defaultProps = {
  as: 'span',
  direction: 'row',
  justifyItems: 'start',
  alignItems: 'center',
  inline: false,
  visualDebug: false,
  wrapItems: false
}, _temp)) || _class);
exports.default = Flex;
