// @instructure/ui-elements/lib/components/Progress/ProgressBar/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _dec, _class, _class2, _temp; /*
                                   * The MIT License (MIT)
                                   *
                                   * Copyright (c) 2015 - present Instructure, Inc.
                                   *
                                   * Permission is hereby granted, free of charge, to any person obtaining a copy
                                   * of this software and associated documentation files (the "Software"), to deal
                                   * in the Software without restriction, including without limitation the rights
                                   * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
                                   * copies of the Software, and to permit persons to whom the Software is
                                   * furnished to do so, subject to the following conditions:
                                   *
                                   * The above copyright notice and this permission notice shall be included in all
                                   * copies or substantial portions of the Software.
                                   *
                                   * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
                                   * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
                                   * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
                                   * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
                                   * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
                                   * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
                                   * SOFTWARE.
                                   */


var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _View = require('@instructure/ui-layout/lib/components/View');

var _View2 = _interopRequireDefault(_View);

var _passthroughProps = require('@instructure/ui-utils/lib/react/passthroughProps');

var _uiThemeable = require('@instructure/ui-themeable');

var _uiThemeable2 = _interopRequireDefault(_uiThemeable);

var _ThemeablePropTypes = require('@instructure/ui-themeable/lib/utils/ThemeablePropTypes');

var _ThemeablePropTypes2 = _interopRequireDefault(_ThemeablePropTypes);

var _CustomPropTypes = require('@instructure/ui-utils/lib/react/CustomPropTypes');

var _CustomPropTypes2 = _interopRequireDefault(_CustomPropTypes);

var _theme = require('./theme');

var _theme2 = _interopRequireDefault(_theme);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var styles = {
  template: function template(theme) {
    var tmpl = function tmpl() {
      return '/*  imported from styles.css  */\n\n/* stylelint-disable selector-max-class, max-nesting-depth, no-descending-specificity */\n\n._1cyLshU {\n  width: 100%;\n  display: flex;\n  align-items: center;\n  font-family: ' + theme.fontFamily + ';\n  font-weight: ' + theme.fontWeight + ';\n  line-height: ' + theme.lineHeight + ';\n  font-size: ' + theme.fontSize + ';\n}\n\n._3z7wYK-[value] {\n  box-sizing: border-box;\n  -webkit-appearance: none;\n     -moz-appearance: none;\n          appearance: none;\n  width: auto;\n  flex: 1;\n  border: none;\n}\n\n._3z7wYK-[value]::-webkit-progress-value {\n    background: linear-gradient(to right, ' + theme.meterColorStart + ' 0%, ' + theme.meterColorEnd + ' 100%);\n  }\n\n._3z7wYK-[value]::-moz-progress-bar {\n    background: linear-gradient(to right, ' + theme.meterColorStart + ' 0%, ' + theme.meterColorEnd + ' 100%);\n  }\n\n._3z7wYK-[value]::-ms-fill {\n    border: none;\n    background: linear-gradient(to right, ' + theme.meterColorStart + ' 0%, ' + theme.meterColorEnd + ' 100%);\n  }\n\n._2-SBGtT {\n  line-height: 1;\n  box-sizing: border-box;\n  -webkit-padding-start: ' + theme.valuePadding + ';\n          padding-inline-start: ' + theme.valuePadding + ';\n  flex: 0 0 5.625rem;\n}\n\n[dir="ltr"] ._2-SBGtT {\n  padding-left: ' + theme.valuePadding + ';\n}\n\n[dir="rtl"] ._2-SBGtT {\n  padding-right: ' + theme.valuePadding + ';\n}\n\n._2nqqLn2 {\n  -webkit-margin-start: 0.1875rem;\n          margin-inline-start: 0.1875rem;\n  -webkit-margin-end: 0.1875rem;\n          margin-inline-end: 0.1875rem;\n}\n\n[dir="ltr"] ._2nqqLn2 {\n  margin-left: 0.1875rem;\n  margin-right: 0.1875rem;\n}\n\n[dir="rtl"] ._2nqqLn2 {\n  margin-right: 0.1875rem;\n  margin-left: 0.1875rem;\n}\n\n._25HfEei ._3z7wYK-[value] {\n    height: ' + theme.xSmallHeight + ';\n  }\n\n._25HfEei ._2-SBGtT {\n    font-size: ' + theme.xSmallValueFontSize + ';\n  }\n\n._2NkI4m5 ._3z7wYK-[value] {\n    /* product design wants = 18px */\n    height: 1.125rem;\n  }\n\n._2NkI4m5 ._2-SBGtT {\n    font-size: ' + theme.smallValueFontSize + ';\n  }\n\n._3EBh7H8 ._3z7wYK-[value] {\n    height: ' + theme.mediumHeight + ';\n  }\n\n._3EBh7H8 ._2-SBGtT {\n    font-size: ' + theme.mediumValueFontSize + ';\n  }\n\n.yIzOdev ._3z7wYK-[value] {\n    height: ' + theme.largeHeight + ';\n  }\n\n.yIzOdev ._2-SBGtT {\n    font-size: ' + theme.largeValueFontSize + ';\n  }\n\n.bJkRkQp ._3z7wYK-[value] {\n    background-color: ' + theme.trackColor + ';\n  }\n\n.bJkRkQp ._3z7wYK-[value]::-webkit-progress-bar {\n      background-color: ' + theme.trackColor + ';\n    }\n\n.bJkRkQp ._2-SBGtT {\n    color: ' + theme.color + ';\n  }\n\n._3YMLqJo ._3z7wYK-[value] {\n    background-color: ' + theme.inverseTrackColor + ';\n  }\n\n._3YMLqJo ._3z7wYK-[value]::-webkit-progress-bar {\n      background-color: ' + theme.inverseTrackColor + ';\n    }\n\n._3YMLqJo ._2-SBGtT {\n    color: ' + theme.inverseColor + ';\n  }\n\n._1QkCPQ8 ._3z7wYK-[value]::-webkit-progress-value {\n    background: linear-gradient(to right, ' + theme.doneMeterColorStart + ' 0%, ' + theme.doneMeterColorEnd + ' 100%);\n  }\n\n._1QkCPQ8 ._3z7wYK-[value]::-moz-progress-bar {\n    background: linear-gradient(to right, ' + theme.doneMeterColorStart + ' 0%, ' + theme.doneMeterColorEnd + ' 100%);\n  }\n\n._1QkCPQ8 ._3z7wYK-[value]::-ms-fill {\n    background: linear-gradient(to right, ' + theme.doneMeterColorStart + ' 0%, ' + theme.doneMeterColorEnd + ' 100%);\n  }\n\n/* stylelint-enable selector-max-class, max-nesting-depth, no-descending-specificity */\n';
    };

    return tmpl.call(theme, theme);
  },
  'root': '_1cyLshU',
  'bar': '_3z7wYK-',
  'value': '_2-SBGtT',
  'slash': '_2nqqLn2',
  'x-small': '_25HfEei',
  'small': '_2NkI4m5',
  'medium': '_3EBh7H8',
  'large': 'yIzOdev',
  'default': 'bJkRkQp',
  'inverse': '_3YMLqJo',
  'done': '_1QkCPQ8'
};


/**
---
parent: Progress
---
**/
var ProgressBar = (_dec = (0, _uiThemeable2.default)(_theme2.default, styles), _dec(_class = (_temp = _class2 = function (_Component) {
  _inherits(ProgressBar, _Component);

  function ProgressBar() {
    _classCallCheck(this, ProgressBar);

    return _possibleConstructorReturn(this, (ProgressBar.__proto__ || Object.getPrototypeOf(ProgressBar)).apply(this, arguments));
  }

  _createClass(ProgressBar, [{
    key: 'render',
    value: function render() {
      var _classes;

      var classes = (_classes = {}, _defineProperty(_classes, styles.root, true), _defineProperty(_classes, styles[this.props.size], true), _defineProperty(_classes, styles[this.props.variant], true), _defineProperty(_classes, styles.done, this.props.successColor && this.props.valueNow / this.props.valueMax >= 1), _classes);

      var _props = this.props,
          formatDisplayedValue = _props.formatDisplayedValue,
          formatValueText = _props.formatValueText,
          valueNow = _props.valueNow,
          valueMax = _props.valueMax,
          label = _props.label;


      var valueText = formatValueText(valueNow, valueMax);
      var value = typeof formatDisplayedValue === 'function' && formatDisplayedValue(valueNow, valueMax);

      /* eslint-disable jsx-a11y/no-redundant-roles, jsx-a11y/no-noninteractive-element-to-interactive-role */
      return _react2.default.createElement(
        _View2.default,
        Object.assign({}, (0, _passthroughProps.omitProps)(this.props, Object.assign({}, ProgressBar.propTypes, _View2.default.propTypes), ['animateOnMount']), {
          as: this.props.as,
          className: (0, _classnames2.default)(classes),
          margin: this.props.margin,
          elementRef: this.props.elementRef
        }),
        _react2.default.createElement('progress', {
          className: styles.bar,
          max: valueMax,
          value: valueNow,
          role: 'progressbar',
          'aria-valuetext': valueText,
          'aria-valuenow': valueNow,
          'aria-valuemax': valueMax,
          'aria-label': label
        }),
        value && _react2.default.createElement(
          'span',
          { className: styles.value, 'aria-hidden': 'true' },
          value
        )
      );
      /* eslint-enable jsx-a11y/no-redundant-roles, jsx-a11y/no-noninteractive-element-to-interactive-role */
    }
  }]);

  ProgressBar.displayName = 'ProgressBar'
  ;
  return ProgressBar;
}(_react.Component), _class2.propTypes = {
  /**
  * A label is required for accessibility
  */
  label: _propTypes2.default.string.isRequired,
  /**
  * Different-sized progress bars and circles
  */
  size: _propTypes2.default.oneOf(['x-small', 'small', 'medium', 'large']),
  /**
  * Maximum value (defaults to 100)
  */
  valueMax: _propTypes2.default.number,
  /**
  * Receives the progress of the event
  */
  valueNow: _propTypes2.default.number,
  /**
  * A function that returns the current value formatted for screen readers
  */
  formatValueText: _propTypes2.default.func,
  /**
  * A function to format the displayed value. If null the value will not display.
  */
  formatDisplayedValue: _propTypes2.default.func,
  /**
  * The bar changes to your theme's success color when complete
  */
  successColor: _propTypes2.default.bool,
  /**
  * Choose either a progress bar or circle. The `-inverse` variants are for
  * when you need the Progress component to appear on inverse backgrounds
  */
  variant: _propTypes2.default.oneOf(['default', 'inverse']),
  /**
  * Valid values are `0`, `none`, `auto`, `xxx-small`, `xx-small`, `x-small`,
  * `small`, `medium`, `large`, `x-large`, `xx-large`. Apply these values via
  * familiar CSS-like shorthand. For example: `margin="small auto large"`.
  */
  margin: _ThemeablePropTypes2.default.spacing,
  elementRef: _propTypes2.default.func,
  as: _CustomPropTypes2.default.elementType
}, _class2.defaultProps = {
  formatValueText: function formatValueText(valueNow, valueMax) {
    return valueNow + ' / ' + valueMax;
  },
  size: 'medium',
  valueMax: 100,
  valueNow: 0,
  variant: 'default',
  successColor: true,
  as: 'div'
}, _temp)) || _class);
exports.default = ProgressBar;
