// @instructure/ui-a11y/es/components/Dialog/index.js
var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _dec, _class, _class2, _temp2;

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import React, { Component } from 'react';
import PropTypes from 'prop-types';

import { omitProps } from '@instructure/ui-utils/lib/react/passthroughProps';
import findDOMNode from '@instructure/ui-utils/lib/dom/findDOMNode';
import deprecated from '@instructure/ui-utils/lib/react/deprecated';
import warning from '@instructure/ui-utils/lib/warning';
import requestAnimationFrame from '@instructure/ui-utils/lib/dom/requestAnimationFrame';
import getElementType from '@instructure/ui-utils/lib/react/getElementType';
import CustomPropTypes from '@instructure/ui-utils/lib/react/CustomPropTypes';

import FocusRegionManager from '../../utils/FocusRegionManager';

/**
---
category: components/utilities
---
**/

var Dialog = (_dec = deprecated('5.0.0', {
  applicationElement: true
}, 'Elements outside of the `<Dialog />` are now hidden from screen readers by default. ' + 'Use the `ignore` prop to specify any elements that should not be hidden'), _dec(_class = (_temp2 = _class2 = function (_Component) {
  _inherits(Dialog, _Component);

  function Dialog() {
    var _ref;

    var _temp, _this, _ret;

    _classCallCheck(this, Dialog);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = Dialog.__proto__ || Object.getPrototypeOf(Dialog)).call.apply(_ref, [this].concat(args))), _this), _this._raf = [], _this._focusRegion = null, _temp), _possibleConstructorReturn(_this, _ret);
  }

  _createClass(Dialog, [{
    key: 'componentDidMount',
    value: function componentDidMount() {
      if (this.props.open) {
        this.focus();
      }
    }
  }, {
    key: 'componentDidUpdate',
    value: function componentDidUpdate(prevProps) {
      var _props = this.props,
          open = _props.open,
          shouldContainFocus = _props.shouldContainFocus;


      if (open && !prevProps.open || open && shouldContainFocus && !prevProps.shouldContainFocus) {
        this.focus();
      } else if (!open && prevProps.open || !open && !shouldContainFocus && prevProps.shouldContainFocus) {
        this.blur();
      }

      if (this._focusRegion) {
        this._focusRegion.updateElement(this.contentElement);
      }
    }
  }, {
    key: 'componentWillUnmount',
    value: function componentWillUnmount() {
      if (this.props.open) {
        this.blur();
      }

      this._raf.forEach(function (request) {
        return request.cancel();
      });
      this._raf = [];
    }
  }, {
    key: 'focus',
    value: function focus() {
      var _this2 = this;

      var _props2 = this.props,
          open = _props2.open,
          contentElement = _props2.contentElement,
          options = _objectWithoutProperties(_props2, ['open', 'contentElement']);

      if (!open || !this.contentElement) {
        warning(false, '[Dialog] Can\'t focus a Dialog that isn\'t open.');
      }

      this._raf.push(requestAnimationFrame(function () {
        _this2._focusRegion = FocusRegionManager.focusRegion(_this2.contentElement, Object.assign({}, options));
      }));
    }
  }, {
    key: 'blur',
    value: function blur() {
      if (this._focusRegion) {
        FocusRegionManager.blurRegion(this.contentElement, this._focusRegion.id);
      }
    }
  }, {
    key: 'render',
    value: function render() {
      var _this3 = this;

      var ElementType = getElementType(Dialog, this.props);
      return this.props.open ? React.createElement(
        ElementType,
        Object.assign({}, omitProps(this.props, Dialog.propTypes), {
          ref: function ref(el) {
            _this3._root = el;
          },
          role: this.props.label ? 'dialog' : null,
          'aria-label': this.props.label,
          className: this.props.className // eslint-disable-line react/prop-types
        }),
        this.props.children
      ) : null;
    }
  }, {
    key: 'contentElement',
    get: function get() {
      var contentElement = findDOMNode(this.props.contentElement);

      if (!contentElement) {
        contentElement = findDOMNode(this._root);
      }

      return contentElement;
    }
  }, {
    key: 'focused',
    get: function get() {
      return this.contentElement && this._focusRegion && FocusRegionManager.isFocused(this.contentElement, this._focusRegion.id);
    }
  }]);

  Dialog.displayName = 'Dialog'
  ;
  return Dialog;
}(Component), _class2.propTypes = {
  /**
   * The children to be rendered within the `<Dialog />`
   */
  children: PropTypes.node,

  /**
  * The element to render as the component root, `span` by default
  */
  as: CustomPropTypes.elementType,

  display: PropTypes.oneOf(['auto', 'block', 'inline-block']),

  label: PropTypes.string,

  /**
   * Whether or not the `<Dialog />` is open
   */
  open: PropTypes.bool,

  /**
   * Function called when tab focus leaves the `<Dialog />` focusable content. This only
   * occurs when `shouldContainFocus` is set to false.
   */
  onBlur: PropTypes.func,

  onDismiss: PropTypes.func,

  /**
   * An element or a function returning an element to focus by default
   */
  defaultFocusElement: PropTypes.oneOfType([PropTypes.element, PropTypes.func]),

  /**
   * An element or a function returning an element that wraps the content of the `<Dialog />`
   */
  contentElement: PropTypes.oneOfType([PropTypes.element, PropTypes.func]),

  /**
   * An element, function returning an element, or array of elements that will not be hidden from
   * the screen reader when the `<Dialog />` is open
   */
  liveRegion: PropTypes.oneOfType([PropTypes.arrayOf(PropTypes.element), PropTypes.element, PropTypes.func]),
  shouldContainFocus: PropTypes.oneOfType([PropTypes.bool, PropTypes.oneOf(['keyboard', 'screenreader'])]),
  shouldReturnFocus: PropTypes.bool,
  shouldCloseOnDocumentClick: PropTypes.bool,
  shouldCloseOnEscape: PropTypes.bool
}, _class2.defaultProps = {
  open: false,
  shouldContainFocus: false,
  shouldReturnFocus: false,
  shouldCloseOnDocumentClick: true,
  shouldCloseOnEscape: true,
  defaultFocusElement: null,
  contentElement: null,
  liveRegion: null,
  onBlur: function onBlur(event) {},
  onDismiss: function onDismiss(event) {}
}, _temp2)) || _class);


export default Dialog;
