// @instructure/ui-portal/lib/components/Portal/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _dec, _dec2, _class, _class2, _temp; /*
                                          * The MIT License (MIT)
                                          *
                                          * Copyright (c) 2015 - present Instructure, Inc.
                                          *
                                          * Permission is hereby granted, free of charge, to any person obtaining a copy
                                          * of this software and associated documentation files (the "Software"), to deal
                                          * in the Software without restriction, including without limitation the rights
                                          * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
                                          * copies of the Software, and to permit persons to whom the Software is
                                          * furnished to do so, subject to the following conditions:
                                          *
                                          * The above copyright notice and this permission notice shall be included in all
                                          * copies or substantial portions of the Software.
                                          *
                                          * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
                                          * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
                                          * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
                                          * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
                                          * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
                                          * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
                                          * SOFTWARE.
                                          */

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _reactDom = require('react-dom');

var _reactDom2 = _interopRequireDefault(_reactDom);

var _bidirectional = require('@instructure/ui-i18n/lib/bidirectional');

var _bidirectional2 = _interopRequireDefault(_bidirectional);

var _shallowEqual = require('@instructure/ui-utils/lib/shallowEqual');

var _shallowEqual2 = _interopRequireDefault(_shallowEqual);

var _CustomPropTypes = require('@instructure/ui-utils/lib/react/CustomPropTypes');

var _CustomPropTypes2 = _interopRequireDefault(_CustomPropTypes);

var _deprecated = require('@instructure/ui-utils/lib/react/deprecated');

var _deprecated2 = _interopRequireDefault(_deprecated);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

/**
---
category: components/utilities
---
**/
var Portal = (_dec = (0, _deprecated2.default)('3.0.0', {
  container: 'mountNode',
  isOpen: 'open',
  onReady: 'onOpen'
}), _dec2 = (0, _bidirectional2.default)(), _dec(_class = _dec2(_class = (_temp = _class2 = function (_Component) {
  _inherits(Portal, _Component);

  function Portal() {
    _classCallCheck(this, Portal);

    return _possibleConstructorReturn(this, (Portal.__proto__ || Object.getPrototypeOf(Portal)).apply(this, arguments));
  }

  _createClass(Portal, [{
    key: 'componentDidMount',
    value: function componentDidMount() {
      this.renderPortal(this.props);
    }
  }, {
    key: 'shouldComponentUpdate',
    value: function shouldComponentUpdate(nextProps, nextState) {
      return !((0, _shallowEqual2.default)(this.props, nextProps) && (0, _shallowEqual2.default)(this.state, nextState));
    }
  }, {
    key: 'componentWillReceiveProps',
    value: function componentWillReceiveProps(nextProps) {
      this.renderPortal(nextProps);
    }
  }, {
    key: 'componentWillUnmount',
    value: function componentWillUnmount() {
      this.removePortal(this.props);
    }
  }, {
    key: 'render',
    value: function render() {
      return null;
    }
  }, {
    key: 'renderPortal',
    value: function renderPortal(props) {
      var _this2 = this;

      var isInitialMount = !this._node;
      var mountNode = this.mountNode;

      var children = props.children;

      // Wrap text in a span since subtree will only render a single top-level node
      if (typeof children === 'string' && children.length > 0) {
        children = _react2.default.createElement(
          'span',
          null,
          children
        );
      }

      // Render subtree if Portal is open and has children to render
      if (props.open && _react2.default.Children.count(children) > 0) {
        // Create node if it doesn't already exist
        if (!this._node) {
          this._node = document.createElement('span');
          this._node.setAttribute('dir', this.dir);
        }

        // Append node to container if it isn't already
        if (this._node.parentNode !== mountNode) {
          if (this.props.insertAt === 'bottom') {
            mountNode.appendChild(this._node);
          } else {
            mountNode.insertBefore(this._node, mountNode.firstChild);
          }
        }

        // Notify that subtree has been rendered if props ask for it
        var handleMount = function handleMount() {
          // Only fire onOpen if Portal was closed and is now open
          if ((isInitialMount || !_this2.props.open && props.open) && typeof props.onOpen === 'function') {
            props.onOpen(_this2._node);
          }
        };

        _reactDom2.default.unstable_renderSubtreeIntoContainer(this, children, this._node, handleMount);
      } else {
        this.removePortal(props);
      }
    }
  }, {
    key: 'removePortal',
    value: function removePortal(props) {
      var unmounted = void 0;

      if (this._node) {
        unmounted = _reactDom2.default.unmountComponentAtNode(this._node);
        this._node.parentNode && this._node.parentNode.removeChild(this._node);
        this._node = null;
      }

      if (unmounted && typeof props.onClose === 'function') {
        props.onClose();
      }
    }
  }, {
    key: 'mountNode',
    get: function get() {
      var mountNode = void 0;

      if (typeof this.props.mountNode === 'function') {
        mountNode = _reactDom2.default.findDOMNode(this.props.mountNode.call(null)); // eslint-disable-line react/no-find-dom-node
      } else if (this.props.mountNode) {
        mountNode = _reactDom2.default.findDOMNode(this.props.mountNode); // eslint-disable-line react/no-find-dom-node
      }

      if (!mountNode || !mountNode.nodeName) {
        mountNode = document.body;
      }

      return mountNode;
    }
  }, {
    key: 'node',
    get: function get() {
      return this._node;
    }
  }]);

  Portal.displayName = 'Portal'
  ;
  return Portal;
}(_react.Component), _class2.propTypes = {
  /**
   * Wheter or not the `<Portal />` is open
   */
  open: _propTypes2.default.bool,

  /**
   * Callback fired when `<Portal />` content has been mounted in the DOM
   */
  onOpen: _propTypes2.default.func,

  /**
   * Callback fired when `<Portal />` has been unmounted from the DOM
   */
  onClose: _propTypes2.default.func,

  /**
   * An element or a function returning an element to use as the mount node
   * for the `<Portal />` (defaults to `document.body`)
   */
  mountNode: _propTypes2.default.oneOfType([_CustomPropTypes2.default.element, _propTypes2.default.func]),
  /**
   * Insert the element at the 'top' of the mountNode or at the 'bottom'
   */
  insertAt: _propTypes2.default.oneOf(['bottom', 'top']),

  /**
   * The children to be rendered within the `<Portal />`
   */
  children: _propTypes2.default.node
}, _class2.defaultProps = {
  open: false,
  insertAt: 'bottom',
  onOpen: function onOpen() {},
  onClose: function onClose() {},
  mountNode: null,
  children: null
}, _temp)) || _class) || _class);
exports.default = Portal;
