// @instructure/ui-forms/es/components/NumberInput/__examples__/NumberInput.examples.js
var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import React from 'react';
/* eslint-disable instructure-ui/no-relative-package-imports */
import Select from '../../../../../ui-forms/lib/components/Select';
import View from '../../../../../ui-layout/lib/components/View';
/* eslint-disable instructure-ui/no-relative-package-imports */
import NumberInput from '../index';

export var standard = function standard() {
  return React.createElement(NumberInput, {
    label: 'Age (in years)',
    placeholder: 'Your age goes here'
  });
};

export var disabled = function disabled() {
  return React.createElement(NumberInput, {
    label: 'Age (in years)',
    placeholder: 'Your age goes here',
    disabled: true
  });
};

export var noArrows = function noArrows() {
  return React.createElement(NumberInput, {
    label: 'Age (in years)',
    placeholder: 'Your age goes here',
    showArrows: false
  });
};

export var layoutInline = function layoutInline() {
  return React.createElement(NumberInput, {
    label: 'Age (in years)',
    placeholder: 'Your age goes here',
    layout: 'inline'
  });
};

export var layoutStacked = function layoutStacked() {
  return React.createElement(NumberInput, {
    label: 'Age (in years)',
    placeholder: 'Your age goes here',
    layout: 'stacked'
  });
};

export var inline = function inline() {
  return React.createElement(NumberInput, {
    label: 'Age (in years)',
    placeholder: 'Your age goes here',
    layout: 'inline',
    width: '12rem',
    inline: true
  });
};

export var withWidth = function withWidth() {
  return React.createElement(NumberInput, {
    label: 'Age (in years)',
    placeholder: 'Your age goes here',
    width: '15rem'
  });
};

export var minNumber = function minNumber() {
  return React.createElement(NumberInput, {
    label: 'Pick a number above 7',
    placeholder: 'Your number goes here',
    min: 7
  });
};

export var maxNumber = function maxNumber() {
  return React.createElement(NumberInput, {
    label: 'Pick a number below 21',
    placeholder: 'Your number goes here',
    max: 21
  });
};

export var step = function step() {
  return React.createElement(NumberInput, {
    label: 'Pick a number divisible by 5',
    placeholder: 'Your number goes here',
    step: 5
  });
};

export var defaultValue = function defaultValue() {
  return React.createElement(NumberInput, {
    label: 'What is your favorite number?',
    placeholder: 'Your number goes here',
    defaultValue: '7'
  });
};

var LocaleExample = function (_React$Component) {
  _inherits(LocaleExample, _React$Component);

  function LocaleExample() {
    var _ref;

    var _temp, _this, _ret;

    _classCallCheck(this, LocaleExample);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = LocaleExample.__proto__ || Object.getPrototypeOf(LocaleExample)).call.apply(_ref, [this].concat(args))), _this), _this.state = { locale: 'de' }, _temp), _possibleConstructorReturn(_this, _ret);
  }

  _createClass(LocaleExample, [{
    key: 'render',
    value: function render() {
      var _this2 = this;

      var label = this.state.locale === 'de' ? "Comma separator" : "Period separator";
      return React.createElement(
        'div',
        null,
        React.createElement(
          Select,
          {
            label: 'Choose locale',
            onChange: function onChange(e, o) {
              return _this2.setState({ locale: o.value });
            } },
          React.createElement(
            'option',
            { key: 'de', value: 'de' },
            'de'
          ),
          React.createElement(
            'option',
            { key: 'en', value: 'en' },
            'en'
          )
        ),
        React.createElement(
          View,
          { padding: 'small' },
          React.createElement(NumberInput, {
            label: label,
            step: 0.1,
            min: 0.1,
            locale: this.state.locale,
            defaultValue: 2.4
          })
        )
      );
    }
  }]);

  LocaleExample.displayName = 'LocaleExample'
  ;
  return LocaleExample;
}(React.Component);

export var locale = function locale() {
  return React.createElement(LocaleExample, null);
};

export var error = function error() {
  return React.createElement(NumberInput, {
    label: 'Pick a number above 7',
    placeholder: 'Your number goes here',
    messages: [{ text: 'You must enter a number', type: 'error' }],
    required: true
  });
};

export var hint = function hint() {
  return React.createElement(NumberInput, {
    label: 'Pick a number above 7',
    placeholder: 'Your number goes here',
    messages: [{ text: 'Your number must be 7 or higher', type: 'hint' }],
    min: 7
  });
};

export var success = function success() {
  return React.createElement(NumberInput, {
    label: 'Pick a number below 21',
    placeholder: 'Your number goes here',
    messages: [{ text: 'Thanks for entering a number below 21', type: 'success' }],
    max: 21
  });
};
