// @instructure/ui-buttons/lib/components/Button/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _typeof = typeof Symbol === "function" && typeof Symbol.iterator === "symbol" ? function (obj) { return typeof obj; } : function (obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _dec, _class, _class2, _temp2; /*
                                    * The MIT License (MIT)
                                    *
                                    * Copyright (c) 2015 - present Instructure, Inc.
                                    *
                                    * Permission is hereby granted, free of charge, to any person obtaining a copy
                                    * of this software and associated documentation files (the "Software"), to deal
                                    * in the Software without restriction, including without limitation the rights
                                    * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
                                    * copies of the Software, and to permit persons to whom the Software is
                                    * furnished to do so, subject to the following conditions:
                                    *
                                    * The above copyright notice and this permission notice shall be included in all
                                    * copies or substantial portions of the Software.
                                    *
                                    * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
                                    * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
                                    * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
                                    * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
                                    * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
                                    * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
                                    * SOFTWARE.
                                    */


var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _keycode = require('keycode');

var _keycode2 = _interopRequireDefault(_keycode);

var _uiThemeable = require('@instructure/ui-themeable');

var _uiThemeable2 = _interopRequireDefault(_uiThemeable);

var _CustomPropTypes = require('@instructure/ui-utils/lib/react/CustomPropTypes');

var _CustomPropTypes2 = _interopRequireDefault(_CustomPropTypes);

var _ThemeablePropTypes = require('@instructure/ui-themeable/lib/utils/ThemeablePropTypes');

var _ThemeablePropTypes2 = _interopRequireDefault(_ThemeablePropTypes);

var _passthroughProps = require('@instructure/ui-utils/lib/react/passthroughProps');

var _getElementType = require('@instructure/ui-utils/lib/react/getElementType');

var _getElementType2 = _interopRequireDefault(_getElementType);

var _isActiveElement = require('@instructure/ui-utils/lib/dom/isActiveElement');

var _isActiveElement2 = _interopRequireDefault(_isActiveElement);

var _findDOMNode = require('@instructure/ui-utils/lib/dom/findDOMNode');

var _findDOMNode2 = _interopRequireDefault(_findDOMNode);

var _hasVisibleChildren = require('@instructure/ui-a11y/lib/utils/hasVisibleChildren');

var _hasVisibleChildren2 = _interopRequireDefault(_hasVisibleChildren);

var _warning = require('@instructure/ui-utils/lib/warning');

var _warning2 = _interopRequireDefault(_warning);

var _View = require('@instructure/ui-layout/lib/components/View');

var _View2 = _interopRequireDefault(_View);

var _Flex = require('@instructure/ui-layout/lib/components/Flex');

var _Flex2 = _interopRequireDefault(_Flex);

var _theme = require('./theme');

var _theme2 = _interopRequireDefault(_theme);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var styles = {
  template: function template(theme) {
    var tmpl = function tmpl() {
      return '/*  imported from styles.css  */\n\n/* stylelint-disable selector-max-class, no-descending-specificity */\n\n/*\n\nNOTE ABOUT ICON STYLES: If you make changes to the Button icon styles,\nbe sure you are not breaking the following deprecated use cases where\nthe icon is inserted as a child:\n\n<Button>\n  <SomeIcon />\n  Button label\n</Button>\n\n<Button variant=[circle-primary, circle-danger, icon, icon-inverse]>\n  <SomeIcon />\n</Button>\n\n*/\n\n._2pBn_SN[type="button"],\n._2pBn_SN[type="reset"],\n._2pBn_SN[type="submit"],\n._2pBn_SN {\n  -webkit-appearance: none;\n     -moz-appearance: none;\n          appearance: none;\n}\n\n._2pBn_SN::-moz-focus-inner {\n  border: 0; /* removes default dotted focus outline in Firefox */\n}\n\n._2pBn_SN {\n  animation: none 0s ease 0s 1 normal none running;\n  backface-visibility: visible;\n  background: transparent none repeat 0 0 / auto auto padding-box border-box scroll;\n  border: medium none currentColor;\n  border-collapse: separate;\n  border-image: none;\n  border-radius: 0;\n  border-spacing: 0;\n  bottom: auto;\n  box-shadow: none;\n  box-sizing: content-box;\n  caption-side: top;\n  clear: none;\n  clip: auto;\n  color: #000;\n  columns: auto;\n  column-count: auto;\n  column-fill: balance;\n  column-gap: normal;\n  column-rule: medium none currentColor;\n  column-span: 1;\n  column-width: auto;\n  content: normal;\n  counter-increment: none;\n  counter-reset: none;\n  cursor: auto;\n  direction: ltr;\n  display: inline;\n  empty-cells: show;\n  float: none;\n  font-family: serif;\n  font-size: medium;\n  font-style: normal;\n  font-variant: normal;\n  font-weight: normal;\n  font-stretch: normal;\n  line-height: normal;\n  height: auto;\n  hyphens: none;\n  left: auto;\n  letter-spacing: normal;\n  list-style: disc outside none;\n  margin: 0;\n  max-height: none;\n  max-width: none;\n  min-height: 0;\n  min-width: 0;\n  opacity: 1;\n  orphans: 2;\n  outline: medium none invert;\n  overflow: visible;\n  overflow-x: visible;\n  overflow-y: visible;\n  padding: 0;\n  page-break-after: auto;\n  page-break-before: auto;\n  page-break-inside: auto;\n  perspective: none;\n  perspective-origin: 50% 50%;\n  position: static;\n  right: auto;\n  tab-size: 8;\n  table-layout: auto;\n  text-align: left;\n  text-align-last: auto;\n  text-decoration: none;\n  text-indent: 0;\n  text-shadow: none;\n  text-transform: none;\n  top: auto;\n  transform: none;\n  transform-origin: 50% 50% 0;\n  transform-style: flat;\n  transition: none 0s ease 0s;\n  unicode-bidi: normal;\n  vertical-align: baseline;\n  visibility: visible;\n  white-space: normal;\n  widows: 2;\n  width: auto;\n  word-spacing: normal;\n  z-index: auto;\n  all: initial;\n  box-sizing: border-box;\n  max-width: 100%;\n  display: inline-block;\n  vertical-align: middle;\n  position: relative;\n  overflow: visible;\n  direction: inherit;\n  margin: 0;\n  text-decoration: none; /* for links styled as buttons */\n  cursor: pointer;\n  -webkit-user-select: none;\n     -moz-user-select: none;\n      -ms-user-select: none;\n          user-select: none;\n  touch-action: manipulation;\n  transition: background 0.2s, transform 0.2s;\n  transform: ' + theme.transform + ';\n  font-family: ' + theme.fontFamily + ';\n  font-weight: ' + theme.fontWeight + ';\n  text-transform: ' + theme.textTransform + ';\n  letter-spacing: ' + theme.letterSpacing + ';\n  border-style: ' + theme.borderStyle + ';\n  border-width: ' + theme.borderWidth + ';\n}\n\n._2pBn_SN._3ysMWZa {\n    cursor: not-allowed;\n    pointer-events: none;\n    opacity: 0.5;\n  }\n\n._2pBn_SN:not(._3-Zl5rJ)::before,\n  ._2pBn_SN._3-Zl5rJ._1LL06hd::before { /* still use focus ring on link variant if just an icon */\n    content: "";\n    position: absolute;\n    top: -0.25rem;\n    left: -0.25rem; /* stylelint-disable-line property-blacklist */\n    right: -0.25rem; /* stylelint-disable-line property-blacklist */\n    bottom: -0.25rem;\n    border: ' + theme.focusBorder + ';\n    opacity: 0;\n    transform: scale(0.01);\n    transition: all 0.2s;\n    pointer-events: none;\n  }\n\n._2pBn_SN:focus {\n    outline: none;\n  }\n\n._2pBn_SN:focus:not(._3-Zl5rJ)::before,\n    ._2pBn_SN:focus._3-Zl5rJ._1LL06hd::before {\n      opacity: 1;\n      transform: scale(1);\n    }\n\n._2pBn_SN:hover {\n    transform: ' + theme.hoverTransform + ';\n  }\n\n._95vdWYg {\n  box-sizing: border-box;\n}\n\n/* Rectangular and square Buttons */\n._30tkGIB {\n  border-radius: ' + theme.borderRadius + ';\n}\n._30tkGIB::before {\n    border-radius: calc(' + theme.focusBorderRadius + ' + 0.125rem);\n  }\n\n/* Circular Buttons */\n.xGLZogw {\n  border-radius: 50%;\n}\n.xGLZogw::before {\n    border-radius: 50%;\n  }\n\n/* Styles for square Buttons w/ only an icon and no visible children */\n._1LL06hd {\n  text-decoration: none;\n  text-align: center;\n  line-height: 1;\n}\n[dir="ltr"] ._1LL06hd {\n  text-align: center;\n}\n[dir="rtl"] ._1LL06hd {\n  text-align: center;\n}\n._1LL06hd:focus,\n  ._1LL06hd:hover,\n  ._1LL06hd:active {\n    text-decoration: none;\n  }\n._1LL06hd._1qX2Ex0 {\n    height: ' + theme.smallHeight + ';\n    width: ' + theme.smallHeight + ';\n  }\n._1LL06hd._1qX2Ex0:not(.u2Zu_fh) {\n      font-size: 1rem;\n    }\n._1LL06hd._1qX2Ex0 ._1s_Wtxo {\n      font-size: ' + theme.iconSizeSmall + ';\n    }\n._1LL06hd._4NO1WYb {\n    height: ' + theme.mediumHeight + ';\n    width: ' + theme.mediumHeight + ';\n  }\n._1LL06hd._4NO1WYb:not(.u2Zu_fh) {\n      font-size: 1.25rem;\n    }\n._1LL06hd._4NO1WYb ._1s_Wtxo {\n      font-size: ' + theme.iconSizeMedium + ';\n    }\n._1LL06hd._3NjWj4C {\n    height: ' + theme.largeHeight + ';\n    width: ' + theme.largeHeight + ';\n  }\n._1LL06hd._3NjWj4C:not(.u2Zu_fh) {\n      font-size: 1.75rem;\n    }\n._1LL06hd._3NjWj4C ._1s_Wtxo {\n      font-size: ' + theme.iconSizeLarge + ';\n    }\n._1LL06hd ._95vdWYg {\n    display: flex;\n    align-items: center;\n    justify-content: center;\n    height: 100%;\n  }\n\n/* Shared styles between auto-width and fluid-width buttons */\n.f4fthw8._1qX2Ex0, ._6Gadv9z._1qX2Ex0 {\n    font-size: ' + theme.smallFontSize + ';\n    padding-left: ' + theme.smallPadding + '; /* stylelint-disable-line property-blacklist */\n    padding-right: ' + theme.smallPadding + '; /* stylelint-disable-line property-blacklist */\n\n    /* subtract the top and bottom border */\n    line-height: calc(' + theme.smallHeight + ' - (' + theme.borderWidth + ' * 2));\n  }\n.f4fthw8._1qX2Ex0 ._1s_Wtxo, ._6Gadv9z._1qX2Ex0 ._1s_Wtxo {\n      font-size: ' + theme.iconSizeSmall + ';\n    }\n.f4fthw8._4NO1WYb, ._6Gadv9z._4NO1WYb {\n    font-size: ' + theme.mediumFontSize + ';\n    padding-left: ' + theme.mediumPadding + '; /* stylelint-disable-line property-blacklist */\n    padding-right: ' + theme.mediumPadding + '; /* stylelint-disable-line property-blacklist */\n\n    /* subtract the top and bottom border */\n    line-height: calc(' + theme.mediumHeight + ' - (' + theme.borderWidth + ' * 2));\n  }\n.f4fthw8._4NO1WYb ._1s_Wtxo, ._6Gadv9z._4NO1WYb ._1s_Wtxo {\n      font-size: ' + theme.iconSizeMedium + ';\n    }\n.f4fthw8._3NjWj4C, ._6Gadv9z._3NjWj4C {\n    font-size: ' + theme.largeFontSize + ';\n    padding-left: ' + theme.largePadding + '; /* stylelint-disable-line property-blacklist */\n    padding-right: ' + theme.largePadding + '; /* stylelint-disable-line property-blacklist */\n\n    /* subtract the top and bottom border */\n    line-height: calc(' + theme.largeHeight + ' - (' + theme.borderWidth + ' * 2));\n  }\n.f4fthw8._3NjWj4C ._1s_Wtxo, ._6Gadv9z._3NjWj4C ._1s_Wtxo {\n      font-size: ' + theme.iconSizeLarge + ';\n    }\n.f4fthw8 ._95vdWYg, ._6Gadv9z ._95vdWYg {\n    display: block;\n  }\n.f4fthw8 ._1s_Wtxo, ._6Gadv9z ._1s_Wtxo {\n    display: inline-block;\n    vertical-align: middle;\n    margin-bottom: 0.1875rem;\n  }\n/* hide this selector from Buttons with icon prop, without breaking existing icons-as-children */\n.f4fthw8:not(.u2Zu_fh) ._95vdWYg > svg, ._6Gadv9z:not(.u2Zu_fh) ._95vdWYg > svg { /* stylelint-disable-line selector-max-type */\n    font-size: ' + theme.iconPlusTextFontSize + ';\n    -webkit-margin-end: ' + theme.iconPlusTextMargin + ';\n            margin-inline-end: ' + theme.iconPlusTextMargin + ';\n    -webkit-margin-start: auto;\n            margin-inline-start: auto;\n    display: inline-block;\n    vertical-align: middle;\n    transform: translateY(-0.0625rem);\n  }\n[dir="ltr"] .f4fthw8:not(.u2Zu_fh) ._95vdWYg > svg,\n[dir="ltr"] ._6Gadv9z:not(.u2Zu_fh) ._95vdWYg > svg {\n    margin-right: ' + theme.iconPlusTextMargin + ';\n    margin-left: auto;\n  }\n[dir="rtl"] .f4fthw8:not(.u2Zu_fh) ._95vdWYg > svg,\n[dir="rtl"] ._6Gadv9z:not(.u2Zu_fh) ._95vdWYg > svg {\n    margin-left: ' + theme.iconPlusTextMargin + ';\n    margin-right: auto;\n  }\n\n.f4fthw8 ._95vdWYg {\n    white-space: nowrap;\n    overflow: hidden;\n    text-overflow: ellipsis;\n  }\n\n._6Gadv9z {\n  display: block;\n  width: 100%;\n  text-align: start;\n}\n\n[dir="ltr"] ._6Gadv9z {\n  text-align: left;\n}\n\n[dir="rtl"] ._6Gadv9z {\n  text-align: right;\n}\n\n._6Gadv9z._1qX2Ex0 ._95vdWYg {\n    padding-top: 0.375rem;\n    padding-bottom: 0.3125rem;\n  }\n\n._6Gadv9z._4NO1WYb ._95vdWYg {\n    padding-top: 0.5625rem;\n    padding-bottom: 0.5625rem;\n  }\n\n._6Gadv9z._3NjWj4C ._95vdWYg {\n    padding-top: 0.6875rem;\n    padding-bottom: 0.6875rem;\n  }\n\n._6Gadv9z ._95vdWYg {\n    word-wrap: break-word;\n    -webkit-hyphens: auto;\n        -ms-hyphens: auto;\n            hyphens: auto;\n    line-height: ' + theme.fluidWidthLineHeight + ';\n  }\n\n._22fpOhA,\n._2N7DysK {\n  -webkit-font-smoothing: antialiased;\n  -moz-osx-font-smoothing: grayscale;\n}\n\n._22fpOhA {\n  background: ' + theme.defaultBackground + ';\n  color: ' + theme.defaultColor + ';\n  border-color: ' + theme.defaultBorderColor + ';\n}\n\n._22fpOhA:hover,\n  ._22fpOhA:active {\n    background: ' + theme.defaultHoverBackground + ';\n  }\n\n._22fpOhA:focus,\n  ._22fpOhA:hover,\n  ._22fpOhA:active {\n    color: ' + theme.defaultColor + ';\n  }\n\n._22fpOhA:active {\n    background: ' + theme.defaultActiveBackground + ';\n    box-shadow: ' + theme.defaultActiveBoxShadow + ';\n  }\n\n._1QGaDdQ {\n  background: ' + theme.primaryBackground + ';\n  color: ' + theme.primaryColor + ';\n  border-color: ' + theme.primaryBorderColor + ';\n}\n\n._1QGaDdQ:hover,\n  ._1QGaDdQ:active {\n    background: ' + theme.primaryHoverBackground + ';\n  }\n\n._1QGaDdQ:focus,\n  ._1QGaDdQ:hover,\n  ._1QGaDdQ:active {\n    color: ' + theme.primaryColor + ';\n  }\n\n._1QGaDdQ:active {\n    background: ' + theme.primaryActiveBackground + ';\n    box-shadow: ' + theme.primaryActiveBoxShadow + ';\n  }\n\n._2trZF2L {\n  background: ' + theme.successBackground + ';\n  color: ' + theme.successColor + ';\n  border-color: ' + theme.successBorderColor + ';\n}\n\n._2trZF2L:hover,\n  ._2trZF2L:active {\n    background: ' + theme.successHoverBackground + ';\n  }\n\n._2trZF2L:focus,\n  ._2trZF2L:hover,\n  ._2trZF2L:active {\n    color: ' + theme.successColor + ';\n  }\n\n._2trZF2L:active {\n    background: ' + theme.successActiveBackground + ';\n    box-shadow: ' + theme.successActiveBoxShadow + ';\n  }\n\n._2wBuj2O {\n  background: ' + theme.dangerBackground + ';\n  color: ' + theme.dangerColor + ';\n  border-color: ' + theme.dangerBorderColor + ';\n}\n\n._2wBuj2O:hover,\n  ._2wBuj2O:active {\n    background: ' + theme.dangerHoverBackground + ';\n  }\n\n._2wBuj2O:focus,\n  ._2wBuj2O:hover,\n  ._2wBuj2O:active {\n    color: ' + theme.dangerColor + ';\n  }\n\n._2wBuj2O:active {\n    background: ' + theme.dangerActiveBackground + ';\n    box-shadow: ' + theme.dangerActiveBoxShadow + ';\n  }\n\n.jM8vyYp {\n  background: ' + theme.lightBackground + ';\n  color: ' + theme.lightColor + ';\n  border-color: ' + theme.lightBorderColor + ';\n}\n\n.jM8vyYp:hover,\n  .jM8vyYp:active {\n    background: ' + theme.lightHoverBackground + ';\n  }\n\n.jM8vyYp:focus,\n  .jM8vyYp:hover,\n  .jM8vyYp:active {\n    color: ' + theme.lightColor + ';\n  }\n\n.jM8vyYp:active {\n    background: ' + theme.lightActiveBackground + ';\n    box-shadow: ' + theme.lightActiveBoxShadow + ';\n  }\n\n._2N7DysK {\n  background: ' + theme.ghostBackground + ';\n  color: ' + theme.ghostColor + ';\n  border-style: ' + theme.ghostBorderStyle + ';\n  border-color: ' + theme.ghostBorderColor + ';\n}\n\n._2N7DysK:hover,\n  ._2N7DysK:active {\n    background: ' + theme.ghostHoverBackground + ';\n  }\n\n._2N7DysK:focus,\n  ._2N7DysK:hover,\n  ._2N7DysK:active {\n    color: ' + theme.ghostColor + ';\n  }\n\n._2N7DysK:active {\n    background: ' + theme.ghostActiveBackground + ';\n    box-shadow: ' + theme.ghostActiveBoxShadow + ';\n  }\n\n._2U962ht {\n  background: ' + theme.ghostInverseBackground + ';\n  color: ' + theme.ghostInverseColor + ';\n  border-style: ' + theme.ghostInverseBorderStyle + ';\n  border-color: ' + theme.ghostInverseBorderColor + ';\n}\n\n._2U962ht:hover,\n  ._2U962ht:active {\n    background: ' + theme.ghostInverseHoverBackground + ';\n  }\n\n._2U962ht:focus,\n  ._2U962ht:hover,\n  ._2U962ht:active {\n    color: ' + theme.ghostInverseColor + ';\n  }\n\n._2U962ht:focus::before {\n      border: ' + theme.ghostInverseFocusBorder + ';\n    }\n\n._2U962ht:active {\n    background: ' + theme.ghostActiveBackground + ';\n    box-shadow: ' + theme.ghostInverseActiveBoxShadow + ';\n  }\n\n._3-Zl5rJ,\n.JNeC0vw {\n  background: transparent;\n  -webkit-text-decoration: ' + theme.linkTextDecoration + ';\n          text-decoration: ' + theme.linkTextDecoration + ';\n  border-color: ' + theme.linkBorderColor + ';\n}\n\n._3-Zl5rJ:focus,\n  ._3-Zl5rJ:hover,\n  ._3-Zl5rJ:active,\n  .JNeC0vw:focus,\n  .JNeC0vw:hover,\n  .JNeC0vw:active {\n    text-decoration: underline;\n  }\n\n._3-Zl5rJ {\n  color: ' + theme.linkColor + ';\n}\n\n._3-Zl5rJ:focus,\n  ._3-Zl5rJ:hover,\n  ._3-Zl5rJ:active {\n    border-color: ' + theme.linkFocusBorderColor + ';\n    color: ' + theme.linkHoverColor + ';\n  }\n\n.JNeC0vw {\n  color: ' + theme.linkInverseColor + ';\n}\n\n.JNeC0vw:focus,\n  .JNeC0vw:hover,\n  .JNeC0vw:active {\n    border-color: ' + theme.linkInverseFocusBorderColor + ';\n    color: ' + theme.linkInverseColor + ';\n  }\n\n._3FUvQ3P {\n  background: ' + theme.circleDefaultBackground + ';\n  color: ' + theme.circleDefaultColor + ';\n  border-color: ' + theme.circleDefaultBorderColor + ';\n}\n\n._3FUvQ3P:hover,\n  ._3FUvQ3P:active {\n    background: ' + theme.circleDefaultHoverBackground + ';\n  }\n\n._3FUvQ3P:focus,\n  ._3FUvQ3P:hover,\n  ._3FUvQ3P:active {\n    color: ' + theme.circleDefaultColor + ';\n  }\n\n._3FUvQ3P:active {\n    background: ' + theme.circleDefaultActiveBackground + ';\n    box-shadow: ' + theme.circleDefaultActiveBoxShadow + ';\n  }\n\n._3YzX29z {\n  background: ' + theme.circlePrimaryBackground + ';\n  color: ' + theme.circlePrimaryColor + ';\n  border-color: ' + theme.circlePrimaryBorderColor + ';\n}\n\n._3YzX29z:hover,\n  ._3YzX29z:active {\n    background: ' + theme.circlePrimaryHoverBackground + ';\n  }\n\n._3YzX29z:focus,\n  ._3YzX29z:hover,\n  ._3YzX29z:active {\n    color: ' + theme.circlePrimaryColor + ';\n  }\n\n._3YzX29z:active {\n    background: ' + theme.circlePrimaryActiveBackground + ';\n    box-shadow: ' + theme.circlePrimaryActiveBoxShadow + ';\n  }\n\n._1gGcRtb {\n  background: ' + theme.circleDangerBackground + ';\n  color: ' + theme.circleDangerColor + ';\n  border-color: ' + theme.circleDangerBorderColor + ';\n}\n\n._1gGcRtb:hover,\n  ._1gGcRtb:active {\n    background: ' + theme.circleDangerHoverBackground + ';\n  }\n\n._1gGcRtb:focus,\n  ._1gGcRtb:hover,\n  ._1gGcRtb:active {\n    color: ' + theme.circleDangerColor + ';\n  }\n\n._1gGcRtb:active {\n    background: ' + theme.circleDangerActiveBackground + ';\n    box-shadow: ' + theme.circleDangerActiveBoxShadow + ';\n  }\n\n._3bmqVQ6,\n.WcDgT7y {\n  border-color: ' + theme.iconBorderColor + ';\n}\n\n.WcDgT7y {\n  color: ' + theme.iconInverseColor + ';\n}\n\n.WcDgT7y:focus {\n    color: ' + theme.iconInverseHoverColor + ';\n  }\n\n.WcDgT7y:focus::before {\n      border-color: ' + theme.ghostInverseBorderColor + ';\n    }\n\n.WcDgT7y:hover,\n  .WcDgT7y:active {\n    color: ' + theme.iconInverseHoverColor + ';\n  }\n\n._3bmqVQ6 {\n  color: ' + theme.iconColor + ';\n}\n\n._3bmqVQ6:focus,\n  ._3bmqVQ6:hover,\n  ._3bmqVQ6:active {\n    color: ' + theme.iconHoverColor + ';\n  }\n\n/* If icons are passed in as elements instead of functions, guarantee\nthe SVG will display block */\n._1s_Wtxo > svg { /* stylelint-disable-line selector-max-type */\n  display: block;\n}\n\n/* stylelint-enable selector-max-class, no-descending-specificity */\n';
    };

    return tmpl.call(theme, theme);
  },
  'root': '_2pBn_SN',
  'disabled': '_3ysMWZa',
  'link': '_3-Zl5rJ',
  'width--icon': '_1LL06hd',
  'content': '_95vdWYg',
  'borderRadius--rounded': '_30tkGIB',
  'borderRadius--circle': 'xGLZogw',
  'small': '_1qX2Ex0',
  'has-icon': 'u2Zu_fh',
  'iconSVG': '_1s_Wtxo',
  'medium': '_4NO1WYb',
  'large': '_3NjWj4C',
  'width--auto': 'f4fthw8',
  'width--fluid': '_6Gadv9z',
  'default': '_22fpOhA',
  'ghost': '_2N7DysK',
  'primary': '_1QGaDdQ',
  'success': '_2trZF2L',
  'danger': '_2wBuj2O',
  'light': 'jM8vyYp',
  'ghost-inverse': '_2U962ht',
  'link-inverse': 'JNeC0vw',
  'circle-default': '_3FUvQ3P',
  'circle-primary': '_3YzX29z',
  'circle-danger': '_1gGcRtb',
  'icon': '_3bmqVQ6',
  'icon-inverse': 'WcDgT7y'
};


/**
---
category: components
---
**/

var Button = (_dec = (0, _uiThemeable2.default)(_theme2.default, styles), _dec(_class = (_temp2 = _class2 = function (_Component) {
  _inherits(Button, _Component);

  function Button() {
    var _ref;

    var _temp, _this, _ret;

    _classCallCheck(this, Button);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = Button.__proto__ || Object.getPrototypeOf(Button)).call.apply(_ref, [this].concat(args))), _this), _this.handleClick = function (e) {
      var _this$props = _this.props,
          disabled = _this$props.disabled,
          readOnly = _this$props.readOnly,
          onClick = _this$props.onClick;


      if (disabled || readOnly) {
        e.preventDefault();
        e.stopPropagation();
      } else if (typeof onClick === 'function') {
        onClick(e);
      }
    }, _this.handleKeyDown = function (e) {
      var _this$props2 = _this.props,
          disabled = _this$props2.disabled,
          readOnly = _this$props2.readOnly,
          onClick = _this$props2.onClick,
          href = _this$props2.href;

      // behave like a button when space key is pressed

      if (_this.elementType !== 'button' && (e.keyCode === _keycode2.default.codes.space || e.keyCode === _keycode2.default.codes.enter)) {
        e.preventDefault();
        e.stopPropagation();

        if (typeof onClick === 'function' && !disabled && !readOnly) {
          onClick(e);
        }
        if (href) {
          (0, _findDOMNode2.default)(_this._button).click(); // eslint-disable-line react/no-find-dom-node
        }
      }
    }, _temp), _possibleConstructorReturn(_this, _ret);
  }

  _createClass(Button, [{
    key: 'focus',
    value: function focus() {
      (0, _findDOMNode2.default)(this._button).focus(); // eslint-disable-line react/no-find-dom-node
    }
  }, {
    key: 'renderIcon',
    value: function renderIcon() {
      var Icon = this.props.icon;
      if (typeof this.props.icon === 'function') {
        return _react2.default.createElement(
          'span',
          { className: styles.iconSVG },
          _react2.default.createElement(Icon, { inline: false })
        );
      } else if (Icon) {
        return _react2.default.createElement(
          'span',
          { className: styles.iconSVG },
          Icon
        );
      } else {
        return null;
      }
    }
  }, {
    key: 'buttonBorderRadius',
    value: function buttonBorderRadius() {
      var circleVariants = ['circle-primary', 'circle-danger', 'circle-default'];

      if (circleVariants.indexOf(this.props.variant) !== -1) {
        return 'circle';
      } else {
        return 'rounded';
      }
    }
  }, {
    key: 'buttonWidth',
    value: function buttonWidth() {
      var _props = this.props,
          variant = _props.variant,
          fluidWidth = _props.fluidWidth;


      var squareVariants = ['circle-default', 'circle-primary', 'circle-danger', 'icon', 'icon-inverse'];

      if (this.hasOnlyIcon || squareVariants.indexOf(variant) !== -1) {
        return 'icon';
      } else if (fluidWidth) {
        return 'fluid';
      } else {
        return 'auto';
      }
    }
  }, {
    key: 'renderContent',
    value: function renderContent() {
      var _props2 = this.props,
          children = _props2.children,
          icon = _props2.icon;

      // show warning if icon is added as a child

      if (this.hasVisibleChildren) {
        _react2.default.Children.forEach(children, function (child) {
          var icon = (typeof child === 'undefined' ? 'undefined' : _typeof(child)) === 'object' && child.type.glyphName !== undefined; // eslint-disable-line no-undefined
          (0, _warning2.default)(!icon, 'Adding icons to Button as children is deprecated. Please use the icon prop instead.');
        });
      }

      if (this.hasTextAndIcon) {
        return _react2.default.createElement(
          _Flex2.default,
          { height: '100%', width: '100%' },
          _react2.default.createElement(
            _Flex.FlexItem,
            { padding: '0 x-small 0 0' },
            this.renderIcon()
          ),
          _react2.default.createElement(
            _Flex.FlexItem,
            { grow: true, shrink: true },
            _react2.default.createElement(
              'span',
              { className: styles.content },
              children
            )
          )
        );
      } else {
        // all other button layouts (icon only and text only)
        return _react2.default.createElement(
          'span',
          { className: styles.content },
          icon && this.renderIcon(),
          children
        );
      }
    }
  }, {
    key: 'render',
    value: function render() {
      var _classes,
          _this2 = this;

      var _props3 = this.props,
          variant = _props3.variant,
          size = _props3.size,
          disabled = _props3.disabled,
          readOnly = _props3.readOnly,
          href = _props3.href,
          type = _props3.type,
          onClick = _props3.onClick,
          buttonRef = _props3.buttonRef,
          margin = _props3.margin,
          icon = _props3.icon;


      var classes = (_classes = {}, _defineProperty(_classes, styles.root, true), _defineProperty(_classes, styles[variant], true), _defineProperty(_classes, styles[size], size), _defineProperty(_classes, styles['width--' + this.buttonWidth()], true), _defineProperty(_classes, styles['borderRadius--' + this.buttonBorderRadius()], true), _defineProperty(_classes, styles.disabled, disabled), _defineProperty(_classes, styles['has-icon'], icon), _classes);

      var props = Object.assign({}, (0, _passthroughProps.omitProps)(this.props, Object.assign({}, Button.propTypes, _View2.default.propTypes)), {
        elementRef: function elementRef(c) {
          for (var _len2 = arguments.length, args = Array(_len2 > 1 ? _len2 - 1 : 0), _key2 = 1; _key2 < _len2; _key2++) {
            args[_key2 - 1] = arguments[_key2];
          }

          _this2._button = c;
          buttonRef.apply(_this2, [c].concat(args));
        },
        className: (0, _classnames2.default)(classes),
        disabled: disabled || readOnly,
        'aria-disabled': disabled || readOnly ? 'true' : null,
        onClick: this.handleClick,
        onKeyDown: this.handleKeyDown,
        href: href,
        type: href ? null : type,
        role: onClick && this.props.as ? 'button' : null,
        tabIndex: onClick && this.props.as ? '0' : null
      });

      var ElementType = this.elementType;

      return _react2.default.createElement(
        _View2.default,
        Object.assign({}, props, {
          display: null,
          as: ElementType,
          margin: margin
        }),
        this.renderContent()
      );
    }
  }, {
    key: 'elementType',
    get: function get() {
      return (0, _getElementType2.default)(Button, this.props);
    }
  }, {
    key: 'focused',
    get: function get() {
      return (0, _isActiveElement2.default)(this._button);
    }
  }, {
    key: 'hasVisibleChildren',
    get: function get() {
      return (0, _hasVisibleChildren2.default)(this.props.children);
    }
  }, {
    key: 'hasTextAndIcon',
    get: function get() {
      return this.hasVisibleChildren && this.props.icon; // any button with an icon + text label
    }
  }, {
    key: 'hasOnlyIcon',
    get: function get() {
      return !this.hasVisibleChildren && this.props.icon; // any button with just an icon visible
    }
  }]);

  Button.displayName = 'Button'
  ;
  return Button;
}(_react.Component), _class2.propTypes = {
  children: _propTypes2.default.node.isRequired,
  type: _propTypes2.default.oneOf(['button', 'submit', 'reset']),
  variant: _propTypes2.default.oneOf(['default', 'primary', 'success', 'danger', 'light', 'ghost', 'ghost-inverse', 'link', 'link-inverse', 'circle-default', 'circle-primary', 'circle-danger', 'icon', 'icon-inverse']),
  /**
  * provides a reference to the underlying focusable (`button` or `a`) element
  */
  buttonRef: _propTypes2.default.func,
  /**
  * the element type to render as (will be `<a>` if href is provided)
  */
  as: _CustomPropTypes2.default.elementType,
  size: _propTypes2.default.oneOf(['small', 'medium', 'large']),
  /**
  * should the `<Button/>` fill the width of its container
  */
  fluidWidth: _propTypes2.default.bool,
  disabled: _propTypes2.default.bool,
  readOnly: _propTypes2.default.bool,
  href: _propTypes2.default.string,
  onClick: _propTypes2.default.func,
  /**
  * Valid values are `0`, `none`, `auto`, `xxx-small`, `xx-small`, `x-small`,
  * `small`, `medium`, `large`, `x-large`, `xx-large`. Apply these values via
  * familiar CSS-like shorthand. For example: `margin="small auto large"`.
  */
  margin: _ThemeablePropTypes2.default.spacing,
  /**
  * Add an SVG icon to the button. Do not add icons directly as
  * children.
  */
  icon: _propTypes2.default.oneOfType([_propTypes2.default.func, _propTypes2.default.element])
}, _class2.defaultProps = {
  as: 'button',
  type: 'button',
  variant: 'default',
  size: 'medium',
  margin: '0',
  fluidWidth: false,
  buttonRef: function buttonRef(button) {}
}, _temp2)) || _class);
exports.default = Button;
