// @instructure/ui-elements/lib/components/Text/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _dec, _class, _class2, _temp; /*
                                   * The MIT License (MIT)
                                   *
                                   * Copyright (c) 2015 - present Instructure, Inc.
                                   *
                                   * Permission is hereby granted, free of charge, to any person obtaining a copy
                                   * of this software and associated documentation files (the "Software"), to deal
                                   * in the Software without restriction, including without limitation the rights
                                   * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
                                   * copies of the Software, and to permit persons to whom the Software is
                                   * furnished to do so, subject to the following conditions:
                                   *
                                   * The above copyright notice and this permission notice shall be included in all
                                   * copies or substantial portions of the Software.
                                   *
                                   * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
                                   * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
                                   * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
                                   * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
                                   * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
                                   * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
                                   * SOFTWARE.
                                   */


var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _uiThemeable = require('@instructure/ui-themeable');

var _uiThemeable2 = _interopRequireDefault(_uiThemeable);

var _CustomPropTypes = require('@instructure/ui-utils/lib/react/CustomPropTypes');

var _CustomPropTypes2 = _interopRequireDefault(_CustomPropTypes);

var _getElementType = require('@instructure/ui-utils/lib/react/getElementType');

var _getElementType2 = _interopRequireDefault(_getElementType);

var _passthroughProps = require('@instructure/ui-utils/lib/react/passthroughProps');

var _theme = require('./theme');

var _theme2 = _interopRequireDefault(_theme);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var styles = {
  template: function template(theme) {
    var tmpl = function tmpl() {
      return '/*  imported from styles.css  */\n\n._2Wjlmr5 {\n  font-family: ' + theme.fontFamily + ';\n  /* stylelint-enable selector-max-type */\n}\n\n  /* stylelint-disable selector-max-type */\n\n  ._2Wjlmr5 sub,\n  ._2Wjlmr5 sup {\n    font-size: 75%;\n    line-height: 0;\n    position: relative;\n    vertical-align: baseline;\n  }\n\n  ._2Wjlmr5 sup {\n    top: -0.4em;\n  }\n\n  ._2Wjlmr5 sub {\n    bottom: -0.4em;\n  }\n\n  ._2Wjlmr5 pre,\n  ._2Wjlmr5 code {\n    animation: none 0s ease 0s 1 normal none running;\n    backface-visibility: visible;\n    background: transparent none repeat 0 0 / auto auto padding-box border-box scroll;\n    border: medium none currentColor;\n    border-collapse: separate;\n    border-image: none;\n    border-radius: 0;\n    border-spacing: 0;\n    bottom: auto;\n    box-shadow: none;\n    box-sizing: content-box;\n    caption-side: top;\n    clear: none;\n    clip: auto;\n    color: #000;\n    columns: auto;\n    column-count: auto;\n    column-fill: balance;\n    column-gap: normal;\n    column-rule: medium none currentColor;\n    column-span: 1;\n    column-width: auto;\n    content: normal;\n    counter-increment: none;\n    counter-reset: none;\n    cursor: auto;\n    direction: ltr;\n    display: inline;\n    empty-cells: show;\n    float: none;\n    font-family: serif;\n    font-size: medium;\n    font-style: normal;\n    font-variant: normal;\n    font-weight: normal;\n    font-stretch: normal;\n    line-height: normal;\n    height: auto;\n    hyphens: none;\n    left: auto;\n    letter-spacing: normal;\n    list-style: disc outside none;\n    margin: 0;\n    max-height: none;\n    max-width: none;\n    min-height: 0;\n    min-width: 0;\n    opacity: 1;\n    orphans: 2;\n    outline: medium none invert;\n    overflow: visible;\n    overflow-x: visible;\n    overflow-y: visible;\n    padding: 0;\n    page-break-after: auto;\n    page-break-before: auto;\n    page-break-inside: auto;\n    perspective: none;\n    perspective-origin: 50% 50%;\n    position: static;\n    right: auto;\n    tab-size: 8;\n    table-layout: auto;\n    text-align: left;\n    text-align-last: auto;\n    text-decoration: none;\n    text-indent: 0;\n    text-shadow: none;\n    text-transform: none;\n    top: auto;\n    transform: none;\n    transform-origin: 50% 50% 0;\n    transform-style: flat;\n    transition: none 0s ease 0s;\n    unicode-bidi: normal;\n    vertical-align: baseline;\n    visibility: visible;\n    white-space: normal;\n    widows: 2;\n    width: auto;\n    word-spacing: normal;\n    z-index: auto;\n    all: initial;\n    font-family: ' + theme.fontFamilyMonospace + ';\n  }\n\n  ._2Wjlmr5 i,\n  ._2Wjlmr5 em {\n    font-style: italic;\n  }\n\n  ._2Wjlmr5 b,\n  ._2Wjlmr5 strong {\n    font-weight: ' + theme.fontWeightBold + ';\n  }\n\n  ._2Wjlmr5 p {\n    display: block;\n    padding: 0;\n    margin: ' + theme.paragraphMargin + ';\n  }\n\n._1XwaWGz {\n  color: ' + theme.primaryColor + ';\n}\n\n._1JreiLw {\n  color: ' + theme.secondaryColor + ';\n}\n\n._1fvgkDL {\n  color: ' + theme.primaryInverseColor + ';\n}\n\n._3QfYETf {\n  color: ' + theme.secondaryInverseColor + ';\n}\n\n._1fTUf4Z {\n  color: ' + theme.successColor + ';\n}\n\n._29cN8CR {\n  color: ' + theme.brandColor + ';\n}\n\n._3m9ZEwr {\n  color: ' + theme.warningColor + ';\n}\n\n._1rQrgsh {\n  color: ' + theme.errorColor + ';\n}\n\n._3h5Gydv {\n  font-weight: ' + theme.fontWeightNormal + ';\n}\n\n.VtqvaIe {\n  font-weight: ' + theme.fontWeightLight + ';\n}\n\n._2Kyj5T5 {\n  font-weight: ' + theme.fontWeightBold + ';\n}\n\n.RtN0GRN {\n  font-style: normal;\n}\n\n._3_bEcMY {\n  font-style: italic;\n}\n\n._1BZqwlq {\n  font-size: ' + theme.fontSizeXSmall + ';\n}\n\n.Jejr-nq {\n  font-size: ' + theme.fontSizeSmall + ';\n}\n\n.Fa88R4_ {\n  font-size: ' + theme.fontSizeMedium + ';\n}\n\n._23e23O8 {\n  font-size: ' + theme.fontSizeLarge + ';\n}\n\n.pwrmoHi {\n  font-size: ' + theme.fontSizeXLarge + ';\n}\n\n._3YlzA95 {\n  font-size: ' + theme.fontSizeXXLarge + ';\n}\n\n._3zA_0Eo {\n  line-height: ' + theme.lineHeight + ';\n}\n\n._3SX-Pzu {\n  line-height: ' + theme.lineHeightFit + ';\n}\n\n._3Da9qSZ {\n  line-height: ' + theme.lineHeightCondensed + ';\n}\n\n._2eKOEi_ {\n  line-height: ' + theme.lineHeightDouble + ';\n}\n\n.NE52rMV {\n  letter-spacing: ' + theme.letterSpacingNormal + ';\n}\n\n.ClGCdaD {\n  letter-spacing: ' + theme.letterSpacingCondensed + ';\n}\n\n.uAf-8Ue {\n  letter-spacing: ' + theme.letterSpacingExpanded + ';\n}\n\n._2cMx7HZ {\n  text-transform: none;\n}\n\n._2yfvfeC {\n  text-transform: capitalize;\n}\n\n._1OUNCvu {\n  text-transform: uppercase;\n}\n\n.EUTvPID {\n  text-transform: lowercase;\n}\n';
    };

    return tmpl.call(theme, theme);
  },
  'root': '_2Wjlmr5',
  'color-primary': '_1XwaWGz',
  'color-secondary': '_1JreiLw',
  'color-primary-inverse': '_1fvgkDL',
  'color-secondary-inverse': '_3QfYETf',
  'color-success': '_1fTUf4Z',
  'color-brand': '_29cN8CR',
  'color-warning': '_3m9ZEwr',
  'color-error': '_1rQrgsh',
  'weight-normal': '_3h5Gydv',
  'weight-light': 'VtqvaIe',
  'weight-bold': '_2Kyj5T5',
  'style-normal': 'RtN0GRN',
  'style-italic': '_3_bEcMY',
  'x-small': '_1BZqwlq',
  'small': 'Jejr-nq',
  'medium': 'Fa88R4_',
  'large': '_23e23O8',
  'x-large': 'pwrmoHi',
  'xx-large': '_3YlzA95',
  'lineHeight-default': '_3zA_0Eo',
  'lineHeight-fit': '_3SX-Pzu',
  'lineHeight-condensed': '_3Da9qSZ',
  'lineHeight-double': '_2eKOEi_',
  'letterSpacing-normal': 'NE52rMV',
  'letterSpacing-condensed': 'ClGCdaD',
  'letterSpacing-expanded': 'uAf-8Ue',
  'transform-none': '_2cMx7HZ',
  'transform-capitalize': '_2yfvfeC',
  'transform-uppercase': '_1OUNCvu',
  'transform-lowercase': 'EUTvPID'
};


/**
---
category: components/typography
---
**/
var Text = (_dec = (0, _uiThemeable2.default)(_theme2.default, styles), _dec(_class = (_temp = _class2 = function (_Component) {
  _inherits(Text, _Component);

  function Text() {
    _classCallCheck(this, Text);

    return _possibleConstructorReturn(this, (Text.__proto__ || Object.getPrototypeOf(Text)).apply(this, arguments));
  }

  _createClass(Text, [{
    key: 'render',
    value: function render() {
      var _classes;

      var _props = this.props,
          weight = _props.weight,
          fontStyle = _props.fontStyle,
          size = _props.size,
          lineHeight = _props.lineHeight,
          letterSpacing = _props.letterSpacing,
          transform = _props.transform,
          color = _props.color;


      var classes = (_classes = {}, _defineProperty(_classes, styles.root, true), _defineProperty(_classes, styles[size], size), _defineProperty(_classes, styles['weight-' + weight], weight), _defineProperty(_classes, styles['style-' + fontStyle], fontStyle), _defineProperty(_classes, styles['transform-' + transform], transform), _defineProperty(_classes, styles['lineHeight-' + lineHeight], lineHeight), _defineProperty(_classes, styles['letterSpacing-' + letterSpacing], letterSpacing), _defineProperty(_classes, styles['color-' + color], color), _classes);

      var props = Object.assign({}, (0, _passthroughProps.omitProps)(this.props, Text.propTypes), {
        className: (0, _classnames2.default)(classes)
      });

      var ElementType = (0, _getElementType2.default)(Text, this.props);

      return _react2.default.createElement(
        ElementType,
        props,
        this.props.children
      );
    }
  }]);

  Text.displayName = 'Text'
  ;
  return Text;
}(_react.Component), _class2.propTypes = {
  /**
  * the element type to render as
  */
  as: _CustomPropTypes2.default.elementType,
  weight: _propTypes2.default.oneOf(['normal', 'light', 'bold']),
  fontStyle: _propTypes2.default.oneOf(['italic', 'normal']),
  size: _propTypes2.default.oneOf(['x-small', 'small', 'medium', 'large', 'x-large', 'xx-large']),
  lineHeight: _propTypes2.default.oneOf(['default', 'fit', 'condensed', 'double']),
  letterSpacing: _propTypes2.default.oneOf(['normal', 'condensed', 'expanded']),
  transform: _propTypes2.default.oneOf(['none', 'capitalize', 'uppercase', 'lowercase']),
  color: _propTypes2.default.oneOf(['primary', 'secondary', 'primary-inverse', 'secondary-inverse', 'success', 'error', 'warning', 'brand']),
  children: _propTypes2.default.node
}, _class2.defaultProps = {
  as: 'span',
  size: 'medium',
  letterSpacing: 'normal'
}, _temp)) || _class);
exports.default = Text;
