// @instructure/ui-elements/es/components/Tag/index.js
var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _dec, _class, _class2, _temp2;

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import IconX from '@instructure/ui-icons/lib/Solid/IconX';

import View from '@instructure/ui-layout/lib/components/View';

import themeable from '@instructure/ui-themeable';
import ThemeablePropTypes from '@instructure/ui-themeable/lib/utils/ThemeablePropTypes';
import { omitProps } from '@instructure/ui-utils/lib/react/passthroughProps';
import isActiveElement from '@instructure/ui-utils/lib/dom/isActiveElement';

var styles = {
  template: function template(theme) {
    var tmpl = function tmpl() {
      return '/*  imported from styles.css  */\n\n/* stylelint-disable selector-max-class, no-descending-specificity, max-nesting-depth */\n\n._3LSqa2T {\n  box-sizing: border-box;\n  font-family: ' + theme.fontFamily + ';\n  display: inline-flex;\n  align-items: center;\n  text-align: center;\n  vertical-align: middle;\n  -webkit-user-select: none;\n     -moz-user-select: none;\n      -ms-user-select: none;\n          user-select: none;\n}\n\n[dir="ltr"] ._3LSqa2T {\n  text-align: center;\n}\n\n[dir="rtl"] ._3LSqa2T {\n  text-align: center;\n}\n\n._2jUI8fW {\n  height: ' + theme.heightSmall + ';\n  padding: ' + theme.paddingSmall + ';\n  font-size: ' + theme.fontSizeSmall + ';\n}\n\n._2jUI8fW .STVW0RX {\n    font-size: 0.75rem;\n  }\n\n._2AJ1rRZ {\n  height: ' + theme.heightMedium + ';\n  padding: ' + theme.padding + ';\n  font-size: calc(' + theme.fontSizeMedium + ' - 0.0625rem);\n}\n\n._2AJ1rRZ .STVW0RX {\n    font-size: 0.75rem;\n  }\n\n._3U-tYoW {\n  height: ' + theme.heightLarge + ';\n  padding: ' + theme.padding + ';\n  font-size: calc(' + theme.fontSizeLarge + ' - 0.0625rem);\n}\n\n._3U-tYoW .STVW0RX {\n    font-size: 0.875rem;\n  }\n\n.e2QdnXI {\n  touch-action: manipulation;\n  outline: none;\n  position: relative;\n  overflow: visible;\n  transition: background-color ' + theme.transitionTiming + ';\n}\n\n.e2QdnXI::before {\n    content: "";\n    box-sizing: border-box;\n    width: calc(100% + 0.5rem);\n    height: calc(100% + 0.5rem);\n    border: ' + theme.borderWidth + ' ' + theme.borderStyle + ' ' + theme.focusOutlineColor + ';\n    border-radius: ' + theme.borderRadius + ';\n    position: absolute;\n    top: -0.25rem;\n    left: -0.25rem; /* stylelint-disable-line property-blacklist */\n    opacity: 0;\n    transform: scale(0.01);\n    transition: all ' + theme.transitionTiming + ';\n    pointer-events: none;\n  }\n\n.e2QdnXI:not(._3EoEaU7) {\n    cursor: pointer;\n  }\n\n.e2QdnXI:focus::before {\n      opacity: 1;\n      transform: scale(1);\n    }\n\n.e2QdnXI._1-0_evd {\n    cursor: not-allowed;\n    pointer-events: none;\n    opacity: 0.5;\n  }\n\n.STVW0RX {\n  -webkit-margin-start: ' + theme.iconMargin + ';\n          margin-inline-start: ' + theme.iconMargin + ';\n  -webkit-margin-end: 0;\n          margin-inline-end: 0;\n  transform: translateY(0.0625rem);\n  transition: all ' + theme.transitionTiming + ';\n  cursor: pointer;\n}\n\n[dir="ltr"] .STVW0RX {\n  margin-left: ' + theme.iconMargin + ';\n  margin-right: 0;\n}\n\n[dir="rtl"] .STVW0RX {\n  margin-right: ' + theme.iconMargin + ';\n  margin-left: 0;\n}\n\n.O7a8vll {\n  white-space: nowrap;\n  overflow: hidden;\n  text-overflow: ellipsis;\n  max-width: ' + theme.maxWidth + ';\n}\n\n._3lFUoY9 {\n  background-color: ' + theme.defaultBackground + ';\n  border: ' + theme.defaultBorderWidth + ' ' + theme.defaultBorderStyle + ' ' + theme.defaultBorderColor + ';\n  border-radius: ' + theme.defaultBorderRadius + ';\n  color: ' + theme.defaultColor + ';\n}\n\n._3lFUoY9.e2QdnXI::before {\n      border: ' + theme.defaultBorderWidth + ' ' + theme.defaultBorderStyle + ' ' + theme.focusOutlineColor + ';\n      border-radius: ' + theme.defaultBorderRadius + ';\n    }\n\n._3lFUoY9.e2QdnXI:hover {\n      background-color: ' + theme.defaultBackgroundHover + ';\n    }\n\n._3lFUoY9._3EoEaU7 .STVW0RX {\n      color: ' + theme.defaultIconColor + ';\n    }\n\n._3lFUoY9._3EoEaU7:hover .STVW0RX {\n        color: ' + theme.defaultIconHoverColor + ';\n      }\n\n._5dMZQgp {\n  background-color: ' + theme.inlineBackground + ';\n  border: ' + theme.inlineBorderWidth + ' ' + theme.inlineBorderStyle + ' ' + theme.inlineBorderColor + ';\n  border-radius: ' + theme.inlineBorderRadius + ';\n  color: ' + theme.inlineColor + ';\n  cursor: text;\n  margin: 0 0.1875rem 0.1875rem 0.1875rem;\n}\n\n._5dMZQgp.e2QdnXI::before {\n      border: ' + theme.inlineBorderWidth + ' ' + theme.inlineBorderStyle + ' ' + theme.focusOutlineColor + ';\n      border-radius: ' + theme.inlineBorderRadius + ';\n    }\n\n._5dMZQgp.e2QdnXI:hover {\n      background-color: ' + theme.inlineBackgroundHover + ';\n    }\n\n._5dMZQgp._3EoEaU7 .STVW0RX {\n      background-color: ' + theme.inlineIconColor + ';\n      border-radius: 50%;\n      color: ' + theme.inlineBackground + ';\n      font-size: 0.75rem;\n      padding: 0.25rem;\n      position: absolute;\n      offset-inline-end: 0;\n      offset-inline-start: auto;\n      top: 0;\n      transform: translate(40%, -40%);\n    }\n\n[dir="ltr"] ._5dMZQgp._3EoEaU7 .STVW0RX {\n      right: 0;\n      left: auto;\n    }\n\n[dir="rtl"] ._5dMZQgp._3EoEaU7 .STVW0RX {\n      left: 0;\n      right: auto;\n    }\n\n._5dMZQgp._3EoEaU7:hover .STVW0RX {\n        background-color: ' + theme.inlineIconHoverColor + ';\n      }\n\n[dir="rtl"] ._5dMZQgp._3EoEaU7 .STVW0RX {\n  transform: translate(-40%, -40%);\n}\n/* stylelint-enable selector-max-class, no-descending-specificity, max-nesting-depth */\n';
    };

    return tmpl.call(theme, theme);
  },
  'root': '_3LSqa2T',
  'small': '_2jUI8fW',
  'icon': 'STVW0RX',
  'medium': '_2AJ1rRZ',
  'large': '_3U-tYoW',
  'button': 'e2QdnXI',
  'dismissible': '_3EoEaU7',
  'disabled': '_1-0_evd',
  'text': 'O7a8vll',
  'default': '_3lFUoY9',
  'inline': '_5dMZQgp'
};

import theme from './theme';

/**
---
category: components
---
**/

var Tag = (_dec = themeable(theme, styles), _dec(_class = (_temp2 = _class2 = function (_Component) {
  _inherits(Tag, _Component);

  function Tag() {
    var _ref;

    var _temp, _this, _ret;

    _classCallCheck(this, Tag);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = Tag.__proto__ || Object.getPrototypeOf(Tag)).call.apply(_ref, [this].concat(args))), _this), _this.focus = function () {
      _this._container && _this._container.focus();
    }, _this.handleClick = function (e) {
      var _this$props = _this.props,
          disabled = _this$props.disabled,
          readOnly = _this$props.readOnly,
          onClick = _this$props.onClick;


      if (disabled || readOnly) {
        e.preventDefault();
        e.stopPropagation();
      } else if (typeof onClick === 'function') {
        onClick(e);
      }
    }, _this.handleRef = function (node) {
      _this._container = node;
    }, _temp), _possibleConstructorReturn(_this, _ret);
  }

  _createClass(Tag, [{
    key: 'render',
    value: function render() {
      var _classes;

      var _props = this.props,
          className = _props.className,
          dismissible = _props.dismissible,
          disabled = _props.disabled,
          readOnly = _props.readOnly,
          size = _props.size,
          text = _props.text,
          title = _props.title,
          onClick = _props.onClick,
          margin = _props.margin,
          variant = _props.variant;


      var props = omitProps(this.props, Object.assign({}, Tag.propTypes, View.propTypes));

      var classes = (_classes = {}, _defineProperty(_classes, styles.root, true), _defineProperty(_classes, styles[variant], true), _defineProperty(_classes, styles[size], size), _defineProperty(_classes, styles.dismissible, dismissible), _defineProperty(_classes, styles.button, onClick), _defineProperty(_classes, styles.disabled, disabled), _classes);

      return React.createElement(
        View,
        Object.assign({}, props, {
          ref: this.handleRef,
          elementRef: this.props.elementRef,
          className: classNames(className, classes),
          as: onClick ? 'button' : 'span',
          margin: margin,
          type: onClick ? 'button' : null,
          onClick: onClick ? this.handleClick : null,
          disabled: disabled || readOnly,
          'aria-disabled': onClick && (disabled || readOnly) ? 'true' : null,
          display: null,
          title: title || (typeof text === 'string' ? text : null)
        }),
        React.createElement(
          'span',
          { className: styles.text },
          text
        ),
        onClick && dismissible ? React.createElement(IconX, { className: styles.icon }) : null
      );
    }
  }, {
    key: 'focused',
    get: function get() {
      return isActiveElement(this._container);
    }
  }]);

  Tag.displayName = 'Tag'
  ;
  return Tag;
}(Component), _class2.propTypes = {
  className: PropTypes.string,
  text: PropTypes.oneOfType([PropTypes.string, PropTypes.node]).isRequired,
  title: PropTypes.string,
  /**
   * Whether or not to disable the tag
   */
  disabled: PropTypes.bool,
  /**
   * Works just like disabled but keeps the same styles as if it were active
   */
  readOnly: PropTypes.bool,
  dismissible: PropTypes.bool,
  /**
  * Valid values are `0`, `none`, `auto`, `xxxx-small`, `xx-small`, `x-small`,
  * `small`, `medium`, `large`, `x-large`, `xx-large`. Apply these values via
  * familiar CSS-like shorthand. For example: `margin="small auto large"`.
  */
  margin: ThemeablePropTypes.spacing,
  /**
  * If you add an onClick prop, Tag renders as a clickable button
  */
  onClick: PropTypes.func,
  /**
  * Provides a reference to the underlying html root element
  */
  elementRef: PropTypes.func,
  size: PropTypes.oneOf(['small', 'medium', 'large']),
  variant: PropTypes.oneOf(['default', 'inline'])
}, _class2.defaultProps = {
  size: 'medium',
  dismissible: false,
  variant: 'default'
}, _temp2)) || _class);


export default Tag;
