// @instructure/ui-layout/lib/components/View/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _dec, _dec2, _dec3, _class, _class2, _temp; /*
                                                 * The MIT License (MIT)
                                                 *
                                                 * Copyright (c) 2015 - present Instructure, Inc.
                                                 *
                                                 * Permission is hereby granted, free of charge, to any person obtaining a copy
                                                 * of this software and associated documentation files (the "Software"), to deal
                                                 * in the Software without restriction, including without limitation the rights
                                                 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
                                                 * copies of the Software, and to permit persons to whom the Software is
                                                 * furnished to do so, subject to the following conditions:
                                                 *
                                                 * The above copyright notice and this permission notice shall be included in all
                                                 * copies or substantial portions of the Software.
                                                 *
                                                 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
                                                 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
                                                 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
                                                 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
                                                 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
                                                 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
                                                 * SOFTWARE.
                                                 */

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames2 = require('classnames');

var _classnames3 = _interopRequireDefault(_classnames2);

var _uiThemeable = require('@instructure/ui-themeable');

var _uiThemeable2 = _interopRequireDefault(_uiThemeable);

var _getShorthandPropValue = require('@instructure/ui-themeable/lib/utils/getShorthandPropValue');

var _getShorthandPropValue2 = _interopRequireDefault(_getShorthandPropValue);

var _ThemeablePropTypes = require('@instructure/ui-themeable/lib/utils/ThemeablePropTypes');

var _ThemeablePropTypes2 = _interopRequireDefault(_ThemeablePropTypes);

var _mirrorShorthand = require('@instructure/ui-themeable/lib/utils/mirrorShorthand');

var _bidirectional = require('@instructure/ui-i18n/lib/bidirectional');

var _bidirectional2 = _interopRequireDefault(_bidirectional);

var _warning = require('@instructure/ui-utils/lib/warning');

var _warning2 = _interopRequireDefault(_warning);

var _CustomPropTypes = require('@instructure/ui-utils/lib/react/CustomPropTypes');

var _CustomPropTypes2 = _interopRequireDefault(_CustomPropTypes);

var _deprecated = require('@instructure/ui-utils/lib/react/deprecated');

var _deprecated2 = _interopRequireDefault(_deprecated);

var _getDisplayName = require('@instructure/ui-utils/lib/react/getDisplayName');

var _getDisplayName2 = _interopRequireDefault(_getDisplayName);

var _getElementType = require('@instructure/ui-utils/lib/react/getElementType');

var _getElementType2 = _interopRequireDefault(_getElementType);

var _passthroughProps = require('@instructure/ui-utils/lib/react/passthroughProps');

var _theme = require('./theme');

var _theme2 = _interopRequireDefault(_theme);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var styles = {
  template: function template(theme) {
    var tmpl = function tmpl() {
      return '/*  imported from styles.css  */\n\n._363uIlg {\n  box-sizing: border-box;\n  font-family: ' + theme.fontFamily + ';\n}\n\n._2y5m9zy {\n  display: block;\n}\n\n._2BenJlN {\n  display: inline-block;\n  vertical-align: middle;\n}\n\n.-AGjWtH {\n  display: flex;\n}\n\n._1nexD3g {\n  display: inline-flex;\n  vertical-align: middle;\n}\n\n._1mgbjsV {\n  text-align: start;\n}\n\n[dir="ltr"] ._1mgbjsV {\n  text-align: left;\n}\n\n[dir="rtl"] ._1mgbjsV {\n  text-align: right;\n}\n\n._3EAL0Vy {\n  text-align: center;\n}\n\n[dir="ltr"] ._3EAL0Vy {\n  text-align: center;\n}\n\n[dir="rtl"] ._3EAL0Vy {\n  text-align: center;\n}\n\n._2DincID {\n  text-align: end;\n}\n\n[dir="ltr"] ._2DincID {\n  text-align: right;\n}\n\n[dir="rtl"] ._2DincID {\n  text-align: left;\n}\n\n._3Y_0kws {\n  outline: 0.0625rem dashed ' + theme.debugOutlineColor + ';\n}\n\n._1wWuL2e {\n  border-style: ' + theme.borderStyle + ';\n  border-color: ' + theme.borderColor + ';\n}\n\n._1wWuL2e._3BX4paj {\n    border-color: ' + theme.borderColorInverse + ';\n  }\n\n._2N_bJx5 {\n  color: ' + theme.color + ';\n  background: ' + theme.background + ';\n}\n\n._3BX4paj {\n  color: ' + theme.colorInverse + ';\n  background: ' + theme.backgroundInverse + ';\n}\n\n._2i7pspc {\n  max-width: ' + theme.xSmallMaxWidth + ';\n}\n\n.fADEtzk {\n  max-width: ' + theme.smallMaxWidth + ';\n}\n\n._3PWFTCm {\n  max-width: ' + theme.mediumMaxWidth + ';\n}\n\n._1sX5u9f {\n  max-width: ' + theme.largeMaxWidth + ';\n}\n\n._1t3N0ht {\n  max-width: 100vw;\n}\n\n.qG0OFE1 {\n  z-index: ' + theme.stackingTopmost + ';\n}\n\n._2D63jCg {\n  z-index: ' + theme.stackingAbove + ';\n}\n\n._1apXXYX {}\n\n._1afK_Gp {\n  z-index: ' + theme.stackingBelow + ';\n}\n\n.xwUVzYE {\n  z-index: ' + theme.stackingDeepest + ';\n}\n\n._1HeS82X {\n  box-shadow: ' + theme.shadowTopmost + ';\n}\n\n._1wzHede {\n  box-shadow: ' + theme.shadowResting + ';\n}\n\n._3yelDGE {\n  box-shadow: ' + theme.shadowAbove + ';\n}\n';
    };

    return tmpl.call(theme, theme);
  },
  'root': '_363uIlg',
  'display--block': '_2y5m9zy',
  'display--inline-block': '_2BenJlN',
  'display--flex': '-AGjWtH',
  'display--inline-flex': '_1nexD3g',
  'textAlign--start': '_1mgbjsV',
  'textAlign--center': '_3EAL0Vy',
  'textAlign--end': '_2DincID',
  'debug': '_3Y_0kws',
  'border': '_1wWuL2e',
  'background--inverse': '_3BX4paj',
  'background--default': '_2N_bJx5',
  'size--x-small': '_2i7pspc',
  'size--small': 'fADEtzk',
  'size--medium': '_3PWFTCm',
  'size--large': '_1sX5u9f',
  'size--fullscreen': '_1t3N0ht',
  'stacking--topmost': 'qG0OFE1',
  'stacking--above': '_2D63jCg',
  'stacking--resting': '_1apXXYX',
  'stacking--below': '_1afK_Gp',
  'stacking--deepest': 'xwUVzYE',
  'shadow--topmost': '_1HeS82X',
  'shadow--resting': '_1wzHede',
  'shadow--above': '_3yelDGE'
};


/**
---
category: components/layout
---
**/
var View = (_dec = (0, _deprecated2.default)('5.4.0', {
  size: 'maxWidth'
}), _dec2 = (0, _bidirectional2.default)(), _dec3 = (0, _uiThemeable2.default)(_theme2.default, styles), _dec(_class = _dec2(_class = _dec3(_class = (_temp = _class2 = function (_Component) {
  _inherits(View, _Component);

  function View(props) {
    _classCallCheck(this, View);

    var _this = _possibleConstructorReturn(this, (View.__proto__ || Object.getPrototypeOf(View)).call(this, props));

    _this.handleElementRef = function (el) {
      if (typeof _this.props.elementRef === 'function') {
        _this.props.elementRef(el);
      }
      _this._element = el;
    };

    _this.verifySpanMargin(props);
    _this._element = null;
    return _this;
  }

  _createClass(View, [{
    key: 'componentWillReceiveProps',
    value: function componentWillReceiveProps(nextProps) {
      this.verifySpanMargin(nextProps);
    }
  }, {
    key: 'verifySpanMargin',
    value: function verifySpanMargin(props) {
      var as = props.as,
          display = props.display,
          margin = props.margin;


      (0, _warning2.default)(!((as === 'span' || typeof as === 'undefined') && display === 'auto' && margin && margin !== 'none' && margin !== '0'), '[' + this.displayName + '] element of type \'span\' and display \'auto\' is inline ' + 'and will allow for horizontal margins only');
    }
  }, {
    key: 'render',
    value: function render() {
      var _classnames;

      var _props = this.props,
          children = _props.children,
          textAlign = _props.textAlign,
          background = _props.background,
          display = _props.display,
          debug = _props.debug,
          width = _props.width,
          height = _props.height,
          minWidth = _props.minWidth,
          minHeight = _props.minHeight,
          maxWidth = _props.maxWidth,
          maxHeight = _props.maxHeight,
          stacking = _props.stacking,
          shadow = _props.shadow,
          size = _props.size,
          className = _props.className;


      var ElementType = (0, _getElementType2.default)(View, this.props);

      return _react2.default.createElement(
        ElementType,
        Object.assign({}, (0, _passthroughProps.omitProps)(this.props, View.propTypes), {
          className: (0, _classnames3.default)((_classnames = {}, _defineProperty(_classnames, styles.root, true), _defineProperty(_classnames, styles.border, this.hasBorder), _defineProperty(_classnames, styles.debug, debug), _defineProperty(_classnames, styles['textAlign--' + textAlign], textAlign), _defineProperty(_classnames, styles['background--' + background], background), _defineProperty(_classnames, styles['display--' + display], display && display !== 'auto'), _defineProperty(_classnames, styles['size--' + size], size && size !== 'auto'), _defineProperty(_classnames, styles['stacking--' + stacking], stacking), _defineProperty(_classnames, styles['shadow--' + shadow], shadow), _defineProperty(_classnames, className, className), _classnames)),
          style: Object.assign({}, this.spacingStyle, this.borderStyle, {
            width: width,
            height: height,
            minWidth: minWidth,
            minHeight: minHeight,
            maxWidth: maxWidth,
            maxHeight: maxHeight
          }, this.styleProps),
          ref: this.handleElementRef
        }),
        children
      );
    }
  }, {
    key: 'displayName',
    get: function get() {
      return (0, _getDisplayName2.default)(View);
    }
  }, {
    key: 'hasBorder',
    get: function get() {
      var borderWidth = this.props.borderWidth;

      return borderWidth && borderWidth !== '0' && borderWidth !== 'none';
    }
  }, {
    key: 'borderStyle',
    get: function get() {
      var _props2 = this.props,
          borderRadius = _props2.borderRadius,
          borderWidth = _props2.borderWidth;


      if (this.dir === _bidirectional.DIRECTION.rtl) {
        borderRadius = (0, _mirrorShorthand.mirrorShorthandCorners)(borderRadius);
        borderWidth = (0, _mirrorShorthand.mirrorShorthandEdges)(borderWidth);
      }

      return {
        borderRadius: (0, _getShorthandPropValue2.default)(this.displayName, this.theme, borderRadius, 'borderRadius'),
        borderWidth: (0, _getShorthandPropValue2.default)(this.displayName, this.theme, borderWidth, 'borderWidth')
      };
    }
  }, {
    key: 'spacingStyle',
    get: function get() {
      var _props3 = this.props,
          margin = _props3.margin,
          padding = _props3.padding;


      if (this.dir === 'rtl') {
        margin = (0, _mirrorShorthand.mirrorShorthandEdges)(margin);
        padding = (0, _mirrorShorthand.mirrorShorthandEdges)(padding);
      }

      return {
        margin: (0, _getShorthandPropValue2.default)(this.displayName, this.theme, margin, 'margin'),
        padding: (0, _getShorthandPropValue2.default)(this.displayName, this.theme, padding, 'padding')
      };
    }
  }, {
    key: 'styleProps',
    get: function get() {
      var style = this.props.style; // eslint-disable-line react/prop-types

      return (0, _passthroughProps.pickProps)(style, {}, [
      // Position/calculateElementPosition:
      'top', 'left', 'position', 'display', 'transform', 'overflow', 'minWidth', 'minHeight',
      // Img:
      'filter',
      // FlexItem:
      'flexBasis',
      // Avatar:
      'backgroundImage']);
    }
  }]);

  View.displayName = 'View'
  ;
  return View;
}(_react.Component), _class2.propTypes = {
  /**
  * The element to render as the component root, `span` by default
  */
  as: _CustomPropTypes2.default.elementType,

  /**
  * provides a reference to the underlying html element
  */
  elementRef: _propTypes2.default.func,

  /**
  * By default the display prop is 'auto', meaning it takes on the
  * display rules of the html element it's rendered as (see `as` prop).
  */
  display: _propTypes2.default.oneOf(['auto', 'block', 'inline-block', 'flex', 'inline-flex']),

  /**
  * Set the margin using familiar CSS shorthand
  */
  margin: _ThemeablePropTypes2.default.spacing,

  /**
  * Set the padding using familiar CSS shorthand
  */
  padding: _ThemeablePropTypes2.default.spacing,

  height: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.number]),
  width: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.number]),
  maxHeight: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.number]),
  maxWidth: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.number]),
  minHeight: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.number]),
  minWidth: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.number]),

  /**
  * The children to render inside the `<View />`
  */
  children: _propTypes2.default.node,

  /**
  * Designates the text alignment within the `<View />`
  */
  textAlign: _propTypes2.default.oneOf(['start', 'center', 'end']),

  /**
  * Accepts the familiar CSS shorthand to designate border widths corresponding
  * to edges
  */
  borderWidth: _ThemeablePropTypes2.default.borderWidth,

  /**
  * Accepts the familiar CSS shorthand to designate border radii corresponding
  * to corners
  */
  borderRadius: _ThemeablePropTypes2.default.borderWidth,

  /**
  * Designates the background style of the `<View />`
  */
  background: _propTypes2.default.oneOf(['default', 'inverse', 'transparent']),

  /**
  * Controls the shadow depth for the `<View />`
  */
  shadow: _ThemeablePropTypes2.default.shadow,

  /**
  * Controls the z-index depth for the `<View />`
  */
  stacking: _ThemeablePropTypes2.default.stacking,

  /**
  * Activate a dotted outline around the component to make building your
  * layout easier
  */
  debug: _propTypes2.default.bool
}, _class2.defaultProps = {
  display: 'auto'
  // Note:
  // - `as` will default to type span via getElementType, so for consistency and
  // compatibility with Container we are leaving it undefined here. Otherwise
  // it modifies behavior for consuming components because of the logic around
  // default props in getElementType
  // - `textAlign` is undefined by default so that View can inherit text alignment
  // from parents
  // - Any props used to set inline styles should be undefined so that they
  // don't break consuming components' CSS
}, _temp)) || _class) || _class) || _class);
exports.default = View;
