// @instructure/ui-themeable/es/utils/transformCss.js
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import Browser from '@instructure/ui-utils/lib/Browser';
import parseCss, { ruleTypes } from './parseCss';

/**
* ---
* category: utilities/themes
* ---
* Transform a string of CSS
* @module transformCss
* @param {String} cssText CSS to parse and transform
* @param {Function} transform a transform function to apply to each rule
* @returns {String} the transformed CSS string
*/
export default function transformCss(cssText, transform) {
  var node = parseCss(cssText);

  if (typeof transform === 'function') {
    node = transformNode(node, transform);
  }

  return toCssText(node);
}

/**
* Checks if a rule is a keyframes selector/block
* @param {Object} rule
* @returns {Boolean} true if the rule is a keyframes rule
*/
export function isKeyframesSelector(rule) {
  return rule.parent && rule.parent.type === ruleTypes.keyframes;
}

/**
* Parses a string of CSS into an array of rules objects (selector + declaration block)
* Filters out any vendor rules that don't apply to the current browser
* @param {String} CSS text to parse
* @returns {Array} an array of rules objects
*/
export function toRules(cssText) {
  var node = parseCss(cssText);
  var rules = [];

  if (node.rules && node.rules.length > 0) {
    rules = node.rules.filter(function (rule) {
      return filterUnusedVendorRule(rule.selector);
    }).map(function (rule) {
      return toCssText(rule);
    });
  } else {
    var _cssText = toCssText(node);
    if (_cssText) {
      rules = [_cssText];
    }
  }

  return rules;
}

function transformNode(node, transform) {
  if (!node) {
    return;
  }

  if (node.type === ruleTypes.style) {
    return transform(node);
  }

  var rules = node.rules || [];
  var transformed = Object.assign({}, node);

  transformed.rules = rules.map(function (rule) {
    return transformNode(rule, transform);
  });

  return transformed;
}

function toCssText(node, text) {
  var cssText = '';
  var result = text || '';

  if (node.rules && node.rules.length > 0) {
    cssText = node.rules.map(function (rule) {
      return toCssText(rule, cssText);
    }).join('\n');
  } else {
    cssText = node.cssText.trim();

    if (cssText) {
      cssText = '  ' + cssText + '\n';
    }
  }

  if (cssText) {
    var prefix = node.selector ? node.selector + ' {\n' : '';
    var suffix = node.selector ? '}\n' : '';
    result += '' + prefix + cssText + suffix;
  }

  return result;
}

function filterUnusedVendorRule(selector) {
  if (!(Browser.msedge || Browser.msie) && selector.indexOf('-ms-') > -1) {
    return false;
  }

  if (!(Browser.webkit || Browser.msedge || Browser.blink) && selector.indexOf('-webkit-') > -1) {
    return false;
  }

  if (!Browser.gecko && selector.indexOf('-moz-') > -1) {
    return false;
  }

  return true;
}
