// @instructure/ui-layout/lib/components/Flex/FlexItem/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _dec, _class, _class2, _temp; /*
                                  * The MIT License (MIT)
                                  *
                                  * Copyright (c) 2015 - present Instructure, Inc.
                                  *
                                  * Permission is hereby granted, free of charge, to any person obtaining a copy
                                  * of this software and associated documentation files (the "Software"), to deal
                                  * in the Software without restriction, including without limitation the rights
                                  * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
                                  * copies of the Software, and to permit persons to whom the Software is
                                  * furnished to do so, subject to the following conditions:
                                  *
                                  * The above copyright notice and this permission notice shall be included in all
                                  * copies or substantial portions of the Software.
                                  *
                                  * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
                                  * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
                                  * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
                                  * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
                                  * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
                                  * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
                                  * SOFTWARE.
                                  */

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _uiThemeable = require('@instructure/ui-themeable');

var _uiThemeable2 = _interopRequireDefault(_uiThemeable);

var _CustomPropTypes = require('@instructure/ui-utils/lib/react/CustomPropTypes');

var _CustomPropTypes2 = _interopRequireDefault(_CustomPropTypes);

var _passthroughProps = require('@instructure/ui-utils/lib/react/passthroughProps');

var _ThemeablePropTypes = require('@instructure/ui-themeable/lib/utils/ThemeablePropTypes');

var _ThemeablePropTypes2 = _interopRequireDefault(_ThemeablePropTypes);

var _View = require('../../View');

var _View2 = _interopRequireDefault(_View);

var _theme = require('./theme');

var _theme2 = _interopRequireDefault(_theme);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var styles = {
  template: function template(theme) {
    var tmpl = function tmpl() {
      return '/*  imported from styles.css  */\n\n._1ON24IJ {\n  box-sizing: border-box;\n  min-width: 0.0625rem;\n\n  /* initial value is 1, but we want 0 as our default,\n  so users can opt in to shrink like they do grow */\n  flex-shrink: 0;\n}\n\n._3EyCz7f {\n  /* useful defaults for column layouts */\n  overflow-y: auto;\n}\n\n._35YAFNl {\n  align-self: flex-start;\n}\n\n._1mOX7GW {\n  align-self: flex-end;\n}\n\n._1fjR2aN {\n  align-self: center;\n}\n\n.GlAPhXi {\n  align-self: stretch;\n}\n\n._1AJdHPh {\n  flex-grow: 1;\n}\n\n._12L956q {\n  flex-shrink: 1;\n}\n\n.q3lxVGP {\n  outline: 0.0625rem dashed ' + theme.debugOutlineColor + ';\n}\n';
    };

    return tmpl.call(theme, theme);
  },
  'root': '_1ON24IJ',
  'column': '_3EyCz7f',
  'align--start': '_35YAFNl',
  'align--end': '_1mOX7GW',
  'align--center': '_1fjR2aN',
  'align--stretch': 'GlAPhXi',
  'grow': '_1AJdHPh',
  'shrink': '_12L956q',
  'visualDebug': 'q3lxVGP'
};


/**
---
parent: Flex
---
**/
var FlexItem = (_dec = (0, _uiThemeable2.default)(_theme2.default, styles), _dec(_class = (_temp = _class2 = function (_Component) {
  _inherits(FlexItem, _Component);

  function FlexItem() {
    _classCallCheck(this, FlexItem);

    return _possibleConstructorReturn(this, (FlexItem.__proto__ || Object.getPrototypeOf(FlexItem)).apply(this, arguments));
  }

  _createClass(FlexItem, [{
    key: 'render',
    value: function render() {
      var _classes;

      var props = (0, _passthroughProps.omitProps)(this.props, FlexItem.propTypes);

      var _props = this.props,
          align = _props.align,
          as = _props.as,
          children = _props.children,
          direction = _props.direction,
          grow = _props.grow,
          margin = _props.margin,
          padding = _props.padding,
          shrink = _props.shrink,
          size = _props.size,
          textAlign = _props.textAlign,
          visualDebug = _props.visualDebug;


      var dirColumn = direction === 'column';

      var style = {
        flexBasis: size
      };

      var classes = (_classes = {}, _defineProperty(_classes, styles.root, true), _defineProperty(_classes, styles.visualDebug, visualDebug), _defineProperty(_classes, styles.grow, grow), _defineProperty(_classes, styles.shrink, shrink), _defineProperty(_classes, styles['align--' + align], align), _defineProperty(_classes, styles.column, dirColumn), _classes);

      return _react2.default.createElement(
        _View2.default,
        Object.assign({}, props, {
          className: (0, _classnames2.default)(classes),
          style: style,
          as: as,
          width: dirColumn ? '100%' : undefined // eslint-disable-line no-undefined
          , minHeight: dirColumn ? size : undefined // eslint-disable-line no-undefined
          , minWidth: direction === 'row' ? size : undefined // eslint-disable-line no-undefined
          , textAlign: textAlign,
          margin: margin,
          padding: padding
        }),
        children
      );
    }
  }]);

  FlexItem.displayName = 'FlexItem'
  ;
  return FlexItem;
}(_react.Component), _class2.propTypes = {
  /**
  * overrides the parent Flex's alignItems prop, if needed
  */
  align: _propTypes2.default.oneOf(['center', 'start', 'end', 'stretch']),
  /**
  * the element type to render as
  */
  as: _CustomPropTypes2.default.elementType,
  children: _propTypes2.default.node,
  /**
  * Inherits from the parent Flex component
  */
  direction: _propTypes2.default.oneOf(['row', 'column']),
  /**
  * Should the FlexItem grow to fill any available space?
  */
  grow: _propTypes2.default.bool,
  /**
  * Should the FlexItem shrink (stopping at its `size`)?
  */
  shrink: _propTypes2.default.bool,
  /**
  * Sets the base size of the FlexItem (width if direction is `row`; height if direction is `column`)
  */
  size: _propTypes2.default.string,
  textAlign: _propTypes2.default.oneOf(['start', 'center', 'end']),
  /**
  * Valid values are `0`, `none`, `auto`, `xxx-small`, `xx-small`, `x-small`,
  * `small`, `medium`, `large`, `x-large`, `xx-large`. Apply these values via
  * familiar CSS-like shorthand. For example: `margin="small auto large"`.
  */

  margin: _ThemeablePropTypes2.default.spacing,
  /**
  * Valid values are `0`, `none`, `xxx-small`, `xx-small`, `x-small`,
  * `small`, `medium`, `large`, `x-large`, `xx-large`. Apply these values via
  * familiar CSS-like shorthand. For example: `padding="small x-large large"`.
  */
  padding: _ThemeablePropTypes2.default.spacing,
  /**
  * Places dashed lines around the component's borders to help debug your layout
  */
  visualDebug: _propTypes2.default.bool
}, _class2.defaultProps = {
  as: 'span',
  grow: false,
  shrink: false
}, _temp)) || _class);
exports.default = FlexItem;
