// @instructure/ui-layout/lib/components/Grid/GridRow/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _dec, _class, _class2, _temp; /*
                                   * The MIT License (MIT)
                                   *
                                   * Copyright (c) 2015 - present Instructure, Inc.
                                   *
                                   * Permission is hereby granted, free of charge, to any person obtaining a copy
                                   * of this software and associated documentation files (the "Software"), to deal
                                   * in the Software without restriction, including without limitation the rights
                                   * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
                                   * copies of the Software, and to permit persons to whom the Software is
                                   * furnished to do so, subject to the following conditions:
                                   *
                                   * The above copyright notice and this permission notice shall be included in all
                                   * copies or substantial portions of the Software.
                                   *
                                   * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
                                   * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
                                   * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
                                   * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
                                   * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
                                   * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
                                   * SOFTWARE.
                                   */

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _CustomPropTypes = require('@instructure/ui-utils/lib/react/CustomPropTypes');

var _CustomPropTypes2 = _interopRequireDefault(_CustomPropTypes);

var _safeCloneElement = require('@instructure/ui-utils/lib/react/safeCloneElement');

var _safeCloneElement2 = _interopRequireDefault(_safeCloneElement);

var _capitalizeFirstLetter = require('@instructure/ui-utils/lib/capitalizeFirstLetter');

var _capitalizeFirstLetter2 = _interopRequireDefault(_capitalizeFirstLetter);

var _uiThemeable = require('@instructure/ui-themeable');

var _uiThemeable2 = _interopRequireDefault(_uiThemeable);

var _matchComponentTypes = require('@instructure/ui-utils/lib/react/matchComponentTypes');

var _matchComponentTypes2 = _interopRequireDefault(_matchComponentTypes);

var _passthroughProps = require('@instructure/ui-utils/lib/react/passthroughProps');

var _GridCol = require('../GridCol');

var _GridCol2 = _interopRequireDefault(_GridCol);

var _theme = require('./theme');

var _theme2 = _interopRequireDefault(_theme);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var styles = {
  template: function template(theme) {
    var tmpl = function tmpl() {
      return '/*  imported from styles.css  */\n\n/* stylelint-disable at-rule-no-unknown */\n\n.kk5HawT {\n  display: block;\n  box-sizing: border-box;\n}\n\n._3np8nh4 {\n  display: flex;\n  flex-flow: row nowrap;\n}\n\n._3np8nh4._1OKfZLw {\n    justify-content: center;\n  }\n\n._3np8nh4._2rQKdL9 {\n    justify-content: flex-start;\n  }\n\n._3np8nh4._1OTva3F {\n    justify-content: flex-end;\n  }\n\n._3np8nh4._22481Ru {\n    justify-content: space-around;\n  }\n\n._3np8nh4._2Hmrzw_ {\n    justify-content: space-between;\n  }\n\n._3np8nh4.GzAAZ9m {\n    align-items: flex-start;\n  }\n\n._3np8nh4.O4VIIXu {\n    align-items: center;\n  }\n\n._3np8nh4._1zQ-2ec {\n    align-items: flex-end;\n  }\n\n._3np8nh4._5cWr83K {\n    margin: 0 calc(-1 * ' + theme.spacingSmall + ' / 2);\n  }\n\n._3np8nh4._1fKCQsQ {\n    margin: 0 calc(-1 * ' + theme.spacingMedium + ' / 2);\n  }\n\n._3np8nh4.BtTjeiv {\n    margin: 0 calc(-1 * ' + theme.spacingLarge + ' / 2);\n  }\n\n._3np8nh4.s_Wx-Xw {\n    margin-bottom: ' + theme.spacingSmall + ';\n  }\n\n._3np8nh4._1LKIxSg {\n    margin-bottom: ' + theme.spacingMedium + ';\n  }\n\n._3np8nh4._1dcq1F- {\n    margin-bottom: ' + theme.spacingLarge + ';\n  }\n\n._3np8nh4._2nUSL6N,\n  ._3np8nh4.N2IM_yx {\n    margin-bottom: 0;\n  }\n\n@media screen and (--mediumMin) {\n  ._3OegJQZ {\n    display: flex;\n    flex-flow: row nowrap;\n  }\n    ._3OegJQZ._1OKfZLw {\n    justify-content: center;\n  }\n    ._3OegJQZ._2rQKdL9 {\n    justify-content: flex-start;\n  }\n    ._3OegJQZ._1OTva3F {\n    justify-content: flex-end;\n  }\n    ._3OegJQZ._22481Ru {\n    justify-content: space-around;\n  }\n    ._3OegJQZ._2Hmrzw_ {\n    justify-content: space-between;\n  }\n    ._3OegJQZ.GzAAZ9m {\n    align-items: flex-start;\n  }\n    ._3OegJQZ.O4VIIXu {\n    align-items: center;\n  }\n    ._3OegJQZ._1zQ-2ec {\n    align-items: flex-end;\n  }\n    ._3OegJQZ._5cWr83K {\n    margin: 0 calc(-1 * ' + theme.spacingSmall + ' / 2);\n  }\n    ._3OegJQZ._1fKCQsQ {\n    margin: 0 calc(-1 * ' + theme.spacingMedium + ' / 2);\n  }\n    ._3OegJQZ.BtTjeiv {\n    margin: 0 calc(-1 * ' + theme.spacingLarge + ' / 2);\n  }\n    ._3OegJQZ.s_Wx-Xw {\n    margin-bottom: ' + theme.spacingSmall + ';\n  }\n    ._3OegJQZ._1LKIxSg {\n    margin-bottom: ' + theme.spacingMedium + ';\n  }\n    ._3OegJQZ._1dcq1F- {\n    margin-bottom: ' + theme.spacingLarge + ';\n  }\n    ._3OegJQZ._2nUSL6N,\n  ._3OegJQZ.N2IM_yx {\n    margin-bottom: 0;\n  }\n}\n\n@media screen and (--largeMin) {\n  ._3brUNKH {\n    display: flex;\n    flex-flow: row nowrap;\n  }\n    ._3brUNKH._1OKfZLw {\n    justify-content: center;\n  }\n    ._3brUNKH._2rQKdL9 {\n    justify-content: flex-start;\n  }\n    ._3brUNKH._1OTva3F {\n    justify-content: flex-end;\n  }\n    ._3brUNKH._22481Ru {\n    justify-content: space-around;\n  }\n    ._3brUNKH._2Hmrzw_ {\n    justify-content: space-between;\n  }\n    ._3brUNKH.GzAAZ9m {\n    align-items: flex-start;\n  }\n    ._3brUNKH.O4VIIXu {\n    align-items: center;\n  }\n    ._3brUNKH._1zQ-2ec {\n    align-items: flex-end;\n  }\n    ._3brUNKH._5cWr83K {\n    margin: 0 calc(-1 * ' + theme.spacingSmall + ' / 2);\n  }\n    ._3brUNKH._1fKCQsQ {\n    margin: 0 calc(-1 * ' + theme.spacingMedium + ' / 2);\n  }\n    ._3brUNKH.BtTjeiv {\n    margin: 0 calc(-1 * ' + theme.spacingLarge + ' / 2);\n  }\n    ._3brUNKH.s_Wx-Xw {\n    margin-bottom: ' + theme.spacingSmall + ';\n  }\n    ._3brUNKH._1LKIxSg {\n    margin-bottom: ' + theme.spacingMedium + ';\n  }\n    ._3brUNKH._1dcq1F- {\n    margin-bottom: ' + theme.spacingLarge + ';\n  }\n    ._3brUNKH._2nUSL6N,\n  ._3brUNKH.N2IM_yx {\n    margin-bottom: 0;\n  }\n}\n\n@media screen and (--xLargeMin) {\n  ._3brUNKH {\n    display: flex;\n    flex-flow: row nowrap;\n  }\n    ._3brUNKH._1OKfZLw {\n    justify-content: center;\n  }\n    ._3brUNKH._2rQKdL9 {\n    justify-content: flex-start;\n  }\n    ._3brUNKH._1OTva3F {\n    justify-content: flex-end;\n  }\n    ._3brUNKH._22481Ru {\n    justify-content: space-around;\n  }\n    ._3brUNKH._2Hmrzw_ {\n    justify-content: space-between;\n  }\n    ._3brUNKH.GzAAZ9m {\n    align-items: flex-start;\n  }\n    ._3brUNKH.O4VIIXu {\n    align-items: center;\n  }\n    ._3brUNKH._1zQ-2ec {\n    align-items: flex-end;\n  }\n    ._3brUNKH._5cWr83K {\n    margin: 0 calc(-1 * ' + theme.spacingSmall + ' / 2);\n  }\n    ._3brUNKH._1fKCQsQ {\n    margin: 0 calc(-1 * ' + theme.spacingMedium + ' / 2);\n  }\n    ._3brUNKH.BtTjeiv {\n    margin: 0 calc(-1 * ' + theme.spacingLarge + ' / 2);\n  }\n    ._3brUNKH.s_Wx-Xw {\n    margin-bottom: ' + theme.spacingSmall + ';\n  }\n    ._3brUNKH._1LKIxSg {\n    margin-bottom: ' + theme.spacingMedium + ';\n  }\n    ._3brUNKH._1dcq1F- {\n    margin-bottom: ' + theme.spacingLarge + ';\n  }\n    ._3brUNKH._2nUSL6N,\n  ._3brUNKH.N2IM_yx {\n    margin-bottom: 0;\n  }\n}\n\n._1OXyd-f {\n  /* stylelint-disable-next-line color-named */\n  outline: 0.0625rem dashed blue;\n}\n\n/* stylelint-enable at-rule-no-unknown */\n';
    };

    return tmpl.call(theme, theme);
  },
  'root': 'kk5HawT',
  'startAtSmall': '_3np8nh4',
  'hAlign--center': '_1OKfZLw',
  'hAlign--start': '_2rQKdL9',
  'hAlign--end': '_1OTva3F',
  'hAlign--space-around': '_22481Ru',
  'hAlign--space-between': '_2Hmrzw_',
  'vAlign--top': 'GzAAZ9m',
  'vAlign--middle': 'O4VIIXu',
  'vAlign--bottom': '_1zQ-2ec',
  'colSpacingSmall': '_5cWr83K',
  'colSpacingMedium': '_1fKCQsQ',
  'colSpacingLarge': 'BtTjeiv',
  'rowSpacingSmall': 's_Wx-Xw',
  'rowSpacingMedium': '_1LKIxSg',
  'rowSpacingLarge': '_1dcq1F-',
  'rowSpacingNone': '_2nUSL6N',
  'lastRow': 'N2IM_yx',
  'startAtMedium': '_3OegJQZ',
  'startAtLarge': '_3brUNKH',
  'visualDebug': '_1OXyd-f'
};


/**
---
parent: Grid
---
**/
var GridRow = (_dec = (0, _uiThemeable2.default)(_theme2.default, styles), _dec(_class = (_temp = _class2 = function (_Component) {
  _inherits(GridRow, _Component);

  function GridRow() {
    _classCallCheck(this, GridRow);

    return _possibleConstructorReturn(this, (GridRow.__proto__ || Object.getPrototypeOf(GridRow)).apply(this, arguments));
  }

  _createClass(GridRow, [{
    key: 'startAtClass',
    value: function startAtClass() {
      return !!this.props.startAt && 'startAt' + (0, _capitalizeFirstLetter2.default)(this.props.startAt);
    }
  }, {
    key: 'rowSpacingClass',
    value: function rowSpacingClass() {
      return 'rowSpacing' + (0, _capitalizeFirstLetter2.default)(this.props.rowSpacing);
    }
  }, {
    key: 'colSpacingClass',
    value: function colSpacingClass() {
      return 'colSpacing' + (0, _capitalizeFirstLetter2.default)(this.props.colSpacing);
    }
  }, {
    key: 'renderChildren',
    value: function renderChildren() {
      var _this2 = this;

      var _props = this.props,
          children = _props.children,
          props = _objectWithoutProperties(_props, ['children']);

      return _react.Children.map(children, function (child, index) {
        if ((0, _matchComponentTypes2.default)(child, [_GridCol2.default])) {
          return (0, _safeCloneElement2.default)(child, Object.assign({}, (0, _passthroughProps.pickProps)(_this2.props, GridRow.propTypes), child.props, { /* child props should override parent */
            isLastRow: props.isLastRow,
            isLastCol: index + 1 === _react.Children.count(children)
          }));
        } else {
          return child; // PropType validation should handle errors
        }
      });
    }
  }, {
    key: 'render',
    value: function render() {
      var _classes;

      var classes = (_classes = {}, _defineProperty(_classes, styles.root, true), _defineProperty(_classes, styles.lastRow, this.props.isLastRow), _defineProperty(_classes, styles['hAlign--' + this.props.hAlign], true), _defineProperty(_classes, styles['vAlign--' + this.props.vAlign], true), _defineProperty(_classes, styles[this.rowSpacingClass()], true), _defineProperty(_classes, styles[this.colSpacingClass()], this.props.colSpacing !== 'none'), _defineProperty(_classes, styles[this.startAtClass()], !!this.props.startAt), _classes);

      var props = (0, _passthroughProps.omitProps)(this.props, GridRow.propTypes);

      return _react2.default.createElement(
        'span',
        Object.assign({}, props, {
          className: (0, _classnames2.default)(classes)
        }),
        this.renderChildren()
      );
    }
  }]);

  GridRow.displayName = 'GridRow'
  ;
  return GridRow;
}(_react.Component), _class2.propTypes = {
  children: _CustomPropTypes2.default.Children.oneOf([_GridCol2.default]),
  rowSpacing: _propTypes2.default.oneOf(['none', 'small', 'medium', 'large']),
  colSpacing: _propTypes2.default.oneOf(['none', 'small', 'medium', 'large']),
  hAlign: _propTypes2.default.oneOf(['start', 'center', 'end', 'space-around', 'space-between']),
  vAlign: _propTypes2.default.oneOf(['top', 'middle', 'bottom']),
  startAt: _propTypes2.default.oneOf(['small', 'medium', 'large', 'x-large', null]),
  visualDebug: _propTypes2.default.bool,
  isLastRow: _propTypes2.default.bool
}, _temp)) || _class);
exports.default = GridRow;
