// @instructure/ui-buttons/lib/components/CloseButton/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _dec, _class, _class2, _temp; /*
                                   * The MIT License (MIT)
                                   *
                                   * Copyright (c) 2015 - present Instructure, Inc.
                                   *
                                   * Permission is hereby granted, free of charge, to any person obtaining a copy
                                   * of this software and associated documentation files (the "Software"), to deal
                                   * in the Software without restriction, including without limitation the rights
                                   * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
                                   * copies of the Software, and to permit persons to whom the Software is
                                   * furnished to do so, subject to the following conditions:
                                   *
                                   * The above copyright notice and this permission notice shall be included in all
                                   * copies or substantial portions of the Software.
                                   *
                                   * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
                                   * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
                                   * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
                                   * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
                                   * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
                                   * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
                                   * SOFTWARE.
                                   */

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames2 = require('classnames');

var _classnames3 = _interopRequireDefault(_classnames2);

var _IconX = require('@instructure/ui-icons/lib/Solid/IconX');

var _IconX2 = _interopRequireDefault(_IconX);

var _ScreenReaderContent = require('@instructure/ui-a11y/lib/components/ScreenReaderContent');

var _ScreenReaderContent2 = _interopRequireDefault(_ScreenReaderContent);

var _uiThemeable = require('@instructure/ui-themeable');

var _uiThemeable2 = _interopRequireDefault(_uiThemeable);

var _ThemeablePropTypes = require('@instructure/ui-themeable/lib/utils/ThemeablePropTypes');

var _ThemeablePropTypes2 = _interopRequireDefault(_ThemeablePropTypes);

var _passthroughProps = require('@instructure/ui-utils/lib/react/passthroughProps');

var _Button = require('../Button');

var _Button2 = _interopRequireDefault(_Button);

var _theme = require('./theme');

var _theme2 = _interopRequireDefault(_theme);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var styles = {
  template: function template(theme) {
    var tmpl = function tmpl() {
      return '/*  imported from styles.css  */\n\n._3m_usIo {\n  z-index: ' + theme.zIndex + ';\n  display: inline-block;\n}\n\n._36gt8Il {\n  position: static;\n}\n\n._1LmX8O6,\n._1CQpU_E {\n  position: absolute;\n}\n\n._1m7W3cg {\n  top: ' + theme.offsetXSmall + ';\n}\n\n._2NJu6Lc {\n  top: ' + theme.offsetSmall + ';\n}\n\n.uaV7z0y {\n  top: ' + theme.offsetMedium + ';\n}\n\n._1LmX8O6 {\n  offset-inline-end: 0;\n  offset-inline-start: auto;\n}\n\n[dir="ltr"] ._1LmX8O6 {\n  right: 0;\n  left: auto;\n}\n\n[dir="rtl"] ._1LmX8O6 {\n  left: 0;\n  right: auto;\n}\n\n._1LmX8O6._1m7W3cg {\n    offset-inline-end: ' + theme.offsetXSmall + ';\n    offset-inline-start: auto;\n  }\n\n[dir="ltr"] ._1LmX8O6._1m7W3cg {\n    right: ' + theme.offsetXSmall + ';\n    left: auto;\n  }\n\n[dir="rtl"] ._1LmX8O6._1m7W3cg {\n    left: ' + theme.offsetXSmall + ';\n    right: auto;\n  }\n\n._1LmX8O6._2NJu6Lc {\n    offset-inline-end: ' + theme.offsetSmall + ';\n    offset-inline-start: auto;\n  }\n\n[dir="ltr"] ._1LmX8O6._2NJu6Lc {\n    right: ' + theme.offsetSmall + ';\n    left: auto;\n  }\n\n[dir="rtl"] ._1LmX8O6._2NJu6Lc {\n    left: ' + theme.offsetSmall + ';\n    right: auto;\n  }\n\n._1LmX8O6.uaV7z0y {\n    offset-inline-end: ' + theme.offsetMedium + ';\n    offset-inline-start: auto;\n  }\n\n[dir="ltr"] ._1LmX8O6.uaV7z0y {\n    right: ' + theme.offsetMedium + ';\n    left: auto;\n  }\n\n[dir="rtl"] ._1LmX8O6.uaV7z0y {\n    left: ' + theme.offsetMedium + ';\n    right: auto;\n  }\n\n._1CQpU_E {\n  offset-inline-start: 0;\n  offset-inline-end: auto;\n}\n\n[dir="ltr"] ._1CQpU_E {\n  left: 0;\n  right: auto;\n}\n\n[dir="rtl"] ._1CQpU_E {\n  right: 0;\n  left: auto;\n}\n\n._1CQpU_E._1m7W3cg {\n    offset-inline-start: ' + theme.offsetXSmall + ';\n    offset-inline-end: auto;\n  }\n\n[dir="ltr"] ._1CQpU_E._1m7W3cg {\n    left: ' + theme.offsetXSmall + ';\n    right: auto;\n  }\n\n[dir="rtl"] ._1CQpU_E._1m7W3cg {\n    right: ' + theme.offsetXSmall + ';\n    left: auto;\n  }\n\n._1CQpU_E._2NJu6Lc {\n    offset-inline-start: ' + theme.offsetSmall + ';\n    offset-inline-end: auto;\n  }\n\n[dir="ltr"] ._1CQpU_E._2NJu6Lc {\n    left: ' + theme.offsetSmall + ';\n    right: auto;\n  }\n\n[dir="rtl"] ._1CQpU_E._2NJu6Lc {\n    right: ' + theme.offsetSmall + ';\n    left: auto;\n  }\n\n._1CQpU_E.uaV7z0y {\n    offset-inline-start: ' + theme.offsetMedium + ';\n    offset-inline-end: auto;\n  }\n\n[dir="ltr"] ._1CQpU_E.uaV7z0y {\n    left: ' + theme.offsetMedium + ';\n    right: auto;\n  }\n\n[dir="rtl"] ._1CQpU_E.uaV7z0y {\n    right: ' + theme.offsetMedium + ';\n    left: auto;\n  }\n';
    };

    return tmpl.call(theme, theme);
  },
  'root': '_3m_usIo',
  'placement--static': '_36gt8Il',
  'placement--end': '_1LmX8O6',
  'placement--start': '_1CQpU_E',
  'offset--x-small': '_1m7W3cg',
  'offset--small': '_2NJu6Lc',
  'offset--medium': 'uaV7z0y'
};


/**
---
category: components/utilities
---
**/
var CloseButton = (_dec = (0, _uiThemeable2.default)(_theme2.default, styles), _dec(_class = (_temp = _class2 = function (_Component) {
  _inherits(CloseButton, _Component);

  function CloseButton() {
    _classCallCheck(this, CloseButton);

    return _possibleConstructorReturn(this, (CloseButton.__proto__ || Object.getPrototypeOf(CloseButton)).apply(this, arguments));
  }

  _createClass(CloseButton, [{
    key: 'render',
    value: function render() {
      var _classnames;

      var _props = this.props,
          placement = _props.placement,
          offset = _props.offset;


      return _react2.default.createElement(
        'span',
        Object.assign({}, (0, _passthroughProps.omitProps)(this.props, CloseButton.propTypes), {
          className: (0, _classnames3.default)((_classnames = {}, _defineProperty(_classnames, styles.root, true), _defineProperty(_classnames, styles['placement--' + placement], placement), _defineProperty(_classnames, styles['offset--' + offset], offset), _classnames))
        }),
        _react2.default.createElement(
          _Button2.default,
          Object.assign({}, (0, _passthroughProps.pickProps)(this.props, _Button2.default.propTypes), { icon: _IconX2.default }),
          _react2.default.createElement(
            _ScreenReaderContent2.default,
            null,
            this.props.children
          )
        )
      );
    }
  }]);

  CloseButton.displayName = 'CloseButton'
  ;
  return CloseButton;
}(_react.Component), _class2.propTypes = {
  children: _propTypes2.default.node.isRequired,
  buttonRef: _propTypes2.default.func,
  size: _propTypes2.default.oneOf(['small', 'medium', 'large']),
  onClick: _propTypes2.default.func,
  margin: _ThemeablePropTypes2.default.spacing,
  placement: _propTypes2.default.oneOf(['start', 'end', 'static']),
  offset: _propTypes2.default.oneOf(['none', 'x-small', 'small', 'medium']),
  variant: _propTypes2.default.oneOf(['icon', 'icon-inverse'])
}, _class2.defaultProps = {
  onClick: function onClick(event) {},
  buttonRef: function buttonRef(el) {},
  variant: 'icon',
  placement: 'static',
  offset: 'x-small',
  size: 'small',
  margin: '0'
}, _temp)) || _class);
exports.default = CloseButton;
