// @instructure/ui-a11y/lib/components/ScreenReaderContent/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _dec, _class, _class2, _temp; /*
                                   * The MIT License (MIT)
                                   *
                                   * Copyright (c) 2015 - present Instructure, Inc.
                                   *
                                   * Permission is hereby granted, free of charge, to any person obtaining a copy
                                   * of this software and associated documentation files (the "Software"), to deal
                                   * in the Software without restriction, including without limitation the rights
                                   * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
                                   * copies of the Software, and to permit persons to whom the Software is
                                   * furnished to do so, subject to the following conditions:
                                   *
                                   * The above copyright notice and this permission notice shall be included in all
                                   * copies or substantial portions of the Software.
                                   *
                                   * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
                                   * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
                                   * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
                                   * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
                                   * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
                                   * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
                                   * SOFTWARE.
                                   */

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _uiThemeable = require('@instructure/ui-themeable');

var _uiThemeable2 = _interopRequireDefault(_uiThemeable);

var _CustomPropTypes = require('@instructure/ui-utils/lib/react/CustomPropTypes');

var _CustomPropTypes2 = _interopRequireDefault(_CustomPropTypes);

var _getElementType = require('@instructure/ui-utils/lib/react/getElementType');

var _getElementType2 = _interopRequireDefault(_getElementType);

var _passthroughProps = require('@instructure/ui-utils/lib/react/passthroughProps');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var styles = {
  template: function template(theme) {
    var tmpl = function tmpl() {
      return '/*  imported from styles.css  */\n\n.P20pAtR {\n  width: 0.0625rem;\n  height: 0.0625rem;\n  margin: -0.0625rem;\n  padding: 0;\n  position: absolute;\n  top: 0;\n  offset-inline-start: 0;\n  overflow: hidden;\n  clip: rect(0 0 0 0);\n  border: 0;\n}\n\n[dir="ltr"] .P20pAtR {\n  left: 0;\n}\n\n[dir="rtl"] .P20pAtR {\n  right: 0;\n}\n';
    };

    return tmpl.call(theme, theme);
  },
  'root': 'P20pAtR'
};

/**
---
category: components/utilities
---
**/

var ScreenReaderContent = (_dec = (0, _uiThemeable2.default)(null, styles), _dec(_class = (_temp = _class2 = function (_Component) {
  _inherits(ScreenReaderContent, _Component);

  function ScreenReaderContent() {
    _classCallCheck(this, ScreenReaderContent);

    return _possibleConstructorReturn(this, (ScreenReaderContent.__proto__ || Object.getPrototypeOf(ScreenReaderContent)).apply(this, arguments));
  }

  _createClass(ScreenReaderContent, [{
    key: 'render',
    value: function render() {
      var props = Object.assign({}, (0, _passthroughProps.omitProps)(this.props, ScreenReaderContent.propTypes), {
        className: styles.root
      });

      var ElementType = (0, _getElementType2.default)(ScreenReaderContent, this.props);

      return _react2.default.createElement(
        ElementType,
        props,
        this.props.children
      );
    }
  }]);

  ScreenReaderContent.displayName = 'ScreenReaderContent'
  ;
  return ScreenReaderContent;
}(_react.Component), _class2.propTypes = {
  /**
  * the element type to render as
  */
  as: _CustomPropTypes2.default.elementType,
  /**
  * content meant for screen readers only
  */
  children: _propTypes2.default.node
}, _class2.defaultProps = {
  as: 'span'
}, _temp)) || _class);
exports.default = ScreenReaderContent;
