// @instructure/ui-overlays/lib/components/Tooltip/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _dec, _class, _class2, _temp; /*
                                   * The MIT License (MIT)
                                   *
                                   * Copyright (c) 2015 - present Instructure, Inc.
                                   *
                                   * Permission is hereby granted, free of charge, to any person obtaining a copy
                                   * of this software and associated documentation files (the "Software"), to deal
                                   * in the Software without restriction, including without limitation the rights
                                   * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
                                   * copies of the Software, and to permit persons to whom the Software is
                                   * furnished to do so, subject to the following conditions:
                                   *
                                   * The above copyright notice and this permission notice shall be included in all
                                   * copies or substantial portions of the Software.
                                   *
                                   * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
                                   * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
                                   * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
                                   * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
                                   * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
                                   * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
                                   * SOFTWARE.
                                   */

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames2 = require('classnames');

var _classnames3 = _interopRequireDefault(_classnames2);

var _getElementType = require('@instructure/ui-utils/lib/react/getElementType');

var _getElementType2 = _interopRequireDefault(_getElementType);

var _CustomPropTypes = require('@instructure/ui-utils/lib/react/CustomPropTypes');

var _CustomPropTypes2 = _interopRequireDefault(_CustomPropTypes);

var _LayoutPropTypes = require('@instructure/ui-layout/lib/utils/LayoutPropTypes');

var _LayoutPropTypes2 = _interopRequireDefault(_LayoutPropTypes);

var _passthroughProps = require('@instructure/ui-utils/lib/react/passthroughProps');

var _ensureSingleChild = require('@instructure/ui-utils/lib/react/ensureSingleChild');

var _ensureSingleChild2 = _interopRequireDefault(_ensureSingleChild);

var _generateElementId = require('@instructure/ui-utils/lib/dom/generateElementId');

var _generateElementId2 = _interopRequireDefault(_generateElementId);

var _uiThemeable = require('@instructure/ui-themeable');

var _uiThemeable2 = _interopRequireDefault(_uiThemeable);

var _Popover = require('../Popover');

var _Popover2 = _interopRequireDefault(_Popover);

var _theme = require('./theme');

var _theme2 = _interopRequireDefault(_theme);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var styles = {
  template: function template(theme) {
    var tmpl = function tmpl() {
      return '/*  imported from styles.css  */\n\n._1-1Be03 {\n  font-family: ' + theme.fontFamily + ';\n  font-weight: ' + theme.fontWeight + ';\n  box-sizing: border-box;\n  display: block;\n}\n\n._3s5W1Nr {\n  font-size: ' + theme.smallFontSize + ';\n  padding: ' + theme.smallPadding + ';\n}\n\n._8RKQcke {\n  font-size: ' + theme.mediumFontSize + ';\n  padding: ' + theme.mediumPadding + ';\n}\n\n.Whw3920 {\n  font-size: ' + theme.largeFontSize + ';\n  padding: ' + theme.largePadding + ';\n}\n';
    };

    return tmpl.call(theme, theme);
  },
  'root': '_1-1Be03',
  'small': '_3s5W1Nr',
  'medium': '_8RKQcke',
  'large': 'Whw3920'
};


/**
---
category: components/dialogs
---
**/
var Tooltip = (_dec = (0, _uiThemeable2.default)(_theme2.default, styles), _dec(_class = (_temp = _class2 = function (_Component) {
  _inherits(Tooltip, _Component);

  function Tooltip(props) {
    _classCallCheck(this, Tooltip);

    var _this = _possibleConstructorReturn(this, (Tooltip.__proto__ || Object.getPrototypeOf(Tooltip)).call(this));

    _this._id = (0, _generateElementId2.default)('Tooltip');
    return _this;
  }

  _createClass(Tooltip, [{
    key: 'renderTrigger',
    value: function renderTrigger() {
      if (this.props.as) {
        var Trigger = (0, _getElementType2.default)(Tooltip, this.props);
        var props = (0, _passthroughProps.omitProps)(this.props, Tooltip.propTypes);
        return _react2.default.createElement(
          Trigger,
          props,
          this.props.children
        );
      } else {
        return (0, _ensureSingleChild2.default)(this.props.children);
      }
    }
  }, {
    key: 'render',
    value: function render() {
      var _classnames;

      var trigger = this.renderTrigger();
      return _react2.default.createElement(
        _Popover2.default,
        {
          on: this.props.on,
          shouldRenderOffscreen: true,
          shouldReturnFocus: false,
          placement: this.props.placement,
          variant: this.props.variant,
          mountNode: this.props.mountNode
        },
        _react2.default.createElement(
          _Popover.PopoverTrigger,
          {
            'aria-describedby': this._id,
            'aria-controls': this._id
          },
          trigger
        ),
        _react2.default.createElement(
          _Popover.PopoverContent,
          null,
          _react2.default.createElement(
            'span',
            {
              id: this._id,
              className: (0, _classnames3.default)((_classnames = {}, _defineProperty(_classnames, styles.root, true), _defineProperty(_classnames, styles[this.props.size], this.props.size), _classnames)),
              role: 'tooltip'
            },
            this.props.tip
          )
        )
      );
    }
  }]);

  Tooltip.displayName = 'Tooltip'
  ;
  return Tooltip;
}(_react.Component), _class2.propTypes = {
  tip: _propTypes2.default.node.isRequired,
  children: _propTypes2.default.node.isRequired,
  /**
  * the element type to render as (assumes a single child if 'as' is undefined)
  */
  as: _CustomPropTypes2.default.elementType,
  /**
   * The action that causes the Content to display (`click`, `hover`, `focus`)
   */
  on: _propTypes2.default.oneOfType([_propTypes2.default.oneOf(['click', 'hover', 'focus']), _propTypes2.default.arrayOf(_propTypes2.default.oneOf(['click', 'hover', 'focus']))]),
  variant: _propTypes2.default.oneOf(['default', 'inverse']),
  placement: _LayoutPropTypes2.default.placement,
  size: _propTypes2.default.oneOf(['small', 'medium', 'large']),
  /**
   * An element or a function returning an element to use as the mount node
   * for the `<Tooltip />` (defaults to `document.body`)
   */
  mountNode: _propTypes2.default.oneOfType([_CustomPropTypes2.default.element, _propTypes2.default.func])
}, _class2.defaultProps = {
  variant: 'default',
  placement: 'top',
  size: 'small',
  mountNode: null
}, _temp)) || _class);
exports.default = Tooltip;
