// @instructure/ui-layout/lib/components/ContextView/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _dec, _class, _class2, _temp; /*
                                   * The MIT License (MIT)
                                   *
                                   * Copyright (c) 2015 - present Instructure, Inc.
                                   *
                                   * Permission is hereby granted, free of charge, to any person obtaining a copy
                                   * of this software and associated documentation files (the "Software"), to deal
                                   * in the Software without restriction, including without limitation the rights
                                   * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
                                   * copies of the Software, and to permit persons to whom the Software is
                                   * furnished to do so, subject to the following conditions:
                                   *
                                   * The above copyright notice and this permission notice shall be included in all
                                   * copies or substantial portions of the Software.
                                   *
                                   * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
                                   * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
                                   * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
                                   * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
                                   * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
                                   * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
                                   * SOFTWARE.
                                   */

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames2 = require('classnames');

var _classnames3 = _interopRequireDefault(_classnames2);

var _uiThemeable = require('@instructure/ui-themeable');

var _uiThemeable2 = _interopRequireDefault(_uiThemeable);

var _ThemeablePropTypes = require('@instructure/ui-themeable/lib/utils/ThemeablePropTypes');

var _ThemeablePropTypes2 = _interopRequireDefault(_ThemeablePropTypes);

var _CustomPropTypes = require('@instructure/ui-utils/lib/react/CustomPropTypes');

var _CustomPropTypes2 = _interopRequireDefault(_CustomPropTypes);

var _passthroughProps = require('@instructure/ui-utils/lib/react/passthroughProps');

var _LayoutPropTypes = require('../../utils/LayoutPropTypes');

var _LayoutPropTypes2 = _interopRequireDefault(_LayoutPropTypes);

var _mirrorPlacement = require('../../utils/mirrorPlacement');

var _mirrorPlacement2 = _interopRequireDefault(_mirrorPlacement);

var _View = require('../View');

var _View2 = _interopRequireDefault(_View);

var _theme = require('./theme');

var _theme2 = _interopRequireDefault(_theme);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var styles = {
  template: function template(theme) {
    var tmpl = function tmpl() {
      return '/*  imported from styles.css  */\n\n._3aid-7P {\n  box-sizing: border-box;\n  min-height: calc(' + theme.arrowSize + ' * 2);\n}\n\n._1VlDtsa {\n  position: relative;\n}\n\n._11Fu-Zf,\n._1zBuxrq,\n.IanSaCn,\n._2tPG6e6,\n._1-WSn2P {\n  /* arrow on the left side pointing left */\n  -webkit-padding-start: ' + theme.arrowSize + ';\n          padding-inline-start: ' + theme.arrowSize + ';\n  -webkit-padding-end: 0;\n          padding-inline-end: 0;\n}\n\n[dir="ltr"] ._11Fu-Zf,\n[dir="ltr"] ._1zBuxrq,\n[dir="ltr"] .IanSaCn,\n[dir="ltr"] ._2tPG6e6,\n[dir="ltr"] ._1-WSn2P {\n  padding-left: ' + theme.arrowSize + ';\n  padding-right: 0;\n}\n\n[dir="rtl"] ._11Fu-Zf,\n[dir="rtl"] ._1zBuxrq,\n[dir="rtl"] .IanSaCn,\n[dir="rtl"] ._2tPG6e6,\n[dir="rtl"] ._1-WSn2P {\n  padding-right: ' + theme.arrowSize + ';\n  padding-left: 0;\n}\n\n.n6FPIKb,\n._11_p55d,\n._3iMQ1lt,\n._2t5X_r8,\n._2A89Fl7 {\n  /* arrow on the right side pointing right */\n  -webkit-padding-end: ' + theme.arrowSize + ';\n          padding-inline-end: ' + theme.arrowSize + ';\n  -webkit-padding-start: 0;\n          padding-inline-start: 0;\n}\n\n[dir="ltr"] .n6FPIKb,\n[dir="ltr"] ._11_p55d,\n[dir="ltr"] ._3iMQ1lt,\n[dir="ltr"] ._2t5X_r8,\n[dir="ltr"] ._2A89Fl7 {\n  padding-right: ' + theme.arrowSize + ';\n  padding-left: 0;\n}\n\n[dir="rtl"] .n6FPIKb,\n[dir="rtl"] ._11_p55d,\n[dir="rtl"] ._3iMQ1lt,\n[dir="rtl"] ._2t5X_r8,\n[dir="rtl"] ._2A89Fl7 {\n  padding-left: ' + theme.arrowSize + ';\n  padding-right: 0;\n}\n\n._3vnyOqc,\n._37hbT0_,\n._3-TtFGR,\n._2N351Rc {\n  /* arrow on the top side pointing up */\n  padding-top: ' + theme.arrowSize + ';\n}\n\n.r4NIxYE,\n._125pPog,\n._24Ps7mR,\n.fSy47W_ {\n  /* arrow on the bottom side pointing down */\n  padding-bottom: ' + theme.arrowSize + ';\n}\n\n.rD3lE20 {\n  position: absolute;\n  left: -999em; /* stylelint-disable-line property-blacklist */\n}\n\n.p4vTYe1,\n.p4vTYe1::after {\n  content: "";\n  height: 0;\n  width: 0;\n  position: absolute;\n  display: block;\n  box-sizing: border-box;\n  pointer-events: none;\n  border-style: solid;\n}\n\n.p4vTYe1 {\n  display: block;\n  border-width: calc(' + theme.arrowSize + ' + ' + theme.arrowBorderWidth + ');\n}\n\n.p4vTYe1::after {\n    border-width: ' + theme.arrowSize + ';\n  }\n\n._2UnpZoN,\n._2UnpZoN::after {\n  border-color: ' + theme.arrowBorderColor + ';\n}\n\n._2uV2QSD,\n._2uV2QSD::after {\n  border-color: ' + theme.arrowBorderColorInverse + ';\n}\n\n._2WoFmeI,\n._2t3kahU,\n.Ffgx2wh,\n._2fQ9WtH,\n._15YFE_k {\n  /* arrow on the left side pointing left */\n  top: 50%;\n  offset-inline-end: 100%;\n  offset-inline-start: auto;\n  margin-top: calc(-1 * (' + theme.arrowSize + ' + ' + theme.arrowBorderWidth + '));\n  border-inline-start-width: 0;\n  border-inline-start-color: transparent;\n  border-inline-end-color: ' + theme.arrowBorderColor + ';\n  border-top-color: transparent;\n  border-bottom-color: transparent;\n  border-inline-end-width: ' + theme.arrowSize + ';\n}\n\n[dir="ltr"] ._2WoFmeI,\n[dir="ltr"] ._2t3kahU,\n[dir="ltr"] .Ffgx2wh,\n[dir="ltr"] ._2fQ9WtH,\n[dir="ltr"] ._15YFE_k {\n  right: 100%;\n  left: auto;\n  border-left-width: 0;\n  border-left-color: transparent;\n  border-right-color: ' + theme.arrowBorderColor + ';\n  border-right-width: ' + theme.arrowSize + ';\n}\n\n[dir="rtl"] ._2WoFmeI,\n[dir="rtl"] ._2t3kahU,\n[dir="rtl"] .Ffgx2wh,\n[dir="rtl"] ._2fQ9WtH,\n[dir="rtl"] ._15YFE_k {\n  left: 100%;\n  right: auto;\n  border-right-width: 0;\n  border-right-color: transparent;\n  border-left-color: ' + theme.arrowBorderColor + ';\n  border-left-width: ' + theme.arrowSize + ';\n}\n\n._2WoFmeI::after, ._2t3kahU::after, .Ffgx2wh::after, ._2fQ9WtH::after, ._15YFE_k::after {\n    offset-inline-start: ' + theme.arrowBorderWidth + ';\n    offset-inline-end: auto;\n    margin-top: calc(-1 * ' + theme.arrowSize + ');\n    border-inline-start-width: 0;\n    border-inline-start-color: transparent;\n    border-top-color: transparent;\n    border-bottom-color: transparent;\n    border-inline-end-width: ' + theme.arrowSize + ';\n    border-inline-end-color: ' + theme.arrowBackgroundColor + ';\n  }\n\n[dir="ltr"] ._2WoFmeI::after,\n[dir="ltr"] ._2t3kahU::after,\n[dir="ltr"] .Ffgx2wh::after,\n[dir="ltr"] ._2fQ9WtH::after,\n[dir="ltr"] ._15YFE_k::after {\n    left: ' + theme.arrowBorderWidth + ';\n    right: auto;\n    border-left-width: 0;\n    border-left-color: transparent;\n    border-right-width: ' + theme.arrowSize + ';\n    border-right-color: ' + theme.arrowBackgroundColor + ';\n  }\n\n[dir="rtl"] ._2WoFmeI::after,\n[dir="rtl"] ._2t3kahU::after,\n[dir="rtl"] .Ffgx2wh::after,\n[dir="rtl"] ._2fQ9WtH::after,\n[dir="rtl"] ._15YFE_k::after {\n    right: ' + theme.arrowBorderWidth + ';\n    left: auto;\n    border-right-width: 0;\n    border-right-color: transparent;\n    border-left-width: ' + theme.arrowSize + ';\n    border-left-color: ' + theme.arrowBackgroundColor + ';\n  }\n\n._2WoFmeI._2uV2QSD, ._2t3kahU._2uV2QSD, .Ffgx2wh._2uV2QSD, ._2fQ9WtH._2uV2QSD, ._15YFE_k._2uV2QSD {\n    border-inline-end-color: ' + theme.arrowBorderColorInverse + ';\n  }\n\n[dir="ltr"] ._2WoFmeI._2uV2QSD,\n[dir="ltr"] ._2t3kahU._2uV2QSD,\n[dir="ltr"] .Ffgx2wh._2uV2QSD,\n[dir="ltr"] ._2fQ9WtH._2uV2QSD,\n[dir="ltr"] ._15YFE_k._2uV2QSD {\n    border-right-color: ' + theme.arrowBorderColorInverse + ';\n  }\n\n[dir="rtl"] ._2WoFmeI._2uV2QSD,\n[dir="rtl"] ._2t3kahU._2uV2QSD,\n[dir="rtl"] .Ffgx2wh._2uV2QSD,\n[dir="rtl"] ._2fQ9WtH._2uV2QSD,\n[dir="rtl"] ._15YFE_k._2uV2QSD {\n    border-left-color: ' + theme.arrowBorderColorInverse + ';\n  }\n\n._2WoFmeI._2uV2QSD::after, ._2t3kahU._2uV2QSD::after, .Ffgx2wh._2uV2QSD::after, ._2fQ9WtH._2uV2QSD::after, ._15YFE_k._2uV2QSD::after {\n      border-inline-end-color: ' + theme.arrowBackgroundColorInverse + ';\n    }\n\n[dir="ltr"] ._2WoFmeI._2uV2QSD::after,\n[dir="ltr"] ._2t3kahU._2uV2QSD::after,\n[dir="ltr"] .Ffgx2wh._2uV2QSD::after,\n[dir="ltr"] ._2fQ9WtH._2uV2QSD::after,\n[dir="ltr"] ._15YFE_k._2uV2QSD::after {\n      border-right-color: ' + theme.arrowBackgroundColorInverse + ';\n    }\n\n[dir="rtl"] ._2WoFmeI._2uV2QSD::after,\n[dir="rtl"] ._2t3kahU._2uV2QSD::after,\n[dir="rtl"] .Ffgx2wh._2uV2QSD::after,\n[dir="rtl"] ._2fQ9WtH._2uV2QSD::after,\n[dir="rtl"] ._15YFE_k._2uV2QSD::after {\n      border-left-color: ' + theme.arrowBackgroundColorInverse + ';\n    }\n\n._18u1QMY,\n._2UMkYNe,\n._3e_qMMG,\n._1CS_HAk,\n._3pEPAWV {\n  /* arrow on the right side pointing right */\n  top: 50%;\n  offset-inline-start: 100%;\n  offset-inline-end: auto;\n  margin-top: calc(-1 * (' + theme.arrowSize + ' + ' + theme.arrowBorderWidth + '));\n  border-inline-end-width: 0;\n  border-inline-end-color: transparent;\n  border-inline-start-color: ' + theme.arrowBorderColor + ';\n  border-top-color: transparent;\n  border-bottom-color: transparent;\n  border-inline-start-width: ' + theme.arrowSize + ';\n}\n\n[dir="ltr"] ._18u1QMY,\n[dir="ltr"] ._2UMkYNe,\n[dir="ltr"] ._3e_qMMG,\n[dir="ltr"] ._1CS_HAk,\n[dir="ltr"] ._3pEPAWV {\n  left: 100%;\n  right: auto;\n  border-right-width: 0;\n  border-right-color: transparent;\n  border-left-color: ' + theme.arrowBorderColor + ';\n  border-left-width: ' + theme.arrowSize + ';\n}\n\n[dir="rtl"] ._18u1QMY,\n[dir="rtl"] ._2UMkYNe,\n[dir="rtl"] ._3e_qMMG,\n[dir="rtl"] ._1CS_HAk,\n[dir="rtl"] ._3pEPAWV {\n  right: 100%;\n  left: auto;\n  border-left-width: 0;\n  border-left-color: transparent;\n  border-right-color: ' + theme.arrowBorderColor + ';\n  border-right-width: ' + theme.arrowSize + ';\n}\n\n._18u1QMY::after, ._2UMkYNe::after, ._3e_qMMG::after, ._1CS_HAk::after, ._3pEPAWV::after {\n    offset-inline-end: ' + theme.arrowBorderWidth + ';\n    offset-inline-start: auto;\n    margin-top: calc(-1 * ' + theme.arrowSize + ');\n    border-inline-end-width: 0;\n    border-inline-end-color: transparent;\n    border-top-color: transparent;\n    border-bottom-color: transparent;\n    border-inline-start-width: ' + theme.arrowSize + ';\n    border-inline-start-color: ' + theme.arrowBackgroundColor + ';\n  }\n\n[dir="ltr"] ._18u1QMY::after,\n[dir="ltr"] ._2UMkYNe::after,\n[dir="ltr"] ._3e_qMMG::after,\n[dir="ltr"] ._1CS_HAk::after,\n[dir="ltr"] ._3pEPAWV::after {\n    right: ' + theme.arrowBorderWidth + ';\n    left: auto;\n    border-right-width: 0;\n    border-right-color: transparent;\n    border-left-width: ' + theme.arrowSize + ';\n    border-left-color: ' + theme.arrowBackgroundColor + ';\n  }\n\n[dir="rtl"] ._18u1QMY::after,\n[dir="rtl"] ._2UMkYNe::after,\n[dir="rtl"] ._3e_qMMG::after,\n[dir="rtl"] ._1CS_HAk::after,\n[dir="rtl"] ._3pEPAWV::after {\n    left: ' + theme.arrowBorderWidth + ';\n    right: auto;\n    border-left-width: 0;\n    border-left-color: transparent;\n    border-right-width: ' + theme.arrowSize + ';\n    border-right-color: ' + theme.arrowBackgroundColor + ';\n  }\n\n._18u1QMY._2uV2QSD, ._2UMkYNe._2uV2QSD, ._3e_qMMG._2uV2QSD, ._1CS_HAk._2uV2QSD, ._3pEPAWV._2uV2QSD {\n    border-inline-start-color: ' + theme.arrowBorderColorInverse + ';\n  }\n\n[dir="ltr"] ._18u1QMY._2uV2QSD,\n[dir="ltr"] ._2UMkYNe._2uV2QSD,\n[dir="ltr"] ._3e_qMMG._2uV2QSD,\n[dir="ltr"] ._1CS_HAk._2uV2QSD,\n[dir="ltr"] ._3pEPAWV._2uV2QSD {\n    border-left-color: ' + theme.arrowBorderColorInverse + ';\n  }\n\n[dir="rtl"] ._18u1QMY._2uV2QSD,\n[dir="rtl"] ._2UMkYNe._2uV2QSD,\n[dir="rtl"] ._3e_qMMG._2uV2QSD,\n[dir="rtl"] ._1CS_HAk._2uV2QSD,\n[dir="rtl"] ._3pEPAWV._2uV2QSD {\n    border-right-color: ' + theme.arrowBorderColorInverse + ';\n  }\n\n._18u1QMY._2uV2QSD::after, ._2UMkYNe._2uV2QSD::after, ._3e_qMMG._2uV2QSD::after, ._1CS_HAk._2uV2QSD::after, ._3pEPAWV._2uV2QSD::after {\n      border-inline-start-color: ' + theme.arrowBackgroundColorInverse + ';\n    }\n\n[dir="ltr"] ._18u1QMY._2uV2QSD::after,\n[dir="ltr"] ._2UMkYNe._2uV2QSD::after,\n[dir="ltr"] ._3e_qMMG._2uV2QSD::after,\n[dir="ltr"] ._1CS_HAk._2uV2QSD::after,\n[dir="ltr"] ._3pEPAWV._2uV2QSD::after {\n      border-left-color: ' + theme.arrowBackgroundColorInverse + ';\n    }\n\n[dir="rtl"] ._18u1QMY._2uV2QSD::after,\n[dir="rtl"] ._2UMkYNe._2uV2QSD::after,\n[dir="rtl"] ._3e_qMMG._2uV2QSD::after,\n[dir="rtl"] ._1CS_HAk._2uV2QSD::after,\n[dir="rtl"] ._3pEPAWV._2uV2QSD::after {\n      border-right-color: ' + theme.arrowBackgroundColorInverse + ';\n    }\n\n.WlXLT6H,\n.oqXmyx-,\n._1mNK-Co,\n.l1-X79P {\n  /* arrow on the top side pointing up */\n  bottom: 100%;\n  -webkit-margin-start: calc(-1 * (' + theme.arrowSize + ' + ' + theme.arrowBorderWidth + '));\n          margin-inline-start: calc(-1 * (' + theme.arrowSize + ' + ' + theme.arrowBorderWidth + '));\n  -webkit-margin-end: 0;\n          margin-inline-end: 0;\n  border-top-width: 0;\n  border-top-color: transparent;\n  border-inline-start-color: transparent;\n  border-inline-end-color: transparent;\n}\n\n[dir="ltr"] .WlXLT6H,\n[dir="ltr"] .oqXmyx-,\n[dir="ltr"] ._1mNK-Co,\n[dir="ltr"] .l1-X79P {\n  margin-left: calc(-1 * (' + theme.arrowSize + ' + ' + theme.arrowBorderWidth + '));\n  margin-right: 0;\n  border-left-color: transparent;\n  border-right-color: transparent;\n}\n\n[dir="rtl"] .WlXLT6H,\n[dir="rtl"] .oqXmyx-,\n[dir="rtl"] ._1mNK-Co,\n[dir="rtl"] .l1-X79P {\n  margin-right: calc(-1 * (' + theme.arrowSize + ' + ' + theme.arrowBorderWidth + '));\n  margin-left: 0;\n  border-right-color: transparent;\n  border-left-color: transparent;\n}\n\n.WlXLT6H::after, .oqXmyx-::after, ._1mNK-Co::after, .l1-X79P::after {\n    top: ' + theme.arrowBorderWidth + ';\n    -webkit-margin-start: calc(-1 * ' + theme.arrowSize + ');\n            margin-inline-start: calc(-1 * ' + theme.arrowSize + ');\n    -webkit-margin-end: 0;\n            margin-inline-end: 0;\n    border-top-width: 0;\n    border-top-color: transparent;\n    border-inline-start-color: transparent;\n    border-inline-end-color: transparent;\n    border-bottom-color: ' + theme.arrowBackgroundColor + ';\n  }\n\n[dir="ltr"] .WlXLT6H::after,\n[dir="ltr"] .oqXmyx-::after,\n[dir="ltr"] ._1mNK-Co::after,\n[dir="ltr"] .l1-X79P::after {\n    margin-left: calc(-1 * ' + theme.arrowSize + ');\n    margin-right: 0;\n    border-left-color: transparent;\n    border-right-color: transparent;\n  }\n\n[dir="rtl"] .WlXLT6H::after,\n[dir="rtl"] .oqXmyx-::after,\n[dir="rtl"] ._1mNK-Co::after,\n[dir="rtl"] .l1-X79P::after {\n    margin-right: calc(-1 * ' + theme.arrowSize + ');\n    margin-left: 0;\n    border-right-color: transparent;\n    border-left-color: transparent;\n  }\n\n.WlXLT6H._2uV2QSD::after, .oqXmyx-._2uV2QSD::after, ._1mNK-Co._2uV2QSD::after, .l1-X79P._2uV2QSD::after {\n      border-bottom-color: ' + theme.arrowBackgroundColorInverse + ';\n    }\n\n._3_ss89c,\n._3-8CjdN,\n._3nnK9Ov,\n.OQ85pNy {\n  /* arrow on the bottom side pointing down */\n  top: 100%;\n  -webkit-margin-start: calc(-1 * (' + theme.arrowSize + ' + ' + theme.arrowBorderWidth + '));\n          margin-inline-start: calc(-1 * (' + theme.arrowSize + ' + ' + theme.arrowBorderWidth + '));\n  -webkit-margin-end: 0;\n          margin-inline-end: 0;\n  border-bottom-width: 0;\n  border-bottom-color: transparent;\n  border-inline-start-color: transparent;\n  border-inline-end-color: transparent;\n}\n\n[dir="ltr"] ._3_ss89c,\n[dir="ltr"] ._3-8CjdN,\n[dir="ltr"] ._3nnK9Ov,\n[dir="ltr"] .OQ85pNy {\n  margin-left: calc(-1 * (' + theme.arrowSize + ' + ' + theme.arrowBorderWidth + '));\n  margin-right: 0;\n  border-left-color: transparent;\n  border-right-color: transparent;\n}\n\n[dir="rtl"] ._3_ss89c,\n[dir="rtl"] ._3-8CjdN,\n[dir="rtl"] ._3nnK9Ov,\n[dir="rtl"] .OQ85pNy {\n  margin-right: calc(-1 * (' + theme.arrowSize + ' + ' + theme.arrowBorderWidth + '));\n  margin-left: 0;\n  border-right-color: transparent;\n  border-left-color: transparent;\n}\n\n._3_ss89c::after, ._3-8CjdN::after, ._3nnK9Ov::after, .OQ85pNy::after {\n    bottom: ' + theme.arrowBorderWidth + ';\n    -webkit-margin-start: calc(-1 * ' + theme.arrowSize + ');\n            margin-inline-start: calc(-1 * ' + theme.arrowSize + ');\n    -webkit-margin-end: 0;\n            margin-inline-end: 0;\n    border-bottom-width: 0;\n    border-bottom-color: transparent;\n    border-inline-start-color: transparent;\n    border-inline-end-color: transparent;\n    border-top-color: ' + theme.arrowBackgroundColor + ';\n  }\n\n[dir="ltr"] ._3_ss89c::after,\n[dir="ltr"] ._3-8CjdN::after,\n[dir="ltr"] ._3nnK9Ov::after,\n[dir="ltr"] .OQ85pNy::after {\n    margin-left: calc(-1 * ' + theme.arrowSize + ');\n    margin-right: 0;\n    border-left-color: transparent;\n    border-right-color: transparent;\n  }\n\n[dir="rtl"] ._3_ss89c::after,\n[dir="rtl"] ._3-8CjdN::after,\n[dir="rtl"] ._3nnK9Ov::after,\n[dir="rtl"] .OQ85pNy::after {\n    margin-right: calc(-1 * ' + theme.arrowSize + ');\n    margin-left: 0;\n    border-right-color: transparent;\n    border-left-color: transparent;\n  }\n\n._3_ss89c._2uV2QSD::after, ._3-8CjdN._2uV2QSD::after, ._3nnK9Ov._2uV2QSD::after, .OQ85pNy._2uV2QSD::after {\n      border-top-color: ' + theme.arrowBackgroundColorInverse + ';\n    }\n\n.WlXLT6H,\n._3_ss89c,\n.l1-X79P,\n.OQ85pNy {\n  offset-inline-start: 50%;\n}\n\n[dir="ltr"] .WlXLT6H,\n[dir="ltr"] ._3_ss89c,\n[dir="ltr"] .l1-X79P,\n[dir="ltr"] .OQ85pNy {\n  left: 50%;\n}\n\n[dir="rtl"] .WlXLT6H,\n[dir="rtl"] ._3_ss89c,\n[dir="rtl"] .l1-X79P,\n[dir="rtl"] .OQ85pNy {\n  right: 50%;\n}\n\n._1mNK-Co,\n._3-8CjdN {\n  offset-inline-start: calc((' + theme.arrowSize + ' + ' + theme.arrowBorderWidth + ') * 2);\n  offset-inline-end: auto;\n}\n\n[dir="ltr"] ._1mNK-Co,\n[dir="ltr"] ._3-8CjdN {\n  left: calc((' + theme.arrowSize + ' + ' + theme.arrowBorderWidth + ') * 2);\n  right: auto;\n}\n\n[dir="rtl"] ._1mNK-Co,\n[dir="rtl"] ._3-8CjdN {\n  right: calc((' + theme.arrowSize + ' + ' + theme.arrowBorderWidth + ') * 2);\n  left: auto;\n}\n\n.oqXmyx-,\n._3nnK9Ov {\n  offset-inline-start: calc(100% - (' + theme.arrowSize + ' + ' + theme.arrowBorderWidth + ') * 2);\n  offset-inline-end: auto;\n}\n\n[dir="ltr"] .oqXmyx-,\n[dir="ltr"] ._3nnK9Ov {\n  left: calc(100% - (' + theme.arrowSize + ' + ' + theme.arrowBorderWidth + ') * 2);\n  right: auto;\n}\n\n[dir="rtl"] .oqXmyx-,\n[dir="rtl"] ._3nnK9Ov {\n  right: calc(100% - (' + theme.arrowSize + ' + ' + theme.arrowBorderWidth + ') * 2);\n  left: auto;\n}\n\n._2t3kahU,\n._2UMkYNe {\n  top: calc((' + theme.arrowSize + ' + ' + theme.arrowBorderWidth + ') * 2);\n}\n\n.Ffgx2wh,\n._3e_qMMG {\n  top: calc(100% - (' + theme.arrowSize + ' + ' + theme.arrowBorderWidth + ') * 2);\n}\n';
    };

    return tmpl.call(theme, theme);
  },
  'root': '_3aid-7P',
  'content': '_1VlDtsa',
  'placement--end-center': '_11Fu-Zf',
  'placement--end-top': '_1zBuxrq',
  'placement--end-bottom': 'IanSaCn',
  'placement--center-end': '_2tPG6e6',
  'placement--end': '_1-WSn2P',
  'placement--start-center': 'n6FPIKb',
  'placement--start-top': '_11_p55d',
  'placement--start-bottom': '_3iMQ1lt',
  'placement--center-start': '_2t5X_r8',
  'placement--start': '_2A89Fl7',
  'placement--bottom': '_3vnyOqc',
  'placement--bottom-end': '_37hbT0_',
  'placement--bottom-start': '_3-TtFGR',
  'placement--bottom-center': '_2N351Rc',
  'placement--top': 'r4NIxYE',
  'placement--top-start': '_125pPog',
  'placement--top-end': '_24Ps7mR',
  'placement--top-center': 'fSy47W_',
  'placement--offscreen': 'rD3lE20',
  'arrow': 'p4vTYe1',
  'arrow--default': '_2UnpZoN',
  'arrow--inverse': '_2uV2QSD',
  'arrow--start-center': '_2WoFmeI',
  'arrow--start-top': '_2t3kahU',
  'arrow--start-bottom': 'Ffgx2wh',
  'arrow--center-start': '_2fQ9WtH',
  'arrow--start': '_15YFE_k',
  'arrow--end-center': '_18u1QMY',
  'arrow--end-top': '_2UMkYNe',
  'arrow--end-bottom': '_3e_qMMG',
  'arrow--center-end': '_1CS_HAk',
  'arrow--end': '_3pEPAWV',
  'arrow--top': 'WlXLT6H',
  'arrow--top-end': 'oqXmyx-',
  'arrow--top-start': '_1mNK-Co',
  'arrow--top-center': 'l1-X79P',
  'arrow--bottom': '_3_ss89c',
  'arrow--bottom-start': '_3-8CjdN',
  'arrow--bottom-end': '_3nnK9Ov',
  'arrow--bottom-center': 'OQ85pNy'
};


/**
---
category: components/layout
---
**/
var ContextView = (_dec = (0, _uiThemeable2.default)(_theme2.default, styles), _dec(_class = (_temp = _class2 = function (_Component) {
  _inherits(ContextView, _Component);

  function ContextView() {
    _classCallCheck(this, ContextView);

    return _possibleConstructorReturn(this, (ContextView.__proto__ || Object.getPrototypeOf(ContextView)).apply(this, arguments));
  }

  _createClass(ContextView, [{
    key: 'render',
    value: function render() {
      var _classes, _classnames;

      var _props = this.props,
          as = _props.as,
          background = _props.background,
          children = _props.children,
          debug = _props.debug,
          elementRef = _props.elementRef,
          height = _props.height,
          width = _props.width,
          maxHeight = _props.maxHeight,
          maxWidth = _props.maxWidth,
          minHeight = _props.minHeight,
          minWidth = _props.minWidth,
          margin = _props.margin,
          padding = _props.padding,
          placement = _props.placement,
          shadow = _props.shadow,
          stacking = _props.stacking,
          style = _props.style,
          textAlign = _props.textAlign;


      var classes = (_classes = {}, _defineProperty(_classes, styles.root, true), _defineProperty(_classes, styles['placement--' + placement.replace(' ', '-')], true), _classes);

      return _react2.default.createElement(
        _View2.default,
        Object.assign({}, (0, _passthroughProps.omitProps)(this.props, ContextView.propTypes), {
          style: style,
          className: (0, _classnames3.default)(classes),
          borderWidth: 'none',
          display: 'inline-block',
          as: as,
          debug: debug,
          elementRef: elementRef,
          margin: margin,
          stacking: stacking
        }),
        _react2.default.createElement(
          _View2.default,
          {
            className: styles.content,
            borderRadius: 'medium',
            borderWidth: 'small',
            display: 'block',
            background: background,
            debug: debug,
            height: height,
            width: width,
            maxHeight: maxHeight,
            maxWidth: maxWidth,
            minHeight: minHeight,
            minWidth: minWidth,
            padding: padding,
            shadow: shadow,
            textAlign: textAlign
          },
          _react2.default.createElement('span', {
            className: (0, _classnames3.default)((_classnames = {}, _defineProperty(_classnames, styles.arrow, true), _defineProperty(_classnames, styles['arrow--' + background], background), _defineProperty(_classnames, styles['arrow--' + this.mirroredPlacement], true), _classnames))
          }),
          children
        )
      );
    }
  }, {
    key: 'mirroredPlacement',
    get: function get() {
      return (0, _mirrorPlacement2.default)(this.props.placement, '-');
    }
  }]);

  ContextView.displayName = 'ContextView'
  ;
  return ContextView;
}(_react.Component), _class2.propTypes = {
  /**
  * The element to render as the component root
  */
  as: _CustomPropTypes2.default.elementType,

  /**
  * provides a reference to the underlying html root element
  */
  elementRef: _propTypes2.default.func,

  /**
  * Set the margin using familiar CSS shorthand
  */
  margin: _ThemeablePropTypes2.default.spacing,

  /**
  * Set the padding using familiar CSS shorthand
  */
  padding: _ThemeablePropTypes2.default.spacing,

  height: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.number]),
  width: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.number]),
  maxHeight: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.number]),
  maxWidth: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.number]),
  minHeight: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.number]),
  minWidth: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.number]),

  /**
  * The children to render inside the `<ContextView />`
  */
  children: _propTypes2.default.node,

  /**
  * Designates the text alignment within the `<ContextView />`
  */
  textAlign: _propTypes2.default.oneOf(['start', 'center', 'end']),

  /**
  * Controls the shadow depth for the `<ContextView />`
  */
  shadow: _ThemeablePropTypes2.default.shadow,

  /**
  * Controls the z-index depth for the `<ContextView />`
  */
  stacking: _ThemeablePropTypes2.default.stacking,

  /**
  * Designates the background style of the `<ContextView />`
  */
  background: _propTypes2.default.oneOf(['default', 'inverse']),

  /**
  * Specifies how the arrow for `<ContextView />` will be rendered.
  * Ex. `placement="top"` will render with an arrow pointing down.
  */
  placement: _LayoutPropTypes2.default.placement,

  /**
  * Activate an outline around the component to make building your
  * layout easier
  */
  debug: _propTypes2.default.bool
}, _class2.defaultProps = {
  as: 'span',
  elementRef: function elementRef(el) {},
  debug: false,
  width: 'auto',
  height: 'auto',
  children: null,
  textAlign: 'start',
  background: 'default',
  shadow: 'resting',
  placement: 'center end'
}, _temp)) || _class);
exports.default = ContextView;
