// @instructure/ui-elements/lib/components/List/ListItem/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _dec, _class, _class2, _temp; /*
                                   * The MIT License (MIT)
                                   *
                                   * Copyright (c) 2015 - present Instructure, Inc.
                                   *
                                   * Permission is hereby granted, free of charge, to any person obtaining a copy
                                   * of this software and associated documentation files (the "Software"), to deal
                                   * in the Software without restriction, including without limitation the rights
                                   * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
                                   * copies of the Software, and to permit persons to whom the Software is
                                   * furnished to do so, subject to the following conditions:
                                   *
                                   * The above copyright notice and this permission notice shall be included in all
                                   * copies or substantial portions of the Software.
                                   *
                                   * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
                                   * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
                                   * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
                                   * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
                                   * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
                                   * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
                                   * SOFTWARE.
                                   */


var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _View = require('@instructure/ui-layout/lib/components/View');

var _View2 = _interopRequireDefault(_View);

var _uiThemeable = require('@instructure/ui-themeable');

var _uiThemeable2 = _interopRequireDefault(_uiThemeable);

var _ThemeablePropTypes = require('@instructure/ui-themeable/lib/utils/ThemeablePropTypes');

var _ThemeablePropTypes2 = _interopRequireDefault(_ThemeablePropTypes);

var _passthroughProps = require('@instructure/ui-utils/lib/react/passthroughProps');

var _warning = require('@instructure/ui-utils/lib/warning');

var _warning2 = _interopRequireDefault(_warning);

var _deprecated = require('@instructure/ui-utils/lib/react/deprecated');

var _theme = require('./theme');

var _theme2 = _interopRequireDefault(_theme);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var styles = {
  template: function template(theme) {
    var tmpl = function tmpl() {
      return '/*  imported from styles.css  */\n\n._6NC1koz {\n  font-weight: ' + theme.fontWeight + ';\n  font-family: ' + theme.fontFamily + ';\n  line-height: ' + theme.lineHeight + ';\n  color: ' + theme.color + ';\n  padding: 0;\n}\n\n.ZHUv50L ._1oSbVkt {\n  -webkit-margin-start: ' + theme.noneSpacing + ';\n          margin-inline-start: ' + theme.noneSpacing + ';\n  -webkit-margin-end: ' + theme.noneSpacing + ';\n          margin-inline-end: ' + theme.noneSpacing + ';\n}\n\n[dir="ltr"] .ZHUv50L ._1oSbVkt {\n  margin-left: ' + theme.noneSpacing + ';\n  margin-right: ' + theme.noneSpacing + ';\n}\n\n[dir="rtl"] .ZHUv50L ._1oSbVkt {\n  margin-right: ' + theme.noneSpacing + ';\n  margin-left: ' + theme.noneSpacing + ';\n}\n\n._3M-tUDq ._1oSbVkt {\n  -webkit-margin-start: ' + theme.pipeSpacing + ';\n          margin-inline-start: ' + theme.pipeSpacing + ';\n  -webkit-margin-end: ' + theme.pipeSpacing + ';\n          margin-inline-end: ' + theme.pipeSpacing + ';\n}\n\n[dir="ltr"] ._3M-tUDq ._1oSbVkt {\n  margin-left: ' + theme.pipeSpacing + ';\n  margin-right: ' + theme.pipeSpacing + ';\n}\n\n[dir="rtl"] ._3M-tUDq ._1oSbVkt {\n  margin-right: ' + theme.pipeSpacing + ';\n  margin-left: ' + theme.pipeSpacing + ';\n}\n\n._3M-tUDq ._1oSbVkt::after {\n    content: "|";\n  }\n\n._109fhiE ._1oSbVkt {\n  -webkit-margin-start: ' + theme.slashSpacing + ';\n          margin-inline-start: ' + theme.slashSpacing + ';\n  -webkit-margin-end: ' + theme.slashSpacing + ';\n          margin-inline-end: ' + theme.slashSpacing + ';\n}\n\n[dir="ltr"] ._109fhiE ._1oSbVkt {\n  margin-left: ' + theme.slashSpacing + ';\n  margin-right: ' + theme.slashSpacing + ';\n}\n\n[dir="rtl"] ._109fhiE ._1oSbVkt {\n  margin-right: ' + theme.slashSpacing + ';\n  margin-left: ' + theme.slashSpacing + ';\n}\n\n._109fhiE ._1oSbVkt::after {\n    content: "/";\n  }\n\n._1zz67xT ._1oSbVkt {\n  -webkit-margin-start: ' + theme.arrowSpacing + ';\n          margin-inline-start: ' + theme.arrowSpacing + ';\n  -webkit-margin-end: ' + theme.arrowSpacing + ';\n          margin-inline-end: ' + theme.arrowSpacing + ';\n}\n\n[dir="ltr"] ._1zz67xT ._1oSbVkt {\n  margin-left: ' + theme.arrowSpacing + ';\n  margin-right: ' + theme.arrowSpacing + ';\n}\n\n[dir="rtl"] ._1zz67xT ._1oSbVkt {\n  margin-right: ' + theme.arrowSpacing + ';\n  margin-left: ' + theme.arrowSpacing + ';\n}\n\n._1zz67xT ._1oSbVkt::after {\n    content: ">";\n  }\n\n._3AXw3km {\n  list-style-position: inside;\n}\n\n._3AXw3km:not(:first-child) {\n    border-top: ' + theme.delimiterDashedBorder + ';\n  }\n\n._2WikxD_ {\n  list-style-position: inside;\n}\n\n._2WikxD_:not(:first-child) {\n    border-top: ' + theme.delimiterSolidBorder + ';\n  }\n\n._1GrsSPS {\n  display: inline-block;\n  vertical-align: middle;\n  max-width: 100%;\n  box-sizing: border-box;\n  line-height: ' + theme.pipeLineHeight + ';\n  color: ' + theme.inlineColor + ';\n  font-size: ' + theme.fontSizeSmall + ';\n}\n\n._1GrsSPS:last-of-type ._1oSbVkt::after {\n    content: none;\n  }\n\n._1GrsSPS._2a0TLnK {\n    -webkit-margin-start: ' + theme.spacingXXXSmall + ';\n            margin-inline-start: ' + theme.spacingXXXSmall + ';\n    -webkit-margin-end: ' + theme.spacingXXXSmall + ';\n            margin-inline-end: ' + theme.spacingXXXSmall + ';\n  }\n\n[dir="ltr"] ._1GrsSPS._2a0TLnK {\n    margin-left: ' + theme.spacingXXXSmall + ';\n    margin-right: ' + theme.spacingXXXSmall + ';\n  }\n\n[dir="rtl"] ._1GrsSPS._2a0TLnK {\n    margin-right: ' + theme.spacingXXXSmall + ';\n    margin-left: ' + theme.spacingXXXSmall + ';\n  }\n\n._1GrsSPS.kd9tW8J {\n    -webkit-margin-start: ' + theme.spacingXXSmall + ';\n            margin-inline-start: ' + theme.spacingXXSmall + ';\n    -webkit-margin-end: ' + theme.spacingXXSmall + ';\n            margin-inline-end: ' + theme.spacingXXSmall + ';\n  }\n\n[dir="ltr"] ._1GrsSPS.kd9tW8J {\n    margin-left: ' + theme.spacingXXSmall + ';\n    margin-right: ' + theme.spacingXXSmall + ';\n  }\n\n[dir="rtl"] ._1GrsSPS.kd9tW8J {\n    margin-right: ' + theme.spacingXXSmall + ';\n    margin-left: ' + theme.spacingXXSmall + ';\n  }\n\n._1GrsSPS.vcxlDcz {\n    -webkit-margin-start: ' + theme.spacingXSmall + ';\n            margin-inline-start: ' + theme.spacingXSmall + ';\n    -webkit-margin-end: ' + theme.spacingXSmall + ';\n            margin-inline-end: ' + theme.spacingXSmall + ';\n  }\n\n[dir="ltr"] ._1GrsSPS.vcxlDcz {\n    margin-left: ' + theme.spacingXSmall + ';\n    margin-right: ' + theme.spacingXSmall + ';\n  }\n\n[dir="rtl"] ._1GrsSPS.vcxlDcz {\n    margin-right: ' + theme.spacingXSmall + ';\n    margin-left: ' + theme.spacingXSmall + ';\n  }\n\n._1GrsSPS._3RlTf2l {\n    -webkit-margin-start: ' + theme.spacingSmall + ';\n            margin-inline-start: ' + theme.spacingSmall + ';\n    -webkit-margin-end: ' + theme.spacingSmall + ';\n            margin-inline-end: ' + theme.spacingSmall + ';\n  }\n\n[dir="ltr"] ._1GrsSPS._3RlTf2l {\n    margin-left: ' + theme.spacingSmall + ';\n    margin-right: ' + theme.spacingSmall + ';\n  }\n\n[dir="rtl"] ._1GrsSPS._3RlTf2l {\n    margin-right: ' + theme.spacingSmall + ';\n    margin-left: ' + theme.spacingSmall + ';\n  }\n\n._1GrsSPS._1Q8Q7Yg {\n    -webkit-margin-start: ' + theme.spacingMedium + ';\n            margin-inline-start: ' + theme.spacingMedium + ';\n    -webkit-margin-end: ' + theme.spacingMedium + ';\n            margin-inline-end: ' + theme.spacingMedium + ';\n  }\n\n[dir="ltr"] ._1GrsSPS._1Q8Q7Yg {\n    margin-left: ' + theme.spacingMedium + ';\n    margin-right: ' + theme.spacingMedium + ';\n  }\n\n[dir="rtl"] ._1GrsSPS._1Q8Q7Yg {\n    margin-right: ' + theme.spacingMedium + ';\n    margin-left: ' + theme.spacingMedium + ';\n  }\n\n._1GrsSPS._3RdSymW {\n    -webkit-margin-start: ' + theme.spacingLarge + ';\n            margin-inline-start: ' + theme.spacingLarge + ';\n    -webkit-margin-end: ' + theme.spacingLarge + ';\n            margin-inline-end: ' + theme.spacingLarge + ';\n  }\n\n[dir="ltr"] ._1GrsSPS._3RdSymW {\n    margin-left: ' + theme.spacingLarge + ';\n    margin-right: ' + theme.spacingLarge + ';\n  }\n\n[dir="rtl"] ._1GrsSPS._3RdSymW {\n    margin-right: ' + theme.spacingLarge + ';\n    margin-left: ' + theme.spacingLarge + ';\n  }\n\n._1GrsSPS.X9LKzdG {\n    -webkit-margin-start: ' + theme.spacingXLarge + ';\n            margin-inline-start: ' + theme.spacingXLarge + ';\n    -webkit-margin-end: ' + theme.spacingXLarge + ';\n            margin-inline-end: ' + theme.spacingXLarge + ';\n  }\n\n[dir="ltr"] ._1GrsSPS.X9LKzdG {\n    margin-left: ' + theme.spacingXLarge + ';\n    margin-right: ' + theme.spacingXLarge + ';\n  }\n\n[dir="rtl"] ._1GrsSPS.X9LKzdG {\n    margin-right: ' + theme.spacingXLarge + ';\n    margin-left: ' + theme.spacingXLarge + ';\n  }\n\n._1GrsSPS._15Rc45h {\n    -webkit-margin-start: ' + theme.spacingXXLarge + ';\n            margin-inline-start: ' + theme.spacingXXLarge + ';\n    -webkit-margin-end: ' + theme.spacingXXLarge + ';\n            margin-inline-end: ' + theme.spacingXXLarge + ';\n  }\n\n[dir="ltr"] ._1GrsSPS._15Rc45h {\n    margin-left: ' + theme.spacingXXLarge + ';\n    margin-right: ' + theme.spacingXXLarge + ';\n  }\n\n[dir="rtl"] ._1GrsSPS._15Rc45h {\n    margin-right: ' + theme.spacingXXLarge + ';\n    margin-left: ' + theme.spacingXXLarge + ';\n  }\n\n._1GrsSPS:first-of-type {\n    -webkit-margin-start: 0;\n            margin-inline-start: 0;\n  }\n\n[dir="ltr"] ._1GrsSPS:first-of-type {\n    margin-left: 0;\n  }\n\n[dir="rtl"] ._1GrsSPS:first-of-type {\n    margin-right: 0;\n  }\n\n._1GrsSPS:last-of-type {\n    -webkit-margin-end: 0;\n            margin-inline-end: 0;\n  }\n\n[dir="ltr"] ._1GrsSPS:last-of-type {\n    margin-right: 0;\n  }\n\n[dir="rtl"] ._1GrsSPS:last-of-type {\n    margin-left: 0;\n  }\n\n._6NC1koz:not(._1GrsSPS)._2a0TLnK {\n    margin-top: ' + theme.spacingXXXSmall + ';\n    margin-bottom: ' + theme.spacingXXXSmall + ';\n  }\n\n._6NC1koz:not(._1GrsSPS).kd9tW8J {\n    margin-top: ' + theme.spacingXXSmall + ';\n    margin-bottom: ' + theme.spacingXXSmall + ';\n  }\n\n._6NC1koz:not(._1GrsSPS).vcxlDcz {\n    margin-top: ' + theme.spacingXSmall + ';\n    margin-bottom: ' + theme.spacingXSmall + ';\n  }\n\n._6NC1koz:not(._1GrsSPS)._3RlTf2l {\n    margin-top: ' + theme.spacingSmall + ';\n    margin-bottom: ' + theme.spacingSmall + ';\n  }\n\n._6NC1koz:not(._1GrsSPS)._1Q8Q7Yg {\n    margin-top: ' + theme.spacingMedium + ';\n    margin-bottom: ' + theme.spacingMedium + ';\n  }\n\n._6NC1koz:not(._1GrsSPS)._3RdSymW {\n    margin-top: ' + theme.spacingLarge + ';\n    margin-bottom: ' + theme.spacingLarge + ';\n  }\n\n._6NC1koz:not(._1GrsSPS).X9LKzdG {\n    margin-top: ' + theme.spacingXLarge + ';\n    margin-bottom: ' + theme.spacingXLarge + ';\n  }\n\n._6NC1koz:not(._1GrsSPS)._15Rc45h {\n    margin-top: ' + theme.spacingXXLarge + ';\n    margin-bottom: ' + theme.spacingXXLarge + ';\n  }\n\n._6NC1koz:not(._1GrsSPS):first-of-type {\n    margin-top: 0;\n  }\n\n._6NC1koz:not(._1GrsSPS):last-of-type {\n    margin-bottom: 0;\n  }\n\n._2X3cSZD {\n  font-size: ' + theme.fontSizeSmall + ';\n}\n\n._1BCOnpY {\n  font-size: ' + theme.fontSizeMedium + ';\n}\n\n._27hzYFw {\n  font-size: ' + theme.fontSizeLarge + ';\n}\n';
    };

    return tmpl.call(theme, theme);
  },
  'root': '_6NC1koz',
  'delimiter--none': 'ZHUv50L',
  'delimiter': '_1oSbVkt',
  'delimiter--pipe': '_3M-tUDq',
  'delimiter--slash': '_109fhiE',
  'delimiter--arrow': '_1zz67xT',
  'delimiter--dashed': '_3AXw3km',
  'delimiter--solid': '_2WikxD_',
  'inline': '_1GrsSPS',
  'spacing--xxx-small': '_2a0TLnK',
  'spacing--xx-small': 'kd9tW8J',
  'spacing--x-small': 'vcxlDcz',
  'spacing--small': '_3RlTf2l',
  'spacing--medium': '_1Q8Q7Yg',
  'spacing--large': '_3RdSymW',
  'spacing--x-large': 'X9LKzdG',
  'spacing--xx-large': '_15Rc45h',
  'small': '_2X3cSZD',
  'medium': '_1BCOnpY',
  'large': '_27hzYFw'
};


/**
---
parent: List
---
**/
var ListItem = (_dec = (0, _uiThemeable2.default)(_theme2.default, styles), _dec(_class = (_temp = _class2 = function (_Component) {
  _inherits(ListItem, _Component);

  function ListItem() {
    _classCallCheck(this, ListItem);

    return _possibleConstructorReturn(this, (ListItem.__proto__ || Object.getPrototypeOf(ListItem)).apply(this, arguments));
  }

  _createClass(ListItem, [{
    key: 'render',
    value: function render() {
      var _classes;

      var props = (0, _passthroughProps.omitProps)(this.props, Object.assign({}, ListItem.propTypes, _View2.default.propTypes));
      var delimiter = this.props.variant === 'pipe' ? 'pipe' : this.props.delimiter;
      var variant = this.props.variant === 'pipe' ? 'inline' : this.props.variant;
      var size = this.props.variant === 'pipe' ? 'small' : this.props.size;

      var noDelimiter = delimiter === 'none' && variant !== 'inline';

      var noSpacing = this.props.delimiter !== 'none' || this.props.variant === 'pipe';
      (0, _warning2.default)(!(noSpacing && this.props.spacing), '[List] `itemSpacing` has no effect inside Lists with the `delimiter`\n      prop set to anything other than `none`, or with a `variant` of `pipe`');

      var classes = (_classes = {}, _defineProperty(_classes, styles.root, true), _defineProperty(_classes, styles.inline, variant === 'inline'), _defineProperty(_classes, styles[size], size), _defineProperty(_classes, styles['delimiter--' + delimiter], true), _defineProperty(_classes, styles['spacing--' + this.props.spacing], this.props.spacing && !noSpacing), _classes);

      return _react2.default.createElement(
        _View2.default,
        Object.assign({}, props, {
          as: 'li',
          margin: this.props.margin,
          padding: this.props.padding,
          className: (0, _classnames2.default)(classes),
          elementRef: this.props.elementRef
        }),
        this.props.children,
        !noDelimiter && _react2.default.createElement('span', {
          className: styles.delimiter,
          'aria-hidden': 'true'
        })
      );
    }
  }]);

  ListItem.displayName = 'ListItem'
  ;
  return ListItem;
}(_react.Component), _class2.propTypes = {
  children: _propTypes2.default.node.isRequired,
  variant: (0, _deprecated.deprecatePropValues)(_propTypes2.default.oneOf(['default', 'unstyled', 'inline']), ['pipe'], 'For the same functionality, use `inline` on the `variant` prop and set the `delimiter` prop to `pipe`.'),
  delimiter: _propTypes2.default.oneOf(['none', 'pipe', 'slash', 'arrow', 'dashed', 'solid']),
  size: _propTypes2.default.oneOf(['small', 'medium', 'large']),
  /**
  * Valid values are `0`, `none`, `auto`, `xxx-small`, `xx-small`, `x-small`,
  * `small`, `medium`, `large`, `x-large`, `xx-large`. Apply these values via
  * familiar CSS-like shorthand. For example: `margin="small auto large"`.
  */
  margin: _ThemeablePropTypes2.default.spacing,
  /**
  * Valid values are `0`, `none`, `xxx-small`, `xx-small`, `x-small`,
  * `small`, `medium`, `large`, `x-large`, `xx-large`. Apply these values via
  * familiar CSS-like shorthand. For example: `padding="small x-large large"`.
  */
  padding: _ThemeablePropTypes2.default.spacing,
  spacing: _propTypes2.default.oneOf(['xxx-small', 'xx-small', 'x-small', 'small', 'medium', 'large', 'x-large', 'xx-large']),
  elementRef: _propTypes2.default.func
}, _temp)) || _class);
exports.default = ListItem;
