// @instructure/ui-elements/es/components/Link/index.js
var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _dec, _class, _class2, _temp2;

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';

import View from '@instructure/ui-layout/lib/components/View';

import themeable from '@instructure/ui-themeable';
import CustomPropTypes from '@instructure/ui-utils/lib/react/CustomPropTypes';
import ThemeablePropTypes from '@instructure/ui-themeable/lib/utils/ThemeablePropTypes';
import getElementType from '@instructure/ui-utils/lib/react/getElementType';
import { omitProps } from '@instructure/ui-utils/lib/react/passthroughProps';
import isActiveElement from '@instructure/ui-utils/lib/dom/isActiveElement';
import findDOMNode from '@instructure/ui-utils/lib/dom/findDOMNode';
import hasVisibleChildren from '@instructure/ui-a11y/lib/utils/hasVisibleChildren';

var styles = {
  template: function template(theme) {
    var tmpl = function tmpl() {
      return '/*  imported from styles.css  */\n\n/* stylelint-disable selector-max-class, no-descending-specificity, selector-max-type */\n\n.smso19r a._1wRO7PZ:link,\n.smso19r a._1wRO7PZ:visited,\n.smso19r button._1wRO7PZ {\n  cursor: pointer;\n  -webkit-font-smoothing: antialiased;\n  -moz-osx-font-smoothing: grayscale;\n  font-family: ' + theme.fontFamily + ';\n  font-weight: ' + theme.fontWeight + ';\n  color: ' + theme.color + ';\n  -webkit-text-decoration: ' + theme.textDecoration + ';\n          text-decoration: ' + theme.textDecoration + ';\n  transition: color 0.2s;\n  outline: ' + theme.outlineWidth + ' ' + theme.outlineStyle + ' ' + theme.outlineColor + ';\n  outline-offset: ' + theme.outlineOffset + ';\n}\n\n/* does not work in Edge or IE */\n\n.smso19r a._1wRO7PZ:link:focus,\n  .smso19r a._1wRO7PZ:link:active,\n  .smso19r a._1wRO7PZ:visited:focus,\n  .smso19r a._1wRO7PZ:visited:active,\n  .smso19r button._1wRO7PZ:focus,\n  .smso19r button._1wRO7PZ:active {\n    outline: ' + theme.outlineWidth + ' ' + theme.outlineStyle + ' ' + theme.focusOutlineColor + ';\n  }\n\n.smso19r a._1wRO7PZ:link._2SYApbw:focus,\n    .smso19r a._1wRO7PZ:link._2SYApbw:active,\n    .smso19r a._1wRO7PZ:visited._2SYApbw:focus,\n    .smso19r a._1wRO7PZ:visited._2SYApbw:active,\n    .smso19r button._1wRO7PZ._2SYApbw:focus,\n    .smso19r button._1wRO7PZ._2SYApbw:active {\n      outline: ' + theme.outlineWidth + ' ' + theme.outlineStyle + ' ' + theme.focusIconOutlineColor + ';\n    }\n\n.smso19r a._1wRO7PZ:link:focus,\n  .smso19r a._1wRO7PZ:link:hover,\n  .smso19r a._1wRO7PZ:link:active,\n  .smso19r a._1wRO7PZ:visited:focus,\n  .smso19r a._1wRO7PZ:visited:hover,\n  .smso19r a._1wRO7PZ:visited:active,\n  .smso19r button._1wRO7PZ:focus,\n  .smso19r button._1wRO7PZ:hover,\n  .smso19r button._1wRO7PZ:active {\n    -webkit-text-decoration: ' + theme.hoverTextDecoration + ';\n            text-decoration: ' + theme.hoverTextDecoration + ';\n    color: ' + theme.hoverColor + ';\n  }\n\n.smso19r a._1wRO7PZ:link[aria-disabled], .smso19r a._1wRO7PZ:visited[aria-disabled], .smso19r button._1wRO7PZ[aria-disabled] {\n    cursor: not-allowed;\n    pointer-events: none;\n    opacity: 0.5;\n  }\n\n.smso19r a._3zr3MxT:link,\n.smso19r a._3zr3MxT:visited,\n.smso19r button._3zr3MxT {\n  color: ' + theme.colorInverse + ';\n}\n\n.smso19r a._3zr3MxT:link:hover,\n  .smso19r a._3zr3MxT:link:focus,\n  .smso19r a._3zr3MxT:link:active,\n  .smso19r a._3zr3MxT:visited:hover,\n  .smso19r a._3zr3MxT:visited:focus,\n  .smso19r a._3zr3MxT:visited:active,\n  .smso19r button._3zr3MxT:hover,\n  .smso19r button._3zr3MxT:focus,\n  .smso19r button._3zr3MxT:active {\n    color: ' + theme.colorInverse + ';\n  }\n\n.smso19r a._3zr3MxT:link:focus,\n  .smso19r a._3zr3MxT:link:active,\n  .smso19r a._3zr3MxT:visited:focus,\n  .smso19r a._3zr3MxT:visited:active,\n  .smso19r button._3zr3MxT:focus,\n  .smso19r button._3zr3MxT:active {\n    outline: ' + theme.outlineWidth + ' ' + theme.outlineStyle + ' ' + theme.focusInverseOutlineColor + ';\n  }\n\n.smso19r a._3zr3MxT:link._2SYApbw:focus,\n    .smso19r a._3zr3MxT:link._2SYApbw:active,\n    .smso19r a._3zr3MxT:visited._2SYApbw:focus,\n    .smso19r a._3zr3MxT:visited._2SYApbw:active,\n    .smso19r button._3zr3MxT._2SYApbw:focus,\n    .smso19r button._3zr3MxT._2SYApbw:active {\n      outline: ' + theme.outlineWidth + ' ' + theme.outlineStyle + ' ' + theme.focusInverseIconOutlineColor + ';\n    }\n\n.smso19r button._1wRO7PZ {\n  -webkit-user-select: text;\n     -moz-user-select: text;\n      -ms-user-select: text;\n          user-select: text;\n  background: none;\n  border: none;\n  cursor: pointer;\n  font-size: 1em;\n  margin: 0;\n  padding: 0;\n  display: inline-flex;\n  align-content: flex-start;\n  text-align: inherit;\n}\n\n[dir="ltr"] .smso19r button._1wRO7PZ {\n  text-align: inherit;\n}\n\n[dir="rtl"] .smso19r button._1wRO7PZ {\n  text-align: inherit;\n}\n\n._1bxdVKt {\n  display: inline-block;\n  vertical-align: middle;\n  width: 100%;\n}\n\n._2zc1zen {\n  box-sizing: border-box;\n  display: block;\n  line-height: 1.2;\n  overflow: hidden;\n  white-space: nowrap;\n  text-overflow: ellipsis;\n  width: 100%;\n}\n\n._2FdFAVm {\n  font-size: 1.125em; /* make icon slightly larger than inherited font-size */\n}\n\n._39OZjTU ._2FdFAVm {\n  -webkit-margin-end: ' + theme.iconPlusTextMargin + ';\n          margin-inline-end: ' + theme.iconPlusTextMargin + ';\n  -webkit-margin-start: 0;\n          margin-inline-start: 0;\n}\n\n[dir="ltr"] ._39OZjTU ._2FdFAVm {\n  margin-right: ' + theme.iconPlusTextMargin + ';\n  margin-left: 0;\n}\n\n[dir="rtl"] ._39OZjTU ._2FdFAVm {\n  margin-left: ' + theme.iconPlusTextMargin + ';\n  margin-right: 0;\n}\n\n._1PBSQg9 ._2FdFAVm {\n  -webkit-margin-end: 0;\n          margin-inline-end: 0;\n  -webkit-margin-start: ' + theme.iconPlusTextMargin + ';\n          margin-inline-start: ' + theme.iconPlusTextMargin + ';\n}\n\n[dir="ltr"] ._1PBSQg9 ._2FdFAVm {\n  margin-right: 0;\n  margin-left: ' + theme.iconPlusTextMargin + ';\n}\n\n[dir="rtl"] ._1PBSQg9 ._2FdFAVm {\n  margin-left: 0;\n  margin-right: ' + theme.iconPlusTextMargin + ';\n}\n\n/* stylelint-enable selector-max-class, no-descending-specificity, selector-max-type */\n';
    };

    return tmpl.call(theme, theme);
  },
  'root': 'smso19r',
  'link': '_1wRO7PZ',
  'iconOnly': '_2SYApbw',
  'inverse': '_3zr3MxT',
  'ellipsis': '_1bxdVKt',
  'text': '_2zc1zen',
  'icon': '_2FdFAVm',
  'iconPlacement--start': '_39OZjTU',
  'iconPlacement--end': '_1PBSQg9'
};

import theme from './theme';

/**
---
category: components/navigation
---
**/
var Link = (_dec = themeable(theme, styles), _dec(_class = (_temp2 = _class2 = function (_Component) {
  _inherits(Link, _Component);

  function Link() {
    var _ref;

    var _temp, _this, _ret;

    _classCallCheck(this, Link);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = Link.__proto__ || Object.getPrototypeOf(Link)).call.apply(_ref, [this].concat(args))), _this), _this.handleClick = function (e) {
      var _this$props = _this.props,
          disabled = _this$props.disabled,
          onClick = _this$props.onClick;


      if (disabled) {
        e.preventDefault();
        e.stopPropagation();
      } else if (typeof onClick === 'function') {
        onClick(e);
      }
    }, _temp), _possibleConstructorReturn(_this, _ret);
  }

  _createClass(Link, [{
    key: 'focus',
    value: function focus() {
      findDOMNode(this._link).focus(); // eslint-disable-line react/no-find-dom-node
    }
  }, {
    key: 'renderIcon',
    value: function renderIcon() {
      var Icon = this.props.icon;
      if (typeof this.props.icon === 'function') {
        return React.createElement(
          'span',
          { className: styles.icon },
          React.createElement(Icon, null)
        );
      } else if (Icon) {
        return React.createElement(
          'span',
          { className: styles.icon },
          Icon
        );
      } else {
        return null;
      }
    }
  }, {
    key: 'render',
    value: function render() {
      var _classes,
          _this2 = this;

      var _props = this.props,
          disabled = _props.disabled,
          children = _props.children,
          onClick = _props.onClick,
          variant = _props.variant,
          linkRef = _props.linkRef,
          href = _props.href,
          margin = _props.margin,
          ellipsis = _props.ellipsis,
          elementRef = _props.elementRef,
          icon = _props.icon,
          iconPlacement = _props.iconPlacement;


      var ElementType = getElementType(Link, this.props);

      var classes = (_classes = {}, _defineProperty(_classes, styles.link, true), _defineProperty(_classes, styles.inverse, variant === 'inverse'), _defineProperty(_classes, styles.ellipsis, ellipsis), _defineProperty(_classes, styles['iconPlacement--' + iconPlacement], icon && this.hasVisibleChildren), _defineProperty(_classes, styles.iconOnly, icon && !this.hasVisibleChildren), _classes);

      var props = Object.assign({
        ref: function ref(c) {
          for (var _len2 = arguments.length, args = Array(_len2 > 1 ? _len2 - 1 : 0), _key2 = 1; _key2 < _len2; _key2++) {
            args[_key2 - 1] = arguments[_key2];
          }

          _this2._link = c;
          linkRef.apply(_this2, [c].concat(args));
        }
      }, omitProps(this.props, Object.assign({}, Link.propTypes, View.propTypes)), {
        className: classnames(classes),
        href: href,
        type: href ? null : 'button',
        'aria-disabled': disabled ? 'true' : null,
        role: onClick && this.props.as ? 'button' : null,
        tabIndex: onClick && this.props.as ? '0' : null,
        onClick: this.handleClick
      });

      return React.createElement(
        View,
        {
          margin: margin,
          className: styles.root,
          elementRef: elementRef
        },
        React.createElement(
          ElementType,
          props,
          React.createElement(
            'span',
            { className: ellipsis ? styles.text : null },
            icon && iconPlacement === 'start' && this.renderIcon(),
            children,
            icon && iconPlacement === 'end' && this.renderIcon()
          )
        )
      );
    }
  }, {
    key: 'focused',
    get: function get() {
      return isActiveElement(this._link);
    }
  }, {
    key: 'hasVisibleChildren',
    get: function get() {
      return hasVisibleChildren(this.props.children);
    }
  }]);

  Link.displayName = 'Link'
  ;
  return Link;
}(Component), _class2.propTypes = {
  href: PropTypes.string,
  children: PropTypes.node.isRequired,
  variant: PropTypes.oneOf(['default', 'inverse']),
  /**
  * provides a reference to the underlying focusable (`button` or `a`) element
  */
  linkRef: PropTypes.func,
  /**
  * the element type to render as (will default to `<a>` if href is provided)
  */
  as: CustomPropTypes.elementType,
  disabled: PropTypes.bool,
  onClick: PropTypes.func,
  /**
  * truncates the text to fit within the parent element (also changes display to block).
  */
  ellipsis: PropTypes.bool,
  /**
  * Valid values are `0`, `none`, `auto`, `xxx-small`, `xx-small`, `x-small`,
  * `small`, `medium`, `large`, `x-large`, `xx-large`. Apply these values via
  * familiar CSS-like shorthand. For example: `margin="small auto large"`.
  */
  margin: ThemeablePropTypes.spacing,
  /**
  * Add an SVG icon to the Link. Do not add icons directly as
  * children.
  */
  icon: PropTypes.oneOfType([PropTypes.func, PropTypes.element]),
  /**
  * Place the icon before or after the text in the Link.
  */
  iconPlacement: PropTypes.oneOf(['start', 'end'])
}, _class2.defaultProps = {
  variant: 'default',
  as: 'button',
  linkRef: function linkRef(link) {},
  ellipsis: false,
  iconPlacement: 'start'
}, _temp2)) || _class);


export default Link;
