// @instructure/ui-forms/lib/components/CheckboxGroup/__examples__/CheckboxGroup.examples.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.disabled = exports.success = exports.hint = exports.error = exports.withToggle = exports.sizeLarge = exports.sizeMedium = exports.sizeSmall = exports.layoutInline = exports.layoutColumns = exports.layoutStacked = undefined;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _index = require('../index');

var _index2 = _interopRequireDefault(_index);

var _Checkbox = require('../../Checkbox');

var _Checkbox2 = _interopRequireDefault(_Checkbox);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var layoutStacked = exports.layoutStacked = function layoutStacked() {
  return _react2.default.createElement(
    _index2.default,
    {
      name: 'sports',
      description: 'Select your favorite sports'
    },
    _react2.default.createElement(_Checkbox2.default, { label: 'Tennis', value: 'tennis' }),
    _react2.default.createElement(_Checkbox2.default, { label: 'Lacrosse', value: 'lacrosse' }),
    _react2.default.createElement(_Checkbox2.default, { label: 'Water Polo', value: 'water-polo' }),
    _react2.default.createElement(_Checkbox2.default, { label: 'Golf', value: 'golf' })
  );
}; /*
    * The MIT License (MIT)
    *
    * Copyright (c) 2015 - present Instructure, Inc.
    *
    * Permission is hereby granted, free of charge, to any person obtaining a copy
    * of this software and associated documentation files (the "Software"), to deal
    * in the Software without restriction, including without limitation the rights
    * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
    * copies of the Software, and to permit persons to whom the Software is
    * furnished to do so, subject to the following conditions:
    *
    * The above copyright notice and this permission notice shall be included in all
    * copies or substantial portions of the Software.
    *
    * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
    * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
    * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
    * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
    * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
    * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
    * SOFTWARE.
    */

var layoutColumns = exports.layoutColumns = function layoutColumns() {
  return _react2.default.createElement(
    _index2.default,
    {
      layout: 'columns',
      name: 'sports',
      description: 'Select your favorite sports'
    },
    _react2.default.createElement(_Checkbox2.default, { label: 'Tennis', value: 'tennis' }),
    _react2.default.createElement(_Checkbox2.default, { label: 'Lacrosse', value: 'lacrosse' }),
    _react2.default.createElement(_Checkbox2.default, { label: 'Water Polo', value: 'water-polo' }),
    _react2.default.createElement(_Checkbox2.default, { label: 'Golf', value: 'golf' })
  );
};

var layoutInline = exports.layoutInline = function layoutInline() {
  return _react2.default.createElement(
    _index2.default,
    {
      layout: 'inline',
      name: 'sports',
      description: 'Select your favorite sports'
    },
    _react2.default.createElement(_Checkbox2.default, { label: 'Tennis', value: 'tennis' }),
    _react2.default.createElement(_Checkbox2.default, { label: 'Lacrosse', value: 'lacrosse' }),
    _react2.default.createElement(_Checkbox2.default, { label: 'Water Polo', value: 'water-polo' }),
    _react2.default.createElement(_Checkbox2.default, { label: 'Golf', value: 'golf' })
  );
};

var sizeSmall = exports.sizeSmall = function sizeSmall() {
  return _react2.default.createElement(
    _index2.default,
    {
      size: 'small',
      name: 'sports',
      description: 'Select your favorite sports'
    },
    _react2.default.createElement(_Checkbox2.default, { label: 'Tennis', value: 'tennis' }),
    _react2.default.createElement(_Checkbox2.default, { label: 'Lacrosse', value: 'lacrosse' }),
    _react2.default.createElement(_Checkbox2.default, { label: 'Water Polo', value: 'water-polo' }),
    _react2.default.createElement(_Checkbox2.default, { label: 'Golf', value: 'golf' })
  );
};

var sizeMedium = exports.sizeMedium = function sizeMedium() {
  return _react2.default.createElement(
    _index2.default,
    {
      size: 'medium',
      name: 'sports',
      description: 'Select your favorite sports'
    },
    _react2.default.createElement(_Checkbox2.default, { label: 'Tennis', value: 'tennis' }),
    _react2.default.createElement(_Checkbox2.default, { label: 'Lacrosse', value: 'lacrosse' }),
    _react2.default.createElement(_Checkbox2.default, { label: 'Water Polo', value: 'water-polo' }),
    _react2.default.createElement(_Checkbox2.default, { label: 'Golf', value: 'golf' })
  );
};

var sizeLarge = exports.sizeLarge = function sizeLarge() {
  return _react2.default.createElement(
    _index2.default,
    {
      size: 'large',
      name: 'sports',
      description: 'Select your favorite sports'
    },
    _react2.default.createElement(_Checkbox2.default, { label: 'Tennis', value: 'tennis' }),
    _react2.default.createElement(_Checkbox2.default, { label: 'Lacrosse', value: 'lacrosse' }),
    _react2.default.createElement(_Checkbox2.default, { label: 'Water Polo', value: 'water-polo' }),
    _react2.default.createElement(_Checkbox2.default, { label: 'Golf', value: 'golf' })
  );
};

var withToggle = exports.withToggle = function withToggle() {
  return _react2.default.createElement(
    _index2.default,
    {
      name: 'sports',
      description: 'Select your favorite sports',
      size: 'small'
    },
    _react2.default.createElement(_Checkbox2.default, { variant: 'toggle', label: 'Tennis', value: 'tennis' }),
    _react2.default.createElement(_Checkbox2.default, { variant: 'toggle', label: 'Lacrosse', value: 'lacrosse' }),
    _react2.default.createElement(_Checkbox2.default, { variant: 'toggle', label: 'Water Polo', value: 'water-polo' }),
    _react2.default.createElement(_Checkbox2.default, { variant: 'toggle', label: 'Golf', value: 'golf' })
  );
};

var error = exports.error = function error() {
  return _react2.default.createElement(
    _index2.default,
    {
      name: 'sports',
      description: 'Select your favorite sports',
      messages: [{ text: 'Please make a selection', type: 'error' }]
    },
    _react2.default.createElement(_Checkbox2.default, { label: 'Tennis', value: 'tennis' }),
    _react2.default.createElement(_Checkbox2.default, { label: 'Lacrosse', value: 'lacrosse' }),
    _react2.default.createElement(_Checkbox2.default, { label: 'Water Polo', value: 'water-polo' }),
    _react2.default.createElement(_Checkbox2.default, { label: 'Golf', value: 'golf' })
  );
};

var hint = exports.hint = function hint() {
  return _react2.default.createElement(
    _index2.default,
    {
      name: 'sports',
      description: 'Select your favorite sports',
      messages: [{ text: 'This is to help us decide on course offerings', type: 'hint' }]
    },
    _react2.default.createElement(_Checkbox2.default, { label: 'Tennis', value: 'tennis' }),
    _react2.default.createElement(_Checkbox2.default, { label: 'Lacrosse', value: 'lacrosse' }),
    _react2.default.createElement(_Checkbox2.default, { label: 'Water Polo', value: 'water-polo' }),
    _react2.default.createElement(_Checkbox2.default, { label: 'Golf', value: 'golf' })
  );
};

var success = exports.success = function success() {
  return _react2.default.createElement(
    _index2.default,
    {
      name: 'sports',
      description: 'Select your favorite sports',
      messages: [{ text: 'Woot... you did it', type: 'success' }]
    },
    _react2.default.createElement(_Checkbox2.default, { label: 'Tennis', value: 'tennis' }),
    _react2.default.createElement(_Checkbox2.default, { label: 'Lacrosse', value: 'lacrosse' }),
    _react2.default.createElement(_Checkbox2.default, { label: 'Water Polo', value: 'water-polo' }),
    _react2.default.createElement(_Checkbox2.default, { label: 'Golf', value: 'golf' })
  );
};

var disabled = exports.disabled = function disabled() {
  return _react2.default.createElement(
    _index2.default,
    {
      disabled: true,
      name: 'sports',
      description: 'Select your favorite sports'
    },
    _react2.default.createElement(_Checkbox2.default, { label: 'Tennis', value: 'tennis' }),
    _react2.default.createElement(_Checkbox2.default, { label: 'Lacrosse', value: 'lacrosse' }),
    _react2.default.createElement(_Checkbox2.default, { label: 'Water Polo', value: 'water-polo' }),
    _react2.default.createElement(_Checkbox2.default, { label: 'Golf', value: 'golf' })
  );
};
