// @instructure/ui-forms/es/components/RadioInputGroup/__examples__/RadioInputGroup.examples.js
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import React from 'react';
import RadioInput from '../../RadioInput';
import RadioInputGroup from '../index';

export var layoutStacked = function layoutStacked() {
  return React.createElement(
    RadioInputGroup,
    {
      name: 'example1',
      description: 'Select something' },
    React.createElement(RadioInput, {
      key: 'minnie',
      value: 'minnie',
      label: 'Minnie'
    }),
    React.createElement(RadioInput, {
      key: 'mickey',
      value: 'mickey',
      label: 'Mickey'
    }),
    React.createElement(RadioInput, {
      key: 'goofy',
      value: 'goofy',
      label: 'Goofy'
    })
  );
};

export var layoutColumns = function layoutColumns() {
  return React.createElement(
    RadioInputGroup,
    {
      name: 'example2',
      description: 'Select something',
      layout: 'columns' },
    React.createElement(RadioInput, {
      key: 'minnie',
      value: 'minnie',
      label: 'Minnie'
    }),
    React.createElement(RadioInput, {
      key: 'mickey',
      value: 'mickey',
      label: 'Mickey'
    }),
    React.createElement(RadioInput, {
      key: 'goofy',
      value: 'goofy',
      label: 'Goofy'
    })
  );
};

export var layoutInline = function layoutInline() {
  return React.createElement(
    RadioInputGroup,
    {
      name: 'example3',
      description: 'Select something',
      layout: 'inline' },
    React.createElement(RadioInput, {
      key: 'minnie',
      value: 'minnie',
      label: 'Minnie'
    }),
    React.createElement(RadioInput, {
      key: 'mickey',
      value: 'mickey',
      label: 'Mickey'
    }),
    React.createElement(RadioInput, {
      key: 'goofy',
      value: 'goofy',
      label: 'Goofy'
    })
  );
};

export var defaultValue = function defaultValue() {
  return React.createElement(
    RadioInputGroup,
    {
      name: 'example3',
      defaultValue: 'mickey',
      description: 'Select something',
      layout: 'inline' },
    React.createElement(RadioInput, {
      key: 'minnie',
      value: 'minnie',
      label: 'Minnie'
    }),
    React.createElement(RadioInput, {
      key: 'mickey',
      value: 'mickey',
      label: 'Mickey'
    }),
    React.createElement(RadioInput, {
      key: 'goofy',
      value: 'goofy',
      label: 'Goofy'
    })
  );
};

export var variantToggle = function variantToggle() {
  return React.createElement(
    RadioInputGroup,
    {
      name: 'example4',
      defaultValue: 'goofy',
      description: 'Select something',
      variant: 'toggle' },
    React.createElement(RadioInput, {
      key: 'minnie',
      value: 'minnie',
      label: 'Minnie'
    }),
    React.createElement(RadioInput, {
      key: 'mickey',
      value: 'mickey',
      label: 'Mickey'
    }),
    React.createElement(RadioInput, {
      key: 'goofy',
      value: 'goofy',
      label: 'Goofy'
    })
  );
};

export var sizeSmall = function sizeSmall() {
  return React.createElement(
    RadioInputGroup,
    {
      name: 'example5',
      defaultValue: 'minnie',
      description: 'Select something',
      size: 'small' },
    React.createElement(RadioInput, {
      key: 'minnie',
      value: 'minnie',
      label: 'Minnie'
    }),
    React.createElement(RadioInput, {
      key: 'mickey',
      value: 'mickey',
      label: 'Mickey'
    }),
    React.createElement(RadioInput, {
      key: 'goofy',
      value: 'goofy',
      label: 'Goofy'
    })
  );
};

export var sizeMedium = function sizeMedium() {
  return React.createElement(
    RadioInputGroup,
    {
      name: 'example6',
      defaultValue: 'minnie',
      description: 'Select something',
      size: 'medium' },
    React.createElement(RadioInput, {
      key: 'minnie',
      value: 'minnie',
      label: 'Minnie'
    }),
    React.createElement(RadioInput, {
      key: 'mickey',
      value: 'mickey',
      label: 'Mickey'
    }),
    React.createElement(RadioInput, {
      key: 'goofy',
      value: 'goofy',
      label: 'Goofy'
    })
  );
};

export var sizeLarge = function sizeLarge() {
  return React.createElement(
    RadioInputGroup,
    {
      name: 'example7',
      defaultValue: 'minnie',
      description: 'Select something',
      size: 'large' },
    React.createElement(RadioInput, {
      key: 'minnie',
      value: 'minnie',
      label: 'Minnie'
    }),
    React.createElement(RadioInput, {
      key: 'mickey',
      value: 'mickey',
      label: 'Mickey'
    }),
    React.createElement(RadioInput, {
      key: 'goofy',
      value: 'goofy',
      label: 'Goofy'
    })
  );
};

export var error = function error() {
  return React.createElement(
    RadioInputGroup,
    {
      name: 'example8',
      description: 'Select something',
      messages: [{ text: 'Please make a selection.', type: 'error' }]
    },
    React.createElement(RadioInput, {
      key: 'minnie',
      value: 'minnie',
      label: 'Minnie'
    }),
    React.createElement(RadioInput, {
      key: 'mickey',
      value: 'mickey',
      label: 'Mickey'
    }),
    React.createElement(RadioInput, {
      key: 'goofy',
      value: 'goofy',
      label: 'Goofy'
    })
  );
};

export var hint = function hint() {
  return React.createElement(
    RadioInputGroup,
    {
      name: 'example9',
      description: 'Select something',
      messages: [{ text: 'Who would you like to meet at Disneyland?', type: 'hint' }]
    },
    React.createElement(RadioInput, {
      key: 'minnie',
      value: 'minnie',
      label: 'Minnie'
    }),
    React.createElement(RadioInput, {
      key: 'mickey',
      value: 'mickey',
      label: 'Mickey'
    }),
    React.createElement(RadioInput, {
      key: 'goofy',
      value: 'goofy',
      label: 'Goofy'
    })
  );
};

export var success = function success() {
  return React.createElement(
    RadioInputGroup,
    {
      name: 'example10',
      description: 'Select something',
      messages: [{ text: 'Great choice, thanks for participating', type: 'success' }]
    },
    React.createElement(RadioInput, {
      key: 'minnie',
      value: 'minnie',
      label: 'Minnie'
    }),
    React.createElement(RadioInput, {
      key: 'mickey',
      value: 'mickey',
      label: 'Mickey'
    }),
    React.createElement(RadioInput, {
      key: 'goofy',
      value: 'goofy',
      label: 'Goofy'
    })
  );
};
