// @instructure/ui-elements/lib/components/Spinner/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _dec, _class, _class2, _temp; /*
                                   * The MIT License (MIT)
                                   *
                                   * Copyright (c) 2015 - present Instructure, Inc.
                                   *
                                   * Permission is hereby granted, free of charge, to any person obtaining a copy
                                   * of this software and associated documentation files (the "Software"), to deal
                                   * in the Software without restriction, including without limitation the rights
                                   * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
                                   * copies of the Software, and to permit persons to whom the Software is
                                   * furnished to do so, subject to the following conditions:
                                   *
                                   * The above copyright notice and this permission notice shall be included in all
                                   * copies or substantial portions of the Software.
                                   *
                                   * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
                                   * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
                                   * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
                                   * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
                                   * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
                                   * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
                                   * SOFTWARE.
                                   */


var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _View = require('@instructure/ui-layout/lib/components/View');

var _View2 = _interopRequireDefault(_View);

var _passthroughProps = require('@instructure/ui-utils/lib/react/passthroughProps');

var _uiThemeable = require('@instructure/ui-themeable');

var _uiThemeable2 = _interopRequireDefault(_uiThemeable);

var _Browser = require('@instructure/ui-utils/lib/Browser');

var _Browser2 = _interopRequireDefault(_Browser);

var _ThemeablePropTypes = require('@instructure/ui-themeable/lib/utils/ThemeablePropTypes');

var _ThemeablePropTypes2 = _interopRequireDefault(_ThemeablePropTypes);

var _generateElementId = require('@instructure/ui-utils/lib/dom/generateElementId');

var _generateElementId2 = _interopRequireDefault(_generateElementId);

var _CustomPropTypes = require('@instructure/ui-utils/lib/react/CustomPropTypes');

var _CustomPropTypes2 = _interopRequireDefault(_CustomPropTypes);

var _theme = require('./theme');

var _theme2 = _interopRequireDefault(_theme);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var styles = {
  template: function template(theme) {
    var tmpl = function tmpl() {
      return '/*  imported from styles.css  */\n\n/* stylelint-disable selector-max-class, no-descending-specificity */\n\n@keyframes j5BRzpC {\n  to {\n    transform: rotate(360deg);\n  }\n}\n\n@keyframes cMReJdD {\n  0% {\n    stroke-dashoffset: 190%;\n  }\n\n  50% {\n    stroke-dashoffset: 50%;\n    transform: rotate(90deg);\n  }\n\n  100% {\n    stroke-dashoffset: 190%;\n    transform: rotate(360deg);\n  }\n}\n\n._3smcUP4 {\n  display: inline-block;\n  vertical-align: middle;\n  position: relative;\n  box-sizing: border-box;\n  overflow: hidden;\n}\n\n._1LsthOe {\n  width: ' + theme.xSmallSize + ';\n  height: ' + theme.xSmallSize + ';\n}\n\n._1LsthOe .bQHCS3l {\n    width: ' + theme.xSmallSize + ';\n    height: ' + theme.xSmallSize + ';\n  }\n\n._1LsthOe .QydiXYn {\n    stroke-width: calc(' + theme.xSmallBorderWidth + ' * 1.75);\n  }\n\n._1LsthOe ._3G7OCgZ,\n  ._1LsthOe .dEhHjQA {\n    stroke-width: ' + theme.xSmallBorderWidth + ';\n  }\n\n._1LsthOe .dEhHjQA {\n    stroke-dasharray: 3em;\n    transform-origin: calc(' + theme.xSmallSize + ' / 2) calc(' + theme.xSmallSize + ' / 2);\n  }\n\n.TnUdSEw {\n  width: ' + theme.smallSize + ';\n  height: ' + theme.smallSize + ';\n}\n\n.TnUdSEw .bQHCS3l {\n    width: ' + theme.smallSize + ';\n    height: ' + theme.smallSize + ';\n  }\n\n.TnUdSEw .QydiXYn {\n    stroke-width: calc(' + theme.smallBorderWidth + ' * 1.75);\n  }\n\n.TnUdSEw ._3G7OCgZ,\n  .TnUdSEw .dEhHjQA {\n    stroke-width: ' + theme.smallBorderWidth + ';\n  }\n\n.TnUdSEw .dEhHjQA {\n    stroke-dasharray: 6em;\n    transform-origin: calc(' + theme.smallSize + ' / 2) calc(' + theme.smallSize + ' / 2);\n  }\n\n._3j49KJJ {\n  width: ' + theme.mediumSize + ';\n  height: ' + theme.mediumSize + ';\n}\n\n._3j49KJJ .bQHCS3l {\n    stroke-width: ' + theme.mediumBorderWidth + ';\n    width: ' + theme.mediumSize + ';\n    height: ' + theme.mediumSize + ';\n  }\n\n._3j49KJJ .QydiXYn {\n    stroke-width: calc(' + theme.mediumBorderWidth + ' * 1.75);\n  }\n\n._3j49KJJ ._3G7OCgZ,\n  ._3j49KJJ .dEhHjQA {\n    stroke-width: ' + theme.mediumBorderWidth + ';\n  }\n\n._3j49KJJ .dEhHjQA {\n    stroke-dasharray: 10.5em;\n    transform-origin: calc(' + theme.mediumSize + ' / 2) calc(' + theme.mediumSize + ' / 2);\n  }\n\n._2OCQJvA {\n  width: ' + theme.largeSize + ';\n  height: ' + theme.largeSize + ';\n}\n\n._2OCQJvA .bQHCS3l {\n    stroke-width: ' + theme.largeBorderWidth + ';\n    width: ' + theme.largeSize + ';\n    height: ' + theme.largeSize + ';\n  }\n\n._2OCQJvA .QydiXYn {\n    stroke-width: calc(' + theme.largeBorderWidth + ' * 1.75);\n  }\n\n._2OCQJvA ._3G7OCgZ,\n  ._2OCQJvA .dEhHjQA {\n    stroke-width: ' + theme.largeBorderWidth + ';\n  }\n\n._2OCQJvA .dEhHjQA {\n    stroke-dasharray: 14em;\n    transform-origin: calc(' + theme.largeSize + ' / 2) calc(' + theme.largeSize + ' / 2);\n  }\n\n.bQHCS3l {\n  display: block;\n  position: absolute;\n  top: 0;\n  left: 0; /* stylelint-disable-line property-blacklist */\n  animation-name: j5BRzpC;\n  animation-duration: 2.25s;\n  animation-iteration-count: infinite;\n  animation-timing-function: linear;\n}\n\n._3G7OCgZ,\n.QydiXYn,\n.dEhHjQA {\n  fill: none;\n}\n\n.dEhHjQA {\n  stroke-linecap: round;\n}\n\n._3smcUP4:not(._3Dzm7eU) .dEhHjQA {\n  animation-name: cMReJdD;\n  animation-duration: 1.75s;\n  animation-iteration-count: infinite;\n  animation-timing-function: ease;\n}\n\n._3smcUP4._3Dzm7eU .dEhHjQA {\n  stroke-dashoffset: 100%;\n}\n\n.bX0TiyA .QydiXYn {\n    stroke: none;\n  }\n\n.bX0TiyA ._3G7OCgZ {\n    stroke: ' + theme.trackColor + ';\n  }\n\n.bX0TiyA .dEhHjQA {\n    stroke: ' + theme.color + ';\n  }\n\n._9AxYjlM .QydiXYn {\n    stroke: rgba(0, 0, 0, 0.3);\n  }\n\n._9AxYjlM ._3G7OCgZ {\n    stroke: ' + theme.inverseTrackColor + ';\n  }\n\n._9AxYjlM .dEhHjQA {\n    stroke: ' + theme.inverseColor + ';\n  }\n\n/* stylelint-enable selector-max-class, no-descending-specificity */\n';
    };

    return tmpl.call(theme, theme);
  },
  'root': '_3smcUP4',
  'x-small': '_1LsthOe',
  'circle': 'bQHCS3l',
  'circleShadow': 'QydiXYn',
  'circleTrack': '_3G7OCgZ',
  'circleSpin': 'dEhHjQA',
  'small': 'TnUdSEw',
  'medium': '_3j49KJJ',
  'large': '_2OCQJvA',
  'rotate': 'j5BRzpC',
  'ie11': '_3Dzm7eU',
  'morph': 'cMReJdD',
  'default': 'bX0TiyA',
  'inverse': '_9AxYjlM'
};


/**
---
category: components
---
**/
var Spinner = (_dec = (0, _uiThemeable2.default)(_theme2.default, styles), _dec(_class = (_temp = _class2 = function (_Component) {
  _inherits(Spinner, _Component);

  function Spinner(props) {
    _classCallCheck(this, Spinner);

    var _this = _possibleConstructorReturn(this, (Spinner.__proto__ || Object.getPrototypeOf(Spinner)).call(this));

    _this.titleId = (0, _generateElementId2.default)('Spinner');
    return _this;
  }

  _createClass(Spinner, [{
    key: 'radius',
    value: function radius() {
      switch (this.props.size) {
        case 'x-small':
          return '0.5em';
        case 'small':
          return '1em';
        case 'large':
          return '2.25em';
        default:
          return '1.75em';
      }
    }
  }, {
    key: 'render',
    value: function render() {
      var _classes;

      var ie11 = _Browser2.default.msie && _Browser2.default.version > 10;

      var classes = (_classes = {}, _defineProperty(_classes, styles.root, true), _defineProperty(_classes, styles[this.props.size], true), _defineProperty(_classes, styles[this.props.variant], true), _defineProperty(_classes, styles.ie11, ie11), _classes);
      return _react2.default.createElement(
        _View2.default,
        Object.assign({}, (0, _passthroughProps.omitProps)(this.props, Object.assign({}, Spinner.propTypes, _View2.default.propTypes)), {
          as: this.props.as,
          elementRef: this.props.elementRef,
          className: (0, _classnames2.default)(classes),
          margin: this.props.margin
        }),
        _react2.default.createElement(
          'svg',
          {
            className: styles.circle,
            role: 'img',
            'aria-labelledby': this.titleId,
            focusable: 'false'
          },
          _react2.default.createElement(
            'title',
            { id: this.titleId },
            this.props.title
          ),
          _react2.default.createElement(
            'g',
            { role: 'presentation' },
            _react2.default.createElement('circle', { className: styles.circleShadow, cx: '50%', cy: '50%', r: this.radius() }),
            _react2.default.createElement('circle', { className: styles.circleTrack, cx: '50%', cy: '50%', r: this.radius() }),
            _react2.default.createElement('circle', { className: styles.circleSpin, cx: '50%', cy: '50%', r: this.radius() })
          )
        )
      );
    }
  }]);

  Spinner.displayName = 'Spinner'
  ;
  return Spinner;
}(_react.Component), _class2.propTypes = {
  /**
  * Give the spinner a title to be read by screenreaders
  */
  title: _propTypes2.default.string.isRequired,
  /**
  * Different-sized spinners
  */
  size: _propTypes2.default.oneOf(['x-small', 'small', 'medium', 'large']),
  /**
  * Different color schemes for use with light or dark backgrounds
  */
  variant: _propTypes2.default.oneOf(['default', 'inverse']),
  /**
  * Valid values are `0`, `none`, `auto`, `xxx-small`, `xx-small`, `x-small`,
  * `small`, `medium`, `large`, `x-large`, `xx-large`. Apply these values via
  * familiar CSS-like shorthand. For example: `margin="small auto large"`.
  */
  margin: _ThemeablePropTypes2.default.spacing,
  elementRef: _propTypes2.default.func,
  as: _CustomPropTypes2.default.elementType
}, _class2.defaultProps = {
  as: 'div',
  size: 'medium',
  variant: 'default'
}, _temp)) || _class);
exports.default = Spinner;
