// @instructure/ui-forms/es/components/CheckboxGroup/index.js
var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _class, _temp;

function _toConsumableArray(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } else { return Array.from(arr); } }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import React, { Children, Component } from 'react';
import PropTypes from 'prop-types';

import generateElementId from '@instructure/ui-utils/lib/dom/generateElementId';
import CustomPropTypes from '@instructure/ui-utils/lib/react/CustomPropTypes';
import safeCloneElement from '@instructure/ui-utils/lib/react/safeCloneElement';
import { pickProps, omitProps } from '@instructure/ui-utils/lib/react/passthroughProps';
import matchComponentTypes from '@instructure/ui-utils/lib/react/matchComponentTypes';

import FormPropTypes from '../../utils/FormPropTypes';
import Checkbox from '../Checkbox';
import FormFieldGroup from '../FormFieldGroup';

/**
---
category: components/forms
---
**/
var CheckboxGroup = (_temp = _class = function (_Component) {
  _inherits(CheckboxGroup, _Component);

  function CheckboxGroup(props) {
    _classCallCheck(this, CheckboxGroup);

    var _this = _possibleConstructorReturn(this, (CheckboxGroup.__proto__ || Object.getPrototypeOf(CheckboxGroup)).call(this));

    _this.handleChange = function (e) {
      var newValue = _this.value || [];

      if (_this.props.disabled || _this.props.readOnly) {
        e.preventDefault();
        return;
      }

      if (e.target.checked) {
        newValue.push(e.target.value);
      } else {
        newValue.splice(newValue.indexOf(e.target.value), 1);
      }

      if (typeof _this.props.value === 'undefined') {
        _this.setState({ value: newValue });
      }

      if (typeof _this.props.onChange === 'function') {
        _this.props.onChange(newValue);
      }
    };

    if (typeof props.value === 'undefined') {
      _this.state = {
        value: props.defaultValue
      };
    }

    _this._messagesId = generateElementId('CheckboxGroup-messages');
    return _this;
  }

  _createClass(CheckboxGroup, [{
    key: 'renderChildren',
    value: function renderChildren() {
      var _this2 = this;

      var _props = this.props,
          children = _props.children,
          name = _props.name,
          size = _props.size,
          disabled = _props.disabled,
          readOnly = _props.readOnly;


      return Children.map(children, function (child, index) {
        if (matchComponentTypes(child, [Checkbox])) {
          return safeCloneElement(child, {
            key: '' + child.props.name,
            name: name,
            disabled: disabled || child.props.disabled,
            readOnly: readOnly || child.props.readOnly,
            size: size,
            checked: _this2.value.indexOf(child.props.value) > -1,
            onChange: _this2.handleChange,
            width: child.props.width || 'auto',
            'aria-describedby': _this2.hasMessages && _this2._messagesId
          });
        } else {
          return child;
        }
      });
    }
  }, {
    key: 'render',
    value: function render() {
      return React.createElement(
        FormFieldGroup,
        Object.assign({}, omitProps(this.props, CheckboxGroup.propTypes), pickProps(this.props, FormFieldGroup.propTypes), {
          rowSpacing: 'small',
          vAlign: 'top',
          messagesId: this._messagesId
        }),
        this.renderChildren()
      );
    }
  }, {
    key: 'hasMessages',
    get: function get() {
      return this.props.messages && this.props.messages.length > 0;
    }
  }, {
    key: 'value',
    get: function get() {
      if (typeof this.props.value === 'undefined' && typeof this.state.value === 'undefined') {
        return [];
      } else {
        return typeof this.props.value === 'undefined' ? [].concat(_toConsumableArray(this.state.value)) : [].concat(_toConsumableArray(this.props.value));
      }
    }
  }]);

  CheckboxGroup.displayName = 'CheckboxGroup'
  ;
  return CheckboxGroup;
}(Component), _class.propTypes = {
  name: PropTypes.string.isRequired,
  description: PropTypes.node.isRequired,
  /**
  * value to set on initial render
  */
  defaultValue: PropTypes.array,
  /**
  * the selected values (must be accompanied by an `onChange` prop)
  */
  value: CustomPropTypes.controllable(PropTypes.array),
  /**
  * when used with the `value` prop, the component will not control its own state
  */
  onChange: PropTypes.func,
  disabled: PropTypes.bool,
  readOnly: PropTypes.bool,
  /**
  * object with shape: `{
  text: PropTypes.string,
  type: PropTypes.oneOf(['error', 'hint', 'success', 'screenreader-only'])
    }`
  */
  messages: PropTypes.arrayOf(FormPropTypes.message),
  /**
  * children of type `Checkbox`
  */
  children: CustomPropTypes.Children.oneOf([Checkbox]),
  size: PropTypes.oneOf(['small', 'medium', 'large']),
  layout: PropTypes.oneOf(['stacked', 'columns', 'inline'])
}, _class.defaultProps = {
  disabled: false,
  readOnly: false,
  size: 'medium',
  layout: 'stacked'
}, _temp);
export { CheckboxGroup as default };
