// @instructure/ui-layout/lib/components/Position/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PositionContent = exports.PositionTarget = undefined;

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _class, _temp, _class2, _temp2, _dec, _class3, _class4, _temp3; /*
                                                                     * The MIT License (MIT)
                                                                     *
                                                                     * Copyright (c) 2015 - present Instructure, Inc.
                                                                     *
                                                                     * Permission is hereby granted, free of charge, to any person obtaining a copy
                                                                     * of this software and associated documentation files (the "Software"), to deal
                                                                     * in the Software without restriction, including without limitation the rights
                                                                     * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
                                                                     * copies of the Software, and to permit persons to whom the Software is
                                                                     * furnished to do so, subject to the following conditions:
                                                                     *
                                                                     * The above copyright notice and this permission notice shall be included in all
                                                                     * copies or substantial portions of the Software.
                                                                     *
                                                                     * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
                                                                     * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
                                                                     * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
                                                                     * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
                                                                     * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
                                                                     * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
                                                                     * SOFTWARE.
                                                                     */

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames2 = require('classnames');

var _classnames3 = _interopRequireDefault(_classnames2);

var _uiThemeable = require('@instructure/ui-themeable');

var _uiThemeable2 = _interopRequireDefault(_uiThemeable);

var _ComponentIdentifier3 = require('@instructure/ui-utils/lib/react/ComponentIdentifier');

var _ComponentIdentifier4 = _interopRequireDefault(_ComponentIdentifier3);

var _safeCloneElement = require('@instructure/ui-utils/lib/react/safeCloneElement');

var _safeCloneElement2 = _interopRequireDefault(_safeCloneElement);

var _addPositionChangeListener = require('@instructure/ui-utils/lib/dom/addPositionChangeListener');

var _addPositionChangeListener2 = _interopRequireDefault(_addPositionChangeListener);

var _shallowEqual = require('@instructure/ui-utils/lib/shallowEqual');

var _shallowEqual2 = _interopRequireDefault(_shallowEqual);

var _CustomPropTypes = require('@instructure/ui-utils/lib/react/CustomPropTypes');

var _CustomPropTypes2 = _interopRequireDefault(_CustomPropTypes);

var _debounce = require('@instructure/ui-utils/lib/debounce');

var _debounce2 = _interopRequireDefault(_debounce);

var _deepEqual = require('@instructure/ui-utils/lib/deepEqual');

var _deepEqual2 = _interopRequireDefault(_deepEqual);

var _Portal = require('@instructure/ui-portal/lib/components/Portal');

var _Portal2 = _interopRequireDefault(_Portal);

var _calculateElementPosition = require('../../utils/calculateElementPosition');

var _calculateElementPosition2 = _interopRequireDefault(_calculateElementPosition);

var _LayoutPropTypes = require('../../utils/LayoutPropTypes');

var _LayoutPropTypes2 = _interopRequireDefault(_LayoutPropTypes);

var _theme = require('./theme');

var _theme2 = _interopRequireDefault(_theme);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var styles = {
  template: function template(theme) {
    var tmpl = function tmpl() {
      return '/*  imported from styles.css  */\n\n._3lyFG_K {\n  z-index: ' + theme.zIndex + ';\n  box-sizing: border-box;\n}\n';
    };

    return tmpl.call(theme, theme);
  },
  'root': '_3lyFG_K'
};
var PositionTarget = (_temp = _class = function (_ComponentIdentifier) {
  _inherits(PositionTarget, _ComponentIdentifier);

  function PositionTarget() {
    _classCallCheck(this, PositionTarget);

    return _possibleConstructorReturn(this, (PositionTarget.__proto__ || Object.getPrototypeOf(PositionTarget)).apply(this, arguments));
  }

  return PositionTarget;
}(_ComponentIdentifier4.default), _class.displayName = 'PositionTarget', _temp);
var PositionContent = (_temp2 = _class2 = function (_ComponentIdentifier2) {
  _inherits(PositionContent, _ComponentIdentifier2);

  function PositionContent() {
    _classCallCheck(this, PositionContent);

    return _possibleConstructorReturn(this, (PositionContent.__proto__ || Object.getPrototypeOf(PositionContent)).apply(this, arguments));
  }

  return PositionContent;
}(_ComponentIdentifier4.default), _class2.displayName = 'PositionContent', _class2.propTypes = {
  children: _propTypes2.default.node,
  placement: _LayoutPropTypes2.default.placement
}, _temp2);

/**
---
category: components/utilities
---
**/

var Position = (_dec = (0, _uiThemeable2.default)(_theme2.default, styles), _dec(_class3 = (_temp3 = _class4 = function (_Component) {
  _inherits(Position, _Component);

  function Position(props) {
    _classCallCheck(this, Position);

    var _this3 = _possibleConstructorReturn(this, (Position.__proto__ || Object.getPrototypeOf(Position)).call(this, props));

    _this3._timeouts = [];

    _this3.handlePortalOpen = function () {
      _this3.position();

      if (_this3.props.trackPosition) {
        _this3.startTracking();
      }

      _this3._timeouts.push(setTimeout(function () {
        if (_this3.state.positioned && typeof _this3.props.onPositioned === 'function') {
          _this3.props.onPositioned({
            top: _this3.state.style.top,
            left: _this3.state.style.left,
            placement: _this3.state.placement
          });
        }
      }, 0));
    };

    _this3.position = function () {
      _this3.setState(Object.assign({
        positioned: true
      }, _this3.calculatePosition(_this3.props)));
    };

    _this3.state = Object.assign({
      positioned: false
    }, _this3.calculatePosition(props));

    _this3.position = (0, _debounce2.default)(_this3.position, 0, { leading: false, trailing: true });
    return _this3;
  }

  _createClass(Position, [{
    key: 'shouldComponentUpdate',
    value: function shouldComponentUpdate(nextProps, nextState, nextContext) {
      return !(0, _deepEqual2.default)(this.state, nextState) || !(0, _shallowEqual2.default)(this.props, nextProps) || !(0, _shallowEqual2.default)(this.context, nextContext);
    }
  }, {
    key: 'componentDidUpdate',
    value: function componentDidUpdate(prevProps, prevState) {
      this.position();

      if (this.props.trackPosition !== prevProps.trackPosition) {
        this.props.trackPosition ? this.startTracking() : this.stopTracking();
      }

      var _state = this.state,
          style = _state.style,
          placement = _state.placement;


      if (style && prevState.style && (placement !== prevState.placement || style.top !== prevState.style.top || style.left !== prevState.style.left)) {
        this.props.onPositionChanged({
          top: style.top,
          left: style.left,
          placement: placement
        });
      }
    }
  }, {
    key: 'componentWillUnmount',
    value: function componentWillUnmount() {
      this.position.cancel();
      this.stopTracking();
      this._timeouts.forEach(function (timeout) {
        return clearTimeout(timeout);
      });
    }
  }, {
    key: 'calculatePosition',
    value: function calculatePosition(props) {
      return (0, _calculateElementPosition2.default)(this._content, this._target, {
        placement: props.placement,
        offsetX: props.offsetX,
        offsetY: props.offsetY,
        constrain: props.constrain,
        container: props.mountNode,
        over: props.over
      });
    }
  }, {
    key: 'startTracking',
    value: function startTracking() {
      this._listener = this._listener || (0, _addPositionChangeListener2.default)(this._target, this.position);
    }
  }, {
    key: 'stopTracking',
    value: function stopTracking() {
      if (this._listener) {
        this._listener.remove();
        this._listener = null;
      }
    }
  }, {
    key: 'renderContent',
    value: function renderContent() {
      var _this4 = this;

      var content = (0, _ComponentIdentifier3.pick)(Position.Content, this.props.children);

      if (content && _react2.default.Children.count(content.props.children) > 0) {
        var _classnames;

        content = (0, _safeCloneElement2.default)(content, {
          ref: function ref(el) {
            _this4._content = el;
          },
          style: Object.assign({}, content.props.style, this.state.style),
          className: (0, _classnames3.default)((_classnames = {}, _defineProperty(_classnames, styles.root, true), _defineProperty(_classnames, content.props.className, content.props.className), _classnames))
        });

        content = _react2.default.createElement(
          _Portal2.default,
          { open: true, onOpen: this.handlePortalOpen, mountNode: this.props.mountNode, insertAt: this.props.insertAt },
          content
        );
      }

      return content;
    }
  }, {
    key: 'renderTarget',
    value: function renderTarget() {
      var _this5 = this;

      var target = (0, _ComponentIdentifier3.pick)(Position.Target, this.props.children);
      if (target) {
        target = (0, _safeCloneElement2.default)(target, {
          ref: function ref(el) {
            _this5._target = el;
          }
        });
      } else if (this.props.target) {
        this._target = this.props.target;
      }

      return target;
    }
  }, {
    key: 'render',
    value: function render() {
      return _react2.default.createElement(
        'span',
        null,
        this.renderTarget(),
        this.renderContent()
      );
    }
  }]);

  Position.displayName = 'Position'
  ;
  return Position;
}(_react.Component), _class4.Target = PositionTarget, _class4.Content = PositionContent, _class4.propTypes = {
  /**
   * The children to be rendered within the `<Position />`
   */
  children: _propTypes2.default.node,

  /**
   * The target to be used when not using `<PositionTarget />`
   */
  target: _propTypes2.default.oneOfType([_CustomPropTypes2.default.element, _propTypes2.default.func]),

  /**
   * Whether or not you want the content to position over the `<PositionTarget />`
   */
  over: _propTypes2.default.bool,

  /**
   * The placement of the content in relation to the trigger
   */
  placement: _LayoutPropTypes2.default.placement,

  /**
   * The horizontal offset for the positioned content
   */
  offsetX: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.number]),

  /**
   * The vertical offset for the positioned content
   */
  offsetY: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.number]),

  /**
   * Callback fired when the position changes
   */
  onPositionChanged: _propTypes2.default.func,

  /**
   * Callback fired when `<Position />` content has been mounted and is initially positioned
   */
  onPositioned: _propTypes2.default.func,

  /**
   * Whether or not position of the target should be tracked or just set statically on render
   */
  trackPosition: _propTypes2.default.bool,

  /**
   * An element or a function returning an element to use as the mount node
   * for the `<Position />` (defaults to `document.body`)
   */
  mountNode: _propTypes2.default.oneOfType([_CustomPropTypes2.default.element, _propTypes2.default.func]),

  /**
   * Insert the element at the 'top' of the mountNode or at the 'bottom'
   */
  insertAt: _propTypes2.default.oneOf(['bottom', 'top']),

  /**
   * The parent in which to constrain the placement.
   * One of: 'window', 'scroll-parent', 'parent', 'none', an element,
   * or a function returning an element
   */
  constrain: _propTypes2.default.oneOfType([_CustomPropTypes2.default.element, _propTypes2.default.func, _propTypes2.default.oneOf(['window', 'scroll-parent', 'parent', 'none'])])
}, _class4.defaultProps = {
  trackPosition: true,
  placement: 'bottom center',
  offsetX: 0,
  offsetY: 0,
  mountNode: null,
  target: null,
  insertAt: 'bottom',
  over: false,
  onPositioned: function onPositioned(position) {},
  onPositionChanged: function onPositionChanged(position) {},
  constrain: 'window'
}, _temp3)) || _class3);
exports.default = Position;
exports.PositionTarget = PositionTarget;
exports.PositionContent = PositionContent;
