// @instructure/ui-elements/lib/components/Tag/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _dec, _class, _class2, _temp2; /*
                                    * The MIT License (MIT)
                                    *
                                    * Copyright (c) 2015 - present Instructure, Inc.
                                    *
                                    * Permission is hereby granted, free of charge, to any person obtaining a copy
                                    * of this software and associated documentation files (the "Software"), to deal
                                    * in the Software without restriction, including without limitation the rights
                                    * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
                                    * copies of the Software, and to permit persons to whom the Software is
                                    * furnished to do so, subject to the following conditions:
                                    *
                                    * The above copyright notice and this permission notice shall be included in all
                                    * copies or substantial portions of the Software.
                                    *
                                    * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
                                    * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
                                    * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
                                    * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
                                    * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
                                    * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
                                    * SOFTWARE.
                                    */


var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _IconX = require('@instructure/ui-icons/lib/Solid/IconX');

var _IconX2 = _interopRequireDefault(_IconX);

var _View = require('@instructure/ui-layout/lib/components/View');

var _View2 = _interopRequireDefault(_View);

var _uiThemeable = require('@instructure/ui-themeable');

var _uiThemeable2 = _interopRequireDefault(_uiThemeable);

var _ThemeablePropTypes = require('@instructure/ui-themeable/lib/utils/ThemeablePropTypes');

var _ThemeablePropTypes2 = _interopRequireDefault(_ThemeablePropTypes);

var _passthroughProps = require('@instructure/ui-utils/lib/react/passthroughProps');

var _isActiveElement = require('@instructure/ui-utils/lib/dom/isActiveElement');

var _isActiveElement2 = _interopRequireDefault(_isActiveElement);

var _theme = require('./theme');

var _theme2 = _interopRequireDefault(_theme);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var styles = {
  template: function template(theme) {
    var tmpl = function tmpl() {
      return '/*  imported from styles.css  */\n\n/* stylelint-disable selector-max-class, no-descending-specificity, max-nesting-depth */\n\n._3DA9iZU {\n  box-sizing: border-box;\n  font-family: ' + theme.fontFamily + ';\n  display: inline-flex;\n  align-items: center;\n  text-align: center;\n  vertical-align: middle;\n  -webkit-user-select: none;\n     -moz-user-select: none;\n      -ms-user-select: none;\n          user-select: none;\n}\n\n[dir="ltr"] ._3DA9iZU {\n  text-align: center;\n}\n\n[dir="rtl"] ._3DA9iZU {\n  text-align: center;\n}\n\n._1sYT5vc {\n  height: ' + theme.heightSmall + ';\n  padding: ' + theme.paddingSmall + ';\n  font-size: ' + theme.fontSizeSmall + ';\n}\n\n._1sYT5vc ._1K7MTMD {\n    font-size: 0.75rem;\n  }\n\n._1Ff3jIc {\n  height: ' + theme.heightMedium + ';\n  padding: ' + theme.padding + ';\n  font-size: calc(' + theme.fontSizeMedium + ' - 0.0625rem);\n}\n\n._1Ff3jIc ._1K7MTMD {\n    font-size: 0.75rem;\n  }\n\n._2A62EGm {\n  height: ' + theme.heightLarge + ';\n  padding: ' + theme.padding + ';\n  font-size: calc(' + theme.fontSizeLarge + ' - 0.0625rem);\n}\n\n._2A62EGm ._1K7MTMD {\n    font-size: 0.875rem;\n  }\n\n._3XDPeeO {\n  touch-action: manipulation;\n  outline: none;\n  position: relative;\n  overflow: visible;\n  transition: background-color ' + theme.transitionTiming + ';\n}\n\n._3XDPeeO::before {\n    content: "";\n    box-sizing: border-box;\n    width: calc(100% + 0.5rem);\n    height: calc(100% + 0.5rem);\n    border: ' + theme.borderWidth + ' ' + theme.borderStyle + ' ' + theme.focusOutlineColor + ';\n    border-radius: ' + theme.borderRadius + ';\n    position: absolute;\n    top: -0.25rem;\n    left: -0.25rem; /* stylelint-disable-line property-blacklist */\n    opacity: 0;\n    transform: scale(0.01);\n    transition: all ' + theme.transitionTiming + ';\n    pointer-events: none;\n  }\n\n._3XDPeeO:not(.E5Sb0v7) {\n    cursor: pointer;\n  }\n\n._3XDPeeO:focus::before {\n      opacity: 1;\n      transform: scale(1);\n    }\n\n._3XDPeeO._1ZtXg8c {\n    cursor: not-allowed;\n    pointer-events: none;\n    opacity: 0.5;\n  }\n\n._1K7MTMD {\n  -webkit-margin-start: ' + theme.iconMargin + ';\n          margin-inline-start: ' + theme.iconMargin + ';\n  -webkit-margin-end: 0;\n          margin-inline-end: 0;\n  transform: translateY(0.0625rem);\n  transition: all ' + theme.transitionTiming + ';\n  cursor: pointer;\n}\n\n[dir="ltr"] ._1K7MTMD {\n  margin-left: ' + theme.iconMargin + ';\n  margin-right: 0;\n}\n\n[dir="rtl"] ._1K7MTMD {\n  margin-right: ' + theme.iconMargin + ';\n  margin-left: 0;\n}\n\n.w16BDTX {\n  white-space: nowrap;\n  overflow: hidden;\n  text-overflow: ellipsis;\n  max-width: ' + theme.maxWidth + ';\n}\n\n._2wRyEZO {\n  background-color: ' + theme.defaultBackground + ';\n  border: ' + theme.defaultBorderWidth + ' ' + theme.defaultBorderStyle + ' ' + theme.defaultBorderColor + ';\n  border-radius: ' + theme.defaultBorderRadius + ';\n  color: ' + theme.defaultColor + ';\n}\n\n._2wRyEZO._3XDPeeO::before {\n      border: ' + theme.defaultBorderWidth + ' ' + theme.defaultBorderStyle + ' ' + theme.focusOutlineColor + ';\n      border-radius: ' + theme.defaultBorderRadius + ';\n    }\n\n._2wRyEZO._3XDPeeO:hover {\n      background-color: ' + theme.defaultBackgroundHover + ';\n    }\n\n._2wRyEZO.E5Sb0v7 ._1K7MTMD {\n      color: ' + theme.defaultIconColor + ';\n    }\n\n._2wRyEZO.E5Sb0v7:hover ._1K7MTMD {\n        color: ' + theme.defaultIconHoverColor + ';\n      }\n\n._1xsywnr {\n  background-color: ' + theme.inlineBackground + ';\n  border: ' + theme.inlineBorderWidth + ' ' + theme.inlineBorderStyle + ' ' + theme.inlineBorderColor + ';\n  border-radius: ' + theme.inlineBorderRadius + ';\n  color: ' + theme.inlineColor + ';\n  cursor: text;\n  margin: 0 0.1875rem 0.1875rem 0.1875rem;\n}\n\n._1xsywnr._3XDPeeO::before {\n      border: ' + theme.inlineBorderWidth + ' ' + theme.inlineBorderStyle + ' ' + theme.focusOutlineColor + ';\n      border-radius: ' + theme.inlineBorderRadius + ';\n    }\n\n._1xsywnr._3XDPeeO:hover {\n      background-color: ' + theme.inlineBackgroundHover + ';\n    }\n\n._1xsywnr.E5Sb0v7 ._1K7MTMD {\n      background-color: ' + theme.inlineIconColor + ';\n      border-radius: 50%;\n      color: ' + theme.inlineBackground + ';\n      font-size: 0.75rem;\n      padding: 0.25rem;\n      position: absolute;\n      offset-inline-end: 0;\n      offset-inline-start: auto;\n      top: 0;\n      transform: translate(40%, -40%);\n    }\n\n[dir="ltr"] ._1xsywnr.E5Sb0v7 ._1K7MTMD {\n      right: 0;\n      left: auto;\n    }\n\n[dir="rtl"] ._1xsywnr.E5Sb0v7 ._1K7MTMD {\n      left: 0;\n      right: auto;\n    }\n\n._1xsywnr.E5Sb0v7:hover ._1K7MTMD {\n        background-color: ' + theme.inlineIconHoverColor + ';\n      }\n\n[dir="rtl"] ._1xsywnr.E5Sb0v7 ._1K7MTMD {\n  transform: translate(-40%, -40%);\n}\n/* stylelint-enable selector-max-class, no-descending-specificity, max-nesting-depth */\n';
    };

    return tmpl.call(theme, theme);
  },
  'root': '_3DA9iZU',
  'small': '_1sYT5vc',
  'icon': '_1K7MTMD',
  'medium': '_1Ff3jIc',
  'large': '_2A62EGm',
  'button': '_3XDPeeO',
  'dismissible': 'E5Sb0v7',
  'disabled': '_1ZtXg8c',
  'text': 'w16BDTX',
  'default': '_2wRyEZO',
  'inline': '_1xsywnr'
};


/**
---
category: components
---
**/

var Tag = (_dec = (0, _uiThemeable2.default)(_theme2.default, styles), _dec(_class = (_temp2 = _class2 = function (_Component) {
  _inherits(Tag, _Component);

  function Tag() {
    var _ref;

    var _temp, _this, _ret;

    _classCallCheck(this, Tag);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = Tag.__proto__ || Object.getPrototypeOf(Tag)).call.apply(_ref, [this].concat(args))), _this), _this.focus = function () {
      _this._container && _this._container.focus();
    }, _this.handleClick = function (e) {
      var _this$props = _this.props,
          disabled = _this$props.disabled,
          readOnly = _this$props.readOnly,
          onClick = _this$props.onClick;


      if (disabled || readOnly) {
        e.preventDefault();
        e.stopPropagation();
      } else if (typeof onClick === 'function') {
        onClick(e);
      }
    }, _this.handleRef = function (node) {
      _this._container = node;
    }, _temp), _possibleConstructorReturn(_this, _ret);
  }

  _createClass(Tag, [{
    key: 'render',
    value: function render() {
      var _classes;

      var _props = this.props,
          className = _props.className,
          dismissible = _props.dismissible,
          disabled = _props.disabled,
          readOnly = _props.readOnly,
          size = _props.size,
          text = _props.text,
          title = _props.title,
          onClick = _props.onClick,
          margin = _props.margin,
          variant = _props.variant;


      var props = (0, _passthroughProps.omitProps)(this.props, Object.assign({}, Tag.propTypes, _View2.default.propTypes));

      var classes = (_classes = {}, _defineProperty(_classes, styles.root, true), _defineProperty(_classes, styles[variant], true), _defineProperty(_classes, styles[size], size), _defineProperty(_classes, styles.dismissible, dismissible), _defineProperty(_classes, styles.button, onClick), _defineProperty(_classes, styles.disabled, disabled), _classes);

      return _react2.default.createElement(
        _View2.default,
        Object.assign({}, props, {
          ref: this.handleRef,
          elementRef: this.props.elementRef,
          className: (0, _classnames2.default)(className, classes),
          as: onClick ? 'button' : 'span',
          margin: margin,
          type: onClick ? 'button' : null,
          onClick: onClick ? this.handleClick : null,
          disabled: disabled || readOnly,
          'aria-disabled': onClick && (disabled || readOnly) ? 'true' : null,
          display: null,
          title: title || (typeof text === 'string' ? text : null)
        }),
        _react2.default.createElement(
          'span',
          { className: styles.text },
          text
        ),
        onClick && dismissible ? _react2.default.createElement(_IconX2.default, { className: styles.icon }) : null
      );
    }
  }, {
    key: 'focused',
    get: function get() {
      return (0, _isActiveElement2.default)(this._container);
    }
  }]);

  Tag.displayName = 'Tag'
  ;
  return Tag;
}(_react.Component), _class2.propTypes = {
  className: _propTypes2.default.string,
  text: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.node]).isRequired,
  title: _propTypes2.default.string,
  /**
   * Whether or not to disable the tag
   */
  disabled: _propTypes2.default.bool,
  /**
   * Works just like disabled but keeps the same styles as if it were active
   */
  readOnly: _propTypes2.default.bool,
  dismissible: _propTypes2.default.bool,
  /**
  * Valid values are `0`, `none`, `auto`, `xxxx-small`, `xx-small`, `x-small`,
  * `small`, `medium`, `large`, `x-large`, `xx-large`. Apply these values via
  * familiar CSS-like shorthand. For example: `margin="small auto large"`.
  */
  margin: _ThemeablePropTypes2.default.spacing,
  /**
  * If you add an onClick prop, Tag renders as a clickable button
  */
  onClick: _propTypes2.default.func,
  /**
  * Provides a reference to the underlying html root element
  */
  elementRef: _propTypes2.default.func,
  size: _propTypes2.default.oneOf(['small', 'medium', 'large']),
  variant: _propTypes2.default.oneOf(['default', 'inline'])
}, _class2.defaultProps = {
  size: 'medium',
  dismissible: false,
  variant: 'default'
}, _temp2)) || _class);
exports.default = Tag;
