// @instructure/ui-elements/es/components/index.js
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
import _Avatar from './Avatar';
export { _Avatar as Avatar };
import _Badge from './Badge';
export { _Badge as Badge };
import _Heading from './Heading';
export { _Heading as Heading };
import _Img from './Img';
export { _Img as Img };
import _Link from './Link';
export { _Link as Link };
import _List from './List';
export { _List as List };

export { ListItem } from './List';
import _MetricsList from './MetricsList';
export { _MetricsList as MetricsList };
export { MetricsListItem } from './MetricsList';
import _Pill from './Pill';
export { _Pill as Pill };
import _Progress from './Progress';
export { _Progress as Progress };

export { ProgressBar, ProgressCircle } from './Progress';
import _Rating from './Rating';
export { _Rating as Rating };
import _Spinner from './Spinner';
export { _Spinner as Spinner };
import _Table from './Table';
export { _Table as Table };
import _Tag from './Tag';
export { _Tag as Tag };
import _Text from './Text';
export { _Text as Text };
import _TruncateText from './TruncateText';
export { _TruncateText as TruncateText };
