// @instructure/ui-layout/es/components/ContextView/index.js
var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _dec, _class, _class2, _temp;

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import themeable from '@instructure/ui-themeable';
import ThemeablePropTypes from '@instructure/ui-themeable/lib/utils/ThemeablePropTypes';
import CustomPropTypes from '@instructure/ui-utils/lib/react/CustomPropTypes';
import { omitProps } from '@instructure/ui-utils/lib/react/passthroughProps';

import LayoutPropTypes from '../../utils/LayoutPropTypes';
import mirrorPlacement from '../../utils/mirrorPlacement';
import View from '../View';

var styles = {
  template: function template(theme) {
    var tmpl = function tmpl() {
      return '/*  imported from styles.css  */\n\n.riLv0Xm {\n  box-sizing: border-box;\n  min-height: calc(' + theme.arrowSize + ' * 2);\n}\n\n._1GXlG2S {\n  position: relative;\n}\n\n._2h8cLA8,\n._2XCKaix,\n.y0RX1FA,\n._2_l8bvH,\n.Jmp0xyj {\n  /* arrow on the left side pointing left */\n  -webkit-padding-start: ' + theme.arrowSize + ';\n          padding-inline-start: ' + theme.arrowSize + ';\n  -webkit-padding-end: 0;\n          padding-inline-end: 0;\n}\n\n[dir="ltr"] ._2h8cLA8,\n[dir="ltr"] ._2XCKaix,\n[dir="ltr"] .y0RX1FA,\n[dir="ltr"] ._2_l8bvH,\n[dir="ltr"] .Jmp0xyj {\n  padding-left: ' + theme.arrowSize + ';\n  padding-right: 0;\n}\n\n[dir="rtl"] ._2h8cLA8,\n[dir="rtl"] ._2XCKaix,\n[dir="rtl"] .y0RX1FA,\n[dir="rtl"] ._2_l8bvH,\n[dir="rtl"] .Jmp0xyj {\n  padding-right: ' + theme.arrowSize + ';\n  padding-left: 0;\n}\n\n.oimidfn,\n.GaKnfOq,\n._2jUCcec,\n._3cWt3lO,\n._32n9pA0 {\n  /* arrow on the right side pointing right */\n  -webkit-padding-end: ' + theme.arrowSize + ';\n          padding-inline-end: ' + theme.arrowSize + ';\n  -webkit-padding-start: 0;\n          padding-inline-start: 0;\n}\n\n[dir="ltr"] .oimidfn,\n[dir="ltr"] .GaKnfOq,\n[dir="ltr"] ._2jUCcec,\n[dir="ltr"] ._3cWt3lO,\n[dir="ltr"] ._32n9pA0 {\n  padding-right: ' + theme.arrowSize + ';\n  padding-left: 0;\n}\n\n[dir="rtl"] .oimidfn,\n[dir="rtl"] .GaKnfOq,\n[dir="rtl"] ._2jUCcec,\n[dir="rtl"] ._3cWt3lO,\n[dir="rtl"] ._32n9pA0 {\n  padding-left: ' + theme.arrowSize + ';\n  padding-right: 0;\n}\n\n._3pV6imZ,\n._13mTzFl,\n.LCr3MJt,\n._1OoaREZ {\n  /* arrow on the top side pointing up */\n  padding-top: ' + theme.arrowSize + ';\n}\n\n._3GR6G7Q,\n._3wJPhQm,\n.bo-LxUg,\n._1WP3iQ9 {\n  /* arrow on the bottom side pointing down */\n  padding-bottom: ' + theme.arrowSize + ';\n}\n\n._3gN6MaU {\n  position: absolute;\n  left: -999em; /* stylelint-disable-line property-blacklist */\n}\n\n._1SyqiCu,\n._1SyqiCu::after {\n  content: "";\n  height: 0;\n  width: 0;\n  position: absolute;\n  display: block;\n  box-sizing: border-box;\n  pointer-events: none;\n  border-style: solid;\n}\n\n._1SyqiCu {\n  display: block;\n  border-width: calc(' + theme.arrowSize + ' + ' + theme.arrowBorderWidth + ');\n}\n\n._1SyqiCu::after {\n    border-width: ' + theme.arrowSize + ';\n  }\n\n._2zEUZqf,\n._2zEUZqf::after {\n  border-color: ' + theme.arrowBorderColor + ';\n}\n\n._2uCMbDE,\n._2uCMbDE::after {\n  border-color: ' + theme.arrowBorderColorInverse + ';\n}\n\n._3MRbEp2,\n.xPdm2D7,\n.OoAy-T7,\n._3dA3mpv,\n._1-Sy53l {\n  /* arrow on the left side pointing left */\n  top: 50%;\n  offset-inline-end: 100%;\n  offset-inline-start: auto;\n  margin-top: calc(-1 * (' + theme.arrowSize + ' + ' + theme.arrowBorderWidth + '));\n  border-inline-start-width: 0;\n  border-inline-start-color: transparent;\n  border-inline-end-color: ' + theme.arrowBorderColor + ';\n  border-top-color: transparent;\n  border-bottom-color: transparent;\n  border-inline-end-width: ' + theme.arrowSize + ';\n}\n\n[dir="ltr"] ._3MRbEp2,\n[dir="ltr"] .xPdm2D7,\n[dir="ltr"] .OoAy-T7,\n[dir="ltr"] ._3dA3mpv,\n[dir="ltr"] ._1-Sy53l {\n  right: 100%;\n  left: auto;\n  border-left-width: 0;\n  border-left-color: transparent;\n  border-right-color: ' + theme.arrowBorderColor + ';\n  border-right-width: ' + theme.arrowSize + ';\n}\n\n[dir="rtl"] ._3MRbEp2,\n[dir="rtl"] .xPdm2D7,\n[dir="rtl"] .OoAy-T7,\n[dir="rtl"] ._3dA3mpv,\n[dir="rtl"] ._1-Sy53l {\n  left: 100%;\n  right: auto;\n  border-right-width: 0;\n  border-right-color: transparent;\n  border-left-color: ' + theme.arrowBorderColor + ';\n  border-left-width: ' + theme.arrowSize + ';\n}\n\n._3MRbEp2::after, .xPdm2D7::after, .OoAy-T7::after, ._3dA3mpv::after, ._1-Sy53l::after {\n    offset-inline-start: ' + theme.arrowBorderWidth + ';\n    offset-inline-end: auto;\n    margin-top: calc(-1 * ' + theme.arrowSize + ');\n    border-inline-start-width: 0;\n    border-inline-start-color: transparent;\n    border-top-color: transparent;\n    border-bottom-color: transparent;\n    border-inline-end-width: ' + theme.arrowSize + ';\n    border-inline-end-color: ' + theme.arrowBackgroundColor + ';\n  }\n\n[dir="ltr"] ._3MRbEp2::after,\n[dir="ltr"] .xPdm2D7::after,\n[dir="ltr"] .OoAy-T7::after,\n[dir="ltr"] ._3dA3mpv::after,\n[dir="ltr"] ._1-Sy53l::after {\n    left: ' + theme.arrowBorderWidth + ';\n    right: auto;\n    border-left-width: 0;\n    border-left-color: transparent;\n    border-right-width: ' + theme.arrowSize + ';\n    border-right-color: ' + theme.arrowBackgroundColor + ';\n  }\n\n[dir="rtl"] ._3MRbEp2::after,\n[dir="rtl"] .xPdm2D7::after,\n[dir="rtl"] .OoAy-T7::after,\n[dir="rtl"] ._3dA3mpv::after,\n[dir="rtl"] ._1-Sy53l::after {\n    right: ' + theme.arrowBorderWidth + ';\n    left: auto;\n    border-right-width: 0;\n    border-right-color: transparent;\n    border-left-width: ' + theme.arrowSize + ';\n    border-left-color: ' + theme.arrowBackgroundColor + ';\n  }\n\n._3MRbEp2._2uCMbDE, .xPdm2D7._2uCMbDE, .OoAy-T7._2uCMbDE, ._3dA3mpv._2uCMbDE, ._1-Sy53l._2uCMbDE {\n    border-inline-end-color: ' + theme.arrowBorderColorInverse + ';\n  }\n\n[dir="ltr"] ._3MRbEp2._2uCMbDE,\n[dir="ltr"] .xPdm2D7._2uCMbDE,\n[dir="ltr"] .OoAy-T7._2uCMbDE,\n[dir="ltr"] ._3dA3mpv._2uCMbDE,\n[dir="ltr"] ._1-Sy53l._2uCMbDE {\n    border-right-color: ' + theme.arrowBorderColorInverse + ';\n  }\n\n[dir="rtl"] ._3MRbEp2._2uCMbDE,\n[dir="rtl"] .xPdm2D7._2uCMbDE,\n[dir="rtl"] .OoAy-T7._2uCMbDE,\n[dir="rtl"] ._3dA3mpv._2uCMbDE,\n[dir="rtl"] ._1-Sy53l._2uCMbDE {\n    border-left-color: ' + theme.arrowBorderColorInverse + ';\n  }\n\n._3MRbEp2._2uCMbDE::after, .xPdm2D7._2uCMbDE::after, .OoAy-T7._2uCMbDE::after, ._3dA3mpv._2uCMbDE::after, ._1-Sy53l._2uCMbDE::after {\n      border-inline-end-color: ' + theme.arrowBackgroundColorInverse + ';\n    }\n\n[dir="ltr"] ._3MRbEp2._2uCMbDE::after,\n[dir="ltr"] .xPdm2D7._2uCMbDE::after,\n[dir="ltr"] .OoAy-T7._2uCMbDE::after,\n[dir="ltr"] ._3dA3mpv._2uCMbDE::after,\n[dir="ltr"] ._1-Sy53l._2uCMbDE::after {\n      border-right-color: ' + theme.arrowBackgroundColorInverse + ';\n    }\n\n[dir="rtl"] ._3MRbEp2._2uCMbDE::after,\n[dir="rtl"] .xPdm2D7._2uCMbDE::after,\n[dir="rtl"] .OoAy-T7._2uCMbDE::after,\n[dir="rtl"] ._3dA3mpv._2uCMbDE::after,\n[dir="rtl"] ._1-Sy53l._2uCMbDE::after {\n      border-left-color: ' + theme.arrowBackgroundColorInverse + ';\n    }\n\n._2hrY4Gz,\n._2K5HgnA,\n.Udjdsal,\n._2W8XOPC,\n.WH16m8C {\n  /* arrow on the right side pointing right */\n  top: 50%;\n  offset-inline-start: 100%;\n  offset-inline-end: auto;\n  margin-top: calc(-1 * (' + theme.arrowSize + ' + ' + theme.arrowBorderWidth + '));\n  border-inline-end-width: 0;\n  border-inline-end-color: transparent;\n  border-inline-start-color: ' + theme.arrowBorderColor + ';\n  border-top-color: transparent;\n  border-bottom-color: transparent;\n  border-inline-start-width: ' + theme.arrowSize + ';\n}\n\n[dir="ltr"] ._2hrY4Gz,\n[dir="ltr"] ._2K5HgnA,\n[dir="ltr"] .Udjdsal,\n[dir="ltr"] ._2W8XOPC,\n[dir="ltr"] .WH16m8C {\n  left: 100%;\n  right: auto;\n  border-right-width: 0;\n  border-right-color: transparent;\n  border-left-color: ' + theme.arrowBorderColor + ';\n  border-left-width: ' + theme.arrowSize + ';\n}\n\n[dir="rtl"] ._2hrY4Gz,\n[dir="rtl"] ._2K5HgnA,\n[dir="rtl"] .Udjdsal,\n[dir="rtl"] ._2W8XOPC,\n[dir="rtl"] .WH16m8C {\n  right: 100%;\n  left: auto;\n  border-left-width: 0;\n  border-left-color: transparent;\n  border-right-color: ' + theme.arrowBorderColor + ';\n  border-right-width: ' + theme.arrowSize + ';\n}\n\n._2hrY4Gz::after, ._2K5HgnA::after, .Udjdsal::after, ._2W8XOPC::after, .WH16m8C::after {\n    offset-inline-end: ' + theme.arrowBorderWidth + ';\n    offset-inline-start: auto;\n    margin-top: calc(-1 * ' + theme.arrowSize + ');\n    border-inline-end-width: 0;\n    border-inline-end-color: transparent;\n    border-top-color: transparent;\n    border-bottom-color: transparent;\n    border-inline-start-width: ' + theme.arrowSize + ';\n    border-inline-start-color: ' + theme.arrowBackgroundColor + ';\n  }\n\n[dir="ltr"] ._2hrY4Gz::after,\n[dir="ltr"] ._2K5HgnA::after,\n[dir="ltr"] .Udjdsal::after,\n[dir="ltr"] ._2W8XOPC::after,\n[dir="ltr"] .WH16m8C::after {\n    right: ' + theme.arrowBorderWidth + ';\n    left: auto;\n    border-right-width: 0;\n    border-right-color: transparent;\n    border-left-width: ' + theme.arrowSize + ';\n    border-left-color: ' + theme.arrowBackgroundColor + ';\n  }\n\n[dir="rtl"] ._2hrY4Gz::after,\n[dir="rtl"] ._2K5HgnA::after,\n[dir="rtl"] .Udjdsal::after,\n[dir="rtl"] ._2W8XOPC::after,\n[dir="rtl"] .WH16m8C::after {\n    left: ' + theme.arrowBorderWidth + ';\n    right: auto;\n    border-left-width: 0;\n    border-left-color: transparent;\n    border-right-width: ' + theme.arrowSize + ';\n    border-right-color: ' + theme.arrowBackgroundColor + ';\n  }\n\n._2hrY4Gz._2uCMbDE, ._2K5HgnA._2uCMbDE, .Udjdsal._2uCMbDE, ._2W8XOPC._2uCMbDE, .WH16m8C._2uCMbDE {\n    border-inline-start-color: ' + theme.arrowBorderColorInverse + ';\n  }\n\n[dir="ltr"] ._2hrY4Gz._2uCMbDE,\n[dir="ltr"] ._2K5HgnA._2uCMbDE,\n[dir="ltr"] .Udjdsal._2uCMbDE,\n[dir="ltr"] ._2W8XOPC._2uCMbDE,\n[dir="ltr"] .WH16m8C._2uCMbDE {\n    border-left-color: ' + theme.arrowBorderColorInverse + ';\n  }\n\n[dir="rtl"] ._2hrY4Gz._2uCMbDE,\n[dir="rtl"] ._2K5HgnA._2uCMbDE,\n[dir="rtl"] .Udjdsal._2uCMbDE,\n[dir="rtl"] ._2W8XOPC._2uCMbDE,\n[dir="rtl"] .WH16m8C._2uCMbDE {\n    border-right-color: ' + theme.arrowBorderColorInverse + ';\n  }\n\n._2hrY4Gz._2uCMbDE::after, ._2K5HgnA._2uCMbDE::after, .Udjdsal._2uCMbDE::after, ._2W8XOPC._2uCMbDE::after, .WH16m8C._2uCMbDE::after {\n      border-inline-start-color: ' + theme.arrowBackgroundColorInverse + ';\n    }\n\n[dir="ltr"] ._2hrY4Gz._2uCMbDE::after,\n[dir="ltr"] ._2K5HgnA._2uCMbDE::after,\n[dir="ltr"] .Udjdsal._2uCMbDE::after,\n[dir="ltr"] ._2W8XOPC._2uCMbDE::after,\n[dir="ltr"] .WH16m8C._2uCMbDE::after {\n      border-left-color: ' + theme.arrowBackgroundColorInverse + ';\n    }\n\n[dir="rtl"] ._2hrY4Gz._2uCMbDE::after,\n[dir="rtl"] ._2K5HgnA._2uCMbDE::after,\n[dir="rtl"] .Udjdsal._2uCMbDE::after,\n[dir="rtl"] ._2W8XOPC._2uCMbDE::after,\n[dir="rtl"] .WH16m8C._2uCMbDE::after {\n      border-right-color: ' + theme.arrowBackgroundColorInverse + ';\n    }\n\n._24rlfN3,\n.uLUVWry,\n._3OTAntJ,\n._2QaHU8w {\n  /* arrow on the top side pointing up */\n  bottom: 100%;\n  -webkit-margin-start: calc(-1 * (' + theme.arrowSize + ' + ' + theme.arrowBorderWidth + '));\n          margin-inline-start: calc(-1 * (' + theme.arrowSize + ' + ' + theme.arrowBorderWidth + '));\n  -webkit-margin-end: 0;\n          margin-inline-end: 0;\n  border-top-width: 0;\n  border-top-color: transparent;\n  border-inline-start-color: transparent;\n  border-inline-end-color: transparent;\n}\n\n[dir="ltr"] ._24rlfN3,\n[dir="ltr"] .uLUVWry,\n[dir="ltr"] ._3OTAntJ,\n[dir="ltr"] ._2QaHU8w {\n  margin-left: calc(-1 * (' + theme.arrowSize + ' + ' + theme.arrowBorderWidth + '));\n  margin-right: 0;\n  border-left-color: transparent;\n  border-right-color: transparent;\n}\n\n[dir="rtl"] ._24rlfN3,\n[dir="rtl"] .uLUVWry,\n[dir="rtl"] ._3OTAntJ,\n[dir="rtl"] ._2QaHU8w {\n  margin-right: calc(-1 * (' + theme.arrowSize + ' + ' + theme.arrowBorderWidth + '));\n  margin-left: 0;\n  border-right-color: transparent;\n  border-left-color: transparent;\n}\n\n._24rlfN3::after, .uLUVWry::after, ._3OTAntJ::after, ._2QaHU8w::after {\n    top: ' + theme.arrowBorderWidth + ';\n    -webkit-margin-start: calc(-1 * ' + theme.arrowSize + ');\n            margin-inline-start: calc(-1 * ' + theme.arrowSize + ');\n    -webkit-margin-end: 0;\n            margin-inline-end: 0;\n    border-top-width: 0;\n    border-top-color: transparent;\n    border-inline-start-color: transparent;\n    border-inline-end-color: transparent;\n    border-bottom-color: ' + theme.arrowBackgroundColor + ';\n  }\n\n[dir="ltr"] ._24rlfN3::after,\n[dir="ltr"] .uLUVWry::after,\n[dir="ltr"] ._3OTAntJ::after,\n[dir="ltr"] ._2QaHU8w::after {\n    margin-left: calc(-1 * ' + theme.arrowSize + ');\n    margin-right: 0;\n    border-left-color: transparent;\n    border-right-color: transparent;\n  }\n\n[dir="rtl"] ._24rlfN3::after,\n[dir="rtl"] .uLUVWry::after,\n[dir="rtl"] ._3OTAntJ::after,\n[dir="rtl"] ._2QaHU8w::after {\n    margin-right: calc(-1 * ' + theme.arrowSize + ');\n    margin-left: 0;\n    border-right-color: transparent;\n    border-left-color: transparent;\n  }\n\n._24rlfN3._2uCMbDE::after, .uLUVWry._2uCMbDE::after, ._3OTAntJ._2uCMbDE::after, ._2QaHU8w._2uCMbDE::after {\n      border-bottom-color: ' + theme.arrowBackgroundColorInverse + ';\n    }\n\n.ZPNYQkL,\n._3a1Tpch,\n._2e8MVjW,\n._1DdE097 {\n  /* arrow on the bottom side pointing down */\n  top: 100%;\n  -webkit-margin-start: calc(-1 * (' + theme.arrowSize + ' + ' + theme.arrowBorderWidth + '));\n          margin-inline-start: calc(-1 * (' + theme.arrowSize + ' + ' + theme.arrowBorderWidth + '));\n  -webkit-margin-end: 0;\n          margin-inline-end: 0;\n  border-bottom-width: 0;\n  border-bottom-color: transparent;\n  border-inline-start-color: transparent;\n  border-inline-end-color: transparent;\n}\n\n[dir="ltr"] .ZPNYQkL,\n[dir="ltr"] ._3a1Tpch,\n[dir="ltr"] ._2e8MVjW,\n[dir="ltr"] ._1DdE097 {\n  margin-left: calc(-1 * (' + theme.arrowSize + ' + ' + theme.arrowBorderWidth + '));\n  margin-right: 0;\n  border-left-color: transparent;\n  border-right-color: transparent;\n}\n\n[dir="rtl"] .ZPNYQkL,\n[dir="rtl"] ._3a1Tpch,\n[dir="rtl"] ._2e8MVjW,\n[dir="rtl"] ._1DdE097 {\n  margin-right: calc(-1 * (' + theme.arrowSize + ' + ' + theme.arrowBorderWidth + '));\n  margin-left: 0;\n  border-right-color: transparent;\n  border-left-color: transparent;\n}\n\n.ZPNYQkL::after, ._3a1Tpch::after, ._2e8MVjW::after, ._1DdE097::after {\n    bottom: ' + theme.arrowBorderWidth + ';\n    -webkit-margin-start: calc(-1 * ' + theme.arrowSize + ');\n            margin-inline-start: calc(-1 * ' + theme.arrowSize + ');\n    -webkit-margin-end: 0;\n            margin-inline-end: 0;\n    border-bottom-width: 0;\n    border-bottom-color: transparent;\n    border-inline-start-color: transparent;\n    border-inline-end-color: transparent;\n    border-top-color: ' + theme.arrowBackgroundColor + ';\n  }\n\n[dir="ltr"] .ZPNYQkL::after,\n[dir="ltr"] ._3a1Tpch::after,\n[dir="ltr"] ._2e8MVjW::after,\n[dir="ltr"] ._1DdE097::after {\n    margin-left: calc(-1 * ' + theme.arrowSize + ');\n    margin-right: 0;\n    border-left-color: transparent;\n    border-right-color: transparent;\n  }\n\n[dir="rtl"] .ZPNYQkL::after,\n[dir="rtl"] ._3a1Tpch::after,\n[dir="rtl"] ._2e8MVjW::after,\n[dir="rtl"] ._1DdE097::after {\n    margin-right: calc(-1 * ' + theme.arrowSize + ');\n    margin-left: 0;\n    border-right-color: transparent;\n    border-left-color: transparent;\n  }\n\n.ZPNYQkL._2uCMbDE::after, ._3a1Tpch._2uCMbDE::after, ._2e8MVjW._2uCMbDE::after, ._1DdE097._2uCMbDE::after {\n      border-top-color: ' + theme.arrowBackgroundColorInverse + ';\n    }\n\n._24rlfN3,\n.ZPNYQkL,\n._2QaHU8w,\n._1DdE097 {\n  offset-inline-start: 50%;\n}\n\n[dir="ltr"] ._24rlfN3,\n[dir="ltr"] .ZPNYQkL,\n[dir="ltr"] ._2QaHU8w,\n[dir="ltr"] ._1DdE097 {\n  left: 50%;\n}\n\n[dir="rtl"] ._24rlfN3,\n[dir="rtl"] .ZPNYQkL,\n[dir="rtl"] ._2QaHU8w,\n[dir="rtl"] ._1DdE097 {\n  right: 50%;\n}\n\n._3OTAntJ,\n._3a1Tpch {\n  offset-inline-start: calc((' + theme.arrowSize + ' + ' + theme.arrowBorderWidth + ') * 2);\n  offset-inline-end: auto;\n}\n\n[dir="ltr"] ._3OTAntJ,\n[dir="ltr"] ._3a1Tpch {\n  left: calc((' + theme.arrowSize + ' + ' + theme.arrowBorderWidth + ') * 2);\n  right: auto;\n}\n\n[dir="rtl"] ._3OTAntJ,\n[dir="rtl"] ._3a1Tpch {\n  right: calc((' + theme.arrowSize + ' + ' + theme.arrowBorderWidth + ') * 2);\n  left: auto;\n}\n\n.uLUVWry,\n._2e8MVjW {\n  offset-inline-start: calc(100% - (' + theme.arrowSize + ' + ' + theme.arrowBorderWidth + ') * 2);\n  offset-inline-end: auto;\n}\n\n[dir="ltr"] .uLUVWry,\n[dir="ltr"] ._2e8MVjW {\n  left: calc(100% - (' + theme.arrowSize + ' + ' + theme.arrowBorderWidth + ') * 2);\n  right: auto;\n}\n\n[dir="rtl"] .uLUVWry,\n[dir="rtl"] ._2e8MVjW {\n  right: calc(100% - (' + theme.arrowSize + ' + ' + theme.arrowBorderWidth + ') * 2);\n  left: auto;\n}\n\n.xPdm2D7,\n._2K5HgnA {\n  top: calc((' + theme.arrowSize + ' + ' + theme.arrowBorderWidth + ') * 2);\n}\n\n.OoAy-T7,\n.Udjdsal {\n  top: calc(100% - (' + theme.arrowSize + ' + ' + theme.arrowBorderWidth + ') * 2);\n}\n';
    };

    return tmpl.call(theme, theme);
  },
  'root': 'riLv0Xm',
  'content': '_1GXlG2S',
  'placement--end-center': '_2h8cLA8',
  'placement--end-top': '_2XCKaix',
  'placement--end-bottom': 'y0RX1FA',
  'placement--center-end': '_2_l8bvH',
  'placement--end': 'Jmp0xyj',
  'placement--start-center': 'oimidfn',
  'placement--start-top': 'GaKnfOq',
  'placement--start-bottom': '_2jUCcec',
  'placement--center-start': '_3cWt3lO',
  'placement--start': '_32n9pA0',
  'placement--bottom': '_3pV6imZ',
  'placement--bottom-end': '_13mTzFl',
  'placement--bottom-start': 'LCr3MJt',
  'placement--bottom-center': '_1OoaREZ',
  'placement--top': '_3GR6G7Q',
  'placement--top-start': '_3wJPhQm',
  'placement--top-end': 'bo-LxUg',
  'placement--top-center': '_1WP3iQ9',
  'placement--offscreen': '_3gN6MaU',
  'arrow': '_1SyqiCu',
  'arrow--default': '_2zEUZqf',
  'arrow--inverse': '_2uCMbDE',
  'arrow--start-center': '_3MRbEp2',
  'arrow--start-top': 'xPdm2D7',
  'arrow--start-bottom': 'OoAy-T7',
  'arrow--center-start': '_3dA3mpv',
  'arrow--start': '_1-Sy53l',
  'arrow--end-center': '_2hrY4Gz',
  'arrow--end-top': '_2K5HgnA',
  'arrow--end-bottom': 'Udjdsal',
  'arrow--center-end': '_2W8XOPC',
  'arrow--end': 'WH16m8C',
  'arrow--top': '_24rlfN3',
  'arrow--top-end': 'uLUVWry',
  'arrow--top-start': '_3OTAntJ',
  'arrow--top-center': '_2QaHU8w',
  'arrow--bottom': 'ZPNYQkL',
  'arrow--bottom-start': '_3a1Tpch',
  'arrow--bottom-end': '_2e8MVjW',
  'arrow--bottom-center': '_1DdE097'
};

import theme from './theme';

/**
---
category: components/layout
---
**/
var ContextView = (_dec = themeable(theme, styles), _dec(_class = (_temp = _class2 = function (_Component) {
  _inherits(ContextView, _Component);

  function ContextView() {
    _classCallCheck(this, ContextView);

    return _possibleConstructorReturn(this, (ContextView.__proto__ || Object.getPrototypeOf(ContextView)).apply(this, arguments));
  }

  _createClass(ContextView, [{
    key: 'render',
    value: function render() {
      var _classes, _classnames;

      var _props = this.props,
          as = _props.as,
          background = _props.background,
          children = _props.children,
          debug = _props.debug,
          elementRef = _props.elementRef,
          height = _props.height,
          width = _props.width,
          maxHeight = _props.maxHeight,
          maxWidth = _props.maxWidth,
          minHeight = _props.minHeight,
          minWidth = _props.minWidth,
          margin = _props.margin,
          padding = _props.padding,
          placement = _props.placement,
          shadow = _props.shadow,
          stacking = _props.stacking,
          style = _props.style,
          textAlign = _props.textAlign;


      var classes = (_classes = {}, _defineProperty(_classes, styles.root, true), _defineProperty(_classes, styles['placement--' + placement.replace(' ', '-')], true), _classes);

      return React.createElement(
        View,
        Object.assign({}, omitProps(this.props, ContextView.propTypes), {
          style: style,
          className: classnames(classes),
          borderWidth: 'none',
          display: 'inline-block',
          as: as,
          debug: debug,
          elementRef: elementRef,
          margin: margin,
          stacking: stacking
        }),
        React.createElement(
          View,
          {
            className: styles.content,
            borderRadius: 'medium',
            borderWidth: 'small',
            display: 'block',
            background: background,
            debug: debug,
            height: height,
            width: width,
            maxHeight: maxHeight,
            maxWidth: maxWidth,
            minHeight: minHeight,
            minWidth: minWidth,
            padding: padding,
            shadow: shadow,
            textAlign: textAlign
          },
          React.createElement('span', {
            className: classnames((_classnames = {}, _defineProperty(_classnames, styles.arrow, true), _defineProperty(_classnames, styles['arrow--' + background], background), _defineProperty(_classnames, styles['arrow--' + this.mirroredPlacement], true), _classnames))
          }),
          children
        )
      );
    }
  }, {
    key: 'mirroredPlacement',
    get: function get() {
      return mirrorPlacement(this.props.placement, '-');
    }
  }]);

  ContextView.displayName = 'ContextView'
  ;
  return ContextView;
}(Component), _class2.propTypes = {
  /**
  * The element to render as the component root
  */
  as: CustomPropTypes.elementType,

  /**
  * provides a reference to the underlying html root element
  */
  elementRef: PropTypes.func,

  /**
  * Set the margin using familiar CSS shorthand
  */
  margin: ThemeablePropTypes.spacing,

  /**
  * Set the padding using familiar CSS shorthand
  */
  padding: ThemeablePropTypes.spacing,

  height: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  width: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  maxHeight: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  maxWidth: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  minHeight: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  minWidth: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),

  /**
  * The children to render inside the `<ContextView />`
  */
  children: PropTypes.node,

  /**
  * Designates the text alignment within the `<ContextView />`
  */
  textAlign: PropTypes.oneOf(['start', 'center', 'end']),

  /**
  * Controls the shadow depth for the `<ContextView />`
  */
  shadow: ThemeablePropTypes.shadow,

  /**
  * Controls the z-index depth for the `<ContextView />`
  */
  stacking: ThemeablePropTypes.stacking,

  /**
  * Designates the background style of the `<ContextView />`
  */
  background: PropTypes.oneOf(['default', 'inverse']),

  /**
  * Specifies how the arrow for `<ContextView />` will be rendered.
  * Ex. `placement="top"` will render with an arrow pointing down.
  */
  placement: LayoutPropTypes.placement,

  /**
  * Activate an outline around the component to make building your
  * layout easier
  */
  debug: PropTypes.bool
}, _class2.defaultProps = {
  as: 'span',
  elementRef: function elementRef(el) {},
  debug: false,
  width: 'auto',
  height: 'auto',
  children: null,
  textAlign: 'start',
  background: 'default',
  shadow: 'resting',
  placement: 'center end'
}, _temp)) || _class);


export default ContextView;
