// @instructure/ui-forms/lib/components/RadioInputGroup/__examples__/RadioInputGroup.examples.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.success = exports.hint = exports.error = exports.sizeLarge = exports.sizeMedium = exports.sizeSmall = exports.variantToggle = exports.defaultValue = exports.layoutInline = exports.layoutColumns = exports.layoutStacked = undefined;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _RadioInput = require('../../RadioInput');

var _RadioInput2 = _interopRequireDefault(_RadioInput);

var _index = require('../index');

var _index2 = _interopRequireDefault(_index);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var layoutStacked = exports.layoutStacked = function layoutStacked() {
  return _react2.default.createElement(
    _index2.default,
    {
      name: 'example1',
      description: 'Select something' },
    _react2.default.createElement(_RadioInput2.default, {
      key: 'minnie',
      value: 'minnie',
      label: 'Minnie'
    }),
    _react2.default.createElement(_RadioInput2.default, {
      key: 'mickey',
      value: 'mickey',
      label: 'Mickey'
    }),
    _react2.default.createElement(_RadioInput2.default, {
      key: 'goofy',
      value: 'goofy',
      label: 'Goofy'
    })
  );
}; /*
    * The MIT License (MIT)
    *
    * Copyright (c) 2015 - present Instructure, Inc.
    *
    * Permission is hereby granted, free of charge, to any person obtaining a copy
    * of this software and associated documentation files (the "Software"), to deal
    * in the Software without restriction, including without limitation the rights
    * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
    * copies of the Software, and to permit persons to whom the Software is
    * furnished to do so, subject to the following conditions:
    *
    * The above copyright notice and this permission notice shall be included in all
    * copies or substantial portions of the Software.
    *
    * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
    * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
    * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
    * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
    * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
    * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
    * SOFTWARE.
    */

var layoutColumns = exports.layoutColumns = function layoutColumns() {
  return _react2.default.createElement(
    _index2.default,
    {
      name: 'example2',
      description: 'Select something',
      layout: 'columns' },
    _react2.default.createElement(_RadioInput2.default, {
      key: 'minnie',
      value: 'minnie',
      label: 'Minnie'
    }),
    _react2.default.createElement(_RadioInput2.default, {
      key: 'mickey',
      value: 'mickey',
      label: 'Mickey'
    }),
    _react2.default.createElement(_RadioInput2.default, {
      key: 'goofy',
      value: 'goofy',
      label: 'Goofy'
    })
  );
};

var layoutInline = exports.layoutInline = function layoutInline() {
  return _react2.default.createElement(
    _index2.default,
    {
      name: 'example3',
      description: 'Select something',
      layout: 'inline' },
    _react2.default.createElement(_RadioInput2.default, {
      key: 'minnie',
      value: 'minnie',
      label: 'Minnie'
    }),
    _react2.default.createElement(_RadioInput2.default, {
      key: 'mickey',
      value: 'mickey',
      label: 'Mickey'
    }),
    _react2.default.createElement(_RadioInput2.default, {
      key: 'goofy',
      value: 'goofy',
      label: 'Goofy'
    })
  );
};

var defaultValue = exports.defaultValue = function defaultValue() {
  return _react2.default.createElement(
    _index2.default,
    {
      name: 'example3',
      defaultValue: 'mickey',
      description: 'Select something',
      layout: 'inline' },
    _react2.default.createElement(_RadioInput2.default, {
      key: 'minnie',
      value: 'minnie',
      label: 'Minnie'
    }),
    _react2.default.createElement(_RadioInput2.default, {
      key: 'mickey',
      value: 'mickey',
      label: 'Mickey'
    }),
    _react2.default.createElement(_RadioInput2.default, {
      key: 'goofy',
      value: 'goofy',
      label: 'Goofy'
    })
  );
};

var variantToggle = exports.variantToggle = function variantToggle() {
  return _react2.default.createElement(
    _index2.default,
    {
      name: 'example4',
      defaultValue: 'goofy',
      description: 'Select something',
      variant: 'toggle' },
    _react2.default.createElement(_RadioInput2.default, {
      key: 'minnie',
      value: 'minnie',
      label: 'Minnie'
    }),
    _react2.default.createElement(_RadioInput2.default, {
      key: 'mickey',
      value: 'mickey',
      label: 'Mickey'
    }),
    _react2.default.createElement(_RadioInput2.default, {
      key: 'goofy',
      value: 'goofy',
      label: 'Goofy'
    })
  );
};

var sizeSmall = exports.sizeSmall = function sizeSmall() {
  return _react2.default.createElement(
    _index2.default,
    {
      name: 'example5',
      defaultValue: 'minnie',
      description: 'Select something',
      size: 'small' },
    _react2.default.createElement(_RadioInput2.default, {
      key: 'minnie',
      value: 'minnie',
      label: 'Minnie'
    }),
    _react2.default.createElement(_RadioInput2.default, {
      key: 'mickey',
      value: 'mickey',
      label: 'Mickey'
    }),
    _react2.default.createElement(_RadioInput2.default, {
      key: 'goofy',
      value: 'goofy',
      label: 'Goofy'
    })
  );
};

var sizeMedium = exports.sizeMedium = function sizeMedium() {
  return _react2.default.createElement(
    _index2.default,
    {
      name: 'example6',
      defaultValue: 'minnie',
      description: 'Select something',
      size: 'medium' },
    _react2.default.createElement(_RadioInput2.default, {
      key: 'minnie',
      value: 'minnie',
      label: 'Minnie'
    }),
    _react2.default.createElement(_RadioInput2.default, {
      key: 'mickey',
      value: 'mickey',
      label: 'Mickey'
    }),
    _react2.default.createElement(_RadioInput2.default, {
      key: 'goofy',
      value: 'goofy',
      label: 'Goofy'
    })
  );
};

var sizeLarge = exports.sizeLarge = function sizeLarge() {
  return _react2.default.createElement(
    _index2.default,
    {
      name: 'example7',
      defaultValue: 'minnie',
      description: 'Select something',
      size: 'large' },
    _react2.default.createElement(_RadioInput2.default, {
      key: 'minnie',
      value: 'minnie',
      label: 'Minnie'
    }),
    _react2.default.createElement(_RadioInput2.default, {
      key: 'mickey',
      value: 'mickey',
      label: 'Mickey'
    }),
    _react2.default.createElement(_RadioInput2.default, {
      key: 'goofy',
      value: 'goofy',
      label: 'Goofy'
    })
  );
};

var error = exports.error = function error() {
  return _react2.default.createElement(
    _index2.default,
    {
      name: 'example8',
      description: 'Select something',
      messages: [{ text: 'Please make a selection.', type: 'error' }]
    },
    _react2.default.createElement(_RadioInput2.default, {
      key: 'minnie',
      value: 'minnie',
      label: 'Minnie'
    }),
    _react2.default.createElement(_RadioInput2.default, {
      key: 'mickey',
      value: 'mickey',
      label: 'Mickey'
    }),
    _react2.default.createElement(_RadioInput2.default, {
      key: 'goofy',
      value: 'goofy',
      label: 'Goofy'
    })
  );
};

var hint = exports.hint = function hint() {
  return _react2.default.createElement(
    _index2.default,
    {
      name: 'example9',
      description: 'Select something',
      messages: [{ text: 'Who would you like to meet at Disneyland?', type: 'hint' }]
    },
    _react2.default.createElement(_RadioInput2.default, {
      key: 'minnie',
      value: 'minnie',
      label: 'Minnie'
    }),
    _react2.default.createElement(_RadioInput2.default, {
      key: 'mickey',
      value: 'mickey',
      label: 'Mickey'
    }),
    _react2.default.createElement(_RadioInput2.default, {
      key: 'goofy',
      value: 'goofy',
      label: 'Goofy'
    })
  );
};

var success = exports.success = function success() {
  return _react2.default.createElement(
    _index2.default,
    {
      name: 'example10',
      description: 'Select something',
      messages: [{ text: 'Great choice, thanks for participating', type: 'success' }]
    },
    _react2.default.createElement(_RadioInput2.default, {
      key: 'minnie',
      value: 'minnie',
      label: 'Minnie'
    }),
    _react2.default.createElement(_RadioInput2.default, {
      key: 'mickey',
      value: 'mickey',
      label: 'Mickey'
    }),
    _react2.default.createElement(_RadioInput2.default, {
      key: 'goofy',
      value: 'goofy',
      label: 'Goofy'
    })
  );
};
