// @instructure/ui-forms/lib/components/Checkbox/ToggleFacade/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _dec, _class, _class2, _temp; /*
                                   * The MIT License (MIT)
                                   *
                                   * Copyright (c) 2015 - present Instructure, Inc.
                                   *
                                   * Permission is hereby granted, free of charge, to any person obtaining a copy
                                   * of this software and associated documentation files (the "Software"), to deal
                                   * in the Software without restriction, including without limitation the rights
                                   * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
                                   * copies of the Software, and to permit persons to whom the Software is
                                   * furnished to do so, subject to the following conditions:
                                   *
                                   * The above copyright notice and this permission notice shall be included in all
                                   * copies or substantial portions of the Software.
                                   *
                                   * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
                                   * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
                                   * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
                                   * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
                                   * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
                                   * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
                                   * SOFTWARE.
                                   */

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _IconCheck = require('@instructure/ui-icons/lib/Solid/IconCheck');

var _IconCheck2 = _interopRequireDefault(_IconCheck);

var _IconX = require('@instructure/ui-icons/lib/Solid/IconX');

var _IconX2 = _interopRequireDefault(_IconX);

var _uiThemeable = require('@instructure/ui-themeable');

var _uiThemeable2 = _interopRequireDefault(_uiThemeable);

var _theme = require('./theme');

var _theme2 = _interopRequireDefault(_theme);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var styles = {
  template: function template(theme) {
    var tmpl = function tmpl() {
      return '/*  imported from styles.css  */\n\n/* stylelint-disable selector-max-class, no-descending-specificity */\n\n.JJsCoei {\n  display: flex;\n  align-items: center;\n}\n\n.JJsCoei .QAEjK0o.OOcDbEj {\n    background: ' + theme.checkedBackground + ';\n    box-shadow: none;\n  }\n\n.JJsCoei .QAEjK0o.OOcDbEj ._1aGvloF {\n      transform: translate3d(50%, 0, 0);\n    }\n\n.JJsCoei .QAEjK0o.OOcDbEj ._2TlEmpz {\n      color: ' + theme.checkedIconColor + ';\n    }\n\n.JJsCoei .QAEjK0o._2CYIFqT::before {\n      transform: scale(1) translate3d(0, 0, 0);\n      opacity: 1;\n    }\n\n[dir="rtl"] .JJsCoei .QAEjK0o.OOcDbEj ._1aGvloF {\n      transform: translate3d(-50%, 0, 0);\n    }\n\n.QAEjK0o {\n  background: ' + theme.background + ';\n  border-color: ' + theme.borderColor + ';\n  display: inline-block;\n  -webkit-user-select: none;\n     -moz-user-select: none;\n      -ms-user-select: none;\n          user-select: none;\n  position: relative;\n  border-radius: 3rem;\n  -webkit-margin-end: ' + theme.marginEnd + ';\n          margin-inline-end: ' + theme.marginEnd + ';\n  -webkit-margin-start: 0;\n          margin-inline-start: 0;\n  vertical-align: middle;\n  box-shadow: inset 0 0 0 ' + theme.borderWidth + ' ' + theme.borderColor + ';\n}\n\n[dir="ltr"] .QAEjK0o {\n  margin-right: ' + theme.marginEnd + ';\n  margin-left: 0;\n}\n\n[dir="rtl"] .QAEjK0o {\n  margin-left: ' + theme.marginEnd + ';\n  margin-right: 0;\n}\n\n.QAEjK0o::before {\n    content: "";\n    position: absolute;\n    top: -0.25rem;\n    left: -0.25rem; /* stylelint-disable-line property-blacklist */\n    width: calc(100% + 0.5rem);\n    height: calc(100% + 0.5rem);\n    box-sizing: border-box;\n    border-radius: calc(' + theme.borderRadius + ' - 0.25rem);\n    border: ' + theme.focusBorderWidth + ' ' + theme.focusBorderStyle + ' ' + theme.focusOutlineColor + ';\n    transition: all 0.2s;\n    transform: scale(0.75) translate3d(0, 0, 0);\n    opacity: 0;\n    pointer-events: none;\n  }\n\n._1aGvloF {\n  display: block;\n  text-align: center;\n  position: absolute;\n  top: 0;\n  offset-inline-start: 0;\n  offset-inline-end: auto;\n  transition: all 0.2s;\n  transform: translate3d(0, 0, 0);\n}\n\n[dir="ltr"] ._1aGvloF {\n  text-align: center;\n  left: 0;\n  right: auto;\n}\n\n[dir="rtl"] ._1aGvloF {\n  text-align: center;\n  right: 0;\n  left: auto;\n}\n\n._1XthnGd {\n  position: relative;\n  height: 100%;\n  width: 100%;\n  display: flex;\n  align-items: center;\n  justify-content: center;\n}\n\n._1XthnGd::before {\n  content: "";\n  position: absolute;\n  top: ' + theme.borderWidth + ';\n  left: ' + theme.borderWidth + '; /* stylelint-disable-line property-blacklist */\n  height: calc(100% - (' + theme.borderWidth + ' * 2));\n  width: calc(100% - (' + theme.borderWidth + ' * 2));\n  background: ' + theme.toggleBackground + ';\n  box-shadow: ' + theme.toggleShadow + ';\n  border-radius: 100%;\n}\n\n._2TlEmpz {\n  display: block;\n  color: ' + theme.uncheckedIconColor + ';\n  position: relative;\n  /* stylelint-disable-next-line scale-unlimited/declaration-strict-value */\n  z-index: 1;\n}\n\n._30I5rOD {\n  flex: 1;\n  min-width: 0.0625rem;\n  color: ' + theme.labelColor + ';\n  font-family: ' + theme.labelFontFamily + ';\n  font-weight: ' + theme.labelFontWeight + ';\n  line-height: ' + theme.labelLineHeight + ';\n}\n\n._3WmopB- {\n  height: ' + theme.baseSizeSmall + ';\n  width: calc(' + theme.baseSizeSmall + ' * 1.5);\n}\n\n._3WmopB- ._1aGvloF {\n    font-size: 0.875rem;\n    height: ' + theme.baseSizeSmall + ';\n    width: ' + theme.baseSizeSmall + ';\n  }\n\n._3WmopB- + ._30I5rOD {\n    font-size: ' + theme.labelFontSizeSmall + ';\n  }\n\n._2J_yk7I {\n  height: ' + theme.baseSizeMedium + ';\n  width: calc(' + theme.baseSizeMedium + ' * 1.5);\n}\n\n._2J_yk7I ._1aGvloF {\n    font-size: 1.125rem;\n    height: ' + theme.baseSizeMedium + ';\n    width: ' + theme.baseSizeMedium + ';\n  }\n\n._2J_yk7I + ._30I5rOD {\n    font-size: ' + theme.labelFontSizeMedium + ';\n  }\n\n._1GbmDW7 {\n  height: ' + theme.baseSizeLarge + ';\n  width: calc(' + theme.baseSizeLarge + ' * 1.5);\n}\n\n._1GbmDW7 ._1aGvloF {\n    font-size: 1.5rem;\n    height: ' + theme.baseSizeLarge + ';\n    width: ' + theme.baseSizeLarge + ';\n  }\n\n._1GbmDW7 + ._30I5rOD {\n    font-size: ' + theme.labelFontSizeLarge + ';\n  }\n\n/* stylelint-enable selector-max-class, no-descending-specificity */\n';
    };

    return tmpl.call(theme, theme);
  },
  'root': 'JJsCoei',
  'facade': 'QAEjK0o',
  'checked': 'OOcDbEj',
  'icon': '_1aGvloF',
  'iconSVG': '_2TlEmpz',
  'focused': '_2CYIFqT',
  'iconToggle': '_1XthnGd',
  'label': '_30I5rOD',
  'small': '_3WmopB-',
  'medium': '_2J_yk7I',
  'large': '_1GbmDW7'
};


/**
---
parent: Checkbox
---
**/
var ToggleFacade = (_dec = (0, _uiThemeable2.default)(_theme2.default, styles), _dec(_class = (_temp = _class2 = function (_Component) {
  _inherits(ToggleFacade, _Component);

  function ToggleFacade() {
    _classCallCheck(this, ToggleFacade);

    return _possibleConstructorReturn(this, (ToggleFacade.__proto__ || Object.getPrototypeOf(ToggleFacade)).apply(this, arguments));
  }

  _createClass(ToggleFacade, [{
    key: 'renderIcon',
    value: function renderIcon() {
      if (this.props.checked) {
        return _react2.default.createElement(_IconCheck2.default, { className: styles.iconSVG });
      } else {
        return _react2.default.createElement(_IconX2.default, { className: styles.iconSVG });
      }
    }
  }, {
    key: 'render',
    value: function render() {
      var _classes;

      var _props = this.props,
          size = _props.size,
          checked = _props.checked,
          disabled = _props.disabled,
          focused = _props.focused;


      var classes = (_classes = {}, _defineProperty(_classes, styles.facade, true), _defineProperty(_classes, styles.checked, checked), _defineProperty(_classes, styles.disabled, disabled), _defineProperty(_classes, styles.focused, focused), _defineProperty(_classes, styles[size], true), _classes);

      return _react2.default.createElement(
        'span',
        { className: styles.root },
        _react2.default.createElement(
          'span',
          { className: (0, _classnames2.default)(classes), 'aria-hidden': 'true' },
          _react2.default.createElement(
            'span',
            { className: styles.icon },
            _react2.default.createElement(
              'span',
              { className: styles.iconToggle },
              this.renderIcon()
            )
          )
        ),
        _react2.default.createElement(
          'span',
          { className: styles.label },
          this.props.children
        )
      );
    }
  }]);

  ToggleFacade.displayName = 'ToggleFacade'
  ;
  return ToggleFacade;
}(_react.Component), _class2.propTypes = {
  children: _propTypes2.default.node.isRequired,
  checked: _propTypes2.default.bool,
  disabled: _propTypes2.default.bool,
  readOnly: _propTypes2.default.bool,
  focused: _propTypes2.default.bool,
  size: _propTypes2.default.oneOf(['small', 'medium', 'large'])
}, _class2.defaultProps = {
  checked: false,
  focused: false,
  size: 'medium'
}, _temp)) || _class);
exports.default = ToggleFacade;
