// @instructure/ui-utils/es/react/index.js
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

/* list utils in alphabetical order */
import _ComponentIdentifier from './ComponentIdentifier';
export { _ComponentIdentifier as ComponentIdentifier };
import _containerQuery from './containerQuery';
export { _containerQuery as containerQuery };
import _CustomPropTypes from './CustomPropTypes';
export { _CustomPropTypes as CustomPropTypes };
import _deprecated from './deprecated';
export { _deprecated as deprecated };
import _ensureSingleChild from './ensureSingleChild';
export { _ensureSingleChild as ensureSingleChild };
import _getDisplayName from './getDisplayName';
export { _getDisplayName as getDisplayName };
import _getElementType from './getElementType';
export { _getElementType as getElementType };
import _matchComponentTypes from './matchComponentTypes';
export { _matchComponentTypes as matchComponentTypes };

export { omitProps, pickProps } from './passthroughProps';
import _safeCloneElement from './safeCloneElement';
export { _safeCloneElement as safeCloneElement };
import _windowMessageListener from './windowMessageListener';
export { _windowMessageListener as windowMessageListener };
