// @instructure/ui-overlays/es/components/Modal/ModalBody/index.js
var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _dec, _class, _class2, _temp;

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import React, { Component } from 'react';
import PropTypes from 'prop-types';

import View from '@instructure/ui-layout/lib/components/View';

import { omitProps } from '@instructure/ui-utils/lib/react/passthroughProps';
import themeable from '@instructure/ui-themeable';
import ThemeablePropTypes from '@instructure/ui-themeable/lib/utils/ThemeablePropTypes';
import CustomPropTypes from '@instructure/ui-utils/lib/react/CustomPropTypes';

var styles = {
  template: function template(theme) {
    var tmpl = function tmpl() {
      return '/*  imported from styles.css  */\n\n.wka3cQQ {\n  box-sizing: border-box;\n  overflow-y: auto;\n  flex: 1 1 auto;\n}\n\n._2byYX0X {\n  box-sizing: border-box;\n  max-width: ' + theme.contentMaxWidth + ';\n  margin: 0 auto;\n}\n';
    };

    return tmpl.call(theme, theme);
  },
  'root': 'wka3cQQ',
  'content': '_2byYX0X'
};

import theme from './theme';

/**
---
parent: Modal
---
**/
var ModalBody = (_dec = themeable(theme, styles), _dec(_class = (_temp = _class2 = function (_Component) {
  _inherits(ModalBody, _Component);

  function ModalBody() {
    _classCallCheck(this, ModalBody);

    return _possibleConstructorReturn(this, (ModalBody.__proto__ || Object.getPrototypeOf(ModalBody)).apply(this, arguments));
  }

  _createClass(ModalBody, [{
    key: 'render',
    value: function render() {
      return React.createElement(
        View,
        Object.assign({}, omitProps(this.props, Object.assign({}, ModalBody.propTypes, View.propTypes)), {
          elementRef: this.props.elementRef,
          as: this.props.as,
          className: styles.root,
          padding: this.props.padding
        }),
        React.createElement(
          'div',
          { className: styles.content },
          this.props.children
        )
      );
    }
  }]);

  ModalBody.displayName = 'ModalBody'
  ;
  return ModalBody;
}(Component), _class2.propTypes = {
  children: PropTypes.node,
  padding: ThemeablePropTypes.spacing,
  elementRef: PropTypes.func,
  as: CustomPropTypes.elementType
}, _class2.defaultProps = {
  padding: 'medium',
  as: 'div'
}, _temp)) || _class);
export { ModalBody as default };
