// @instructure/ui-themes/lib/canvas/base/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.variables = exports.brandVariables = exports.baseVariables = exports.key = undefined;

var _registry = require('@instructure/ui-themeable/lib/registry');

var _keys = require('../../keys');

var _keys2 = _interopRequireDefault(_keys);

var _borders = require('./borders');

var _borders2 = _interopRequireDefault(_borders);

var _colors = require('./colors');

var _colors2 = _interopRequireDefault(_colors);

var _transitions = require('./transitions');

var _transitions2 = _interopRequireDefault(_transitions);

var _typography = require('./typography');

var _typography2 = _interopRequireDefault(_typography);

var _spacing = require('./spacing');

var _spacing2 = _interopRequireDefault(_spacing);

var _forms = require('./forms');

var _forms2 = _interopRequireDefault(_forms);

var _media = require('./media');

var _media2 = _interopRequireDefault(_media);

var _breakpoints = require('./breakpoints');

var _breakpoints2 = _interopRequireDefault(_breakpoints);

var _shadows = require('./shadows');

var _shadows2 = _interopRequireDefault(_shadows);

var _stacking = require('./stacking');

var _stacking2 = _interopRequireDefault(_stacking);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

var key = exports.key = _keys2.default.CANVAS;

var baseVariables = exports.baseVariables = {
  borders: _borders2.default,
  colors: _colors2.default,
  transitions: _transitions2.default,
  typography: _typography2.default,
  spacing: _spacing2.default,
  forms: _forms2.default,
  media: _media2.default,
  breakpoints: _breakpoints2.default,
  shadows: _shadows2.default,
  stacking: _stacking2.default
};

var brandVariables = exports.brandVariables = {
  /* Defaults for Canvas account branding variables: */
  'ic-brand-primary': _colors2.default.textBrand,
  'ic-brand-font-color-dark': _colors2.default.textDarkest,

  'ic-link-color': _colors2.default.textBrand,
  'ic-link-decoration': 'none',

  'ic-brand-button--primary-bgd': _colors2.default.backgroundBrand,
  'ic-brand-button--primary-text': _colors2.default.textLightest,
  'ic-brand-button--secondary-bgd': _colors2.default.backgroundDarkest,
  'ic-brand-button--secondary-text': _colors2.default.textLightest,

  'ic-brand-global-nav-bgd': _colors2.default.backgroundBrandSecondary,
  'ic-global-nav-link-hover': _colors2.default.backgroundDarkest,
  'ic-brand-global-nav-ic-icon-svg-fill': _colors2.default.textLightest,
  'ic-brand-global-nav-ic-icon-svg-fill--active': _colors2.default.textBrand,
  'ic-brand-global-nav-menu-item__text-color': _colors2.default.textLightest,
  'ic-brand-global-nav-menu-item__text-color--active': _colors2.default.textBrand
};

var variables = exports.variables = Object.assign({}, baseVariables, brandVariables);

exports.default = (0, _registry.registerTheme)({
  key: _keys2.default.CANVAS,
  variables: variables
});
