// @instructure/ui-forms/lib/components/FormField/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.FormFieldMessages = exports.FormFieldMessage = exports.FormFieldLabel = exports.FormFieldLayout = undefined;

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _class, _temp; /*
                    * The MIT License (MIT)
                    *
                    * Copyright (c) 2015 - present Instructure, Inc.
                    *
                    * Permission is hereby granted, free of charge, to any person obtaining a copy
                    * of this software and associated documentation files (the "Software"), to deal
                    * in the Software without restriction, including without limitation the rights
                    * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
                    * copies of the Software, and to permit persons to whom the Software is
                    * furnished to do so, subject to the following conditions:
                    *
                    * The above copyright notice and this permission notice shall be included in all
                    * copies or substantial portions of the Software.
                    *
                    * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
                    * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
                    * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
                    * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
                    * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
                    * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
                    * SOFTWARE.
                    */

var _FormFieldLayout = require('./FormFieldLayout');

Object.defineProperty(exports, 'FormFieldLayout', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_FormFieldLayout).default;
  }
});

var _FormFieldLabel = require('./FormFieldLabel');

Object.defineProperty(exports, 'FormFieldLabel', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_FormFieldLabel).default;
  }
});

var _FormFieldMessage = require('./FormFieldMessage');

Object.defineProperty(exports, 'FormFieldMessage', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_FormFieldMessage).default;
  }
});

var _FormFieldMessages = require('./FormFieldMessages');

Object.defineProperty(exports, 'FormFieldMessages', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_FormFieldMessages).default;
  }
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _passthroughProps = require('@instructure/ui-utils/lib/react/passthroughProps');

var _FormPropTypes = require('../../utils/FormPropTypes');

var _FormPropTypes2 = _interopRequireDefault(_FormPropTypes);

var _FormFieldLayout2 = _interopRequireDefault(_FormFieldLayout);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

/**
---
category: components/forms
---
**/
var FormField = (_temp = _class = function (_Component) {
  _inherits(FormField, _Component);

  function FormField() {
    _classCallCheck(this, FormField);

    return _possibleConstructorReturn(this, (FormField.__proto__ || Object.getPrototypeOf(FormField)).apply(this, arguments));
  }

  _createClass(FormField, [{
    key: 'render',
    value: function render() {
      return _react2.default.createElement(_FormFieldLayout2.default, Object.assign({}, (0, _passthroughProps.pickProps)(this.props, _FormFieldLayout2.default.propTypes), {
        vAlign: this.props.vAlign,
        as: 'label',
        htmlFor: this.props.id
      }));
    }
  }]);

  FormField.displayName = 'FormField'
  ;
  return FormField;
}(_react.Component), _class.propTypes = {
  label: _propTypes2.default.node.isRequired,
  /**
  * the id of the input (to link it to its label for a11y)
  */
  id: _propTypes2.default.string.isRequired,
  /**
  * object with shape: `{
  * text: PropTypes.string,
  * type: PropTypes.oneOf(['error', 'hint', 'success', 'screenreader-only'])
  *   }`
  */
  messages: _propTypes2.default.arrayOf(_FormPropTypes2.default.message),
  messagesId: _propTypes2.default.string,
  children: _propTypes2.default.node,
  inline: _propTypes2.default.bool,
  layout: _propTypes2.default.oneOf(['stacked', 'inline']),
  labelAlign: _propTypes2.default.oneOf(['start', 'end']),
  vAlign: _propTypes2.default.oneOf(['top', 'middle', 'bottom']),
  width: _propTypes2.default.string
}, _class.defaultProps = {
  inline: false,
  layout: 'stacked',
  labelAlign: 'end',
  vAlign: 'middle'
}, _temp);
exports.default = FormField;
