// @instructure/ui-overlays/lib/components/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Tray = exports.Tooltip = exports.PopoverContent = exports.PopoverTrigger = exports.Popover = exports.Overlay = exports.ModalFooter = exports.ModalBody = exports.ModalHeader = exports.Modal = exports.Mask = undefined;

var _Modal2 = require('./Modal');

Object.defineProperty(exports, 'ModalHeader', {
  enumerable: true,
  get: function get() {
    return _Modal2.ModalHeader;
  }
});
Object.defineProperty(exports, 'ModalBody', {
  enumerable: true,
  get: function get() {
    return _Modal2.ModalBody;
  }
});
Object.defineProperty(exports, 'ModalFooter', {
  enumerable: true,
  get: function get() {
    return _Modal2.ModalFooter;
  }
});

var _Popover2 = require('./Popover');

Object.defineProperty(exports, 'PopoverTrigger', {
  enumerable: true,
  get: function get() {
    return _Popover2.PopoverTrigger;
  }
});
Object.defineProperty(exports, 'PopoverContent', {
  enumerable: true,
  get: function get() {
    return _Popover2.PopoverContent;
  }
});

var _Mask2 = require('./Mask');

var _Mask3 = _interopRequireDefault(_Mask2);

var _Modal3 = _interopRequireDefault(_Modal2);

var _Overlay2 = require('./Overlay');

var _Overlay3 = _interopRequireDefault(_Overlay2);

var _Popover3 = _interopRequireDefault(_Popover2);

var _Tooltip2 = require('./Tooltip');

var _Tooltip3 = _interopRequireDefault(_Tooltip2);

var _Tray2 = require('./Tray');

var _Tray3 = _interopRequireDefault(_Tray2);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.Mask = _Mask3.default; /*
                                * The MIT License (MIT)
                                *
                                * Copyright (c) 2015 - present Instructure, Inc.
                                *
                                * Permission is hereby granted, free of charge, to any person obtaining a copy
                                * of this software and associated documentation files (the "Software"), to deal
                                * in the Software without restriction, including without limitation the rights
                                * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
                                * copies of the Software, and to permit persons to whom the Software is
                                * furnished to do so, subject to the following conditions:
                                *
                                * The above copyright notice and this permission notice shall be included in all
                                * copies or substantial portions of the Software.
                                *
                                * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
                                * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
                                * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
                                * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
                                * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
                                * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
                                * SOFTWARE.
                                */

exports.Modal = _Modal3.default;
exports.Overlay = _Overlay3.default;
exports.Popover = _Popover3.default;
exports.Tooltip = _Tooltip3.default;
exports.Tray = _Tray3.default;
