// @instructure/ui-layout/es/components/Grid/index.js
var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _dec, _class, _class2, _temp;

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import React, { Component, Children } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';

import themeable from '@instructure/ui-themeable';
import CustomPropTypes from '@instructure/ui-utils/lib/react/CustomPropTypes';
import capitalizeFirstLetter from '@instructure/ui-utils/lib/capitalizeFirstLetter';
import safeCloneElement from '@instructure/ui-utils/lib/react/safeCloneElement';
import matchComponentTypes from '@instructure/ui-utils/lib/react/matchComponentTypes';
import { omitProps, pickProps } from '@instructure/ui-utils/lib/react/passthroughProps';

import GridRow from './GridRow';

var styles = {
  template: function template(theme) {
    var tmpl = function tmpl() {
      return '/*  imported from styles.css  */\n\n/* stylelint-disable at-rule-no-unknown */\n\n._1oH_9-R {\n  display: block;\n}\n\n._1CTWUft {\n  /* stylelint-disable-next-line color-named */\n  outline: 0.0625rem dashed red;\n}\n\n._2sTeRaf {\n  box-sizing: border-box;\n}\n\n@media screen and (--mediumMin) {\n  .nEL4exv {\n    box-sizing: border-box;\n  }\n}\n\n@media screen and (--largeMin) {\n  ._2zcKYzE {\n    box-sizing: border-box;\n  }\n}\n\n@media screen and (--xLargeMin) {\n  .mDN2NVJ {\n    box-sizing: border-box;\n  }\n}\n\n/* stylelint-enable at-rule-no-unknown */\n';
    };

    return tmpl.call(theme, theme);
  },
  'root': '_1oH_9-R',
  'visualDebug': '_1CTWUft',
  'startAtSmall': '_2sTeRaf',
  'startAtMedium': 'nEL4exv',
  'startAtLarge': '_2zcKYzE',
  'startAtXLarge': 'mDN2NVJ'
};

import theme from './theme';

/**
---
category: components/layout
---
**/
var Grid = (_dec = themeable(theme, styles), _dec(_class = (_temp = _class2 = function (_Component) {
  _inherits(Grid, _Component);

  function Grid() {
    _classCallCheck(this, Grid);

    return _possibleConstructorReturn(this, (Grid.__proto__ || Object.getPrototypeOf(Grid)).apply(this, arguments));
  }

  _createClass(Grid, [{
    key: 'startAtClass',
    value: function startAtClass() {
      return !!this.props.startAt && 'startAt' + capitalizeFirstLetter(this.props.startAt);
    }
  }, {
    key: 'renderChildren',
    value: function renderChildren() {
      var _this2 = this;

      var children = Children.toArray(this.props.children);

      return children.map(function (child, index) {
        if (matchComponentTypes(child, [GridRow])) {
          return safeCloneElement(child, Object.assign({}, pickProps(_this2.props, Grid.propTypes), child.props, { /* child props should override parent */
            isLastRow: index + 1 === children.length
          }));
        } else {
          return child; // PropType validation should handle errors
        }
      });
    }
  }, {
    key: 'render',
    value: function render() {
      var _classes;

      var classes = (_classes = {}, _defineProperty(_classes, styles.root, true), _defineProperty(_classes, styles[this.startAtClass()], !!this.props.startAt), _defineProperty(_classes, styles.visualDebug, this.props.visualDebug), _classes);

      var props = omitProps(this.props, Grid.propTypes);

      return React.createElement(
        'span',
        Object.assign({}, props, {
          className: classnames(classes)
        }),
        this.renderChildren()
      );
    }
  }]);

  Grid.displayName = 'Grid'
  ;
  return Grid;
}(Component), _class2.propTypes = {
  children: CustomPropTypes.Children.oneOf([GridRow]),
  colSpacing: PropTypes.oneOf(['none', 'small', 'medium', 'large']),
  rowSpacing: PropTypes.oneOf(['none', 'small', 'medium', 'large']),
  hAlign: PropTypes.oneOf(['start', 'center', 'end', 'space-around', 'space-between']),
  vAlign: PropTypes.oneOf(['top', 'middle', 'bottom']),
  startAt: PropTypes.oneOf(['small', 'medium', 'large', 'x-large', null]),
  visualDebug: PropTypes.bool
}, _class2.defaultProps = {
  colSpacing: 'medium',
  rowSpacing: 'medium',
  hAlign: 'start',
  startAt: 'small',
  vAlign: 'top',
  visualDebug: false
}, _temp)) || _class);
export { Grid as default };


export { default as GridCol } from './GridCol';
export { default as GridRow } from './GridRow';
