// @instructure/ui-layout/lib/components/Media/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _dec, _class, _class2, _temp; /*
                                   * The MIT License (MIT)
                                   *
                                   * Copyright (c) 2015 - present Instructure, Inc.
                                   *
                                   * Permission is hereby granted, free of charge, to any person obtaining a copy
                                   * of this software and associated documentation files (the "Software"), to deal
                                   * in the Software without restriction, including without limitation the rights
                                   * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
                                   * copies of the Software, and to permit persons to whom the Software is
                                   * furnished to do so, subject to the following conditions:
                                   *
                                   * The above copyright notice and this permission notice shall be included in all
                                   * copies or substantial portions of the Software.
                                   *
                                   * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
                                   * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
                                   * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
                                   * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
                                   * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
                                   * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
                                   * SOFTWARE.
                                   */

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames2 = require('classnames');

var _classnames3 = _interopRequireDefault(_classnames2);

var _uiThemeable = require('@instructure/ui-themeable');

var _uiThemeable2 = _interopRequireDefault(_uiThemeable);

var _ThemeablePropTypes = require('@instructure/ui-themeable/lib/utils/ThemeablePropTypes');

var _ThemeablePropTypes2 = _interopRequireDefault(_ThemeablePropTypes);

var _passthroughProps = require('@instructure/ui-utils/lib/react/passthroughProps');

var _View = require('../View');

var _View2 = _interopRequireDefault(_View);

var _theme = require('./theme');

var _theme2 = _interopRequireDefault(_theme);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var styles = {
  template: function template(theme) {
    var tmpl = function tmpl() {
      return '/*  imported from styles.css  */\n\nfigure._1OrSYVC { /* stylelint-disable-line selector-max-type */\n  display: flex;\n  background: ' + theme.background + ';\n  margin: 0;\n  padding: 0;\n  font-family: ' + theme.fontFamily + ';\n}\n\n._3rVyYH8 {\n  align-items: flex-start;\n}\n\n._2noCVO5 {\n  align-items: center;\n}\n\nfigcaption._3flshmv { /* stylelint-disable-line selector-max-type */\n  color: ' + theme.color + ';\n  margin: 0;\n  padding: 0;\n}\n\n.tSAhYS5 {\n  -webkit-margin-start: 0;\n          margin-inline-start: 0;\n  -webkit-margin-end: ' + theme.figureMargin + ';\n          margin-inline-end: ' + theme.figureMargin + ';\n}\n\n[dir="ltr"] .tSAhYS5 {\n  margin-left: 0;\n  margin-right: ' + theme.figureMargin + ';\n}\n\n[dir="rtl"] .tSAhYS5 {\n  margin-right: 0;\n  margin-left: ' + theme.figureMargin + ';\n}\n\n.X7d4CZI {\n  text-rendering: optimizeLegibility;\n  -webkit-font-smoothing: antialiased;\n  -moz-osx-font-smoothing: grayscale;\n  display: block;\n  margin: ' + theme.titleMargin + ';\n  font-size: ' + theme.titleFontSize + ';\n  font-weight: ' + theme.titleFontWeight + ';\n  line-height: ' + theme.titleLineHeight + ';\n}\n\n._1zYGvH0 {\n  font-size: ' + theme.descriptionFontSize + ';\n  line-height: ' + theme.descriptionLineHeight + ';\n  font-weight: ' + theme.descriptionFontWeight + ';\n}\n';
    };

    return tmpl.call(theme, theme);
  },
  'root': '_1OrSYVC',
  'top': '_3rVyYH8',
  'center': '_2noCVO5',
  'caption': '_3flshmv',
  'figure': 'tSAhYS5',
  'title': 'X7d4CZI',
  'description': '_1zYGvH0'
};


/**
---
category: components/layout
---
**/
var Media = (_dec = (0, _uiThemeable2.default)(_theme2.default, styles), _dec(_class = (_temp = _class2 = function (_Component) {
  _inherits(Media, _Component);

  function Media() {
    _classCallCheck(this, Media);

    return _possibleConstructorReturn(this, (Media.__proto__ || Object.getPrototypeOf(Media)).apply(this, arguments));
  }

  _createClass(Media, [{
    key: 'render',
    value: function render() {
      var _classnames;

      return _react2.default.createElement(
        _View2.default,
        Object.assign({}, (0, _passthroughProps.omitProps)(this.props, Object.assign({}, Media.propTypes, _View2.default.propTypes)), {
          elementRef: this.props.elementRef,
          className: (0, _classnames3.default)((_classnames = {}, _defineProperty(_classnames, styles.root, true), _defineProperty(_classnames, styles[this.props.alignContent], true), _classnames)),
          as: 'figure',
          margin: this.props.margin,
          maxWidth: this.theme[this.props.size]
        }),
        _react2.default.createElement(
          'div',
          { className: styles.figure },
          this.props.children
        ),
        _react2.default.createElement(
          'figcaption',
          { className: styles.caption },
          this.props.title && _react2.default.createElement(
            'span',
            { className: styles.title },
            this.props.title
          ),
          this.props.description && _react2.default.createElement(
            'div',
            { className: styles.description },
            this.props.description
          )
        )
      );
    }
  }]);

  Media.displayName = 'Media'
  ;
  return Media;
}(_react.Component), _class2.propTypes = {
  /**
  * the media object
  */
  children: _propTypes2.default.node.isRequired,
  /**
  * the media title
  */
  title: _propTypes2.default.node,
  /**
  * the media description
  */
  description: _propTypes2.default.string,
  /**
  * how should the title and description align
  */
  alignContent: _propTypes2.default.oneOf(['top', 'center']),
  /**
  * Valid values are `0`, `none`, `auto`, `xxx-small`, `xx-small`, `x-small`,
  * `small`, `medium`, `large`, `x-large`, `xx-large`. Apply these values via
  * familiar CSS-like shorthand. For example: `margin="small auto large"`.
  */
  margin: _ThemeablePropTypes2.default.spacing,
  size: _propTypes2.default.oneOf(['small', 'medium', 'large']),
  elementRef: _propTypes2.default.func
}, _class2.defaultProps = {
  alignContent: 'center'
}, _temp)) || _class);
exports.default = Media;
