// @instructure/ui-svg-images/es/components/InlineSVG/index.js
var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _dec, _class, _class2, _temp;

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';

import themeable from '@instructure/ui-themeable';
import generateElementId from '@instructure/ui-utils/lib/dom/generateElementId';
import { omitProps } from '@instructure/ui-utils/lib/react/passthroughProps';

var styles = {
  template: function template(theme) {
    var tmpl = function tmpl() {
      return '/*  imported from styles.css  */\n\n._1U0eZJB {\n  color: inherit;\n  fill: currentColor;\n}\n\n.WMYJeH6 {\n  display: inline-block;\n}\n\n._1s-joJZ {\n  display: block;\n}\n\n.P0oI4N9 {\n  color: ' + theme.primaryColor + ';\n}\n\n._3moNjdr {\n  color: ' + theme.secondaryColor + ';\n}\n\n._2Q2uARw {\n  color: ' + theme.primaryInverseColor + ';\n}\n\n._2QQl8Vm {\n  color: ' + theme.secondaryInverseColor + ';\n}\n\n._2m0KZOp {\n  color: ' + theme.successColor + ';\n}\n\n._7UuVa5t {\n  color: ' + theme.brandColor + ';\n}\n\n._1PN6oO_ {\n  color: ' + theme.warningColor + ';\n}\n\n._2A3sMmX {\n  color: ' + theme.errorColor + ';\n}\n';
    };

    return tmpl.call(theme, theme);
  },
  'root': '_1U0eZJB',
  'inline': 'WMYJeH6',
  'block': '_1s-joJZ',
  'color--primary': 'P0oI4N9',
  'color--secondary': '_3moNjdr',
  'color--primary-inverse': '_2Q2uARw',
  'color--secondary-inverse': '_2QQl8Vm',
  'color--success': '_2m0KZOp',
  'color--brand': '_7UuVa5t',
  'color--warning': '_1PN6oO_',
  'color--error': '_2A3sMmX'
};

import theme from './theme';

/**
---
category: components/utilities
---
**/
var InlineSVG = (_dec = themeable(theme, styles), _dec(_class = (_temp = _class2 = function (_Component) {
  _inherits(InlineSVG, _Component);

  function InlineSVG() {
    _classCallCheck(this, InlineSVG);

    var _this = _possibleConstructorReturn(this, (InlineSVG.__proto__ || Object.getPrototypeOf(InlineSVG)).call(this));

    _this.titleId = generateElementId('InlineSVG-title');
    _this.descId = generateElementId('InlineSVG-desc');
    return _this;
  }

  _createClass(InlineSVG, [{
    key: 'renderTitle',
    value: function renderTitle() {
      var title = this.props.title;

      return title ? React.createElement(
        'title',
        { id: this.titleId },
        title
      ) : null;
    }
  }, {
    key: 'renderDesc',
    value: function renderDesc(desc) {
      return desc ? React.createElement(
        'desc',
        { id: this.descId },
        desc
      ) : null;
    }
  }, {
    key: 'renderContent',
    value: function renderContent() {
      if (this.props.src) {
        var src = InlineSVG.prepareSrc(this.props.src);
        return React.createElement('g', {
          role: 'presentation',
          dangerouslySetInnerHTML: { __html: src } // eslint-disable-line react/no-danger
        });
      } else {
        return React.createElement(
          'g',
          { role: 'presentation' },
          this.props.children
        );
      }
    }
  }, {
    key: 'render',
    value: function render() {
      var _classnames;

      var _props = this.props,
          style = _props.style,
          width = _props.width,
          height = _props.height,
          title = _props.title,
          description = _props.description,
          focusable = _props.focusable,
          children = _props.children,
          src = _props.src,
          color = _props.color,
          props = _objectWithoutProperties(_props, ['style', 'width', 'height', 'title', 'description', 'focusable', 'children', 'src', 'color']);

      return React.createElement(
        'svg',
        Object.assign({}, parseAttributes(this.props.src), omitProps(this.props, InlineSVG.propTypes, ['inline']), {
          style: Object.assign({}, style, {
            width: width,
            height: height
          }),
          width: width || '1em',
          height: height || '1em',
          'aria-hidden': title ? null : 'true',
          'aria-labelledby': this.labelledBy,
          role: this.role,
          focusable: focusable ? 'true' : 'false',
          className: classnames((_classnames = {}, _defineProperty(_classnames, styles.root, true), _defineProperty(_classnames, styles.inline, this.props.inline), _defineProperty(_classnames, styles.block, !this.props.inline), _defineProperty(_classnames, props.className, props.className), _defineProperty(_classnames, styles['color--' + color], color), _classnames))
        }),
        this.renderTitle(),
        this.renderDesc(description),
        this.renderContent()
      );
    }
  }, {
    key: 'role',
    get: function get() {
      if (this.props.title) {
        return 'img';
      } else {
        return 'presentation';
      }
    }
  }, {
    key: 'labelledBy',
    get: function get() {
      var ids = [];

      if (this.props.title) {
        ids.push(this.titleId);
      }

      if (this.props.description) {
        ids.push(this.descId);
      }

      return ids.length > 0 ? ids.join(' ') : null;
    }
  }]);

  InlineSVG.displayName = 'InlineSVG'
  ;
  return InlineSVG;
}(Component), _class2.propTypes = {
  children: PropTypes.node, // eslint-disable-line
  src: PropTypes.string,
  title: PropTypes.string,
  description: PropTypes.string,
  focusable: PropTypes.bool,
  width: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  height: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  inline: PropTypes.bool,
  color: PropTypes.oneOf(['primary', 'secondary', 'primary-inverse', 'secondary-inverse', 'success', 'error', 'warning', 'brand'])
}, _class2.defaultProps = {
  focusable: false,
  src: '',
  title: '',
  description: '',
  inline: true
}, _class2.prepareSrc = function (src) {
  var pattern = /<svg[^>]*>((.|[\n\r])*)<\/svg>/;
  var matches = pattern.exec(src);

  return matches ? matches[1] : src;
}, _temp)) || _class);
export { InlineSVG as default };


function parseAttributes(src) {
  var attributes = {};
  var SVGAttributesRegExp = /<svg\s+([^>]*)\s*>/;
  var namesAndValuesRegExp = /(\S+)=["']?((?:.(?!["']?\s+(?:\S+)=|[>"']))+.)["']?/g;

  if (typeof src === 'string') {
    var attributesMatches = SVGAttributesRegExp.exec(src);
    var attributesString = attributesMatches ? attributesMatches[1] : '';
    var excludes = ['xmlns', 'xmlns:xlink', 'version'];

    var match = namesAndValuesRegExp.exec(attributesString);

    while (match != null) {
      if (excludes.indexOf(match[1]) === -1) {
        attributes[match[1]] = match[2] || (match[3] ? match[3] : match[4] ? match[4] : match[5]) || match[1]; // eslint-disable-line no-nested-ternary, max-len
      }
      match = namesAndValuesRegExp.exec(attributesString);
    }
  }

  return attributes;
}
