// @instructure/ui-layout/lib/components/Grid/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.GridRow = exports.GridCol = exports.default = undefined;

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _dec, _class, _class2, _temp; /*
                                   * The MIT License (MIT)
                                   *
                                   * Copyright (c) 2015 - present Instructure, Inc.
                                   *
                                   * Permission is hereby granted, free of charge, to any person obtaining a copy
                                   * of this software and associated documentation files (the "Software"), to deal
                                   * in the Software without restriction, including without limitation the rights
                                   * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
                                   * copies of the Software, and to permit persons to whom the Software is
                                   * furnished to do so, subject to the following conditions:
                                   *
                                   * The above copyright notice and this permission notice shall be included in all
                                   * copies or substantial portions of the Software.
                                   *
                                   * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
                                   * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
                                   * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
                                   * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
                                   * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
                                   * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
                                   * SOFTWARE.
                                   */

var _GridCol = require('./GridCol');

Object.defineProperty(exports, 'GridCol', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_GridCol).default;
  }
});

var _GridRow = require('./GridRow');

Object.defineProperty(exports, 'GridRow', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_GridRow).default;
  }
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _uiThemeable = require('@instructure/ui-themeable');

var _uiThemeable2 = _interopRequireDefault(_uiThemeable);

var _CustomPropTypes = require('@instructure/ui-utils/lib/react/CustomPropTypes');

var _CustomPropTypes2 = _interopRequireDefault(_CustomPropTypes);

var _capitalizeFirstLetter = require('@instructure/ui-utils/lib/capitalizeFirstLetter');

var _capitalizeFirstLetter2 = _interopRequireDefault(_capitalizeFirstLetter);

var _safeCloneElement = require('@instructure/ui-utils/lib/react/safeCloneElement');

var _safeCloneElement2 = _interopRequireDefault(_safeCloneElement);

var _matchComponentTypes = require('@instructure/ui-utils/lib/react/matchComponentTypes');

var _matchComponentTypes2 = _interopRequireDefault(_matchComponentTypes);

var _passthroughProps = require('@instructure/ui-utils/lib/react/passthroughProps');

var _GridRow2 = _interopRequireDefault(_GridRow);

var _theme = require('./theme');

var _theme2 = _interopRequireDefault(_theme);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var styles = {
  template: function template(theme) {
    var tmpl = function tmpl() {
      return '/*  imported from styles.css  */\n\n/* stylelint-disable at-rule-no-unknown */\n\n.SvDtTgU {\n  display: block;\n}\n\n.MLuKPBz {\n  /* stylelint-disable-next-line color-named */\n  outline: 0.0625rem dashed red;\n}\n\n._23-g_2G {\n  box-sizing: border-box;\n}\n\n@media screen and (--mediumMin) {\n  ._291X7ul {\n    box-sizing: border-box;\n  }\n}\n\n@media screen and (--largeMin) {\n  ._1v4Hpk2 {\n    box-sizing: border-box;\n  }\n}\n\n@media screen and (--xLargeMin) {\n  .Nq3l5Lu {\n    box-sizing: border-box;\n  }\n}\n\n/* stylelint-enable at-rule-no-unknown */\n';
    };

    return tmpl.call(theme, theme);
  },
  'root': 'SvDtTgU',
  'visualDebug': 'MLuKPBz',
  'startAtSmall': '_23-g_2G',
  'startAtMedium': '_291X7ul',
  'startAtLarge': '_1v4Hpk2',
  'startAtXLarge': 'Nq3l5Lu'
};


/**
---
category: components/layout
---
**/
var Grid = (_dec = (0, _uiThemeable2.default)(_theme2.default, styles), _dec(_class = (_temp = _class2 = function (_Component) {
  _inherits(Grid, _Component);

  function Grid() {
    _classCallCheck(this, Grid);

    return _possibleConstructorReturn(this, (Grid.__proto__ || Object.getPrototypeOf(Grid)).apply(this, arguments));
  }

  _createClass(Grid, [{
    key: 'startAtClass',
    value: function startAtClass() {
      return !!this.props.startAt && 'startAt' + (0, _capitalizeFirstLetter2.default)(this.props.startAt);
    }
  }, {
    key: 'renderChildren',
    value: function renderChildren() {
      var _this2 = this;

      var children = _react.Children.toArray(this.props.children);

      return children.map(function (child, index) {
        if ((0, _matchComponentTypes2.default)(child, [_GridRow2.default])) {
          return (0, _safeCloneElement2.default)(child, Object.assign({}, (0, _passthroughProps.pickProps)(_this2.props, Grid.propTypes), child.props, { /* child props should override parent */
            isLastRow: index + 1 === children.length
          }));
        } else {
          return child; // PropType validation should handle errors
        }
      });
    }
  }, {
    key: 'render',
    value: function render() {
      var _classes;

      var classes = (_classes = {}, _defineProperty(_classes, styles.root, true), _defineProperty(_classes, styles[this.startAtClass()], !!this.props.startAt), _defineProperty(_classes, styles.visualDebug, this.props.visualDebug), _classes);

      var props = (0, _passthroughProps.omitProps)(this.props, Grid.propTypes);

      return _react2.default.createElement(
        'span',
        Object.assign({}, props, {
          className: (0, _classnames2.default)(classes)
        }),
        this.renderChildren()
      );
    }
  }]);

  Grid.displayName = 'Grid'
  ;
  return Grid;
}(_react.Component), _class2.propTypes = {
  children: _CustomPropTypes2.default.Children.oneOf([_GridRow2.default]),
  colSpacing: _propTypes2.default.oneOf(['none', 'small', 'medium', 'large']),
  rowSpacing: _propTypes2.default.oneOf(['none', 'small', 'medium', 'large']),
  hAlign: _propTypes2.default.oneOf(['start', 'center', 'end', 'space-around', 'space-between']),
  vAlign: _propTypes2.default.oneOf(['top', 'middle', 'bottom']),
  startAt: _propTypes2.default.oneOf(['small', 'medium', 'large', 'x-large', null]),
  visualDebug: _propTypes2.default.bool
}, _class2.defaultProps = {
  colSpacing: 'medium',
  rowSpacing: 'medium',
  hAlign: 'start',
  startAt: 'small',
  vAlign: 'top',
  visualDebug: false
}, _temp)) || _class);
exports.default = Grid;
