// @instructure/ui-elements/es/components/List/index.js
var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _dec, _class, _class2, _temp;

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
import React, { Children, Component } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';

import View from '@instructure/ui-layout/lib/components/View';

import themeable from '@instructure/ui-themeable';
import { omitProps } from '@instructure/ui-utils/lib/react/passthroughProps';
import CustomPropTypes from '@instructure/ui-utils/lib/react/CustomPropTypes';
import ThemeablePropTypes from '@instructure/ui-themeable/lib/utils/ThemeablePropTypes';
import safeCloneElement from '@instructure/ui-utils/lib/react/safeCloneElement';

import ListItem from './ListItem';

var styles = {
  template: function template(theme) {
    var tmpl = function tmpl() {
      return '/*  imported from styles.css  */\n\n._2l3psub {\n  font-weight: ' + theme.fontWeight + ';\n  font-family: ' + theme.fontFamily + ';\n  padding: 0;\n}\n\n._2ObuHyI {\n  -webkit-padding-start: ' + theme.defaultListPadding + ';\n          padding-inline-start: ' + theme.defaultListPadding + ';\n  -webkit-padding-end: 0;\n          padding-inline-end: 0;\n}\n\n[dir="ltr"] ._2ObuHyI {\n  padding-left: ' + theme.defaultListPadding + ';\n  padding-right: 0;\n}\n\n[dir="rtl"] ._2ObuHyI {\n  padding-right: ' + theme.defaultListPadding + ';\n  padding-left: 0;\n}\n\n._2ObuHyI._3Qy6K6m {\n    list-style-type: none;\n    -webkit-padding-start: calc(' + theme.defaultListPadding + ' / 2);\n            padding-inline-start: calc(' + theme.defaultListPadding + ' / 2);\n    -webkit-padding-end: 0;\n            padding-inline-end: 0;\n    counter-reset: ol-counter;\n    /* stylelint-enable selector-max-type */\n  }\n\n[dir="ltr"] ._2ObuHyI._3Qy6K6m {\n    padding-left: calc(' + theme.defaultListPadding + ' / 2);\n    padding-right: 0;\n  }\n\n[dir="rtl"] ._2ObuHyI._3Qy6K6m {\n    padding-right: calc(' + theme.defaultListPadding + ' / 2);\n    padding-left: 0;\n  }\n\n/* stylelint-disable selector-max-type */\n\n._2ObuHyI._3Qy6K6m li::before {\n      font-weight: ' + theme.orderedNumberFontWeight + ';\n      counter-increment: ol-counter;\n      content: counters(ol-counter, ".") ".";\n      -webkit-margin-end: ' + theme.orderedNumberMargin + ';\n              margin-inline-end: ' + theme.orderedNumberMargin + ';\n      -webkit-margin-start: 0;\n              margin-inline-start: 0;\n    }\n\n[dir="ltr"] ._2ObuHyI._3Qy6K6m li::before {\n      margin-right: ' + theme.orderedNumberMargin + ';\n      margin-left: 0;\n    }\n\n[dir="rtl"] ._2ObuHyI._3Qy6K6m li::before {\n      margin-left: ' + theme.orderedNumberMargin + ';\n      margin-right: 0;\n    }\n\n._2-2PrF9,\n._1djVotZ,\n._2d9rysY {\n  list-style-type: none;\n  -webkit-padding-start: 0;\n          padding-inline-start: 0;\n}\n\n[dir="ltr"] ._2-2PrF9,\n[dir="ltr"] ._1djVotZ,\n[dir="ltr"] ._2d9rysY {\n  padding-left: 0;\n}\n\n[dir="rtl"] ._2-2PrF9,\n[dir="rtl"] ._1djVotZ,\n[dir="rtl"] ._2d9rysY {\n  padding-right: 0;\n}\n\n._3wFfR2D {}\n';
    };

    return tmpl.call(theme, theme);
  },
  'root': '_2l3psub',
  'default': '_2ObuHyI',
  'ordered': '_3Qy6K6m',
  'unstyled': '_2-2PrF9',
  'inline': '_1djVotZ',
  'steps': '_2d9rysY',
  'unordered': '_3wFfR2D'
};

import theme from './theme';

/**
---
category: components/typography
---
**/
var List = (_dec = themeable(theme, styles), _dec(_class = (_temp = _class2 = function (_Component) {
  _inherits(List, _Component);

  function List() {
    _classCallCheck(this, List);

    return _possibleConstructorReturn(this, (List.__proto__ || Object.getPrototypeOf(List)).apply(this, arguments));
  }

  _createClass(List, [{
    key: 'renderChildren',
    value: function renderChildren() {
      var _this2 = this;

      return Children.map(this.props.children, function (child) {
        if (!child) return; // ignore null, falsy children

        return safeCloneElement(child, {
          variant: _this2.props.variant,
          delimiter: _this2.props.delimiter,
          size: _this2.props.size,
          as: _this2.props.as,
          spacing: _this2.props.itemSpacing
        });
      });
    }
  }, {
    key: 'render',
    value: function render() {
      var _classes;

      var props = omitProps(this.props, Object.assign({}, List.propTypes, View.propTypes));

      var _props = this.props,
          as = _props.as,
          margin = _props.margin,
          variant = _props.variant,
          elementRef = _props.elementRef;


      var classes = (_classes = {}, _defineProperty(_classes, styles.root, true), _defineProperty(_classes, styles[variant], variant), _defineProperty(_classes, styles.unordered, as === 'ul'), _defineProperty(_classes, styles.ordered, as === 'ol'), _classes);

      return React.createElement(
        View,
        Object.assign({}, props, {
          className: classnames(classes),
          as: as,
          margin: margin,
          elementRef: elementRef
        }),
        this.renderChildren()
      );
    }
  }]);

  List.displayName = 'List'
  ;
  return List;
}(Component), _class2.propTypes = {
  /**
  * Only accepts <ListItem> as a child
  */
  children: CustomPropTypes.Children.oneOf([ListItem]),
  as: PropTypes.oneOf(['ul', 'ol']),
  /**
  * Valid values are `0`, `none`, `auto`, `xxx-small`, `xx-small`, `x-small`,
  * `small`, `medium`, `large`, `x-large`, `xx-large`. Apply these values via
  * familiar CSS-like shorthand. For example: `margin="small auto large"`.
  */
  margin: ThemeablePropTypes.spacing,
  size: PropTypes.oneOf(['small', 'medium', 'large']),
  variant: PropTypes.oneOf(['default', 'unstyled', 'inline']),
  delimiter: PropTypes.oneOf(['none', 'pipe', 'slash', 'arrow', 'dashed', 'solid']),
  /**
  * Sets the margin separating each ListItem.
  */
  itemSpacing: PropTypes.oneOf(['xxx-small', 'xx-small', 'x-small', 'small', 'medium', 'large', 'x-large', 'xx-large']),
  elementRef: PropTypes.func
}, _class2.defaultProps = {
  as: 'ul',
  margin: 'none',
  variant: 'default',
  delimiter: 'none',
  size: 'medium'
}, _temp)) || _class);
export { List as default };


export { default as ListItem } from './ListItem';
