// @instructure/ui-forms/lib/components/FormFieldGroup/__examples__/FormFieldGroup.examples.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.startAtXLarge = exports.startAtLarge = exports.startAtMedium = exports.startAtSmall = exports.colSpacingLarge = exports.colSpacingMedium = exports.colSpacingSmall = exports.rowSpacingLarge = exports.rowSpacingMedium = exports.rowSpacingSmall = exports.success = exports.hint = exports.error = exports.alignBottom = exports.alignMiddle = exports.alignTop = exports.columns = exports.stacked = exports.inline = undefined;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _TextInput = require('../../../../../ui-forms/lib/components/TextInput');

var _TextInput2 = _interopRequireDefault(_TextInput);

var _Select = require('../../../../../ui-forms/lib/components/Select');

var _Select2 = _interopRequireDefault(_Select);

var _Checkbox = require('../../../../../ui-forms/lib/components/Checkbox');

var _Checkbox2 = _interopRequireDefault(_Checkbox);

var _CheckboxGroup = require('../../../../../ui-forms/lib/components/CheckboxGroup');

var _CheckboxGroup2 = _interopRequireDefault(_CheckboxGroup);

var _RadioInput = require('../../../../../ui-forms/lib/components/RadioInput');

var _RadioInput2 = _interopRequireDefault(_RadioInput);

var _RadioInputGroup = require('../../../../../ui-forms/lib/components/RadioInputGroup');

var _RadioInputGroup2 = _interopRequireDefault(_RadioInputGroup);

var _index = require('../index');

var _index2 = _interopRequireDefault(_index);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

var inline = exports.inline = function inline() {
  return _react2.default.createElement(
    _index2.default,
    {
      description: 'Breakfast',
      layout: 'inline'
    },
    _react2.default.createElement(_TextInput2.default, { label: 'Favorite Breakfast Eatery',
      messages: [{ text: 'Invalid name', type: 'error' }]
    }),
    _react2.default.createElement(
      _Select2.default,
      { label: 'Select Your Favorite Side Dish' },
      _react2.default.createElement(
        'option',
        { value: 'apples' },
        'Fresh Fruit'
      ),
      _react2.default.createElement(
        'option',
        { value: 'oranges' },
        'Biscuits and Gravy'
      ),
      _react2.default.createElement(
        'option',
        { value: 'bananas' },
        'Bacon'
      ),
      _react2.default.createElement(
        'option',
        { value: 'watermelon' },
        'English Muffins'
      ),
      _react2.default.createElement(
        'option',
        { value: 'pancakes' },
        'Pancakes'
      )
    ),
    _react2.default.createElement(
      _RadioInputGroup2.default,
      {
        name: 'beverage',
        description: 'Beverage of Choice',
        defaultValue: 'coffee',
        layout: 'columns'
      },
      _react2.default.createElement(_RadioInput2.default, { label: 'Juice', value: 'juice' }),
      _react2.default.createElement(_RadioInput2.default, { label: 'Water', value: 'water' }),
      _react2.default.createElement(_RadioInput2.default, { label: 'Coffee', value: 'coffee' }),
      _react2.default.createElement(_RadioInput2.default, { label: 'Milk', value: 'milk' }),
      _react2.default.createElement(_RadioInput2.default, { label: 'Soda', value: 'soda' }),
      _react2.default.createElement(_RadioInput2.default, { label: 'Hot Tea', value: 'tea' })
    )
  );
};

var stacked = exports.stacked = function stacked() {
  return _react2.default.createElement(
    _index2.default,
    {
      description: 'Morning Preferences',
      layout: 'stacked'
    },
    _react2.default.createElement(_TextInput2.default, { label: 'Favorite Breakfast Eatery',
      messages: [{ text: 'Invalid name', type: 'error' }]
    }),
    _react2.default.createElement(
      _Select2.default,
      { label: 'Select Your Favorite Side Dish' },
      _react2.default.createElement(
        'option',
        { value: 'apples' },
        'Fresh Fruit'
      ),
      _react2.default.createElement(
        'option',
        { value: 'oranges' },
        'Biscuits and Gravy'
      ),
      _react2.default.createElement(
        'option',
        { value: 'bananas' },
        'Bacon'
      ),
      _react2.default.createElement(
        'option',
        { value: 'watermelon' },
        'English Muffins'
      ),
      _react2.default.createElement(
        'option',
        { value: 'pancakes' },
        'Pancakes'
      )
    ),
    _react2.default.createElement(
      _CheckboxGroup2.default,
      {
        name: 'beverage',
        description: 'Favorite Beverages',
        layout: 'columns'
      },
      _react2.default.createElement(_Checkbox2.default, { label: 'Juice', value: 'juice' }),
      _react2.default.createElement(_Checkbox2.default, { label: 'Water', value: 'water' }),
      _react2.default.createElement(_Checkbox2.default, { label: 'Coffee', value: 'coffee' }),
      _react2.default.createElement(_Checkbox2.default, { label: 'Milk', value: 'milk' }),
      _react2.default.createElement(_Checkbox2.default, { label: 'Soda', value: 'soda' })
    )
  );
};

var columns = exports.columns = function columns() {
  return _react2.default.createElement(
    _index2.default,
    {
      description: 'Breakfast',
      layout: 'columns',
      vAlign: 'top'
    },
    _react2.default.createElement(_TextInput2.default, { label: 'Favorite Breakfast Eatery',
      messages: [{ text: 'Invalid name', type: 'error' }]
    }),
    _react2.default.createElement(
      _Select2.default,
      { label: 'Select Your Favorite Side Dish' },
      _react2.default.createElement(
        'option',
        { value: 'apples' },
        'Fresh Fruit'
      ),
      _react2.default.createElement(
        'option',
        { value: 'oranges' },
        'Biscuits and Gravy'
      ),
      _react2.default.createElement(
        'option',
        { value: 'bananas' },
        'Bacon'
      ),
      _react2.default.createElement(
        'option',
        { value: 'watermelon' },
        'English Muffins'
      ),
      _react2.default.createElement(
        'option',
        { value: 'pancakes' },
        'Pancakes'
      )
    ),
    _react2.default.createElement(
      _RadioInputGroup2.default,
      {
        name: 'beverage',
        description: 'Beverage of Choice',
        defaultValue: 'coffee',
        layout: 'columns'
      },
      _react2.default.createElement(_RadioInput2.default, { label: 'Juice', value: 'juice' }),
      _react2.default.createElement(_RadioInput2.default, { label: 'Water', value: 'water' }),
      _react2.default.createElement(_RadioInput2.default, { label: 'Coffee', value: 'coffee' }),
      _react2.default.createElement(_RadioInput2.default, { label: 'Milk', value: 'milk' }),
      _react2.default.createElement(_RadioInput2.default, { label: 'Soda', value: 'soda' }),
      _react2.default.createElement(_RadioInput2.default, { label: 'Hot Tea', value: 'tea' })
    )
  );
};

var alignTop = exports.alignTop = function alignTop() {
  return _react2.default.createElement(
    _index2.default,
    {
      description: 'Breakfast',
      layout: 'inline',
      vAlign: 'top'
    },
    _react2.default.createElement(_TextInput2.default, { label: 'Favorite Breakfast Eatery' }),
    _react2.default.createElement(
      _Select2.default,
      { label: 'Select Your Favorite Side Dish' },
      _react2.default.createElement(
        'option',
        { value: 'apples' },
        'Fresh Fruit'
      ),
      _react2.default.createElement(
        'option',
        { value: 'oranges' },
        'Biscuits and Gravy'
      ),
      _react2.default.createElement(
        'option',
        { value: 'bananas' },
        'Bacon'
      ),
      _react2.default.createElement(
        'option',
        { value: 'watermelon' },
        'English Muffins'
      ),
      _react2.default.createElement(
        'option',
        { value: 'pancakes' },
        'Pancakes'
      )
    )
  );
};

var alignMiddle = exports.alignMiddle = function alignMiddle() {
  return _react2.default.createElement(
    _index2.default,
    {
      description: 'Breakfast',
      layout: 'inline'
    },
    _react2.default.createElement(_TextInput2.default, { label: 'Favorite Breakfast Eatery' }),
    _react2.default.createElement(
      _Select2.default,
      { label: 'Select Your Favorite Side Dish' },
      _react2.default.createElement(
        'option',
        { value: 'apples' },
        'Fresh Fruit'
      ),
      _react2.default.createElement(
        'option',
        { value: 'oranges' },
        'Biscuits and Gravy'
      ),
      _react2.default.createElement(
        'option',
        { value: 'bananas' },
        'Bacon'
      ),
      _react2.default.createElement(
        'option',
        { value: 'watermelon' },
        'English Muffins'
      ),
      _react2.default.createElement(
        'option',
        { value: 'pancakes' },
        'Pancakes'
      )
    )
  );
};

var alignBottom = exports.alignBottom = function alignBottom() {
  return _react2.default.createElement(
    _index2.default,
    {
      description: 'Breakfast',
      layout: 'inline',
      vAlign: 'bottom'
    },
    _react2.default.createElement(_TextInput2.default, { label: 'Favorite Breakfast Eatery' }),
    _react2.default.createElement(
      _Select2.default,
      { label: 'Select Your Favorite Side Dish' },
      _react2.default.createElement(
        'option',
        { value: 'apples' },
        'Fresh Fruit'
      ),
      _react2.default.createElement(
        'option',
        { value: 'oranges' },
        'Biscuits and Gravy'
      ),
      _react2.default.createElement(
        'option',
        { value: 'bananas' },
        'Bacon'
      ),
      _react2.default.createElement(
        'option',
        { value: 'watermelon' },
        'English Muffins'
      ),
      _react2.default.createElement(
        'option',
        { value: 'pancakes' },
        'Pancakes'
      )
    )
  );
};

var error = exports.error = function error() {
  return _react2.default.createElement(
    _index2.default,
    {
      description: 'Breakfast',
      layout: 'columns',
      vAlign: 'top',
      messages: [{ text: 'You must choose', type: 'error' }]
    },
    _react2.default.createElement(_TextInput2.default, { label: 'Favorite Breakfast Eatery' }),
    _react2.default.createElement(
      _Select2.default,
      { label: 'Select Your Favorite Side Dish' },
      _react2.default.createElement(
        'option',
        { value: 'apples' },
        'Fresh Fruit'
      ),
      _react2.default.createElement(
        'option',
        { value: 'oranges' },
        'Biscuits and Gravy'
      ),
      _react2.default.createElement(
        'option',
        { value: 'bananas' },
        'Bacon'
      ),
      _react2.default.createElement(
        'option',
        { value: 'watermelon' },
        'English Muffins'
      ),
      _react2.default.createElement(
        'option',
        { value: 'pancakes' },
        'Pancakes'
      )
    )
  );
};

var hint = exports.hint = function hint() {
  return _react2.default.createElement(
    _index2.default,
    {
      description: 'Breakfast',
      layout: 'columns',
      vAlign: 'top',
      messages: [{ text: 'This will help us choose a caterer', type: 'hint' }]
    },
    _react2.default.createElement(_TextInput2.default, { label: 'Favorite Breakfast Eatery' }),
    _react2.default.createElement(
      _Select2.default,
      { label: 'Select Your Favorite Side Dish' },
      _react2.default.createElement(
        'option',
        { value: 'apples' },
        'Fresh Fruit'
      ),
      _react2.default.createElement(
        'option',
        { value: 'oranges' },
        'Biscuits and Gravy'
      ),
      _react2.default.createElement(
        'option',
        { value: 'bananas' },
        'Bacon'
      ),
      _react2.default.createElement(
        'option',
        { value: 'watermelon' },
        'English Muffins'
      ),
      _react2.default.createElement(
        'option',
        { value: 'pancakes' },
        'Pancakes'
      )
    )
  );
};

var success = exports.success = function success() {
  return _react2.default.createElement(
    _index2.default,
    {
      description: 'Breakfast',
      layout: 'columns',
      vAlign: 'top',
      messages: [{ text: 'Thanks for participating', type: 'success' }]
    },
    _react2.default.createElement(_TextInput2.default, { label: 'Favorite Breakfast Eatery' }),
    _react2.default.createElement(
      _Select2.default,
      { label: 'Select Your Favorite Side Dish' },
      _react2.default.createElement(
        'option',
        { value: 'apples' },
        'Fresh Fruit'
      ),
      _react2.default.createElement(
        'option',
        { value: 'oranges' },
        'Biscuits and Gravy'
      ),
      _react2.default.createElement(
        'option',
        { value: 'bananas' },
        'Bacon'
      ),
      _react2.default.createElement(
        'option',
        { value: 'watermelon' },
        'English Muffins'
      ),
      _react2.default.createElement(
        'option',
        { value: 'pancakes' },
        'Pancakes'
      )
    )
  );
};

var rowSpacingSmall = exports.rowSpacingSmall = function rowSpacingSmall() {
  return _react2.default.createElement(
    _index2.default,
    {
      description: 'Breakfast',
      layout: 'inline',
      rowSpacing: 'small'
    },
    _react2.default.createElement(_TextInput2.default, { label: 'Favorite Breakfast Eatery' }),
    _react2.default.createElement(
      _Select2.default,
      { label: 'Select Your Favorite Side Dish' },
      _react2.default.createElement(
        'option',
        { value: 'apples' },
        'Fresh Fruit'
      ),
      _react2.default.createElement(
        'option',
        { value: 'oranges' },
        'Biscuits and Gravy'
      ),
      _react2.default.createElement(
        'option',
        { value: 'bananas' },
        'Bacon'
      ),
      _react2.default.createElement(
        'option',
        { value: 'watermelon' },
        'English Muffins'
      ),
      _react2.default.createElement(
        'option',
        { value: 'pancakes' },
        'Pancakes'
      )
    )
  );
};

var rowSpacingMedium = exports.rowSpacingMedium = function rowSpacingMedium() {
  return _react2.default.createElement(
    _index2.default,
    {
      description: 'Breakfast',
      layout: 'inline',
      rowSpacing: 'medium'
    },
    _react2.default.createElement(_TextInput2.default, { label: 'Favorite Breakfast Eatery' }),
    _react2.default.createElement(
      _Select2.default,
      { label: 'Select Your Favorite Side Dish' },
      _react2.default.createElement(
        'option',
        { value: 'apples' },
        'Fresh Fruit'
      ),
      _react2.default.createElement(
        'option',
        { value: 'oranges' },
        'Biscuits and Gravy'
      ),
      _react2.default.createElement(
        'option',
        { value: 'bananas' },
        'Bacon'
      ),
      _react2.default.createElement(
        'option',
        { value: 'watermelon' },
        'English Muffins'
      ),
      _react2.default.createElement(
        'option',
        { value: 'pancakes' },
        'Pancakes'
      )
    )
  );
};

var rowSpacingLarge = exports.rowSpacingLarge = function rowSpacingLarge() {
  return _react2.default.createElement(
    _index2.default,
    {
      description: 'Breakfast',
      layout: 'inline',
      rowSpacing: 'large'
    },
    _react2.default.createElement(_TextInput2.default, { label: 'Favorite Breakfast Eatery' }),
    _react2.default.createElement(
      _Select2.default,
      { label: 'Select Your Favorite Side Dish' },
      _react2.default.createElement(
        'option',
        { value: 'apples' },
        'Fresh Fruit'
      ),
      _react2.default.createElement(
        'option',
        { value: 'oranges' },
        'Biscuits and Gravy'
      ),
      _react2.default.createElement(
        'option',
        { value: 'bananas' },
        'Bacon'
      ),
      _react2.default.createElement(
        'option',
        { value: 'watermelon' },
        'English Muffins'
      ),
      _react2.default.createElement(
        'option',
        { value: 'pancakes' },
        'Pancakes'
      )
    )
  );
};

var colSpacingSmall = exports.colSpacingSmall = function colSpacingSmall() {
  return _react2.default.createElement(
    _index2.default,
    {
      description: 'Breakfast',
      layout: 'columns',
      colSpacing: 'small'
    },
    _react2.default.createElement(_TextInput2.default, { label: 'Favorite Breakfast Eatery' }),
    _react2.default.createElement(
      _Select2.default,
      { label: 'Select Your Favorite Side Dish' },
      _react2.default.createElement(
        'option',
        { value: 'apples' },
        'Fresh Fruit'
      ),
      _react2.default.createElement(
        'option',
        { value: 'oranges' },
        'Biscuits and Gravy'
      ),
      _react2.default.createElement(
        'option',
        { value: 'bananas' },
        'Bacon'
      ),
      _react2.default.createElement(
        'option',
        { value: 'watermelon' },
        'English Muffins'
      ),
      _react2.default.createElement(
        'option',
        { value: 'pancakes' },
        'Pancakes'
      )
    )
  );
};

var colSpacingMedium = exports.colSpacingMedium = function colSpacingMedium() {
  return _react2.default.createElement(
    _index2.default,
    {
      description: 'Breakfast',
      layout: 'columns',
      colSpacing: 'medium'
    },
    _react2.default.createElement(_TextInput2.default, { label: 'Favorite Breakfast Eatery' }),
    _react2.default.createElement(
      _Select2.default,
      { label: 'Select Your Favorite Side Dish' },
      _react2.default.createElement(
        'option',
        { value: 'apples' },
        'Fresh Fruit'
      ),
      _react2.default.createElement(
        'option',
        { value: 'oranges' },
        'Biscuits and Gravy'
      ),
      _react2.default.createElement(
        'option',
        { value: 'bananas' },
        'Bacon'
      ),
      _react2.default.createElement(
        'option',
        { value: 'watermelon' },
        'English Muffins'
      ),
      _react2.default.createElement(
        'option',
        { value: 'pancakes' },
        'Pancakes'
      )
    )
  );
};

var colSpacingLarge = exports.colSpacingLarge = function colSpacingLarge() {
  return _react2.default.createElement(
    _index2.default,
    {
      description: 'Breakfast',
      layout: 'columns',
      colSpacing: 'large'
    },
    _react2.default.createElement(_TextInput2.default, { label: 'Favorite Breakfast Eatery' }),
    _react2.default.createElement(
      _Select2.default,
      { label: 'Select Your Favorite Side Dish' },
      _react2.default.createElement(
        'option',
        { value: 'apples' },
        'Fresh Fruit'
      ),
      _react2.default.createElement(
        'option',
        { value: 'oranges' },
        'Biscuits and Gravy'
      ),
      _react2.default.createElement(
        'option',
        { value: 'bananas' },
        'Bacon'
      ),
      _react2.default.createElement(
        'option',
        { value: 'watermelon' },
        'English Muffins'
      ),
      _react2.default.createElement(
        'option',
        { value: 'pancakes' },
        'Pancakes'
      )
    )
  );
};

var startAtSmall = exports.startAtSmall = function startAtSmall() {
  return _react2.default.createElement(
    _index2.default,
    {
      description: 'Breakfast',
      layout: 'columns',
      startAt: 'small'
    },
    _react2.default.createElement(_TextInput2.default, { label: 'Favorite Breakfast Eatery' }),
    _react2.default.createElement(
      _Select2.default,
      { label: 'Select Your Favorite Side Dish' },
      _react2.default.createElement(
        'option',
        { value: 'apples' },
        'Fresh Fruit'
      ),
      _react2.default.createElement(
        'option',
        { value: 'oranges' },
        'Biscuits and Gravy'
      ),
      _react2.default.createElement(
        'option',
        { value: 'bananas' },
        'Bacon'
      ),
      _react2.default.createElement(
        'option',
        { value: 'watermelon' },
        'English Muffins'
      ),
      _react2.default.createElement(
        'option',
        { value: 'pancakes' },
        'Pancakes'
      )
    )
  );
};

var startAtMedium = exports.startAtMedium = function startAtMedium() {
  return _react2.default.createElement(
    _index2.default,
    {
      description: 'Breakfast',
      layout: 'columns',
      startAt: 'medium'
    },
    _react2.default.createElement(_TextInput2.default, { label: 'Favorite Breakfast Eatery' }),
    _react2.default.createElement(
      _Select2.default,
      { label: 'Select Your Favorite Side Dish' },
      _react2.default.createElement(
        'option',
        { value: 'apples' },
        'Fresh Fruit'
      ),
      _react2.default.createElement(
        'option',
        { value: 'oranges' },
        'Biscuits and Gravy'
      ),
      _react2.default.createElement(
        'option',
        { value: 'bananas' },
        'Bacon'
      ),
      _react2.default.createElement(
        'option',
        { value: 'watermelon' },
        'English Muffins'
      ),
      _react2.default.createElement(
        'option',
        { value: 'pancakes' },
        'Pancakes'
      )
    )
  );
};

var startAtLarge = exports.startAtLarge = function startAtLarge() {
  return _react2.default.createElement(
    _index2.default,
    {
      description: 'Breakfast',
      layout: 'columns',
      startAt: 'large'
    },
    _react2.default.createElement(_TextInput2.default, { label: 'Favorite Breakfast Eatery' }),
    _react2.default.createElement(
      _Select2.default,
      { label: 'Select Your Favorite Side Dish' },
      _react2.default.createElement(
        'option',
        { value: 'apples' },
        'Fresh Fruit'
      ),
      _react2.default.createElement(
        'option',
        { value: 'oranges' },
        'Biscuits and Gravy'
      ),
      _react2.default.createElement(
        'option',
        { value: 'bananas' },
        'Bacon'
      ),
      _react2.default.createElement(
        'option',
        { value: 'watermelon' },
        'English Muffins'
      ),
      _react2.default.createElement(
        'option',
        { value: 'pancakes' },
        'Pancakes'
      )
    )
  );
};

var startAtXLarge = exports.startAtXLarge = function startAtXLarge() {
  return _react2.default.createElement(
    _index2.default,
    {
      description: 'Breakfast',
      layout: 'columns',
      startAt: 'x-large'
    },
    _react2.default.createElement(_TextInput2.default, { label: 'Favorite Breakfast Eatery' }),
    _react2.default.createElement(
      _Select2.default,
      { label: 'Select Your Favorite Side Dish' },
      _react2.default.createElement(
        'option',
        { value: 'apples' },
        'Fresh Fruit'
      ),
      _react2.default.createElement(
        'option',
        { value: 'oranges' },
        'Biscuits and Gravy'
      ),
      _react2.default.createElement(
        'option',
        { value: 'bananas' },
        'Bacon'
      ),
      _react2.default.createElement(
        'option',
        { value: 'watermelon' },
        'English Muffins'
      ),
      _react2.default.createElement(
        'option',
        { value: 'pancakes' },
        'Pancakes'
      )
    )
  );
};
