// @instructure/ui-forms/es/components/Checkbox/CheckboxFacade/index.js
var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _dec, _class, _class2, _temp;

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import SVGIcon from '@instructure/ui-svg-images/lib/components/SVGIcon';
import IconCheckMark from '@instructure/ui-icons/lib/Solid/IconCheckMark';
import themeable from '@instructure/ui-themeable';

var styles = {
  template: function template(theme) {
    var tmpl = function tmpl() {
      return '/*  imported from styles.css  */\n\n._3MVMN7s {\n  display: flex;\n  align-items: flex-start;\n}\n\n._3vZ9FB9 {\n  color: ' + theme.color + ';\n  background: ' + theme.background + ';\n  position: relative;\n  display: flex;\n  align-items: center;\n  justify-content: center;\n  box-sizing: border-box;\n  flex-shrink: 0;\n  transition: all 0.2s;\n  border: ' + theme.borderWidth + ' solid ' + theme.borderColor + ';\n  border-radius: ' + theme.borderRadius + ';\n  -webkit-margin-end: ' + theme.marginRight + ';\n          margin-inline-end: ' + theme.marginRight + ';\n  -webkit-margin-start: 0;\n          margin-inline-start: 0;\n  padding: ' + theme.padding + ';\n}\n\n[dir="ltr"] ._3vZ9FB9 {\n  margin-right: ' + theme.marginRight + ';\n  margin-left: 0;\n}\n\n[dir="rtl"] ._3vZ9FB9 {\n  margin-left: ' + theme.marginRight + ';\n  margin-right: 0;\n}\n\n._3vZ9FB9::before {\n    content: "";\n    position: absolute;\n    top: -0.375rem;\n    left: -0.375rem; /* stylelint-disable-line property-blacklist */\n    width: calc(100% + 0.75rem);\n    height: calc(100% + 0.75rem);\n    box-sizing: border-box;\n    border-radius: calc(' + theme.borderRadius + ' + 0.1875rem);\n    border: ' + theme.focusBorderWidth + ' ' + theme.focusBorderStyle + ' ' + theme.focusBorderColor + ';\n    transition: all 0.2s;\n    transform: scale(0.75) translate3d(0, 0, 0);\n    opacity: 0;\n    pointer-events: none;\n  }\n\n._3vfD8yo {\n  flex: 1 1 auto;\n  min-width: 0.0625rem;\n  color: ' + theme.labelColor + ';\n  font-family: ' + theme.labelFontFamily + ';\n  font-weight: ' + theme.labelFontWeight + ';\n  line-height: ' + theme.labelLineHeight + ';\n}\n\n._39Ri4Ab ._3vfD8yo {\n    font-size: ' + theme.labelFontSizeSmall + ';\n  }\n\n._39Ri4Ab ._3vZ9FB9 {\n    font-size: ' + theme.iconSizeSmall + ';\n    width: ' + theme.facadeSizeSmall + ';\n    height: ' + theme.facadeSizeSmall + ';\n  }\n\n._1QfHprh ._3vfD8yo {\n    font-size: ' + theme.labelFontSizeMedium + ';\n  }\n\n._1QfHprh ._3vZ9FB9 {\n    font-size: ' + theme.iconSizeMedium + ';\n    width: ' + theme.facadeSizeMedium + ';\n    height: ' + theme.facadeSizeMedium + ';\n  }\n\n._3QImKqs ._3vfD8yo {\n    font-size: ' + theme.labelFontSizeLarge + ';\n  }\n\n._3QImKqs ._3vZ9FB9 {\n    font-size: ' + theme.iconSizeLarge + ';\n    width: ' + theme.facadeSizeLarge + ';\n    height: ' + theme.facadeSizeLarge + ';\n  }\n\n.bOu16_B ._3vZ9FB9 {\n    background: ' + theme.checkedBackground + ';\n    border-color: ' + theme.checkedBorderColor + ';\n  }\n\n.bOu16_B ._3vfD8yo {\n    color: ' + theme.checkedLabelColor + ';\n  }\n\n._29nSP9v ._3vZ9FB9::before {\n      transform: scale(1) translate3d(0, 0, 0);\n      opacity: 1;\n    }\n\n._12EmQtH ._3vZ9FB9 {\n    border-color: ' + theme.hoverBorderColor + ';\n  }\n';
    };

    return tmpl.call(theme, theme);
  },
  'root': '_3MVMN7s',
  'facade': '_3vZ9FB9',
  'label': '_3vfD8yo',
  'small': '_39Ri4Ab',
  'medium': '_1QfHprh',
  'large': '_3QImKqs',
  'checked': 'bOu16_B',
  'focused': '_29nSP9v',
  'hovered': '_12EmQtH'
};

import theme from './theme';

/**
---
parent: Checkbox
---
**/
var CheckboxFacade = (_dec = themeable(theme, styles), _dec(_class = (_temp = _class2 = function (_Component) {
  _inherits(CheckboxFacade, _Component);

  function CheckboxFacade() {
    _classCallCheck(this, CheckboxFacade);

    return _possibleConstructorReturn(this, (CheckboxFacade.__proto__ || Object.getPrototypeOf(CheckboxFacade)).apply(this, arguments));
  }

  _createClass(CheckboxFacade, [{
    key: 'renderIcon',
    value: function renderIcon() {
      if (this.props.indeterminate) {
        return React.createElement(
          SVGIcon,
          { viewBox: '0 0 1920 1920', inline: false },
          React.createElement('rect', { x: '140', y: '820', width: '1640', height: '280' })
        );
      } else if (this.props.checked) {
        return React.createElement(IconCheckMark, { inline: false });
      } else {
        return null;
      }
    }
  }, {
    key: 'render',
    value: function render() {
      var _classes;

      var _props = this.props,
          size = _props.size,
          checked = _props.checked,
          focused = _props.focused,
          hovered = _props.hovered,
          indeterminate = _props.indeterminate;


      var classes = (_classes = {}, _defineProperty(_classes, styles.root, true), _defineProperty(_classes, styles.checked, checked || indeterminate), _defineProperty(_classes, styles.focused, focused), _defineProperty(_classes, styles.hovered, hovered), _defineProperty(_classes, styles[size], true), _classes);

      return React.createElement(
        'span',
        { className: classnames(classes) },
        React.createElement(
          'span',
          { className: styles.facade, 'aria-hidden': 'true' },
          this.renderIcon()
        ),
        React.createElement(
          'span',
          { className: styles.label },
          this.props.children
        )
      );
    }
  }]);

  CheckboxFacade.displayName = 'CheckboxFacade'
  ;
  return CheckboxFacade;
}(Component), _class2.propTypes = {
  children: PropTypes.node.isRequired,
  checked: PropTypes.bool,
  focused: PropTypes.bool,
  hovered: PropTypes.bool,
  size: PropTypes.oneOf(['small', 'medium', 'large']),
  /**
  * Visual state showing that child checkboxes are a combination of checked and unchecked
  */
  indeterminate: PropTypes.bool
}, _class2.defaultProps = {
  checked: false,
  focused: false,
  hovered: false,
  size: 'medium'
}, _temp)) || _class);
export { CheckboxFacade as default };
