// @instructure/ui-elements/lib/components/Table/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _dec, _dec2, _class, _class2, _temp; /*
                                          * The MIT License (MIT)
                                          *
                                          * Copyright (c) 2015 - present Instructure, Inc.
                                          *
                                          * Permission is hereby granted, free of charge, to any person obtaining a copy
                                          * of this software and associated documentation files (the "Software"), to deal
                                          * in the Software without restriction, including without limitation the rights
                                          * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
                                          * copies of the Software, and to permit persons to whom the Software is
                                          * furnished to do so, subject to the following conditions:
                                          *
                                          * The above copyright notice and this permission notice shall be included in all
                                          * copies or substantial portions of the Software.
                                          *
                                          * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
                                          * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
                                          * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
                                          * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
                                          * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
                                          * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
                                          * SOFTWARE.
                                          */


var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _View = require('@instructure/ui-layout/lib/components/View');

var _View2 = _interopRequireDefault(_View);

var _uiThemeable = require('@instructure/ui-themeable');

var _uiThemeable2 = _interopRequireDefault(_uiThemeable);

var _ThemeablePropTypes = require('@instructure/ui-themeable/lib/utils/ThemeablePropTypes');

var _ThemeablePropTypes2 = _interopRequireDefault(_ThemeablePropTypes);

var _passthroughProps = require('@instructure/ui-utils/lib/react/passthroughProps');

var _deprecated = require('@instructure/ui-utils/lib/react/deprecated');

var _deprecated2 = _interopRequireDefault(_deprecated);

var _theme = require('./theme');

var _theme2 = _interopRequireDefault(_theme);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var styles = {
  template: function template(theme) {
    var tmpl = function tmpl() {
      return '/*  imported from styles.css  */\n\n/* stylelint-disable selector-max-type, selector-max-class, no-descending-specificity */\n\n._1rU9nYN {\n  display: table;\n  width: 100%;\n  max-width: 100%;\n  border-collapse: collapse;\n  border-spacing: 0;\n  font-family: ' + theme.fontFamily + ';\n  font-weight: ' + theme.fontWeight + ';\n  font-size: ' + theme.fontSize + ';\n  color: ' + theme.color + ';\n  background: ' + theme.background + ';\n}\n\n._1rU9nYN caption {\n    text-align: start;\n    background: ' + theme.background + ';\n    font-size: ' + theme.captionFontSize + ';\n    color: ' + theme.captionColor + ';\n  }\n\n[dir="ltr"] ._1rU9nYN caption {\n    text-align: left;\n  }\n\n[dir="rtl"] ._1rU9nYN caption {\n    text-align: right;\n  }\n\n._1rU9nYN tr {\n    box-sizing: border-box;\n  }\n\n._1rU9nYN th,\n  ._1rU9nYN td {\n    text-align: start;\n    vertical-align: middle;\n    border-bottom-style: solid;\n    border-bottom-width: 0.0625rem;\n    box-sizing: border-box;\n    border-bottom-color: ' + theme.borderColor + ';\n  }\n\n[dir="ltr"] ._1rU9nYN th,\n[dir="ltr"] ._1rU9nYN td {\n    text-align: left;\n  }\n\n[dir="rtl"] ._1rU9nYN th,\n[dir="rtl"] ._1rU9nYN td {\n    text-align: right;\n  }\n\n._1rU9nYN th {\n    font-weight: bold;\n  }\n\n._1rU9nYN thead > th {\n    border-bottom-color: ' + theme.headerBorderColor + ';\n  }\n\n.HgJcqDO {\n  font-size: ' + theme.smallFontSize + ';\n}\n\n.HgJcqDO th,\n  .HgJcqDO td {\n    line-height: ' + theme.smallLineHeight + ';\n    padding: ' + theme.smallPadding + ';\n  }\n\n._2tLGfHy {\n  font-size: ' + theme.mediumFontSize + ';\n}\n\n._2tLGfHy th,\n  ._2tLGfHy td {\n    line-height: ' + theme.mediumLineHeight + ';\n    padding: ' + theme.mediumPadding + ';\n  }\n\n.tG9yk9Z {\n  font-size: ' + theme.largeFontSize + ';\n}\n\n.tG9yk9Z th,\n  .tG9yk9Z td {\n    line-height: ' + theme.largeLineHeight + ';\n    padding: ' + theme.largePadding + ';\n  }\n\n._2uASuAY > tbody > tr > td:nth-child(even) {\n    background-color: ' + theme.stripedBackground + ';\n  }\n\n._16Tar1y > tbody > tr:nth-child(odd) {\n    background-color: ' + theme.stripedBackground + ';\n  }\n\n._2ocxFcU > tbody > tr {\n    border-left: 0.1875rem solid transparent; /* stylelint-disable-line property-blacklist */\n    border-right: 0.1875rem solid transparent;\n  }\n\n/* stylelint-disable-line property-blacklist */\n\n._2ocxFcU > tbody > tr:hover {\n      border-color: ' + theme.hoverBorderColor + ';\n    }\n\n/* stylelint-enable selector-max-type, selector-max-class, no-descending-specificity */\n';
    };

    return tmpl.call(theme, theme);
  },
  'root': '_1rU9nYN',
  'small': 'HgJcqDO',
  'medium': '_2tLGfHy',
  'large': 'tG9yk9Z',
  'colStriped': '_2uASuAY',
  'rowStriped': '_16Tar1y',
  'hover': '_2ocxFcU'
};


/**
---
category: components
---
**/

var Table = (_dec = (0, _deprecated2.default)('3.0.0', {
  tableData: true,
  rowHeaders: true,
  colHeaders: true
}), _dec2 = (0, _uiThemeable2.default)(_theme2.default, styles), _dec(_class = _dec2(_class = (_temp = _class2 = function (_Component) {
  _inherits(Table, _Component);

  function Table() {
    _classCallCheck(this, Table);

    return _possibleConstructorReturn(this, (Table.__proto__ || Object.getPrototypeOf(Table)).apply(this, arguments));
  }

  _createClass(Table, [{
    key: 'render',
    value: function render() {
      var _classes;

      var classes = (_classes = {}, _defineProperty(_classes, styles.root, true), _defineProperty(_classes, styles[this.props.size], true), _defineProperty(_classes, styles.rowStriped, this.props.striped === 'rows'), _defineProperty(_classes, styles.colStriped, this.props.striped === 'columns'), _defineProperty(_classes, styles.hover, this.props.hover), _classes);

      return _react2.default.createElement(
        _View2.default,
        Object.assign({}, (0, _passthroughProps.omitProps)(this.props, Object.assign({}, Table.propTypes, _View2.default.propTypes)), {
          elementRef: this.props.elementRef,
          as: 'table',
          className: (0, _classnames2.default)(classes),
          margin: this.props.margin
        }),
        _react2.default.createElement(
          'caption',
          null,
          this.props.caption
        ),
        this.props.children
      );
    }
  }]);

  Table.displayName = 'Table'
  ;
  return Table;
}(_react.Component), _class2.propTypes = {
  /**
  * Set the table's caption element (its label)
  */
  caption: _propTypes2.default.node.isRequired,
  /**
  * Build the table markup via the children prop
  */
  children: _propTypes2.default.node,
  /**
  * Highlight each row on hover
  */
  hover: _propTypes2.default.bool,
  /**
  * Controls the padding and font-size of table cells
  */
  size: _propTypes2.default.oneOf(['small', 'medium', 'large']),
  /**
  * Add a light background color to alternate rows or columns
  */
  striped: _propTypes2.default.oneOf(['columns', 'rows']),
  /**
  * Valid values are `0`, `none`, `auto`, `xxx-small`, `xx-small`, `x-small`,
  * `small`, `medium`, `large`, `x-large`, `xx-large`. Apply these values via
  * familiar CSS-like shorthand. For example: `margin="small auto large"`.
  */
  margin: _ThemeablePropTypes2.default.spacing,
  /**
  * @deprecated
  */
  tableData: _propTypes2.default.array,
  /**
  * @deprecated
  */
  colHeaders: _propTypes2.default.arrayOf(_propTypes2.default.string),
  /**
  * @deprecated
  */
  rowHeaders: _propTypes2.default.bool
}, _class2.defaultProps = {
  hover: false,
  size: 'medium'
}, _temp)) || _class) || _class);
exports.default = Table;
