// @instructure/ui-forms/es/components/CheckboxGroup/__examples__/CheckboxGroup.examples.js
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import React from 'react';
import CheckboxGroup from '../index';
import Checkbox from '../../Checkbox';

export var layoutStacked = function layoutStacked() {
  return React.createElement(
    CheckboxGroup,
    {
      name: 'sports',
      description: 'Select your favorite sports'
    },
    React.createElement(Checkbox, { label: 'Tennis', value: 'tennis' }),
    React.createElement(Checkbox, { label: 'Lacrosse', value: 'lacrosse' }),
    React.createElement(Checkbox, { label: 'Water Polo', value: 'water-polo' }),
    React.createElement(Checkbox, { label: 'Golf', value: 'golf' })
  );
};

export var layoutColumns = function layoutColumns() {
  return React.createElement(
    CheckboxGroup,
    {
      layout: 'columns',
      name: 'sports',
      description: 'Select your favorite sports'
    },
    React.createElement(Checkbox, { label: 'Tennis', value: 'tennis' }),
    React.createElement(Checkbox, { label: 'Lacrosse', value: 'lacrosse' }),
    React.createElement(Checkbox, { label: 'Water Polo', value: 'water-polo' }),
    React.createElement(Checkbox, { label: 'Golf', value: 'golf' })
  );
};

export var layoutInline = function layoutInline() {
  return React.createElement(
    CheckboxGroup,
    {
      layout: 'inline',
      name: 'sports',
      description: 'Select your favorite sports'
    },
    React.createElement(Checkbox, { label: 'Tennis', value: 'tennis' }),
    React.createElement(Checkbox, { label: 'Lacrosse', value: 'lacrosse' }),
    React.createElement(Checkbox, { label: 'Water Polo', value: 'water-polo' }),
    React.createElement(Checkbox, { label: 'Golf', value: 'golf' })
  );
};

export var sizeSmall = function sizeSmall() {
  return React.createElement(
    CheckboxGroup,
    {
      size: 'small',
      name: 'sports',
      description: 'Select your favorite sports'
    },
    React.createElement(Checkbox, { label: 'Tennis', value: 'tennis' }),
    React.createElement(Checkbox, { label: 'Lacrosse', value: 'lacrosse' }),
    React.createElement(Checkbox, { label: 'Water Polo', value: 'water-polo' }),
    React.createElement(Checkbox, { label: 'Golf', value: 'golf' })
  );
};

export var sizeMedium = function sizeMedium() {
  return React.createElement(
    CheckboxGroup,
    {
      size: 'medium',
      name: 'sports',
      description: 'Select your favorite sports'
    },
    React.createElement(Checkbox, { label: 'Tennis', value: 'tennis' }),
    React.createElement(Checkbox, { label: 'Lacrosse', value: 'lacrosse' }),
    React.createElement(Checkbox, { label: 'Water Polo', value: 'water-polo' }),
    React.createElement(Checkbox, { label: 'Golf', value: 'golf' })
  );
};

export var sizeLarge = function sizeLarge() {
  return React.createElement(
    CheckboxGroup,
    {
      size: 'large',
      name: 'sports',
      description: 'Select your favorite sports'
    },
    React.createElement(Checkbox, { label: 'Tennis', value: 'tennis' }),
    React.createElement(Checkbox, { label: 'Lacrosse', value: 'lacrosse' }),
    React.createElement(Checkbox, { label: 'Water Polo', value: 'water-polo' }),
    React.createElement(Checkbox, { label: 'Golf', value: 'golf' })
  );
};

export var withToggle = function withToggle() {
  return React.createElement(
    CheckboxGroup,
    {
      name: 'sports',
      description: 'Select your favorite sports',
      size: 'small'
    },
    React.createElement(Checkbox, { variant: 'toggle', label: 'Tennis', value: 'tennis' }),
    React.createElement(Checkbox, { variant: 'toggle', label: 'Lacrosse', value: 'lacrosse' }),
    React.createElement(Checkbox, { variant: 'toggle', label: 'Water Polo', value: 'water-polo' }),
    React.createElement(Checkbox, { variant: 'toggle', label: 'Golf', value: 'golf' })
  );
};

export var error = function error() {
  return React.createElement(
    CheckboxGroup,
    {
      name: 'sports',
      description: 'Select your favorite sports',
      messages: [{ text: 'Please make a selection', type: 'error' }]
    },
    React.createElement(Checkbox, { label: 'Tennis', value: 'tennis' }),
    React.createElement(Checkbox, { label: 'Lacrosse', value: 'lacrosse' }),
    React.createElement(Checkbox, { label: 'Water Polo', value: 'water-polo' }),
    React.createElement(Checkbox, { label: 'Golf', value: 'golf' })
  );
};

export var hint = function hint() {
  return React.createElement(
    CheckboxGroup,
    {
      name: 'sports',
      description: 'Select your favorite sports',
      messages: [{ text: 'This is to help us decide on course offerings', type: 'hint' }]
    },
    React.createElement(Checkbox, { label: 'Tennis', value: 'tennis' }),
    React.createElement(Checkbox, { label: 'Lacrosse', value: 'lacrosse' }),
    React.createElement(Checkbox, { label: 'Water Polo', value: 'water-polo' }),
    React.createElement(Checkbox, { label: 'Golf', value: 'golf' })
  );
};

export var success = function success() {
  return React.createElement(
    CheckboxGroup,
    {
      name: 'sports',
      description: 'Select your favorite sports',
      messages: [{ text: 'Woot... you did it', type: 'success' }]
    },
    React.createElement(Checkbox, { label: 'Tennis', value: 'tennis' }),
    React.createElement(Checkbox, { label: 'Lacrosse', value: 'lacrosse' }),
    React.createElement(Checkbox, { label: 'Water Polo', value: 'water-polo' }),
    React.createElement(Checkbox, { label: 'Golf', value: 'golf' })
  );
};

export var disabled = function disabled() {
  return React.createElement(
    CheckboxGroup,
    {
      disabled: true,
      name: 'sports',
      description: 'Select your favorite sports'
    },
    React.createElement(Checkbox, { label: 'Tennis', value: 'tennis' }),
    React.createElement(Checkbox, { label: 'Lacrosse', value: 'lacrosse' }),
    React.createElement(Checkbox, { label: 'Water Polo', value: 'water-polo' }),
    React.createElement(Checkbox, { label: 'Golf', value: 'golf' })
  );
};
