// @instructure/ui-elements/lib/components/Badge/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _dec, _class, _class2, _temp; /*
                                   * The MIT License (MIT)
                                   *
                                   * Copyright (c) 2015 - present Instructure, Inc.
                                   *
                                   * Permission is hereby granted, free of charge, to any person obtaining a copy
                                   * of this software and associated documentation files (the "Software"), to deal
                                   * in the Software without restriction, including without limitation the rights
                                   * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
                                   * copies of the Software, and to permit persons to whom the Software is
                                   * furnished to do so, subject to the following conditions:
                                   *
                                   * The above copyright notice and this permission notice shall be included in all
                                   * copies or substantial portions of the Software.
                                   *
                                   * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
                                   * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
                                   * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
                                   * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
                                   * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
                                   * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
                                   * SOFTWARE.
                                   */


var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames2 = require('classnames');

var _classnames3 = _interopRequireDefault(_classnames2);

var _View = require('@instructure/ui-layout/lib/components/View');

var _View2 = _interopRequireDefault(_View);

var _uiThemeable = require('@instructure/ui-themeable');

var _uiThemeable2 = _interopRequireDefault(_uiThemeable);

var _ThemeablePropTypes = require('@instructure/ui-themeable/lib/utils/ThemeablePropTypes');

var _ThemeablePropTypes2 = _interopRequireDefault(_ThemeablePropTypes);

var _LayoutPropTypes = require('@instructure/ui-layout/lib/utils/LayoutPropTypes');

var _LayoutPropTypes2 = _interopRequireDefault(_LayoutPropTypes);

var _safeCloneElement = require('@instructure/ui-utils/lib/react/safeCloneElement');

var _safeCloneElement2 = _interopRequireDefault(_safeCloneElement);

var _generateElementId = require('@instructure/ui-utils/lib/dom/generateElementId');

var _generateElementId2 = _interopRequireDefault(_generateElementId);

var _CustomPropTypes = require('@instructure/ui-utils/lib/react/CustomPropTypes');

var _CustomPropTypes2 = _interopRequireDefault(_CustomPropTypes);

var _theme = require('./theme');

var _theme2 = _interopRequireDefault(_theme);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var styles = {
  template: function template(theme) {
    var tmpl = function tmpl() {
      return '/*  imported from styles.css  */\n\n/* stylelint-disable selector-max-class, max-nesting-depth */\n\n._1cExNS0 {\n  font-family: ' + theme.fontFamily + ';\n  font-weight: ' + theme.fontWeight + ';\n  box-sizing: border-box;\n  pointer-events: none;\n  text-align: center;\n  color: ' + theme.color + ';\n  font-size: ' + theme.fontSize + ';\n  white-space: nowrap;\n  border-radius: ' + theme.borderRadius + ';\n}\n\n[dir="ltr"] ._1cExNS0 {\n  text-align: center;\n}\n\n[dir="rtl"] ._1cExNS0 {\n  text-align: center;\n}\n\n._1cExNS0:not(._1qKTlxt) {\n    position: absolute;\n    /* stylelint-disable-next-line scale-unlimited/declaration-strict-value */\n    z-index: 1;\n  }\n\n._1cExNS0:not(._1qKTlxt)._3eMBRn8.hXO5S-r {\n        top: calc(-1 * ' + theme.countOffset + ');\n      }\n\n._1cExNS0:not(._1qKTlxt)._3eMBRn8._1mH0dBN {\n        top: ' + theme.notificationOffset + ';\n      }\n\n._1cExNS0:not(._1qKTlxt)._20VzxOS.hXO5S-r {\n        bottom: calc(-1 * ' + theme.countOffset + ');\n      }\n\n._1cExNS0:not(._1qKTlxt)._20VzxOS._1mH0dBN {\n        bottom: ' + theme.notificationOffset + ';\n      }\n\n._1cExNS0:not(._1qKTlxt)._22eOKus.hXO5S-r {\n        offset-inline-start: calc(-1 * ' + theme.countOffset + ');\n        offset-inline-end: auto;\n      }\n\n[dir="ltr"] ._1cExNS0:not(._1qKTlxt)._22eOKus.hXO5S-r {\n        left: calc(-1 * ' + theme.countOffset + ');\n        right: auto;\n      }\n\n[dir="rtl"] ._1cExNS0:not(._1qKTlxt)._22eOKus.hXO5S-r {\n        right: calc(-1 * ' + theme.countOffset + ');\n        left: auto;\n      }\n\n._1cExNS0:not(._1qKTlxt)._22eOKus._1mH0dBN {\n        offset-inline-start: ' + theme.notificationOffset + ';\n        offset-inline-end: auto;\n      }\n\n[dir="ltr"] ._1cExNS0:not(._1qKTlxt)._22eOKus._1mH0dBN {\n        left: ' + theme.notificationOffset + ';\n        right: auto;\n      }\n\n[dir="rtl"] ._1cExNS0:not(._1qKTlxt)._22eOKus._1mH0dBN {\n        right: ' + theme.notificationOffset + ';\n        left: auto;\n      }\n\n._1cExNS0:not(._1qKTlxt)._hJFOC0.hXO5S-r {\n        offset-inline-end: calc(-1 * ' + theme.countOffset + ');\n        offset-inline-start: auto;\n      }\n\n[dir="ltr"] ._1cExNS0:not(._1qKTlxt)._hJFOC0.hXO5S-r {\n        right: calc(-1 * ' + theme.countOffset + ');\n        left: auto;\n      }\n\n[dir="rtl"] ._1cExNS0:not(._1qKTlxt)._hJFOC0.hXO5S-r {\n        left: calc(-1 * ' + theme.countOffset + ');\n        right: auto;\n      }\n\n._1cExNS0:not(._1qKTlxt)._hJFOC0._1mH0dBN {\n        offset-inline-end: ' + theme.notificationOffset + ';\n        offset-inline-start: auto;\n      }\n\n[dir="ltr"] ._1cExNS0:not(._1qKTlxt)._hJFOC0._1mH0dBN {\n        right: ' + theme.notificationOffset + ';\n        left: auto;\n      }\n\n[dir="rtl"] ._1cExNS0:not(._1qKTlxt)._hJFOC0._1mH0dBN {\n        left: ' + theme.notificationOffset + ';\n        right: auto;\n      }\n\n._1cExNS0:not(._1qKTlxt)._3n8tP6H._hJFOC0.hXO5S-r, ._1cExNS0:not(._1qKTlxt)._3n8tP6H._22eOKus.hXO5S-r {\n        top: calc(50% - (' + theme.size + ' / 2));\n      }\n\n._1cExNS0:not(._1qKTlxt)._3n8tP6H._hJFOC0._1mH0dBN, ._1cExNS0:not(._1qKTlxt)._3n8tP6H._22eOKus._1mH0dBN {\n        top: calc(50% - (' + theme.sizeNotification + ' / 2));\n      }\n\n/*\n    * The diagonally positioned badges looked better closer to\n    * the button/child, while the centered badges looked happier a\n    * bit further away. Hence the special treatment below.\n    */\n\n._1cExNS0:not(._1qKTlxt)._3n8tP6H._22eOKus.hXO5S-r {\n        offset-inline-start: auto;\n        offset-inline-end: calc(100% - ' + theme.countOffset + ');\n      }\n\n[dir="ltr"] ._1cExNS0:not(._1qKTlxt)._3n8tP6H._22eOKus.hXO5S-r {\n        left: auto;\n        right: calc(100% - ' + theme.countOffset + ');\n      }\n\n[dir="rtl"] ._1cExNS0:not(._1qKTlxt)._3n8tP6H._22eOKus.hXO5S-r {\n        right: auto;\n        left: calc(100% - ' + theme.countOffset + ');\n      }\n\n._1cExNS0:not(._1qKTlxt)._3n8tP6H._22eOKus._1mH0dBN {\n        offset-inline-start: calc(-1 * ' + theme.sizeNotification + ' / 2);\n        offset-inline-end: auto;\n      }\n\n[dir="ltr"] ._1cExNS0:not(._1qKTlxt)._3n8tP6H._22eOKus._1mH0dBN {\n        left: calc(-1 * ' + theme.sizeNotification + ' / 2);\n        right: auto;\n      }\n\n[dir="rtl"] ._1cExNS0:not(._1qKTlxt)._3n8tP6H._22eOKus._1mH0dBN {\n        right: calc(-1 * ' + theme.sizeNotification + ' / 2);\n        left: auto;\n      }\n\n._1cExNS0:not(._1qKTlxt)._3n8tP6H._hJFOC0.hXO5S-r {\n        offset-inline-end: auto;\n        offset-inline-start: calc(100% - ' + theme.countOffset + ');\n      }\n\n[dir="ltr"] ._1cExNS0:not(._1qKTlxt)._3n8tP6H._hJFOC0.hXO5S-r {\n        right: auto;\n        left: calc(100% - ' + theme.countOffset + ');\n      }\n\n[dir="rtl"] ._1cExNS0:not(._1qKTlxt)._3n8tP6H._hJFOC0.hXO5S-r {\n        left: auto;\n        right: calc(100% - ' + theme.countOffset + ');\n      }\n\n._1cExNS0:not(._1qKTlxt)._3n8tP6H._hJFOC0._1mH0dBN {\n        offset-inline-end: calc(-1 * ' + theme.sizeNotification + ' / 2);\n        offset-inline-start: auto;\n      }\n\n[dir="ltr"] ._1cExNS0:not(._1qKTlxt)._3n8tP6H._hJFOC0._1mH0dBN {\n        right: calc(-1 * ' + theme.sizeNotification + ' / 2);\n        left: auto;\n      }\n\n[dir="rtl"] ._1cExNS0:not(._1qKTlxt)._3n8tP6H._hJFOC0._1mH0dBN {\n        left: calc(-1 * ' + theme.sizeNotification + ' / 2);\n        right: auto;\n      }\n\n._1cExNS0.hXO5S-r {\n    line-height: ' + theme.size + ';\n    min-width: ' + theme.size + ';\n    -webkit-padding-start: ' + theme.padding + ';\n            padding-inline-start: ' + theme.padding + ';\n    -webkit-padding-end: ' + theme.padding + ';\n            padding-inline-end: ' + theme.padding + ';\n  }\n\n[dir="ltr"] ._1cExNS0.hXO5S-r {\n    padding-left: ' + theme.padding + ';\n    padding-right: ' + theme.padding + ';\n  }\n\n[dir="rtl"] ._1cExNS0.hXO5S-r {\n    padding-right: ' + theme.padding + ';\n    padding-left: ' + theme.padding + ';\n  }\n\n._1cExNS0._1mH0dBN {\n    width: ' + theme.sizeNotification + ';\n    height: ' + theme.sizeNotification + ';\n  }\n\n._15xUr6V {\n  background-color: ' + theme.colorDanger + ';\n}\n\n._15xUr6V._1SymH0a::before {\n    border-color: ' + theme.colorDanger + ';\n  }\n\n._3dSTM_W {\n  background-color: ' + theme.colorSuccess + ';\n}\n\n._3dSTM_W._1SymH0a::before {\n    border-color: ' + theme.colorSuccess + ';\n  }\n\n._2r6NB93 {\n  background-color: ' + theme.colorPrimary + ';\n}\n\n._2r6NB93._1SymH0a::before {\n    border-color: ' + theme.colorPrimary + ';\n  }\n\n@keyframes _1SymH0a {\n  to {\n    transform: scale(1);\n    opacity: 0.9;\n  }\n}\n\n._1SymH0a {\n  position: relative;\n}\n\n._1SymH0a::before {\n    content: "";\n    width: calc(100% + 0.5rem);\n    height: calc(100% + 0.5rem);\n    border-radius: ' + theme.borderRadius + ';\n    position: absolute;\n    top: -0.25rem;\n    offset-inline-end: auto;\n    offset-inline-start: -0.25rem;\n    box-sizing: border-box;\n    border: ' + theme.pulseBorderThickness + ' solid;\n    opacity: 0;\n    transform: scale(0.75);\n    animation-name: _1SymH0a;\n    animation-duration: 1s;\n    animation-iteration-count: 4;\n    animation-direction: alternate;\n  }\n\n[dir="ltr"] ._1SymH0a::before {\n    right: auto;\n    left: -0.25rem;\n  }\n\n[dir="rtl"] ._1SymH0a::before {\n    left: auto;\n    right: -0.25rem;\n  }\n\n._3o0AmSg {\n  position: relative;\n  box-sizing: border-box;\n}\n\n._3o0AmSg svg { /* stylelint-disable-line selector-max-type */\n    display: block;\n  }\n\n/* stylelint-enable selector-max-class, max-nesting-depth */\n';
    };

    return tmpl.call(theme, theme);
  },
  'badge': '_1cExNS0',
  'standalone': '_1qKTlxt',
  'positioned--top': '_3eMBRn8',
  'count': 'hXO5S-r',
  'notification': '_1mH0dBN',
  'positioned--bottom': '_20VzxOS',
  'positioned--start': '_22eOKus',
  'positioned--end': '_hJFOC0',
  'positioned--center': '_3n8tP6H',
  'danger': '_15xUr6V',
  'pulse': '_1SymH0a',
  'success': '_3dSTM_W',
  'primary': '_2r6NB93',
  'wrapper': '_3o0AmSg'
};


/**
---
category: components
---
**/

var Badge = (_dec = (0, _uiThemeable2.default)(_theme2.default, styles), _dec(_class = (_temp = _class2 = function (_Component) {
  _inherits(Badge, _Component);

  function Badge(props) {
    _classCallCheck(this, Badge);

    var _this = _possibleConstructorReturn(this, (Badge.__proto__ || Object.getPrototypeOf(Badge)).call(this, props));

    _this._defaultId = (0, _generateElementId2.default)('Badge');
    return _this;
  }

  _createClass(Badge, [{
    key: 'countOverflow',
    value: function countOverflow() {
      var _props = this.props,
          count = _props.count,
          countUntil = _props.countUntil;


      if (countUntil > 1 && count >= countUntil) {
        return true;
      } else {
        return false;
      }
    }
  }, {
    key: 'renderOutput',
    value: function renderOutput() {
      var _props2 = this.props,
          count = _props2.count,
          countUntil = _props2.countUntil,
          formatOverflowText = _props2.formatOverflowText,
          formatOutput = _props2.formatOutput,
          type = _props2.type;

      // If the badge count is >= than the countUntil limit, format the badge text
      // via the formatOverflowText function prop

      var formattedCount = type === 'count' && this.countOverflow() ? formatOverflowText(count, countUntil) : count;

      if (typeof formatOutput === 'function') {
        return formatOutput(formattedCount);
      } else {
        return type === 'count' ? formattedCount : null;
      }
    }
  }, {
    key: 'renderBadge',
    value: function renderBadge() {
      var _classnames;

      var _props3 = this.props,
          count = _props3.count,
          margin = _props3.margin,
          pulse = _props3.pulse,
          placement = _props3.placement,
          standalone = _props3.standalone,
          type = _props3.type,
          variant = _props3.variant;


      return _react2.default.createElement(
        _View2.default,
        {
          margin: standalone ? margin : 'none',
          className: (0, _classnames3.default)((_classnames = {}, _defineProperty(_classnames, styles.badge, true), _defineProperty(_classnames, styles[type], type), _defineProperty(_classnames, styles[variant], variant), _defineProperty(_classnames, styles['positioned--top'], placement.indexOf('top') > -1), _defineProperty(_classnames, styles['positioned--bottom'], placement.indexOf('bottom') > -1), _defineProperty(_classnames, styles['positioned--start'], placement.indexOf('start') > -1), _defineProperty(_classnames, styles['positioned--end'], placement.indexOf('end') > -1), _defineProperty(_classnames, styles['positioned--center'], placement.indexOf('center') > -1), _defineProperty(_classnames, styles.standalone, standalone), _defineProperty(_classnames, styles.pulse, pulse), _classnames)),
          title: type === 'count' && this.countOverflow() ? count : null,
          id: !standalone ? this._defaultId : null,
          display: standalone ? 'inline-block' : 'block'
        },
        this.renderOutput()
      );
    }
  }, {
    key: 'renderChildren',
    value: function renderChildren() {
      var _this2 = this;

      return _react.Children.map(this.props.children, function (child) {
        return (0, _safeCloneElement2.default)(child, {
          'aria-describedby': _this2._defaultId
        });
      });
    }
  }, {
    key: 'render',
    value: function render() {
      var _props4 = this.props,
          margin = _props4.margin,
          elementRef = _props4.elementRef,
          standalone = _props4.standalone,
          as = _props4.as;


      if (standalone) {
        return this.renderBadge();
      } else {
        return _react2.default.createElement(
          _View2.default,
          {
            as: as,
            margin: margin,
            elementRef: elementRef,
            className: styles.wrapper,
            display: 'inline-block'
          },
          this.renderChildren(),
          this.renderBadge()
        );
      }
    }
  }]);

  Badge.displayName = 'Badge'
  ;
  return Badge;
}(_react.Component), _class2.propTypes = {
  count: _propTypes2.default.number,
  /**
   * The number at which the count gets truncated by
   * formatOverflowText. For example, a countUntil of 100
   * would stop the count at 99.
  */
  countUntil: _propTypes2.default.number,
  children: _propTypes2.default.element,
  /**
   * Render Badge as a counter (`count`) or as a smaller dot (`notification`) with
   * no count number displayed.
  */
  type: _propTypes2.default.oneOf(['count', 'notification']),
  /**
   * Render Badge as an inline html element that is not positioned relative
   * to a child.
  */
  standalone: _propTypes2.default.bool,
  /**
   * Make the Badge slowly pulse twice to get the user's attention.
  */
  pulse: _propTypes2.default.bool,
  variant: _propTypes2.default.oneOf(['primary', 'success', 'danger']),
  /**
  * Supported values are `top start`, `top end`, `end center`, `bottom end`,
  * `bottom start`, and `start center`
  */
  placement: _LayoutPropTypes2.default.placement,
  /**
  * Valid values are `0`, `none`, `auto`, `xxx-small`, `xx-small`, `x-small`,
  * `small`, `medium`, `large`, `x-large`, `xx-large`. Apply these values via
  * familiar CSS-like shorthand. For example: `margin="small auto large"`.
  */
  margin: _ThemeablePropTypes2.default.spacing,
  /**
  * provides a reference to the underlying html root element
  */
  elementRef: _propTypes2.default.func,
  formatOverflowText: _propTypes2.default.func,
  formatOutput: _propTypes2.default.func,
  as: _CustomPropTypes2.default.elementType
}, _class2.defaultProps = {
  standalone: false,
  type: 'count',
  variant: 'primary',
  pulse: false,
  placement: 'top end',
  elementRef: function elementRef(el) {},
  formatOverflowText: function formatOverflowText(count, countUntil) {
    return countUntil - 1 + ' +';
  }
}, _temp)) || _class);
exports.default = Badge;
