// @instructure/ui-forms/es/components/TextArea/index.js
var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _dec, _class, _class2, _temp;

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';

import CustomPropTypes from '@instructure/ui-utils/lib/react/CustomPropTypes';
import themeable from '@instructure/ui-themeable';
import debounce from '@instructure/ui-utils/lib/debounce';
import addEventListener from '@instructure/ui-utils/lib/dom/addEventListener';
import findDOMNode from '@instructure/ui-utils/lib/dom/findDOMNode';
import requestAnimationFrame from '@instructure/ui-utils/lib/dom/requestAnimationFrame';
import isActiveElement from '@instructure/ui-utils/lib/dom/isActiveElement';
import px from '@instructure/ui-utils/lib/px';
import { pickProps, omitProps } from '@instructure/ui-utils/lib/react/passthroughProps';
import generateElementId from '@instructure/ui-utils/lib/dom/generateElementId';

var styles = {
  template: function template(theme) {
    var tmpl = function tmpl() {
      return '/*  imported from styles.css  */\n\n._16BRrcp {\n  animation: none 0s ease 0s 1 normal none running;\n  backface-visibility: visible;\n  background: transparent none repeat 0 0 / auto auto padding-box border-box scroll;\n  border: medium none currentColor;\n  border-collapse: separate;\n  border-image: none;\n  border-radius: 0;\n  border-spacing: 0;\n  bottom: auto;\n  box-shadow: none;\n  box-sizing: content-box;\n  caption-side: top;\n  clear: none;\n  clip: auto;\n  color: #000;\n  columns: auto;\n  column-count: auto;\n  column-fill: balance;\n  column-gap: normal;\n  column-rule: medium none currentColor;\n  column-span: 1;\n  column-width: auto;\n  content: normal;\n  counter-increment: none;\n  counter-reset: none;\n  cursor: auto;\n  direction: ltr;\n  display: inline;\n  empty-cells: show;\n  float: none;\n  font-family: serif;\n  font-size: medium;\n  font-style: normal;\n  font-variant: normal;\n  font-weight: normal;\n  font-stretch: normal;\n  line-height: normal;\n  height: auto;\n  hyphens: none;\n  left: auto;\n  letter-spacing: normal;\n  list-style: disc outside none;\n  margin: 0;\n  max-height: none;\n  max-width: none;\n  min-height: 0;\n  min-width: 0;\n  opacity: 1;\n  orphans: 2;\n  outline: medium none invert;\n  overflow: visible;\n  overflow-x: visible;\n  overflow-y: visible;\n  padding: 0;\n  page-break-after: auto;\n  page-break-before: auto;\n  page-break-inside: auto;\n  perspective: none;\n  perspective-origin: 50% 50%;\n  position: static;\n  right: auto;\n  tab-size: 8;\n  table-layout: auto;\n  text-align: left;\n  text-align-last: auto;\n  text-decoration: none;\n  text-indent: 0;\n  text-shadow: none;\n  text-transform: none;\n  top: auto;\n  transform: none;\n  transform-origin: 50% 50% 0;\n  transform-style: flat;\n  transition: none 0s ease 0s;\n  unicode-bidi: normal;\n  vertical-align: baseline;\n  visibility: visible;\n  white-space: normal;\n  widows: 2;\n  width: auto;\n  word-spacing: normal;\n  z-index: auto;\n  all: initial;\n  -webkit-font-smoothing: antialiased;\n  -moz-osx-font-smoothing: grayscale;\n  -webkit-appearance: none;\n     -moz-appearance: none;\n          appearance: none;\n  height: auto;\n  margin: 0;\n  width: 100%;\n  display: block;\n  box-sizing: border-box;\n  padding: ' + theme.padding + ';\n  transition: all 0.2s ease-out;\n  outline: 0.1875rem solid transparent;\n  outline-offset: -0.5rem;\n  font-family: ' + theme.fontFamily + ';\n  font-weight: ' + theme.fontWeight + ';\n  border-width: ' + theme.borderWidth + ';\n  border-style: ' + theme.borderStyle + ';\n  border-top-color: ' + theme.borderTopColor + ';\n  border-right-color: ' + theme.borderRightColor + '; /* stylelint-disable-line property-blacklist */\n  border-bottom-color: ' + theme.borderBottomColor + ';\n  border-left-color: ' + theme.borderLeftColor + '; /* stylelint-disable-line property-blacklist */\n  border-radius: ' + theme.borderRadius + ';\n  color: ' + theme.color + ';\n  background: ' + theme.background + ';\n  white-space: pre-wrap;\n  word-wrap: break-word;\n  text-align: start;\n}\n\n[dir="ltr"] ._16BRrcp {\n  text-align: left;\n}\n\n[dir="rtl"] ._16BRrcp {\n  text-align: right;\n}\n\n  ._16BRrcp:focus {\n    outline: 0.1875rem solid ' + theme.focusOutlineColor + ';\n    outline-offset: -0.1875rem;\n    border-top-color: ' + theme.focusBorderTopColor + ';\n    border-right-color: ' + theme.focusBorderRightColor + '; /* stylelint-disable-line property-blacklist */\n    border-bottom-color: ' + theme.focusBorderBottomColor + ';\n    border-left-color: ' + theme.focusBorderLeftColor + '; /* stylelint-disable-line property-blacklist */\n    box-shadow: 0 0.0625rem 0 0 ' + theme.boxShadowColor + ';\n  }\n\n  ._16BRrcp[aria-invalid] {\n    border-color: ' + theme.errorBorderColor + ';\n  }\n\n  ._16BRrcp[aria-invalid]:focus {\n      border-color: ' + theme.errorBorderColor + ';\n      outline-color: ' + theme.errorOutlineColor + ';\n      box-shadow: 0 0.125rem 0 0 ' + theme.boxShadowErrorColor + ';\n    }\n\n  ._16BRrcp:-ms-input-placeholder {\n    color: ' + theme.placeholderColor + ';\n  }\n\n  ._16BRrcp::-ms-input-placeholder {\n    color: ' + theme.placeholderColor + ';\n  }\n\n  ._16BRrcp::placeholder {\n    color: ' + theme.placeholderColor + ';\n  }\n\n  ._16BRrcp.TQJ3s4P {\n    cursor: not-allowed;\n    pointer-events: none;\n    opacity: 0.5;\n  }\n\n._3iB5Mq_ {\n  font-size: ' + theme.smallFontSize + ';\n}\n\n.BgBppdU {\n  font-size: ' + theme.mediumFontSize + ';\n}\n\n.Eqqryt0 {\n  font-size: ' + theme.largeFontSize + ';\n}\n';
    };

    return tmpl.call(theme, theme);
  },
  'textarea': '_16BRrcp',
  'disabled': 'TQJ3s4P',
  'small': '_3iB5Mq_',
  'medium': 'BgBppdU',
  'large': 'Eqqryt0'
};

import theme from './theme';

import FormPropTypes from '../../utils/FormPropTypes';
import FormField from '../FormField';

/**
---
category: components/forms
---
**/
var TextArea = (_dec = themeable(theme, styles), _dec(_class = (_temp = _class2 = function (_Component) {
  _inherits(TextArea, _Component);

  function TextArea() {
    _classCallCheck(this, TextArea);

    var _this = _possibleConstructorReturn(this, (TextArea.__proto__ || Object.getPrototypeOf(TextArea)).call(this));

    _this.grow = function () {
      if (!_this._textarea) {
        return;
      }

      var offset = _this._textarea.offsetHeight - _this._textarea.clientHeight;
      var height = '';

      // Note:
      // 1. height has be reset to `auto` every time this method runs, or scrollHeight will not reset
      _this._textarea.style.height = 'auto';
      // 2. `this._textarea.scrollHeight` will not reset if assigned to a variable; it needs to be written out each time

      _this._textarea.style.overflowY = 'hidden'; // hide scrollbars for autoGrow textareas
      height = _this._textarea.scrollHeight + offset + 'px';

      if (_this.props.maxHeight && _this._textarea.scrollHeight > px(_this.props.maxHeight, findDOMNode(_this))) {
        _this._textarea.style.overflowY = 'auto'; // add scroll if scrollHeight exceeds maxHeight in pixels
      } else if (_this.props.height) {
        if (_this._textarea.value === '') {
          height = _this.props.height;
        } else if (px(_this.props.height, findDOMNode(_this)) > _this._textarea.scrollHeight) {
          _this._textarea.style.overflowY = 'auto'; // add scroll if scrollHeight exceeds height in pixels
          height = _this.props.height;
        }
      }

      _this._textarea.style.height = height;
      _this._textarea.scrollTop = _this._textarea.scrollHeight;
    };

    _this.handleChange = function (event) {
      var _this$props = _this.props,
          onChange = _this$props.onChange,
          value = _this$props.value,
          disabled = _this$props.disabled,
          readOnly = _this$props.readOnly;


      if (disabled || readOnly) {
        event.preventDefault();
        return;
      }

      if (typeof value === 'undefined') {
        // if uncontrolled
        _this.autoGrow();
      }

      if (typeof onChange === 'function') {
        onChange(event);
      }
    };

    _this._defaultId = generateElementId('TextArea');
    return _this;
  }

  _createClass(TextArea, [{
    key: 'componentDidMount',
    value: function componentDidMount() {
      this.autoGrow();
    }
  }, {
    key: 'componentWillReceiveProps',
    value: function componentWillReceiveProps(nextProps) {
      this.autoGrow();
    }
  }, {
    key: 'componentWillUnmount',
    value: function componentWillUnmount() {
      if (this._listener) {
        this._listener.remove();
      }

      if (this._request) {
        this._request.cancel();
      }

      if (this._debounced) {
        this._debounced.cancel();
      }
    }
  }, {
    key: 'autoGrow',
    value: function autoGrow() {
      if (this.props.autoGrow) {
        if (!this._debounced) {
          this._debounced = debounce(this.grow, 200, { leading: false, trailing: true });
        }

        if (!this._listener) {
          this._listener = addEventListener(window, 'resize', this._debounced);
        }

        this._request = requestAnimationFrame(this._debounced);
      }
    }
  }, {
    key: 'focus',
    value: function focus() {
      this._textarea.focus();
    }
  }, {
    key: 'render',
    value: function render() {
      var _classes,
          _this2 = this;

      var _props = this.props,
          autoGrow = _props.autoGrow,
          placeholder = _props.placeholder,
          value = _props.value,
          defaultValue = _props.defaultValue,
          disabled = _props.disabled,
          readOnly = _props.readOnly,
          required = _props.required,
          width = _props.width,
          height = _props.height,
          maxHeight = _props.maxHeight,
          textareaRef = _props.textareaRef,
          resize = _props.resize,
          size = _props.size;


      var props = omitProps(this.props, TextArea.propTypes);

      var classes = (_classes = {}, _defineProperty(_classes, styles.textarea, true), _defineProperty(_classes, styles[size], true), _defineProperty(_classes, styles.disabled, disabled), _classes);

      var style = {
        width: width,
        resize: resize,
        height: !autoGrow ? height : null,
        maxHeight: maxHeight
      };

      return React.createElement(
        FormField,
        Object.assign({}, pickProps(this.props, FormField.propTypes), {
          vAlign: 'top',
          id: this.id,
          ref: function ref(el) {
            _this2._node = el;
          }
        }),
        React.createElement('textarea', Object.assign({}, props, {
          value: value,
          defaultValue: defaultValue,
          placeholder: placeholder,
          ref: function ref(textarea) {
            for (var _len = arguments.length, args = Array(_len > 1 ? _len - 1 : 0), _key = 1; _key < _len; _key++) {
              args[_key - 1] = arguments[_key];
            }

            _this2._textarea = textarea;
            textareaRef.apply(_this2, [textarea].concat(args));
          },
          style: style,
          id: this.id,
          required: required,
          'aria-required': required,
          'aria-invalid': this.invalid ? 'true' : null,
          disabled: disabled || readOnly,
          'aria-disabled': disabled || readOnly ? 'true' : null,
          className: classnames(classes),
          onChange: this.handleChange
        }))
      );
    }
  }, {
    key: 'minHeight',
    get: function get() {
      return this._textarea.style.minHeight;
    }
  }, {
    key: 'invalid',
    get: function get() {
      return this.props.messages && this.props.messages.findIndex(function (message) {
        return message.type === 'error';
      }) >= 0;
    }
  }, {
    key: 'id',
    get: function get() {
      return this.props.id || this._defaultId;
    }
  }, {
    key: 'focused',
    get: function get() {
      return isActiveElement(this._textarea);
    }
  }, {
    key: 'value',
    get: function get() {
      return this._textarea.value;
    }
  }]);

  TextArea.displayName = 'TextArea'
  ;
  return TextArea;
}(Component), _class2.propTypes = {
  label: PropTypes.node.isRequired,
  id: PropTypes.string,
  /**
  * sets the font-size for the textarea
  */
  size: PropTypes.oneOf(['small', 'medium', 'large']),
  layout: PropTypes.oneOf(['stacked', 'inline']),
  /**
  * the textarea will expand vertically to fit the height of the content,
  * unless its content exceeds `maxHeight`
  */
  autoGrow: PropTypes.bool,
  /**
  * is the textarea resizable (in supported browsers)
  */
  resize: PropTypes.oneOf(['none', 'both', 'horizontal', 'vertical']),
  /**
  * a fixed width for the textarea
  */
  width: PropTypes.string,
  /**
  * a initial height for the textarea (if autoGrow is true it will grow vertically)
  */
  height: PropTypes.string,
  /**
  * when autoGrow is true, the textarea will never grow beyond this value
  */
  maxHeight: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
  /**
  * object with shape: `{
  * text: PropTypes.string,
  * type: PropTypes.oneOf(['error', 'hint', 'success', 'screenreader-only'])
  *   }`
  */
  messages: PropTypes.arrayOf(FormPropTypes.message),
  inline: PropTypes.bool,
  /**
  * Html placeholder text to display when the input has no value. This should be hint text, not a label
  * replacement.
  */
  placeholder: PropTypes.string,
  /**
   * Whether or not to disable the textarea
   */
  disabled: PropTypes.bool,
  /**
   * Works just like disabled but keeps the same styles as if it were active
   */
  readOnly: PropTypes.bool,
  required: PropTypes.bool,
  /**
  * a function that provides a reference to the actual textarea element
  */
  textareaRef: PropTypes.func,
  /**
  * value to set on initial render
  */
  defaultValue: PropTypes.string,
  /**
  * the selected value (must be accompanied by an `onChange` prop)
  */
  value: CustomPropTypes.controllable(PropTypes.string),
  /**
  * when used with the `value` prop, the component will not control its own state
  */
  onChange: PropTypes.func
}, _class2.defaultProps = {
  size: 'medium',
  autoGrow: true,
  resize: 'none',
  inline: false,
  messages: [],
  disabled: false,
  readOnly: false,
  textareaRef: function textareaRef(textarea) {},
  layout: 'stacked'
}, _temp)) || _class);


export default TextArea;
