// @instructure/ui-utils/lib/dom/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.transformSelection = exports.supportsObjectFit = exports.scopeTab = exports.requestAnimationFrame = exports.ownerWindow = exports.ownerDocument = exports.isActiveElement = exports.InputModeListener = exports.handleMouseOverOut = exports.getScrollParents = exports.getOffsetParents = exports.getFontSize = exports.getComputedStyle = exports.getClassList = exports.getBoundingClientRect = exports.getActiveElement = exports.generateElementId = exports.findTabbable = exports.findDOMNode = exports.containsActiveElement = exports.contains = exports.canUseDOM = exports.calculateElementPosition = exports.addResizeListener = exports.addPositionChangeListener = exports.addEventListener = undefined;

var _addEventListener2 = require('./addEventListener');

var _addEventListener3 = _interopRequireDefault(_addEventListener2);

var _addPositionChangeListener2 = require('./addPositionChangeListener');

var _addPositionChangeListener3 = _interopRequireDefault(_addPositionChangeListener2);

var _addResizeListener2 = require('./addResizeListener');

var _addResizeListener3 = _interopRequireDefault(_addResizeListener2);

var _calculateElementPosition2 = require('./calculateElementPosition');

var _calculateElementPosition3 = _interopRequireDefault(_calculateElementPosition2);

var _canUseDOM2 = require('./canUseDOM');

var _canUseDOM3 = _interopRequireDefault(_canUseDOM2);

var _contains2 = require('./contains');

var _contains3 = _interopRequireDefault(_contains2);

var _containsActiveElement2 = require('./containsActiveElement');

var _containsActiveElement3 = _interopRequireDefault(_containsActiveElement2);

var _findDOMNode2 = require('./findDOMNode');

var _findDOMNode3 = _interopRequireDefault(_findDOMNode2);

var _findTabbable2 = require('./findTabbable');

var _findTabbable3 = _interopRequireDefault(_findTabbable2);

var _generateElementId2 = require('./generateElementId');

var _generateElementId3 = _interopRequireDefault(_generateElementId2);

var _getActiveElement2 = require('./getActiveElement');

var _getActiveElement3 = _interopRequireDefault(_getActiveElement2);

var _getBoundingClientRect2 = require('./getBoundingClientRect');

var _getBoundingClientRect3 = _interopRequireDefault(_getBoundingClientRect2);

var _getClassList2 = require('./getClassList');

var _getClassList3 = _interopRequireDefault(_getClassList2);

var _getComputedStyle2 = require('./getComputedStyle');

var _getComputedStyle3 = _interopRequireDefault(_getComputedStyle2);

var _getFontSize2 = require('./getFontSize');

var _getFontSize3 = _interopRequireDefault(_getFontSize2);

var _getOffsetParents2 = require('./getOffsetParents');

var _getOffsetParents3 = _interopRequireDefault(_getOffsetParents2);

var _getScrollParents2 = require('./getScrollParents');

var _getScrollParents3 = _interopRequireDefault(_getScrollParents2);

var _handleMouseOverOut2 = require('./handleMouseOverOut');

var _handleMouseOverOut3 = _interopRequireDefault(_handleMouseOverOut2);

var _InputModeListener2 = require('./InputModeListener');

var _InputModeListener3 = _interopRequireDefault(_InputModeListener2);

var _isActiveElement2 = require('./isActiveElement');

var _isActiveElement3 = _interopRequireDefault(_isActiveElement2);

var _ownerDocument2 = require('./ownerDocument');

var _ownerDocument3 = _interopRequireDefault(_ownerDocument2);

var _ownerWindow2 = require('./ownerWindow');

var _ownerWindow3 = _interopRequireDefault(_ownerWindow2);

var _requestAnimationFrame2 = require('./requestAnimationFrame');

var _requestAnimationFrame3 = _interopRequireDefault(_requestAnimationFrame2);

var _scopeTab2 = require('./scopeTab');

var _scopeTab3 = _interopRequireDefault(_scopeTab2);

var _supportsObjectFit2 = require('./supportsObjectFit');

var _supportsObjectFit3 = _interopRequireDefault(_supportsObjectFit2);

var _transformSelection2 = require('./transformSelection');

var _transformSelection3 = _interopRequireDefault(_transformSelection2);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.addEventListener = _addEventListener3.default; /*
                                                        * The MIT License (MIT)
                                                        *
                                                        * Copyright (c) 2015 - present Instructure, Inc.
                                                        *
                                                        * Permission is hereby granted, free of charge, to any person obtaining a copy
                                                        * of this software and associated documentation files (the "Software"), to deal
                                                        * in the Software without restriction, including without limitation the rights
                                                        * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
                                                        * copies of the Software, and to permit persons to whom the Software is
                                                        * furnished to do so, subject to the following conditions:
                                                        *
                                                        * The above copyright notice and this permission notice shall be included in all
                                                        * copies or substantial portions of the Software.
                                                        *
                                                        * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
                                                        * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
                                                        * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
                                                        * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
                                                        * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
                                                        * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
                                                        * SOFTWARE.
                                                        */

/* list utils in alphabetical order */

exports.addPositionChangeListener = _addPositionChangeListener3.default;
exports.addResizeListener = _addResizeListener3.default;
exports.calculateElementPosition = _calculateElementPosition3.default;
exports.canUseDOM = _canUseDOM3.default;
exports.contains = _contains3.default;
exports.containsActiveElement = _containsActiveElement3.default;
exports.findDOMNode = _findDOMNode3.default;
exports.findTabbable = _findTabbable3.default;
exports.generateElementId = _generateElementId3.default;
exports.getActiveElement = _getActiveElement3.default;
exports.getBoundingClientRect = _getBoundingClientRect3.default;
exports.getClassList = _getClassList3.default;
exports.getComputedStyle = _getComputedStyle3.default;
exports.getFontSize = _getFontSize3.default;
exports.getOffsetParents = _getOffsetParents3.default;
exports.getScrollParents = _getScrollParents3.default;
exports.handleMouseOverOut = _handleMouseOverOut3.default;
exports.InputModeListener = _InputModeListener3.default;
exports.isActiveElement = _isActiveElement3.default;
exports.ownerDocument = _ownerDocument3.default;
exports.ownerWindow = _ownerWindow3.default;
exports.requestAnimationFrame = _requestAnimationFrame3.default;
exports.scopeTab = _scopeTab3.default;
exports.supportsObjectFit = _supportsObjectFit3.default;
exports.transformSelection = _transformSelection3.default;
