// @instructure/ui-layout/es/components/DrawerLayout/index.js
var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _dec, _dec2, _class, _class2, _temp2;

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
import React, { Children, Component } from 'react';
import PropTypes from 'prop-types';

import bidirectional from '@instructure/ui-i18n/lib/bidirectional';

import themeable from '@instructure/ui-themeable';
import CustomPropTypes from '@instructure/ui-utils/lib/react/CustomPropTypes';
import safeCloneElement from '@instructure/ui-utils/lib/react/safeCloneElement';
import matchComponentTypes from '@instructure/ui-utils/lib/react/matchComponentTypes';
import getBoundingClientRect from '@instructure/ui-utils/lib/dom/getBoundingClientRect';
import px from '@instructure/ui-utils/lib/px';
import warning from '@instructure/ui-utils/lib/warning';

import { mirrorHorizontalPlacement } from '../../utils/mirrorPlacement';
import DrawerContent from './DrawerContent';
import DrawerTray from './DrawerTray';

var styles = {
  template: function template(theme) {
    var tmpl = function tmpl() {
      return '/*  imported from styles.css  */\n\n.qBBtv5E {\n  position: relative;\n  height: 100%;\n}\n';
    };

    return tmpl.call(theme, theme);
  },
  'root': 'qBBtv5E'
};

import theme from './theme';

/**
---
category: components/layout
---
**/
var DrawerLayout = (_dec = bidirectional(), _dec2 = themeable(theme, styles), _dec(_class = _dec2(_class = (_temp2 = _class2 = function (_Component) {
  _inherits(DrawerLayout, _Component);

  function DrawerLayout() {
    var _ref;

    var _temp, _this, _ret;

    _classCallCheck(this, DrawerLayout);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = DrawerLayout.__proto__ || Object.getPrototypeOf(DrawerLayout)).call.apply(_ref, [this].concat(args))), _this), _this.state = {
      shouldOverlayTray: false,
      trayWidth: 0,
      contentWidth: 0
    }, _this._content = null, _this.handleContentRef = function (el) {
      _this._content = el;
    }, _this.handleTrayContentRef = function (el) {
      _this._tray = el;
    }, _this.handleContentSizeChange = function (_ref2) {
      var width = _ref2.width;

      _this.setState(function (state, props) {
        var nextState = _this.getNextState(props.minWidth, state.trayWidth, width, state.shouldOverlayTray);

        if (state.shouldOverlayTray !== nextState.shouldOverlayTray) {
          _this.notifyOverlayTrayChange(nextState.shouldOverlayTray);
        }

        return nextState;
      });
    }, _this.handleTraySizeChange = function (_ref3) {
      var width = _ref3.width;

      _this.setState(function (state, props) {
        var nextState = _this.getNextState(props.minWidth, width, state.contentWidth, true);

        if (state.shouldOverlayTray !== nextState.shouldOverlayTray) {
          _this.notifyOverlayTrayChange(nextState.shouldOverlayTray);
        }

        return nextState;
      });
    }, _this.handleTrayTransitionEnter = function () {
      var width = 0;

      if (_this._tray) {
        width = getBoundingClientRect(_this._tray).width;
      }

      _this.handleTraySizeChange({ width: width });
    }, _this.handleTrayTransitionExit = function () {
      _this.handleTraySizeChange({ width: 0 });
    }, _temp), _possibleConstructorReturn(_this, _ret);
  }

  _createClass(DrawerLayout, [{
    key: 'getChildContext',
    value: function getChildContext() {
      return {
        shouldOverlayTray: this.state.shouldOverlayTray
      };
    }
  }, {
    key: 'shouldOverlayTray',
    value: function shouldOverlayTray(minWidth, trayWidth, contentWidth, trayIsOverlayed) {
      if (!this._content) return false;

      var minWidthPx = px(minWidth, this._content);

      if (trayIsOverlayed) {
        return contentWidth - trayWidth < minWidthPx;
      } else {
        return contentWidth < minWidthPx;
      }
    }
  }, {
    key: 'getNextState',
    value: function getNextState(minWidth, trayWidth, contentWidth, trayIsOverlayed) {
      var shouldOverlayTray = this.shouldOverlayTray(minWidth, trayWidth, contentWidth, trayIsOverlayed);

      return {
        trayWidth: trayWidth,
        contentWidth: contentWidth,
        shouldOverlayTray: shouldOverlayTray
      };
    }
  }, {
    key: 'notifyOverlayTrayChange',
    value: function notifyOverlayTrayChange(shouldOverlayTray) {
      var onOverlayTrayChange = this.props.onOverlayTrayChange;

      if (typeof onOverlayTrayChange === 'function') {
        onOverlayTrayChange(shouldOverlayTray);
      }
    }
  }, {
    key: 'renderChildren',
    value: function renderChildren() {
      var _this2 = this;

      var trayCount = 0;
      var contentCount = 0;

      var shouldOverlayTray = this.shouldOverlayTray(this.props.minWidth, this.state.trayWidth, this.state.contentWidth, this.state.shouldOverlayTray);

      var children = Children.map(this.props.children, function (child, index) {
        if (matchComponentTypes(child, [DrawerTray])) {
          trayCount++;
          return safeCloneElement(child, {
            key: child.props.label,
            contentRef: _this2.handleTrayContentRef,
            onEnter: _this2.handleTrayTransitionEnter,
            onExit: _this2.handleTrayTransitionExit
          });
        } else if (matchComponentTypes(child, [DrawerContent])) {
          contentCount++;
          return _this2.state.trayWidth !== null ? safeCloneElement(child, {
            key: child.props.label,
            style: _this2.contentStyle,
            onSizeChange: _this2.handleContentSizeChange,
            contentRef: _this2.handleContentRef,
            shouldTransition: !shouldOverlayTray
          }) : null;
        } else {
          return child;
        }
      });

      warning(trayCount <= 1, '[DrawerLayout] Only one DrawerTray per DrawerLayout is supported.');
      warning(contentCount <= 1, '[DrawerLayout] Only one DrawerContent per DrawerLayout is supported.');

      return children;
    }
  }, {
    key: 'render',
    value: function render() {
      return React.createElement(
        'div',
        { className: styles.root },
        this.renderChildren()
      );
    }
  }, {
    key: 'trayProps',
    get: function get() {
      var tray = Children.toArray(this.props.children).filter(function (child) {
        return matchComponentTypes(child, [DrawerTray]);
      })[0];
      return tray.props;
    }
  }, {
    key: 'trayPlacement',
    get: function get() {
      var placement = this.trayProps.placement;

      return this.rtl ? mirrorHorizontalPlacement(placement, ' ') : placement;
    }
  }, {
    key: 'contentMargin',
    get: function get() {
      var trayWidth = this.state.trayWidth || 0;
      return this.state.shouldOverlayTray ? 0 : trayWidth;
    }
  }, {
    key: 'contentStyle',
    get: function get() {
      var shouldOverlayTray = this.shouldOverlayTray(this.props.minWidth, this.state.trayWidth, this.state.contentWidth, this.state.shouldOverlayTray);
      var marginLeft = 0;
      var marginRight = 0;

      if (!shouldOverlayTray) {
        if (this.trayPlacement === 'start') {
          marginLeft = this.contentMargin;
        }

        if (this.trayPlacement === 'end') {
          marginRight = this.contentMargin;
        }
      }

      return {
        marginLeft: marginLeft + 'px',
        marginRight: marginRight + 'px'
      };
    }
  }]);

  DrawerLayout.displayName = 'DrawerLayout'
  ;
  return DrawerLayout;
}(Component), _class2.propTypes = {
  /**
   * Exactly one of each of the following child types: `DrawerContent`, `DrawerTray`
   */
  children: CustomPropTypes.Children.oneOfEach([DrawerContent, DrawerTray]),
  /**
   * Min width for the `<DrawerContent />`
   */
  minWidth: PropTypes.string,
  /**
   * Function called when the `<DrawerContent />` is resized and hits the `minWidth` breakpoint
   * Called with a boolean value, `true` if the tray is now overlaying the content or `false` if
   * it is side by side
   */
  onOverlayTrayChange: PropTypes.func
}, _class2.defaultProps = {
  children: null,
  minWidth: '30rem',
  onOverlayTrayChange: function onOverlayTrayChange(shouldOverlayTray) {}
}, _class2.childContextTypes = {
  shouldOverlayTray: PropTypes.bool
}, _temp2)) || _class) || _class);


export default DrawerLayout;
export { default as DrawerContent } from './DrawerContent';
export { default as DrawerTray } from './DrawerTray';
