// @instructure/ui-forms/lib/components/RangeInput/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _dec, _class, _class2, _temp; /*
                                   * The MIT License (MIT)
                                   *
                                   * Copyright (c) 2015 - present Instructure, Inc.
                                   *
                                   * Permission is hereby granted, free of charge, to any person obtaining a copy
                                   * of this software and associated documentation files (the "Software"), to deal
                                   * in the Software without restriction, including without limitation the rights
                                   * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
                                   * copies of the Software, and to permit persons to whom the Software is
                                   * furnished to do so, subject to the following conditions:
                                   *
                                   * The above copyright notice and this permission notice shall be included in all
                                   * copies or substantial portions of the Software.
                                   *
                                   * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
                                   * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
                                   * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
                                   * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
                                   * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
                                   * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
                                   * SOFTWARE.
                                   */

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _ContextView = require('@instructure/ui-layout/lib/components/ContextView');

var _ContextView2 = _interopRequireDefault(_ContextView);

var _uiThemeable = require('@instructure/ui-themeable');

var _uiThemeable2 = _interopRequireDefault(_uiThemeable);

var _CustomPropTypes = require('@instructure/ui-utils/lib/react/CustomPropTypes');

var _CustomPropTypes2 = _interopRequireDefault(_CustomPropTypes);

var _addEventListener = require('@instructure/ui-utils/lib/dom/addEventListener');

var _addEventListener2 = _interopRequireDefault(_addEventListener);

var _passthroughProps = require('@instructure/ui-utils/lib/react/passthroughProps');

var _generateElementId = require('@instructure/ui-utils/lib/dom/generateElementId');

var _generateElementId2 = _interopRequireDefault(_generateElementId);

var _Browser = require('@instructure/ui-utils/lib/Browser');

var _Browser2 = _interopRequireDefault(_Browser);

var _FormField = require('../FormField');

var _FormField2 = _interopRequireDefault(_FormField);

var _FormPropTypes = require('../../utils/FormPropTypes');

var _FormPropTypes2 = _interopRequireDefault(_FormPropTypes);

var _theme = require('./theme');

var _theme2 = _interopRequireDefault(_theme);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var styles = {
  template: function template(theme) {
    var tmpl = function tmpl() {
      return '/*  imported from styles.css  */\n\n/* stylelint-disable no-descending-specificity */\n\n._2boPe4d {\n  display: flex;\n  justify-content: center;\n  align-items: center;\n}\n\n[dir="rtl"] ._3M9I1hP {\n  direction: rtl;\n}\n\n._3M9I1hP {\n  animation: none 0s ease 0s 1 normal none running;\n  backface-visibility: visible;\n  background: transparent none repeat 0 0 / auto auto padding-box border-box scroll;\n  border: medium none currentColor;\n  border-collapse: separate;\n  border-image: none;\n  border-radius: 0;\n  border-spacing: 0;\n  bottom: auto;\n  box-shadow: none;\n  box-sizing: content-box;\n  caption-side: top;\n  clear: none;\n  clip: auto;\n  color: #000;\n  columns: auto;\n  column-count: auto;\n  column-fill: balance;\n  column-gap: normal;\n  column-rule: medium none currentColor;\n  column-span: 1;\n  column-width: auto;\n  content: normal;\n  counter-increment: none;\n  counter-reset: none;\n  cursor: auto;\n  direction: ltr;\n  display: inline;\n  empty-cells: show;\n  float: none;\n  font-family: serif;\n  font-size: medium;\n  font-style: normal;\n  font-variant: normal;\n  font-weight: normal;\n  font-stretch: normal;\n  line-height: normal;\n  height: auto;\n  hyphens: none;\n  left: auto;\n  letter-spacing: normal;\n  list-style: disc outside none;\n  margin: 0;\n  max-height: none;\n  max-width: none;\n  min-height: 0;\n  min-width: 0;\n  opacity: 1;\n  orphans: 2;\n  outline: medium none invert;\n  overflow: visible;\n  overflow-x: visible;\n  overflow-y: visible;\n  padding: 0;\n  page-break-after: auto;\n  page-break-before: auto;\n  page-break-inside: auto;\n  perspective: none;\n  perspective-origin: 50% 50%;\n  position: static;\n  right: auto;\n  tab-size: 8;\n  table-layout: auto;\n  text-align: left;\n  text-align-last: auto;\n  text-decoration: none;\n  text-indent: 0;\n  text-shadow: none;\n  text-transform: none;\n  top: auto;\n  transform: none;\n  transform-origin: 50% 50% 0;\n  transform-style: flat;\n  transition: none 0s ease 0s;\n  unicode-bidi: normal;\n  vertical-align: baseline;\n  visibility: visible;\n  white-space: normal;\n  widows: 2;\n  width: auto;\n  word-spacing: normal;\n  z-index: auto;\n  all: initial;\n  flex: auto;\n  box-sizing: border-box;\n  -webkit-appearance: none;\n     -moz-appearance: none;\n          appearance: none;\n  width: 100%; /* for Firefox */\n  outline: none;\n  margin: 0;\n  padding: 0;\n}\n\n/* stylelint-disable-next-line selector-pseudo-element-no-unknown */\n\n._3M9I1hP::-moz-range-thumb {\n    -moz-appearance: none;\n         appearance: none;\n    border-radius: 50%;\n    cursor: pointer;\n    transition: all 0.15s ease-in-out;\n    width: ' + theme.handleSize + ';\n    height: ' + theme.handleSize + ';\n    background: ' + theme.handleBackground + ';\n    box-shadow: 0 0.0625rem 0 ' + theme.handleShadowColor + ';\n  }\n\n._3M9I1hP::-moz-range-thumb:hover {\n      background: ' + theme.handleHoverBackground + ';\n    }\n\n/* for IE */\n\n._3M9I1hP::-ms-thumb {\n    appearance: none;\n    border-radius: 50%;\n    cursor: pointer;\n    transition: all 0.15s ease-in-out;\n    width: ' + theme.handleSize + ';\n    height: ' + theme.handleSize + ';\n    background: ' + theme.handleBackground + ';\n    box-shadow: 0 0.0625rem 0 ' + theme.handleShadowColor + ';\n  }\n\n._3M9I1hP::-ms-thumb:hover {\n      background: ' + theme.handleHoverBackground + ';\n    }\n\n._3M9I1hP::-webkit-slider-thumb {\n    -webkit-appearance: none;\n            appearance: none;\n    border-radius: 50%;\n    cursor: pointer;\n    transition: all 0.15s ease-in-out;\n    width: ' + theme.handleSize + ';\n    height: ' + theme.handleSize + ';\n    background: ' + theme.handleBackground + ';\n    box-shadow: 0 0.0625rem 0 ' + theme.handleShadowColor + ';\n  }\n\n._3M9I1hP::-webkit-slider-thumb:hover {\n      background: ' + theme.handleHoverBackground + ';\n    }\n\n._3M9I1hP::-webkit-slider-thumb {\n    margin-top: calc(-1 * ' + theme.handleSize + ' / 4);\n  }\n\n._3M9I1hP:focus,\n  ._3M9I1hP:active {\n    outline: none;\n  }\n\n/* stylelint-disable-next-line selector-pseudo-element-no-unknown */\n\n._3M9I1hP:focus::-moz-range-thumb, ._3M9I1hP:active::-moz-range-thumb {\n      box-shadow:\n        0 0.0625rem 0 ' + theme.handleShadowColor + ',\n        0 0 0 ' + theme.handleFocusOutlineWidth + ' ' + theme.handleFocusOutlineColor + ';\n    }\n\n._3M9I1hP:focus::-ms-thumb, ._3M9I1hP:active::-ms-thumb {\n      box-shadow:\n        0 0.0625rem 0 ' + theme.handleShadowColor + ',\n        0 0 0 ' + theme.handleFocusOutlineWidth + ' ' + theme.handleFocusOutlineColor + ';\n    }\n\n._3M9I1hP:focus::-webkit-slider-thumb, ._3M9I1hP:active::-webkit-slider-thumb {\n      box-shadow:\n        0 0.0625rem 0 ' + theme.handleShadowColor + ',\n        0 0 0 ' + theme.handleFocusOutlineWidth + ' ' + theme.handleFocusOutlineColor + ';\n    }\n\n/* remove outline in FF */\n\n._3M9I1hP::-moz-focus-inner,\n  ._3M9I1hP::-moz-focus-outer {\n    border: 0;\n    outline: none;\n  }\n\n/* stylelint-disable-next-line selector-pseudo-element-no-unknown */\n\n._3M9I1hP::-moz-range-track {\n    border-radius: 0.312em;\n    border-color: transparent;\n    color: transparent;\n    cursor: pointer;\n    background: ' + theme.trackBackground + ';\n    height: calc(' + theme.handleSize + ' / 2);\n  }\n\n._3M9I1hP::-ms-track {\n    border-radius: 0.312em;\n    border-color: transparent;\n    color: transparent;\n    cursor: pointer;\n    background: ' + theme.trackBackground + ';\n    height: calc(' + theme.handleSize + ' / 2);\n  }\n\n._3M9I1hP::-webkit-slider-runnable-track {\n    border-radius: 0.312em;\n    border-color: transparent;\n    color: transparent;\n    cursor: pointer;\n    background: ' + theme.trackBackground + ';\n    height: calc(' + theme.handleSize + ' / 2);\n  }\n\n._3M9I1hP::-ms-track {\n    background: transparent;\n    border-width: ' + theme.handleSize + ' 0;\n  }\n\n._3M9I1hP::-ms-fill-upper,\n  ._3M9I1hP::-ms-fill-lower {\n    border-width: ' + theme.handleSize + ' 0;\n    border-radius: calc(' + theme.handleSize + ' / 2);\n    background: ' + theme.trackBackground + ';\n  }\n\n._3M9I1hP::-ms-tooltip {\n    display: none;\n  }\n\n/* patch broken MSEdge keyboard focus styles in 16 and up */\n._20nM71D ._3M9I1hP:focus {\n  outline: 0.0625rem solid ' + theme.handleBackground + ';\n}\n\n.UOPWGdk {\n  animation: none 0s ease 0s 1 normal none running;\n  backface-visibility: visible;\n  background: transparent none repeat 0 0 / auto auto padding-box border-box scroll;\n  border: medium none currentColor;\n  border-collapse: separate;\n  border-image: none;\n  border-radius: 0;\n  border-spacing: 0;\n  bottom: auto;\n  box-shadow: none;\n  box-sizing: content-box;\n  caption-side: top;\n  clear: none;\n  clip: auto;\n  color: #000;\n  columns: auto;\n  column-count: auto;\n  column-fill: balance;\n  column-gap: normal;\n  column-rule: medium none currentColor;\n  column-span: 1;\n  column-width: auto;\n  content: normal;\n  counter-increment: none;\n  counter-reset: none;\n  cursor: auto;\n  direction: ltr;\n  display: inline;\n  empty-cells: show;\n  float: none;\n  font-family: serif;\n  font-size: medium;\n  font-style: normal;\n  font-variant: normal;\n  font-weight: normal;\n  font-stretch: normal;\n  line-height: normal;\n  height: auto;\n  hyphens: none;\n  left: auto;\n  letter-spacing: normal;\n  list-style: disc outside none;\n  margin: 0;\n  max-height: none;\n  max-width: none;\n  min-height: 0;\n  min-width: 0;\n  opacity: 1;\n  orphans: 2;\n  outline: medium none invert;\n  overflow: visible;\n  overflow-x: visible;\n  overflow-y: visible;\n  padding: 0;\n  page-break-after: auto;\n  page-break-before: auto;\n  page-break-inside: auto;\n  perspective: none;\n  perspective-origin: 50% 50%;\n  position: static;\n  right: auto;\n  tab-size: 8;\n  table-layout: auto;\n  text-align: left;\n  text-align-last: auto;\n  text-decoration: none;\n  text-indent: 0;\n  text-shadow: none;\n  text-transform: none;\n  top: auto;\n  transform: none;\n  transform-origin: 50% 50% 0;\n  transform-style: flat;\n  transition: none 0s ease 0s;\n  unicode-bidi: normal;\n  vertical-align: baseline;\n  visibility: visible;\n  white-space: normal;\n  widows: 2;\n  width: auto;\n  word-spacing: normal;\n  z-index: auto;\n  all: initial;\n  box-sizing: border-box;\n  display: block;\n  color: inherit;\n  font-family: ' + theme.valueFontFamily + ';\n  font-weight: ' + theme.valueFontWeight + ';\n  text-align: center;\n}\n\n[dir="ltr"] .UOPWGdk {\n  text-align: center;\n}\n\n[dir="rtl"] .UOPWGdk {\n  text-align: center;\n}\n\n._3QAuY0O .UOPWGdk {\n    font-size: ' + theme.valueSmallFontSize + ';\n    padding: ' + theme.valueSmallPadding + ';\n    line-height: ' + theme.valueSmallLineHeight + ';\n  }\n\n._33NRL6w .UOPWGdk {\n    font-size: ' + theme.valueMediumFontSize + ';\n    padding: ' + theme.valueMediumPadding + ';\n    line-height: ' + theme.valueMediumLineHeight + ';\n  }\n\n._37bg7Zw .UOPWGdk {\n    font-size: ' + theme.valueLargeFontSize + ';\n    padding: ' + theme.valueLargePadding + ';\n    line-height: ' + theme.valueLargeLineHeight + ';\n  }\n\n/* stylelint-enable no-descending-specificity */\n';
    };

    return tmpl.call(theme, theme);
  },
  'root': '_2boPe4d',
  'input': '_3M9I1hP',
  'edge16Up': '_20nM71D',
  'value': 'UOPWGdk',
  'small': '_3QAuY0O',
  'medium': '_33NRL6w',
  'large': '_37bg7Zw'
};


/**
---
category: components/forms
---
**/
var RangeInput = (_dec = (0, _uiThemeable2.default)(_theme2.default, styles), _dec(_class = (_temp = _class2 = function (_Component) {
  _inherits(RangeInput, _Component);

  function RangeInput(props) {
    _classCallCheck(this, RangeInput);

    var _this = _possibleConstructorReturn(this, (RangeInput.__proto__ || Object.getPrototypeOf(RangeInput)).call(this));

    _this.handleChange = function (event) {
      var _this$props = _this.props,
          onChange = _this$props.onChange,
          value = _this$props.value;


      if (typeof value === 'undefined') {
        _this.setState({ value: event.target.value });
      }

      if (typeof onChange === 'function') {
        onChange(event.target.value);
      }
    };

    _this.noopChange = function () {};

    if (typeof props.value === 'undefined') {
      _this.state = {
        value: props.defaultValue
      };
    }

    _this.defaultId = (0, _generateElementId2.default)('RangeInput');
    return _this;
  }

  /* workaround for https://github.com/facebook/react/issues/554 */


  _createClass(RangeInput, [{
    key: 'componentDidMount',
    value: function componentDidMount() {
      if (!this._input) {
        return;
      }
      // https://connect.microsoft.com/IE/Feedback/Details/856998
      this.inputListener = (0, _addEventListener2.default)(this._input, 'input', this.handleChange);
      this.changeListener = (0, _addEventListener2.default)(this._input, 'change', this.handleChange);
    }
  }, {
    key: 'componentWillUnmount',
    value: function componentWillUnmount() {
      if (!this._input) {
        return;
      }
      this.inputListener.remove();
      this.changeListener.remove();
    }
    /* end workaround */

    // controlled input must have an onChange, but we're handling it with native events

  }, {
    key: 'renderValue',
    value: function renderValue() {
      if (this.props.displayValue) {
        return _react2.default.createElement(
          _ContextView2.default,
          { background: 'inverse', placement: 'end center' },
          _react2.default.createElement(
            'output',
            { htmlFor: this.id, className: styles.value },
            this.props.formatValue(this.value)
          )
        );
      }
    }
  }, {
    key: 'render',
    value: function render() {
      var _classes,
          _this2 = this;

      var _props = this.props,
          formatValue = _props.formatValue,
          size = _props.size,
          disabled = _props.disabled,
          readOnly = _props.readOnly;


      var edge16Up = _Browser2.default.msedge && parseFloat(_Browser2.default.version) >= 16;

      var props = (0, _passthroughProps.omitProps)(this.props, RangeInput.propTypes);

      var classes = (_classes = {}, _defineProperty(_classes, styles.root, true), _defineProperty(_classes, styles[size], size), _defineProperty(_classes, styles.edge16Up, edge16Up), _classes);

      /* eslint-disable jsx-a11y/no-redundant-roles */
      return _react2.default.createElement(
        _FormField2.default,
        Object.assign({}, (0, _passthroughProps.pickProps)(this.props, _FormField2.default.propTypes), {
          id: this.id
        }),
        _react2.default.createElement(
          'div',
          { className: (0, _classnames2.default)(classes) },
          _react2.default.createElement('input', Object.assign({
            className: styles.input,
            ref: function ref(c) {
              _this2._input = c;
            },
            type: 'range',
            role: 'slider',
            id: this.id,
            min: this.props.min,
            max: this.props.max,
            step: this.props.step,
            value: this.value,
            onChange: this.noopChange,
            'aria-valuenow': this.value,
            'aria-valuemin': this.props.min,
            'aria-valuemax': this.props.max,
            'aria-valuetext': formatValue(this.value, this.props.max)
          }, props, {
            disabled: disabled || readOnly,
            'aria-disabled': disabled || readOnly ? 'true' : null
          })),
          this.renderValue()
        )
      );
      /* eslint-enable jsx-a11y/no-redundant-roles */
    }
  }, {
    key: 'value',
    get: function get() {
      return typeof this.props.value === 'undefined' ? this.state.value : this.props.value;
    }
  }, {
    key: 'invalid',
    get: function get() {
      return this.props.messages && this.props.messages.findIndex(function (message) {
        return message.type === 'error';
      }) >= 0;
    }
  }, {
    key: 'id',
    get: function get() {
      return this.props.id || this.defaultId;
    }
  }]);

  RangeInput.displayName = 'RangeInput'
  ;
  return RangeInput;
}(_react.Component), _class2.propTypes = {
  min: _propTypes2.default.number.isRequired,
  max: _propTypes2.default.number.isRequired,
  /**
  * value to set on initial render
  */
  defaultValue: _propTypes2.default.number,
  /**
  * the selected value (must be accompanied by an `onChange` prop)
  */
  value: _CustomPropTypes2.default.controllable(_propTypes2.default.number),
  /**
  * when used with the `value` prop, the component will not control its own state
  */
  onChange: _propTypes2.default.func,
  messages: _propTypes2.default.arrayOf(_FormPropTypes2.default.message),
  size: _propTypes2.default.oneOf(['small', 'medium', 'large']),
  layout: _propTypes2.default.oneOf(['stacked', 'inline']),
  id: _propTypes2.default.string,
  label: _propTypes2.default.node.isRequired,
  /**
  * whether to display the current value
  */
  displayValue: _propTypes2.default.bool,
  step: _propTypes2.default.number,
  /**
  * A function to format the displayed value
  */
  formatValue: _propTypes2.default.func,
  inline: _propTypes2.default.bool,
  disabled: _propTypes2.default.bool,
  readOnly: _propTypes2.default.bool
}, _class2.defaultProps = {
  step: 1,
  formatValue: function formatValue(val) {
    return val;
  },
  max: 0,
  min: 0,
  inline: false,
  size: 'medium',
  layout: 'stacked',
  displayValue: true,
  disabled: false,
  readOnly: false
}, _temp)) || _class);
exports.default = RangeInput;
