// @instructure/ui-forms/lib/components/NumberInput/theme.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = generator;

var _color = require('@instructure/ui-themeable/lib/utils/color');

function generator(_ref) {
  var colors = _ref.colors,
      typography = _ref.typography,
      borders = _ref.borders,
      spacing = _ref.spacing,
      forms = _ref.forms;

  return {
    fontFamily: typography.fontFamily,
    fontWeight: typography.fontWeightNormal,

    borderWidth: borders.widthSmall,
    borderStyle: borders.style,
    borderColor: colors.borderMedium,
    borderRadius: borders.radiusMedium,

    color: colors.textDarkest,
    background: colors.backgroundLightest,

    padding: '0 ' + spacing.small,

    arrowsContainerWidth: '2rem',
    arrowsColor: colors.textDarkest,
    arrowsBackgroundColor: colors.backgroundLight,
    arrowsHoverBackgroundColor: colors.backgroundMedium,
    arrowsBorderColor: colors.borderMedium,
    arrowsActiveBoxShadow: 'inset 0 0 3px 1px ' + colors.borderMedium,

    focusBorderColor: colors.borderBrand,
    focusOutlineColor: (0, _color.alpha)(colors.borderBrand, 50),

    errorBorderColor: colors.borderDanger,
    errorOutlineColor: (0, _color.alpha)(colors.borderDanger, 50),

    placeholderColor: colors.textDark,

    mediumFontSize: typography.fontSizeMedium,
    mediumHeight: forms.inputHeightMedium,

    largeFontSize: typography.fontSizeMedium,
    largeHeight: forms.inputHeightLarge
  };
} /*
   * The MIT License (MIT)
   *
   * Copyright (c) 2015 - present Instructure, Inc.
   *
   * Permission is hereby granted, free of charge, to any person obtaining a copy
   * of this software and associated documentation files (the "Software"), to deal
   * in the Software without restriction, including without limitation the rights
   * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
   * copies of the Software, and to permit persons to whom the Software is
   * furnished to do so, subject to the following conditions:
   *
   * The above copyright notice and this permission notice shall be included in all
   * copies or substantial portions of the Software.
   *
   * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
   * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
   * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
   * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
   * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
   * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
   * SOFTWARE.
   */

generator.canvas = function (variables) {
  return {
    color: variables['ic-brand-font-color-dark'],
    arrowsColor: variables['ic-brand-font-color-dark'],
    focusBorderColor: variables['ic-brand-primary'],
    focusOutlineColor: (0, _color.alpha)(variables['ic-brand-primary'], 50)
  };
};
