// @instructure/ui-forms/lib/components/Select/__examples__/Select.examples.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.persistentOptions = exports.multipleAutocomplete = exports.singleAutocomplete = exports.success = exports.error = exports.withOptGroup = exports.withIcon = exports.layoutDisabled = exports.layoutInline = exports.layoutStacked = exports.sizeLarge = exports.sizeMedium = exports.sizeSmall = undefined;

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _IconHeart = require('@instructure/ui-icons/lib/Line/IconHeart');

var _IconHeart2 = _interopRequireDefault(_IconHeart);

var _IconApple = require('@instructure/ui-icons/lib/Line/IconApple');

var _IconApple2 = _interopRequireDefault(_IconApple);

var _IconBank = require('@instructure/ui-icons/lib/Line/IconBank');

var _IconBank2 = _interopRequireDefault(_IconBank);

var _Button = require('../../../../../ui-buttons/lib/components/Button');

var _Button2 = _interopRequireDefault(_Button);

var _AccessibleContent = require('../../../../../ui-a11y/lib/components/AccessibleContent');

var _AccessibleContent2 = _interopRequireDefault(_AccessibleContent);

var _index = require('../index');

var _index2 = _interopRequireDefault(_index);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _toConsumableArray(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } else { return Array.from(arr); } }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; } /*
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * The MIT License (MIT)
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                *
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * Copyright (c) 2015 - present Instructure, Inc.
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                *
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * Permission is hereby granted, free of charge, to any person obtaining a copy
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * of this software and associated documentation files (the "Software"), to deal
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * in the Software without restriction, including without limitation the rights
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * copies of the Software, and to permit persons to whom the Software is
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * furnished to do so, subject to the following conditions:
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                *
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * The above copyright notice and this permission notice shall be included in all
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * copies or substantial portions of the Software.
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                *
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * SOFTWARE.
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                */

/* eslint-disable instructure-ui/no-relative-package-imports */

/* eslint-disable instructure-ui/no-relative-package-imports */


var sizeSmall = exports.sizeSmall = function sizeSmall() {
  return _react2.default.createElement(
    _index2.default,
    {
      size: 'small',
      label: 'Small',
      assistiveText: '3 options available. Use arrow keys to navigate options.'
    },
    _react2.default.createElement(
      'option',
      { value: 'foo' },
      'Foo'
    ),
    _react2.default.createElement(
      'option',
      { value: 'bar' },
      'Bar'
    ),
    _react2.default.createElement(
      'option',
      { value: 'baz' },
      'Baz'
    )
  );
};

var sizeMedium = exports.sizeMedium = function sizeMedium() {
  return _react2.default.createElement(
    _index2.default,
    {
      size: 'medium',
      label: 'Medium',
      assistiveText: '3 options available. Use arrow keys to navigate options.'
    },
    _react2.default.createElement(
      'option',
      { value: 'foo' },
      'Foo'
    ),
    _react2.default.createElement(
      'option',
      { value: 'bar' },
      'Bar'
    ),
    _react2.default.createElement(
      'option',
      { value: 'baz' },
      'Baz'
    )
  );
};

var sizeLarge = exports.sizeLarge = function sizeLarge() {
  return _react2.default.createElement(
    _index2.default,
    {
      size: 'large',
      label: 'Large',
      assistiveText: '3 options available. Use arrow keys to navigate options.'
    },
    _react2.default.createElement(
      'option',
      { value: 'foo' },
      'Foo'
    ),
    _react2.default.createElement(
      'option',
      { value: 'bar' },
      'Bar'
    ),
    _react2.default.createElement(
      'option',
      { value: 'baz' },
      'Baz'
    )
  );
};

var layoutStacked = exports.layoutStacked = function layoutStacked() {
  return _react2.default.createElement(
    _index2.default,
    {
      label: 'Stacked',
      layout: 'stacked',
      assistiveText: '3 options available. Use arrow keys to navigate options.'
    },
    _react2.default.createElement(
      'option',
      { value: 'foo' },
      'Foo'
    ),
    _react2.default.createElement(
      'option',
      { value: 'bar' },
      'Bar'
    ),
    _react2.default.createElement(
      'option',
      { value: 'baz' },
      'Baz'
    )
  );
};

var layoutInline = exports.layoutInline = function layoutInline() {
  return _react2.default.createElement(
    _index2.default,
    {
      label: 'Inline',
      layout: 'inline',
      assistiveText: '3 options available. Use arrow keys to navigate options.'
    },
    _react2.default.createElement(
      'option',
      { value: 'foo' },
      'Foo'
    ),
    _react2.default.createElement(
      'option',
      { value: 'bar' },
      'Bar'
    ),
    _react2.default.createElement(
      'option',
      { value: 'baz' },
      'Baz'
    )
  );
};

var layoutDisabled = exports.layoutDisabled = function layoutDisabled() {
  return _react2.default.createElement(
    _index2.default,
    {
      disabled: true,
      label: 'Disabled',
      assistiveText: '3 options available. Use arrow keys to navigate options.'
    },
    _react2.default.createElement(
      'option',
      { value: 'foo' },
      'Foo'
    ),
    _react2.default.createElement(
      'option',
      { value: 'bar' },
      'Bar'
    ),
    _react2.default.createElement(
      'option',
      { value: 'baz' },
      'Baz'
    )
  );
};

var withIcon = exports.withIcon = function withIcon() {
  return _react2.default.createElement(
    _index2.default,
    {
      label: 'Options with icons',
      assistiveText: '3 options available. Use arrow keys to navigate options.'
    },
    _react2.default.createElement(
      'option',
      { value: 'foo', icon: _IconHeart2.default },
      'Foo'
    ),
    _react2.default.createElement(
      'option',
      { value: 'bar', icon: _IconApple2.default },
      'Bar'
    ),
    _react2.default.createElement(
      'option',
      { value: 'baz', icon: _IconBank2.default },
      'Baz'
    )
  );
};

var withOptGroup = exports.withOptGroup = function withOptGroup() {
  return _react2.default.createElement(
    _index2.default,
    { label: 'Option Group Select' },
    _react2.default.createElement(
      'optgroup',
      { label: 'Group One' },
      _react2.default.createElement(
        'option',
        { value: 'item1' },
        'Item One'
      ),
      _react2.default.createElement(
        'option',
        { value: 'item2' },
        'Item Two'
      )
    ),
    _react2.default.createElement(
      'optgroup',
      { label: 'Group Two' },
      _react2.default.createElement(
        'option',
        { value: 'item3' },
        'Item Three'
      )
    ),
    _react2.default.createElement(
      'optgroup',
      { label: 'Group Three', icon: _IconHeart2.default },
      _react2.default.createElement(
        'option',
        { value: 'item4' },
        'Item Four'
      ),
      _react2.default.createElement(
        'option',
        { value: 'item5' },
        'Item Five'
      )
    )
  );
};

var error = exports.error = function error() {
  return _react2.default.createElement(
    _index2.default,
    {
      layout: 'inline',
      label: 'Choose a snack',
      messages: [{ text: 'You need to make a selection', type: 'error' }] },
    _react2.default.createElement(
      'option',
      { value: 'apples' },
      'Apples'
    ),
    _react2.default.createElement(
      'option',
      { value: 'oranges' },
      'Oranges'
    ),
    _react2.default.createElement(
      'option',
      { value: 'bananas' },
      'Bananas'
    ),
    _react2.default.createElement(
      'option',
      { value: 'candy', disabled: true },
      'Candy'
    )
  );
};

var success = exports.success = function success() {
  return _react2.default.createElement(
    _index2.default,
    {
      label: 'Choose a snack',
      messages: [{ text: 'Great job choosing something healthy!', type: 'success' }] },
    _react2.default.createElement(
      'option',
      { value: 'apples' },
      'Apples'
    ),
    _react2.default.createElement(
      'option',
      { value: 'oranges' },
      'Oranges'
    ),
    _react2.default.createElement(
      'option',
      { value: 'bananas' },
      'Bananas'
    ),
    _react2.default.createElement(
      'option',
      { value: 'candy', disabled: true },
      'Candy'
    )
  );
};

var AutocompleteExample = function (_React$Component) {
  _inherits(AutocompleteExample, _React$Component);

  function AutocompleteExample() {
    _classCallCheck(this, AutocompleteExample);

    return _possibleConstructorReturn(this, (AutocompleteExample.__proto__ || Object.getPrototypeOf(AutocompleteExample)).apply(this, arguments));
  }

  _createClass(AutocompleteExample, [{
    key: 'render',
    value: function render() {
      var options = ['Alabama', 'Alaska', 'American Samoa', 'Arizona', 'Arkansas', 'California', 'Colorado', 'Connecticut', 'Delaware', 'District Of Columbia', 'Federated States Of Micronesia', 'Florida', 'Georgia', 'Guam', 'Hawaii', 'Idaho', 'Illinois'];

      return _react2.default.createElement(
        _index2.default,
        Object.assign({}, this.props, {
          assistiveText: 'Start typing to search. Press the down arrow to navigate results.',
          formatSelectedOption: function formatSelectedOption(tag) {
            return _react2.default.createElement(
              _AccessibleContent2.default,
              { alt: 'Remove ' + tag.label },
              tag.label
            );
          }
        }),
        options.map(function (label, index) {
          return _react2.default.createElement(
            'option',
            { key: label, value: '' + index },
            label
          );
        })
      );
    }
  }]);

  AutocompleteExample.displayName = 'AutocompleteExample'
  ;
  return AutocompleteExample;
}(_react2.default.Component);

var singleAutocomplete = exports.singleAutocomplete = function singleAutocomplete() {
  return _react2.default.createElement(AutocompleteExample, {
    label: 'Choose a state',
    name: 'state',
    defaultOption: '12',
    editable: true
  });
};

var multipleAutocomplete = exports.multipleAutocomplete = function multipleAutocomplete() {
  return _react2.default.createElement(AutocompleteExample, {
    label: 'Choose a few states',
    name: 'states',
    defaultOption: ["0", "12", "15"],
    editable: true,
    multiple: true
  });
};

var PersistentExample = function (_React$Component2) {
  _inherits(PersistentExample, _React$Component2);

  function PersistentExample() {
    _classCallCheck(this, PersistentExample);

    var _this2 = _possibleConstructorReturn(this, (PersistentExample.__proto__ || Object.getPrototypeOf(PersistentExample)).apply(this, arguments));

    _this2.handleChange = function (e, opts) {
      var options = [].concat(_toConsumableArray(_this2.persistentOptions), _toConsumableArray(opts.slice(_this2.persistentOptions.length)));
      console.log(options); // eslint-disable-line no-console
      _this2.setState({ options: options });
    };

    _this2.clear = function () {
      _this2.setState({
        options: [].concat(_toConsumableArray(_this2.persistentOptions))
      });
    };

    _this2.persistentOptions = [{ label: 'Apples', value: '0', dismissible: false }, { label: 'Bananas', value: '1', dismissible: false }];
    _this2.state = {
      options: [].concat(_toConsumableArray(_this2.persistentOptions))
    };
    return _this2;
  }

  _createClass(PersistentExample, [{
    key: 'getOptionLabel',
    value: function getOptionLabel() {
      var options = this.state.options;

      return options.map(function (o) {
        return o.label;
      }).join(', ');
    }
  }, {
    key: 'render',
    value: function render() {
      return _react2.default.createElement(
        'div',
        null,
        _react2.default.createElement(
          _index2.default,
          Object.assign({
            multiple: true,
            label: 'Choose some fruit',
            editable: true
          }, this.props, {
            selectedOption: this.state.options,
            onChange: this.handleChange
          }),
          _react2.default.createElement(
            'option',
            { value: '0' },
            'Apples'
          ),
          _react2.default.createElement(
            'option',
            { value: '1' },
            'Bananas'
          ),
          _react2.default.createElement(
            'option',
            { value: '2' },
            'Oranges'
          ),
          _react2.default.createElement(
            'option',
            { value: '3' },
            'Mangoes'
          ),
          _react2.default.createElement(
            'option',
            { value: '4' },
            'Pears'
          ),
          _react2.default.createElement(
            'option',
            { value: '5' },
            'Cherries'
          )
        ),
        _react2.default.createElement(
          'div',
          null,
          'Value selected: ',
          this.getOptionLabel()
        ),
        _react2.default.createElement('br', null),
        _react2.default.createElement(
          _Button2.default,
          { onClick: this.clear },
          'Clear Selections'
        )
      );
    }
  }]);

  PersistentExample.displayName = 'PersistentExample'
  ;
  return PersistentExample;
}(_react2.default.Component);

var persistentOptions = exports.persistentOptions = function persistentOptions() {
  return _react2.default.createElement(PersistentExample, null);
};
