// @instructure/ui-elements/es/components/MetricsList/MetricsListItem/index.js
var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _dec, _class, _class2, _temp;

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import themeable from '@instructure/ui-themeable';

var styles = {
  template: function template(theme) {
    var tmpl = function tmpl() {
      return '/*  imported from styles.css  */\n\n.nDs1mGR {\n  display: flex;\n  flex-direction: column-reverse;\n  box-sizing: border-box;\n  padding: ' + theme.padding + ';\n  text-align: ' + theme.textAlign + ';\n  font-family: ' + theme.fontFamily + ';\n}\n\n[dir="ltr"] .nDs1mGR {\n  text-align: ' + theme.textAlign + ';\n}\n\n[dir="rtl"] .nDs1mGR {\n  text-align: ' + theme.textAlign + ';\n}\n\n  .nDs1mGR._3zK5Um9 {\n    text-align: left; /* stylelint-disable-line declaration-property-value-blacklist */\n  }\n\n  [dir="ltr"] .nDs1mGR._3zK5Um9 {\n    text-align: left;\n  }\n\n  [dir="rtl"] .nDs1mGR._3zK5Um9 {\n    text-align: left;\n  }\n\n  .nDs1mGR._1JeWGOU {\n    text-align: center;\n  }\n\n  [dir="ltr"] .nDs1mGR._1JeWGOU {\n    text-align: center;\n  }\n\n  [dir="rtl"] .nDs1mGR._1JeWGOU {\n    text-align: center;\n  }\n\n  .nDs1mGR.ZfslwQi {\n    text-align: right; /* stylelint-disable-line declaration-property-value-blacklist */\n  }\n\n  [dir="ltr"] .nDs1mGR.ZfslwQi {\n    text-align: right;\n  }\n\n  [dir="rtl"] .nDs1mGR.ZfslwQi {\n    text-align: right;\n  }\n\n._1D8MLFB {\n  font-size: ' + theme.labelFontSize + ';\n  color: ' + theme.labelColor + ';\n}\n\n._15LrdLM {\n  font-weight: ' + theme.valueFontWeight + ';\n  font-size: ' + theme.valueFontSize + ';\n  color: ' + theme.valueColor + ';\n}\n';
    };

    return tmpl.call(theme, theme);
  },
  'root': 'nDs1mGR',
  'start': '_3zK5Um9',
  'center': '_1JeWGOU',
  'end': 'ZfslwQi',
  'label': '_1D8MLFB',
  'value': '_15LrdLM'
};

import theme from './theme';

/**
---
parent: MetricsList
---
**/
var MetricsListItem = (_dec = themeable(theme, styles), _dec(_class = (_temp = _class2 = function (_Component) {
  _inherits(MetricsListItem, _Component);

  function MetricsListItem() {
    _classCallCheck(this, MetricsListItem);

    return _possibleConstructorReturn(this, (MetricsListItem.__proto__ || Object.getPrototypeOf(MetricsListItem)).apply(this, arguments));
  }

  _createClass(MetricsListItem, [{
    key: 'render',
    value: function render() {
      var _classes;

      var textAlign = this.props.textAlign;


      var classes = (_classes = {}, _defineProperty(_classes, styles.root, true), _defineProperty(_classes, styles[textAlign], true), _classes);

      return React.createElement(
        'div',
        { role: 'row',
          className: classnames(classes)
        },
        React.createElement(
          'div',
          { role: 'rowheader', className: styles.label },
          this.props.label
        ),
        React.createElement(
          'div',
          { role: 'gridcell', className: styles.value },
          this.props.value
        )
      );
    }
  }]);

  MetricsListItem.displayName = 'MetricsListItem'
  ;
  return MetricsListItem;
}(Component), _class2.propTypes = {
  textAlign: PropTypes.oneOf(['start', 'center', 'end']),
  label: PropTypes.node.isRequired,
  value: PropTypes.node.isRequired
}, _class2.defaultProps = {
  textAlign: 'center'
}, _temp)) || _class);
export { MetricsListItem as default };
