// @instructure/ui-layout/es/components/Flex/FlexItem/index.js
var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _dec, _class, _class2, _temp;

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

/*
* The MIT License (MIT)
*
* Copyright (c) 2015 - present Instructure, Inc.
*
* Permission is hereby granted, free of charge, to any person obtaining a copy
* of this software and associated documentation files (the "Software"), to deal
* in the Software without restriction, including without limitation the rights
* to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
* copies of the Software, and to permit persons to whom the Software is
* furnished to do so, subject to the following conditions:
*
* The above copyright notice and this permission notice shall be included in all
* copies or substantial portions of the Software.
*
* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
* IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
* FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
* AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
* LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
* OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
* SOFTWARE.
*/

import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';

import themeable from '@instructure/ui-themeable';
import CustomPropTypes from '@instructure/ui-utils/lib/react/CustomPropTypes';
import { omitProps } from '@instructure/ui-utils/lib/react/passthroughProps';
import ThemeablePropTypes from '@instructure/ui-themeable/lib/utils/ThemeablePropTypes';

import View from '../../View';

var styles = {
  template: function template(theme) {
    var tmpl = function tmpl() {
      return '/*  imported from styles.css  */\n\n._2POv4X5 {\n  box-sizing: border-box;\n  min-width: 0.0625rem;\n\n  /* initial value is 1, but we want 0 as our default,\n  so users can opt in to shrink like they do grow */\n  flex-shrink: 0;\n}\n\n._1_St6Qj {\n  /* useful defaults for column layouts */\n  overflow-y: auto;\n}\n\n._2qMq6pm {\n  align-self: flex-start;\n}\n\n.DkiK2ov {\n  align-self: flex-end;\n}\n\n.GT4JBNQ {\n  align-self: center;\n}\n\n._1VH1VK9 {\n  align-self: stretch;\n}\n\n.t4HSv-u {\n  flex-grow: 1;\n}\n\n._3MV7uLq {\n  flex-shrink: 1;\n}\n\n._1ryZZG_ {\n  outline: 0.0625rem dashed ' + theme.debugOutlineColor + ';\n}\n';
    };

    return tmpl.call(theme, theme);
  },
  'root': '_2POv4X5',
  'column': '_1_St6Qj',
  'align--start': '_2qMq6pm',
  'align--end': 'DkiK2ov',
  'align--center': 'GT4JBNQ',
  'align--stretch': '_1VH1VK9',
  'grow': 't4HSv-u',
  'shrink': '_3MV7uLq',
  'visualDebug': '_1ryZZG_'
};

import theme from './theme';

/**
---
parent: Flex
---
**/
var FlexItem = (_dec = themeable(theme, styles), _dec(_class = (_temp = _class2 = function (_Component) {
  _inherits(FlexItem, _Component);

  function FlexItem() {
    _classCallCheck(this, FlexItem);

    return _possibleConstructorReturn(this, (FlexItem.__proto__ || Object.getPrototypeOf(FlexItem)).apply(this, arguments));
  }

  _createClass(FlexItem, [{
    key: 'render',
    value: function render() {
      var _classes;

      var props = omitProps(this.props, FlexItem.propTypes);

      var _props = this.props,
          align = _props.align,
          as = _props.as,
          children = _props.children,
          direction = _props.direction,
          grow = _props.grow,
          margin = _props.margin,
          padding = _props.padding,
          shrink = _props.shrink,
          size = _props.size,
          textAlign = _props.textAlign,
          visualDebug = _props.visualDebug;


      var dirColumn = direction === 'column';

      var style = {
        flexBasis: size
      };

      var classes = (_classes = {}, _defineProperty(_classes, styles.root, true), _defineProperty(_classes, styles.visualDebug, visualDebug), _defineProperty(_classes, styles.grow, grow), _defineProperty(_classes, styles.shrink, shrink), _defineProperty(_classes, styles['align--' + align], align), _defineProperty(_classes, styles.column, dirColumn), _classes);

      return React.createElement(
        View,
        Object.assign({}, props, {
          className: classnames(classes),
          style: style,
          as: as,
          width: dirColumn ? '100%' : undefined // eslint-disable-line no-undefined
          , minHeight: dirColumn ? size : undefined // eslint-disable-line no-undefined
          , minWidth: direction === 'row' ? size : undefined // eslint-disable-line no-undefined
          , textAlign: textAlign,
          margin: margin,
          padding: padding
        }),
        children
      );
    }
  }]);

  FlexItem.displayName = 'FlexItem'
  ;
  return FlexItem;
}(Component), _class2.propTypes = {
  /**
  * overrides the parent Flex's alignItems prop, if needed
  */
  align: PropTypes.oneOf(['center', 'start', 'end', 'stretch']),
  /**
  * the element type to render as
  */
  as: CustomPropTypes.elementType,
  children: PropTypes.node,
  /**
  * Inherits from the parent Flex component
  */
  direction: PropTypes.oneOf(['row', 'column']),
  /**
  * Should the FlexItem grow to fill any available space?
  */
  grow: PropTypes.bool,
  /**
  * Should the FlexItem shrink (stopping at its `size`)?
  */
  shrink: PropTypes.bool,
  /**
  * Sets the base size of the FlexItem (width if direction is `row`; height if direction is `column`)
  */
  size: PropTypes.string,
  textAlign: PropTypes.oneOf(['start', 'center', 'end']),
  /**
  * Valid values are `0`, `none`, `auto`, `xxx-small`, `xx-small`, `x-small`,
  * `small`, `medium`, `large`, `x-large`, `xx-large`. Apply these values via
  * familiar CSS-like shorthand. For example: `margin="small auto large"`.
  */

  margin: ThemeablePropTypes.spacing,
  /**
  * Valid values are `0`, `none`, `xxx-small`, `xx-small`, `x-small`,
  * `small`, `medium`, `large`, `x-large`, `xx-large`. Apply these values via
  * familiar CSS-like shorthand. For example: `padding="small x-large large"`.
  */
  padding: ThemeablePropTypes.spacing,
  /**
  * Places dashed lines around the component's borders to help debug your layout
  */
  visualDebug: PropTypes.bool
}, _class2.defaultProps = {
  as: 'span',
  grow: false,
  shrink: false
}, _temp)) || _class);
export { FlexItem as default };
