// @instructure/ui-forms/es/components/RadioInput/index.js
var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _dec, _class, _class2, _temp;

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';

import generateElementId from '@instructure/ui-utils/lib/dom/generateElementId';
import themeable from '@instructure/ui-themeable';
import { omitProps } from '@instructure/ui-utils/lib/react/passthroughProps';
import isActiveElement from '@instructure/ui-utils/lib/dom/isActiveElement';

var styles = {
  template: function template(theme) {
    var tmpl = function tmpl() {
      return '/*  imported from styles.css  */\n\n/* stylelint-disable selector-max-class, max-nesting-depth, no-descending-specificity */\n\n._1OhsTme {\n  position: relative;\n  width: 100%;\n}\n\n._1OhsTme:hover {\n    cursor: default;\n  }\n\n.njnO0Yh {\n  animation: none 0s ease 0s 1 normal none running;\n  backface-visibility: visible;\n  background: transparent none repeat 0 0 / auto auto padding-box border-box scroll;\n  border: medium none currentColor;\n  border-collapse: separate;\n  border-image: none;\n  border-radius: 0;\n  border-spacing: 0;\n  bottom: auto;\n  box-shadow: none;\n  box-sizing: content-box;\n  caption-side: top;\n  clear: none;\n  clip: auto;\n  color: #000;\n  columns: auto;\n  column-count: auto;\n  column-fill: balance;\n  column-gap: normal;\n  column-rule: medium none currentColor;\n  column-span: 1;\n  column-width: auto;\n  content: normal;\n  counter-increment: none;\n  counter-reset: none;\n  cursor: auto;\n  direction: ltr;\n  display: inline;\n  empty-cells: show;\n  float: none;\n  font-family: serif;\n  font-size: medium;\n  font-style: normal;\n  font-variant: normal;\n  font-weight: normal;\n  font-stretch: normal;\n  line-height: normal;\n  height: auto;\n  hyphens: none;\n  left: auto;\n  letter-spacing: normal;\n  list-style: disc outside none;\n  margin: 0;\n  max-height: none;\n  max-width: none;\n  min-height: 0;\n  min-width: 0;\n  opacity: 1;\n  orphans: 2;\n  outline: medium none invert;\n  overflow: visible;\n  overflow-x: visible;\n  overflow-y: visible;\n  padding: 0;\n  page-break-after: auto;\n  page-break-before: auto;\n  page-break-inside: auto;\n  perspective: none;\n  perspective-origin: 50% 50%;\n  position: static;\n  right: auto;\n  tab-size: 8;\n  table-layout: auto;\n  text-align: left;\n  text-align-last: auto;\n  text-decoration: none;\n  text-indent: 0;\n  text-shadow: none;\n  text-transform: none;\n  top: auto;\n  transform: none;\n  transform-origin: 50% 50% 0;\n  transform-style: flat;\n  transition: none 0s ease 0s;\n  unicode-bidi: normal;\n  vertical-align: baseline;\n  visibility: visible;\n  white-space: normal;\n  widows: 2;\n  width: auto;\n  word-spacing: normal;\n  z-index: auto;\n  all: initial;\n  display: block;\n  direction: inherit;\n  text-align: start;\n}\n\n[dir="ltr"] .njnO0Yh {\n  text-align: left;\n}\n\n[dir="rtl"] .njnO0Yh {\n  text-align: right;\n}\n\n/* stylelint-disable selector-max-type */\n\n.Qunk8o_,\ninput.Qunk8o_[type="radio"] {\n  padding: 0;\n  margin: 0;\n  font-size: inherit;\n  line-height: inherit;\n  width: auto;\n  position: absolute;\n  top: 0;\n  left: 0; /* stylelint-disable-line property-blacklist */\n  opacity: 0.0001; /* selenium cannot find fully transparent elements */\n}\n\n/* stylelint-enable selector-max-type */\n\n._25TTlvP {\n  opacity: 0.5;\n}\n\n._25TTlvP:hover {\n    cursor: not-allowed;\n  }\n\n._1_SWJMU {\n  display: inline-block;\n  vertical-align: middle;\n  width: auto;\n}\n\n._2Gozmgm {\n  flex: 1 1 auto;\n  color: ' + theme.labelColor + ';\n  font-family: ' + theme.labelFontFamily + ';\n  font-weight: ' + theme.labelFontWeight + ';\n  line-height: ' + theme.labelLineHeight + ';\n}\n\n._2W83IG7 .njnO0Yh {\n    display: flex;\n    align-items: flex-start;\n  }\n\n._2W83IG7 ._3jMlXvg {\n    box-sizing: border-box;\n    display: block;\n    position: relative;\n    -webkit-margin-end: ' + theme.simpleFacadeMarginEnd + ';\n            margin-inline-end: ' + theme.simpleFacadeMarginEnd + ';\n    -webkit-margin-start: 0;\n            margin-inline-start: 0;\n    flex-shrink: 0;\n    min-width: 1rem;\n    transition: all 0.2s ease-out;\n    border-radius: 100%;\n    border: 0.125rem solid ' + theme.borderColor + ';\n    background: ' + theme.background + ';\n  }\n\n[dir="ltr"] ._2W83IG7 ._3jMlXvg {\n    margin-right: ' + theme.simpleFacadeMarginEnd + ';\n    margin-left: 0;\n  }\n\n[dir="rtl"] ._2W83IG7 ._3jMlXvg {\n    margin-left: ' + theme.simpleFacadeMarginEnd + ';\n    margin-right: 0;\n  }\n\n._2W83IG7 ._3jMlXvg::before {\n      content: "";\n      position: absolute;\n      top: -0.375rem;\n      left: -0.375rem; /* stylelint-disable-line property-blacklist */\n      width: calc(100% + 0.75rem);\n      height: calc(100% + 0.75rem);\n      box-sizing: border-box;\n      border-radius: 100%;\n      border: ' + theme.focusBorderWidth + ' ' + theme.focusBorderStyle + ' ' + theme.focusBorderColor + ';\n      transition: all 0.2s;\n      transform: scale(0.75) translate3d(0, 0, 0);\n      opacity: 0;\n      pointer-events: none;\n    }\n\n._2W83IG7._1hsCAK7 ._3jMlXvg {\n      width: ' + theme.simpleFacadeSmallSize + ';\n      height: ' + theme.simpleFacadeSmallSize + ';\n    }\n\n._2W83IG7._1hsCAK7 ._2Gozmgm {\n      font-size: ' + theme.simpleFontSizeSmall + ';\n    }\n\n._2W83IG7._1hsCAK7 .Qunk8o_:checked + .njnO0Yh ._3jMlXvg {\n      box-shadow: inset 0 0 0 ' + theme.simpleCheckedInsetSmall + ' ' + theme.hoverBorderColor + ';\n    }\n\n._2W83IG7._34CgfEA ._3jMlXvg {\n      width: ' + theme.simpleFacadeMediumSize + ';\n      height: ' + theme.simpleFacadeMediumSize + ';\n    }\n\n._2W83IG7._34CgfEA ._2Gozmgm {\n      font-size: ' + theme.simpleFontSizeMedium + ';\n    }\n\n._2W83IG7._34CgfEA .Qunk8o_:checked + .njnO0Yh ._3jMlXvg {\n      box-shadow: inset 0 0 0 ' + theme.simpleCheckedInsetMedium + ' ' + theme.hoverBorderColor + ';\n    }\n\n._2W83IG7._39G-Tu- ._3jMlXvg {\n      width: ' + theme.simpleFacadeLargeSize + ';\n      height: ' + theme.simpleFacadeLargeSize + ';\n    }\n\n._2W83IG7._39G-Tu- ._2Gozmgm {\n      font-size: ' + theme.simpleFontSizeLarge + ';\n    }\n\n._2W83IG7._39G-Tu- .Qunk8o_:checked + .njnO0Yh ._3jMlXvg {\n      box-shadow: inset 0 0 0 ' + theme.simpleCheckedInsetLarge + ' ' + theme.hoverBorderColor + ';\n    }\n\n._2W83IG7 .Qunk8o_:hover + .njnO0Yh ._3jMlXvg {\n      border-color: ' + theme.hoverBorderColor + ';\n    }\n\n._2W83IG7 .Qunk8o_:checked + .njnO0Yh ._3jMlXvg {\n      background: ' + theme.background + ';\n      box-shadow: inset 0 0 0 ' + theme.controlSize + ' ' + theme.hoverBorderColor + ';\n      border-color: ' + theme.hoverBorderColor + ';\n    }\n\n._2W83IG7 .Qunk8o_:focus + .njnO0Yh ._3jMlXvg {\n      background: ' + theme.background + ';\n    }\n\n._2W83IG7 .Qunk8o_:focus + .njnO0Yh ._3jMlXvg::before {\n        transform: scale(1) translate3d(0, 0, 0);\n        opacity: 1;\n      }\n\n._3PPL_r6 .njnO0Yh {\n    display: block;\n    -webkit-user-select: none;\n       -moz-user-select: none;\n        -ms-user-select: none;\n            user-select: none;\n    box-sizing: border-box;\n    position: relative;\n  }\n\n._3PPL_r6 ._2Gozmgm {\n    position: relative;\n    /* stylelint-disable-next-line scale-unlimited/declaration-strict-value */\n    z-index: 1;\n    text-transform: uppercase;\n    overflow: hidden;\n    white-space: nowrap;\n    text-overflow: ellipsis;\n    line-height: 1;\n    display: flex;\n    align-items: center;\n    min-width: 0.0625rem;\n  }\n\n._3PPL_r6 ._3jMlXvg {\n    box-sizing: border-box;\n    visibility: hidden;\n    display: block;\n    position: absolute;\n    /* stylelint-disable-next-line scale-unlimited/declaration-strict-value */\n    z-index: 1;\n    top: 0;\n    left: 0; /* stylelint-disable-line property-blacklist */\n    width: 100%;\n    height: 100%;\n    box-shadow: ' + theme.toggleShadow + ';\n    border-radius: ' + theme.toggleBorderRadius + ';\n  }\n\n._3PPL_r6 ._3jMlXvg::before {\n      content: "";\n      position: absolute;\n      top: -0.25rem;\n      left: -0.25rem; /* stylelint-disable-line property-blacklist */\n      width: calc(100% + 0.5rem);\n      height: calc(100% + 0.5rem);\n      box-sizing: border-box;\n      border-radius: calc(' + theme.toggleBorderRadius + ' + 0.0625rem);\n      border: ' + theme.focusBorderWidth + ' ' + theme.focusBorderStyle + ' ' + theme.focusBorderColor + ';\n      transition: all 0.2s;\n      transform: scale(0.66) translate3d(0, 0, 0);\n      opacity: 0;\n    }\n\n._3PPL_r6._1i9JKox ._3jMlXvg {\n    background-color: ' + theme.toggleBackgroundSuccess + ';\n  }\n\n._3PPL_r6._12DAKbE ._3jMlXvg {\n    background-color: ' + theme.toggleBackgroundDanger + ';\n  }\n\n._3PPL_r6._4dyU1aP ._3jMlXvg {\n    background-color: ' + theme.toggleBackgroundWarning + ';\n  }\n\n._3PPL_r6._2vReTFx ._3jMlXvg {\n    background-color: ' + theme.toggleBackgroundOff + ';\n  }\n\n._3PPL_r6._1hsCAK7 .njnO0Yh {\n    padding: 0 0.5rem;\n    height: ' + theme.toggleSmallHeight + ';\n  }\n\n._3PPL_r6._1hsCAK7 .njnO0Yh ._2Gozmgm {\n      font-size: ' + theme.toggleSmallFontSize + ';\n      height: ' + theme.toggleSmallHeight + ';\n\n      /* stylelint-enable selector-max-type */\n    }\n\n/* stylelint-disable selector-max-type */\n\n._3PPL_r6._1hsCAK7 .njnO0Yh ._2Gozmgm svg {\n        font-size: calc(' + theme.toggleSmallFontSize + ' + 0.375rem);\n      }\n\n._3PPL_r6._34CgfEA .njnO0Yh {\n    padding: 0 0.875rem;\n    height: ' + theme.toggleMediumHeight + ';\n  }\n\n._3PPL_r6._34CgfEA .njnO0Yh ._2Gozmgm {\n      font-size: ' + theme.toggleMediumFontSize + ';\n      height: ' + theme.toggleMediumHeight + ';\n\n      /* stylelint-enable selector-max-type */\n    }\n\n/* stylelint-disable selector-max-type */\n\n._3PPL_r6._34CgfEA .njnO0Yh ._2Gozmgm svg {\n        font-size: calc(' + theme.toggleMediumFontSize + ' + 0.375rem);\n      }\n\n._3PPL_r6._39G-Tu- .njnO0Yh {\n    padding: 0 1rem;\n    height: ' + theme.toggleLargeHeight + ';\n  }\n\n._3PPL_r6._39G-Tu- .njnO0Yh ._2Gozmgm {\n      font-size: ' + theme.toggleLargeFontSize + ';\n      height: ' + theme.toggleLargeHeight + ';\n\n      /* stylelint-enable selector-max-type */\n    }\n\n/* stylelint-disable selector-max-type */\n\n._3PPL_r6._39G-Tu- .njnO0Yh ._2Gozmgm svg {\n        font-size: calc(' + theme.toggleLargeFontSize + ' + 0.375rem);\n      }\n\n._3PPL_r6 .Qunk8o_:checked + .njnO0Yh ._3jMlXvg {\n      visibility: visible;\n    }\n\n._3PPL_r6 .Qunk8o_:checked + .njnO0Yh ._2Gozmgm {\n      color: ' + theme.toggleHandleText + ';\n    }\n\n._3PPL_r6 .Qunk8o_:focus + .njnO0Yh ._2Gozmgm {\n      text-decoration: underline;\n    }\n\n._3PPL_r6 .Qunk8o_:focus + .njnO0Yh ._3jMlXvg::before {\n        opacity: 1;\n        transform: scale(1) translate3d(0, 0, 0);\n      }\n\n/* stylelint-enable selector-max-class, no-descending-specificity, max-nesting-depth */\n';
    };

    return tmpl.call(theme, theme);
  },
  'root': '_1OhsTme',
  'control': 'njnO0Yh',
  'input': 'Qunk8o_',
  'disabled': '_25TTlvP',
  'inline': '_1_SWJMU',
  'label': '_2Gozmgm',
  'simple': '_2W83IG7',
  'facade': '_3jMlXvg',
  'small': '_1hsCAK7',
  'medium': '_34CgfEA',
  'large': '_39G-Tu-',
  'toggle': '_3PPL_r6',
  'success': '_1i9JKox',
  'danger': '_12DAKbE',
  'warning': '_4dyU1aP',
  'off': '_2vReTFx'
};

import theme from './theme';
/**
---
category: components/forms
---
**/
var RadioInput = (_dec = themeable(theme, styles), _dec(_class = (_temp = _class2 = function (_Component) {
  _inherits(RadioInput, _Component);

  function RadioInput(props) {
    _classCallCheck(this, RadioInput);

    var _this = _possibleConstructorReturn(this, (RadioInput.__proto__ || Object.getPrototypeOf(RadioInput)).call(this, props));

    _this.handleClick = function (e) {
      if (_this.props.disabled || _this.props.readOnly) {
        e.preventDefault();
        return;
      }

      _this.props.onClick(e);
    };

    _this.handleChange = function (e) {
      if (_this.props.disabled || _this.props.readOnly) {
        e.preventDefault();
        return;
      }

      if (typeof _this.props.checked === 'undefined') {
        _this.setState({ checked: !_this.state.checked });
      }

      _this.props.onChange(e);
    };

    _this.state = {};

    if (typeof props.checked === 'undefined') {
      _this.state.checked = false;
    }

    _this._defaultId = generateElementId('RadioInput');
    return _this;
  }

  _createClass(RadioInput, [{
    key: 'focus',
    value: function focus() {
      this._input.focus();
    }
  }, {
    key: 'render',
    value: function render() {
      var _classes,
          _this2 = this;

      var _props = this.props,
          disabled = _props.disabled,
          readOnly = _props.readOnly,
          label = _props.label,
          variant = _props.variant,
          size = _props.size,
          inline = _props.inline,
          context = _props.context,
          value = _props.value,
          name = _props.name;


      var props = omitProps(this.props, RadioInput.propTypes);

      var classes = (_classes = {}, _defineProperty(_classes, styles.root, true), _defineProperty(_classes, styles.disabled, disabled), _defineProperty(_classes, styles[variant], true), _defineProperty(_classes, styles[context], variant === 'toggle'), _defineProperty(_classes, styles[size], true), _defineProperty(_classes, styles['inline'], inline), _classes);

      return React.createElement(
        'div',
        { className: classnames(classes) },
        React.createElement('input', Object.assign({}, props, {
          id: this.id,
          ref: function ref(c) {
            _this2._input = c;
          },
          value: value,
          name: name,
          checked: this.checked,
          type: 'radio',
          className: styles.input,
          disabled: disabled || readOnly ? 'true' : null,
          'aria-disabled': disabled || readOnly ? 'true' : null,
          onChange: this.handleChange,
          onClick: this.handleClick
        })),
        React.createElement(
          'label',
          { className: styles.control, htmlFor: this.id },
          React.createElement('span', { className: styles.facade, 'aria-hidden': 'true' }),
          React.createElement(
            'span',
            { className: styles.label },
            label
          )
        )
      );
    }
  }, {
    key: 'id',
    get: function get() {
      return this.props.id || this._defaultId;
    }
  }, {
    key: 'focused',
    get: function get() {
      return isActiveElement(this._input);
    }
  }, {
    key: 'checked',
    get: function get() {
      return typeof this.props.checked === 'undefined' ? this.state.checked : this.props.checked;
    }
  }]);

  RadioInput.displayName = 'RadioInput'
  ;
  return RadioInput;
}(Component), _class2.propTypes = {
  label: PropTypes.node.isRequired,
  value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  id: PropTypes.string,
  name: PropTypes.string,
  checked: PropTypes.bool,
  /**
   * Whether or not to disable the input
   */
  disabled: PropTypes.bool,
  /**
   * Works just like disabled but keeps the same styles as if it were active
   */
  readOnly: PropTypes.bool,
  variant: PropTypes.oneOf(['simple', 'toggle']),
  size: PropTypes.oneOf(['small', 'medium', 'large']),
  context: PropTypes.oneOf(['success', 'warning', 'danger', 'off']),
  inline: PropTypes.bool,
  onClick: PropTypes.func,
  onChange: PropTypes.func
}, _class2.defaultProps = {
  onClick: function onClick(event) {},
  onChange: function onChange(event) {},
  variant: 'simple',
  size: 'medium',
  disabled: false,
  inline: false,
  context: 'success',
  readOnly: false
}, _temp)) || _class);


export default RadioInput;
