// @instructure/ui-forms/lib/components/TimeInput/__examples__/TimeInput.examples.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.readOnly = exports.disabled = exports.defaultToFirstOption = exports.withLocale = exports.withTimeZone = exports.withStep = exports.withDefaultValue = exports.withValue = exports.basic = undefined;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _index = require('../index');

var _index2 = _interopRequireDefault(_index);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

var basic = exports.basic = function basic() {
  return _react2.default.createElement(_index2.default, { label: 'Time' });
};

var withValue = exports.withValue = function withValue() {
  return _react2.default.createElement(_index2.default, {
    label: 'Time with value',
    value: '1986-05-17T19:00:00.000Z',
    onChange: function onChange(e) {
      console.log('onChange', e);
    } // eslint-disable-line no-console

  });
};

var withDefaultValue = exports.withDefaultValue = function withDefaultValue() {
  return _react2.default.createElement(_index2.default, {
    label: 'Time with defaultValue',
    defaultValue: '1986-05-17T14:00:00.000Z',
    onChange: function onChange(e) {
      console.log('onChange', e);
    } // eslint-disable-line no-console
  });
};

var withStep = exports.withStep = function withStep() {
  return _react2.default.createElement(_index2.default, {
    label: 'Time with step',
    step: 60,
    format: 'LTS'
  });
};

var withTimeZone = exports.withTimeZone = function withTimeZone() {
  return _react2.default.createElement(
    'div',
    { style: { height: 500, width: 500 } },
    _react2.default.createElement(_index2.default, {
      label: 'Eastern',
      value: new Date().toISOString(),
      timezone: 'US/Eastern',
      onChange: function onChange(e) {
        console.log('onChange', e);
      } // eslint-disable-line no-console
    }),
    _react2.default.createElement('br', null),
    _react2.default.createElement(_index2.default, {
      label: 'Central',
      value: new Date().toISOString(),
      timezone: 'US/Central',
      onChange: function onChange(e) {
        console.log('onChange', e);
      } // eslint-disable-line no-console
    }),
    _react2.default.createElement('br', null),
    _react2.default.createElement(_index2.default, {
      label: 'Mountain',
      value: new Date().toISOString(),
      timezone: 'US/Mountain',
      onChange: function onChange(e) {
        console.log('onChange', e);
      } // eslint-disable-line no-console
    }),
    _react2.default.createElement('br', null),
    _react2.default.createElement(_index2.default, {
      label: 'Western',
      value: new Date().toISOString(),
      timezone: 'US/Pacific',
      onChange: function onChange(e) {
        console.log('onChange', e);
      } // eslint-disable-line no-console
    })
  );
};

var withLocale = exports.withLocale = function withLocale() {
  return _react2.default.createElement(_index2.default, {
    label: 'Time with French locale',
    format: 'LTS',
    locale: 'fr'
  });
};

var defaultToFirstOption = exports.defaultToFirstOption = function defaultToFirstOption() {
  return _react2.default.createElement(_index2.default, {
    label: 'Time with defaultToFirstOption',
    defaultToFirstOption: true
  });
};

var disabled = exports.disabled = function disabled() {
  return _react2.default.createElement(_index2.default, {
    label: 'Disabled Time',
    format: 'LTS',
    defaultValue: '1986-05-17T18:00:00.000Z',
    disabled: true
  });
};

var readOnly = exports.readOnly = function readOnly() {
  return _react2.default.createElement(_index2.default, {
    label: 'Readonly Time',
    format: 'LTS',
    defaultValue: '1986-05-17T18:00:00.000Z',
    readOnly: true
  });
};
