// @instructure/ui-elements/lib/components/List/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ListItem = exports.default = undefined;

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _dec, _class, _class2, _temp; /*
                                   * The MIT License (MIT)
                                   *
                                   * Copyright (c) 2015 - present Instructure, Inc.
                                   *
                                   * Permission is hereby granted, free of charge, to any person obtaining a copy
                                   * of this software and associated documentation files (the "Software"), to deal
                                   * in the Software without restriction, including without limitation the rights
                                   * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
                                   * copies of the Software, and to permit persons to whom the Software is
                                   * furnished to do so, subject to the following conditions:
                                   *
                                   * The above copyright notice and this permission notice shall be included in all
                                   * copies or substantial portions of the Software.
                                   *
                                   * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
                                   * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
                                   * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
                                   * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
                                   * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
                                   * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
                                   * SOFTWARE.
                                   */


var _ListItem = require('./ListItem');

Object.defineProperty(exports, 'ListItem', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_ListItem).default;
  }
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _View = require('@instructure/ui-layout/lib/components/View');

var _View2 = _interopRequireDefault(_View);

var _uiThemeable = require('@instructure/ui-themeable');

var _uiThemeable2 = _interopRequireDefault(_uiThemeable);

var _passthroughProps = require('@instructure/ui-utils/lib/react/passthroughProps');

var _CustomPropTypes = require('@instructure/ui-utils/lib/react/CustomPropTypes');

var _CustomPropTypes2 = _interopRequireDefault(_CustomPropTypes);

var _ThemeablePropTypes = require('@instructure/ui-themeable/lib/utils/ThemeablePropTypes');

var _ThemeablePropTypes2 = _interopRequireDefault(_ThemeablePropTypes);

var _safeCloneElement = require('@instructure/ui-utils/lib/react/safeCloneElement');

var _safeCloneElement2 = _interopRequireDefault(_safeCloneElement);

var _ListItem2 = _interopRequireDefault(_ListItem);

var _theme = require('./theme');

var _theme2 = _interopRequireDefault(_theme);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var styles = {
  template: function template(theme) {
    var tmpl = function tmpl() {
      return '/*  imported from styles.css  */\n\n._2l3psub {\n  font-weight: ' + theme.fontWeight + ';\n  font-family: ' + theme.fontFamily + ';\n  padding: 0;\n}\n\n._2ObuHyI {\n  -webkit-padding-start: ' + theme.defaultListPadding + ';\n          padding-inline-start: ' + theme.defaultListPadding + ';\n  -webkit-padding-end: 0;\n          padding-inline-end: 0;\n}\n\n[dir="ltr"] ._2ObuHyI {\n  padding-left: ' + theme.defaultListPadding + ';\n  padding-right: 0;\n}\n\n[dir="rtl"] ._2ObuHyI {\n  padding-right: ' + theme.defaultListPadding + ';\n  padding-left: 0;\n}\n\n._2ObuHyI._3Qy6K6m {\n    list-style-type: none;\n    -webkit-padding-start: calc(' + theme.defaultListPadding + ' / 2);\n            padding-inline-start: calc(' + theme.defaultListPadding + ' / 2);\n    -webkit-padding-end: 0;\n            padding-inline-end: 0;\n    counter-reset: ol-counter;\n    /* stylelint-enable selector-max-type */\n  }\n\n[dir="ltr"] ._2ObuHyI._3Qy6K6m {\n    padding-left: calc(' + theme.defaultListPadding + ' / 2);\n    padding-right: 0;\n  }\n\n[dir="rtl"] ._2ObuHyI._3Qy6K6m {\n    padding-right: calc(' + theme.defaultListPadding + ' / 2);\n    padding-left: 0;\n  }\n\n/* stylelint-disable selector-max-type */\n\n._2ObuHyI._3Qy6K6m li::before {\n      font-weight: ' + theme.orderedNumberFontWeight + ';\n      counter-increment: ol-counter;\n      content: counters(ol-counter, ".") ".";\n      -webkit-margin-end: ' + theme.orderedNumberMargin + ';\n              margin-inline-end: ' + theme.orderedNumberMargin + ';\n      -webkit-margin-start: 0;\n              margin-inline-start: 0;\n    }\n\n[dir="ltr"] ._2ObuHyI._3Qy6K6m li::before {\n      margin-right: ' + theme.orderedNumberMargin + ';\n      margin-left: 0;\n    }\n\n[dir="rtl"] ._2ObuHyI._3Qy6K6m li::before {\n      margin-left: ' + theme.orderedNumberMargin + ';\n      margin-right: 0;\n    }\n\n._2-2PrF9,\n._1djVotZ,\n._2d9rysY {\n  list-style-type: none;\n  -webkit-padding-start: 0;\n          padding-inline-start: 0;\n}\n\n[dir="ltr"] ._2-2PrF9,\n[dir="ltr"] ._1djVotZ,\n[dir="ltr"] ._2d9rysY {\n  padding-left: 0;\n}\n\n[dir="rtl"] ._2-2PrF9,\n[dir="rtl"] ._1djVotZ,\n[dir="rtl"] ._2d9rysY {\n  padding-right: 0;\n}\n\n._3wFfR2D {}\n';
    };

    return tmpl.call(theme, theme);
  },
  'root': '_2l3psub',
  'default': '_2ObuHyI',
  'ordered': '_3Qy6K6m',
  'unstyled': '_2-2PrF9',
  'inline': '_1djVotZ',
  'steps': '_2d9rysY',
  'unordered': '_3wFfR2D'
};


/**
---
category: components/typography
---
**/
var List = (_dec = (0, _uiThemeable2.default)(_theme2.default, styles), _dec(_class = (_temp = _class2 = function (_Component) {
  _inherits(List, _Component);

  function List() {
    _classCallCheck(this, List);

    return _possibleConstructorReturn(this, (List.__proto__ || Object.getPrototypeOf(List)).apply(this, arguments));
  }

  _createClass(List, [{
    key: 'renderChildren',
    value: function renderChildren() {
      var _this2 = this;

      return _react.Children.map(this.props.children, function (child) {
        if (!child) return; // ignore null, falsy children

        return (0, _safeCloneElement2.default)(child, {
          variant: _this2.props.variant,
          delimiter: _this2.props.delimiter,
          size: _this2.props.size,
          as: _this2.props.as,
          spacing: _this2.props.itemSpacing
        });
      });
    }
  }, {
    key: 'render',
    value: function render() {
      var _classes;

      var props = (0, _passthroughProps.omitProps)(this.props, Object.assign({}, List.propTypes, _View2.default.propTypes));

      var _props = this.props,
          as = _props.as,
          margin = _props.margin,
          variant = _props.variant,
          elementRef = _props.elementRef;


      var classes = (_classes = {}, _defineProperty(_classes, styles.root, true), _defineProperty(_classes, styles[variant], variant), _defineProperty(_classes, styles.unordered, as === 'ul'), _defineProperty(_classes, styles.ordered, as === 'ol'), _classes);

      return _react2.default.createElement(
        _View2.default,
        Object.assign({}, props, {
          className: (0, _classnames2.default)(classes),
          as: as,
          margin: margin,
          elementRef: elementRef
        }),
        this.renderChildren()
      );
    }
  }]);

  List.displayName = 'List'
  ;
  return List;
}(_react.Component), _class2.propTypes = {
  /**
  * Only accepts <ListItem> as a child
  */
  children: _CustomPropTypes2.default.Children.oneOf([_ListItem2.default]),
  as: _propTypes2.default.oneOf(['ul', 'ol']),
  /**
  * Valid values are `0`, `none`, `auto`, `xxx-small`, `xx-small`, `x-small`,
  * `small`, `medium`, `large`, `x-large`, `xx-large`. Apply these values via
  * familiar CSS-like shorthand. For example: `margin="small auto large"`.
  */
  margin: _ThemeablePropTypes2.default.spacing,
  size: _propTypes2.default.oneOf(['small', 'medium', 'large']),
  variant: _propTypes2.default.oneOf(['default', 'unstyled', 'inline']),
  delimiter: _propTypes2.default.oneOf(['none', 'pipe', 'slash', 'arrow', 'dashed', 'solid']),
  /**
  * Sets the margin separating each ListItem.
  */
  itemSpacing: _propTypes2.default.oneOf(['xxx-small', 'xx-small', 'x-small', 'small', 'medium', 'large', 'x-large', 'xx-large']),
  elementRef: _propTypes2.default.func
}, _class2.defaultProps = {
  as: 'ul',
  margin: 'none',
  variant: 'default',
  delimiter: 'none',
  size: 'medium'
}, _temp)) || _class);
exports.default = List;
