// @instructure/ui-elements/lib/components/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TruncateText = exports.Text = exports.Tag = exports.Table = exports.Spinner = exports.Rating = exports.ProgressCircle = exports.ProgressBar = exports.Progress = exports.Pill = exports.MetricsListItem = exports.MetricsList = exports.ListItem = exports.List = exports.Link = exports.Img = exports.Heading = exports.Badge = exports.Avatar = undefined;

var _List2 = require('./List');

Object.defineProperty(exports, 'ListItem', {
  enumerable: true,
  get: function get() {
    return _List2.ListItem;
  }
});

var _MetricsList2 = require('./MetricsList');

Object.defineProperty(exports, 'MetricsListItem', {
  enumerable: true,
  get: function get() {
    return _MetricsList2.MetricsListItem;
  }
});

var _Progress2 = require('./Progress');

Object.defineProperty(exports, 'ProgressBar', {
  enumerable: true,
  get: function get() {
    return _Progress2.ProgressBar;
  }
});
Object.defineProperty(exports, 'ProgressCircle', {
  enumerable: true,
  get: function get() {
    return _Progress2.ProgressCircle;
  }
});

var _Avatar2 = require('./Avatar');

var _Avatar3 = _interopRequireDefault(_Avatar2);

var _Badge2 = require('./Badge');

var _Badge3 = _interopRequireDefault(_Badge2);

var _Heading2 = require('./Heading');

var _Heading3 = _interopRequireDefault(_Heading2);

var _Img2 = require('./Img');

var _Img3 = _interopRequireDefault(_Img2);

var _Link2 = require('./Link');

var _Link3 = _interopRequireDefault(_Link2);

var _List3 = _interopRequireDefault(_List2);

var _MetricsList3 = _interopRequireDefault(_MetricsList2);

var _Pill2 = require('./Pill');

var _Pill3 = _interopRequireDefault(_Pill2);

var _Progress3 = _interopRequireDefault(_Progress2);

var _Rating2 = require('./Rating');

var _Rating3 = _interopRequireDefault(_Rating2);

var _Spinner2 = require('./Spinner');

var _Spinner3 = _interopRequireDefault(_Spinner2);

var _Table2 = require('./Table');

var _Table3 = _interopRequireDefault(_Table2);

var _Tag2 = require('./Tag');

var _Tag3 = _interopRequireDefault(_Tag2);

var _Text2 = require('./Text');

var _Text3 = _interopRequireDefault(_Text2);

var _TruncateText2 = require('./TruncateText');

var _TruncateText3 = _interopRequireDefault(_TruncateText2);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.Avatar = _Avatar3.default; /*
                                    * The MIT License (MIT)
                                    *
                                    * Copyright (c) 2015 - present Instructure, Inc.
                                    *
                                    * Permission is hereby granted, free of charge, to any person obtaining a copy
                                    * of this software and associated documentation files (the "Software"), to deal
                                    * in the Software without restriction, including without limitation the rights
                                    * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
                                    * copies of the Software, and to permit persons to whom the Software is
                                    * furnished to do so, subject to the following conditions:
                                    *
                                    * The above copyright notice and this permission notice shall be included in all
                                    * copies or substantial portions of the Software.
                                    *
                                    * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
                                    * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
                                    * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
                                    * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
                                    * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
                                    * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
                                    * SOFTWARE.
                                    */

exports.Badge = _Badge3.default;
exports.Heading = _Heading3.default;
exports.Img = _Img3.default;
exports.Link = _Link3.default;
exports.List = _List3.default;
exports.MetricsList = _MetricsList3.default;
exports.Pill = _Pill3.default;
exports.Progress = _Progress3.default;
exports.Rating = _Rating3.default;
exports.Spinner = _Spinner3.default;
exports.Table = _Table3.default;
exports.Tag = _Tag3.default;
exports.Text = _Text3.default;
exports.TruncateText = _TruncateText3.default;
